\PassOptionsToPackage{unicode=true}{hyperref} % options for packages loaded elsewhere
\PassOptionsToPackage{hyphens}{url}
\PassOptionsToPackage{dvipsnames,svgnames*,x11names*}{xcolor}
%
\documentclass[12pt,]{krantz}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provides euro and other symbols
\else % if luatex or xelatex
  \usepackage{unicode-math}
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
    \setmonofont[Mapping=tex-ansi,Scale=0.7]{Source Code Pro}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage[]{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\usepackage{xcolor}
\usepackage{hyperref}
\hypersetup{
            pdftitle={Computational Genomics With R},
            pdfauthor={Altuna Akalin},
            colorlinks=true,
            linkcolor=Maroon,
            filecolor=Maroon,
            citecolor=Blue,
            urlcolor=Blue,
            breaklinks=true}
\urlstyle{same}  % don't use monospace font for urls
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{longtable,booktabs}
% Fix footnotes in tables (requires footnote package)
\IfFileExists{footnote.sty}{\usepackage{footnote}\makesavenoteenv{longtable}}{}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

% set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother

\usepackage{booktabs}
\usepackage{longtable}
\usepackage[bf,singlelinecheck=off]{caption}

\setmainfont[UprightFeatures={SmallCapsFont=AlegreyaSC-Regular}]{Alegreya}

\usepackage{framed,color}
\definecolor{shadecolor}{RGB}{248,248,248}

\renewcommand{\textfraction}{0.05}
\renewcommand{\topfraction}{0.8}
\renewcommand{\bottomfraction}{0.8}
\renewcommand{\floatpagefraction}{0.75}

\renewenvironment{quote}{\begin{VF}}{\end{VF}}
\let\oldhref\href
\renewcommand{\href}[2]{#2\footnote{\url{#1}}}

\ifxetex
  \usepackage{letltxmacro}
  \setlength{\XeTeXLinkMargin}{1pt}
  \LetLtxMacro\SavedIncludeGraphics\includegraphics
  \def\includegraphics#1#{% #1 catches optional stuff (star/opt. arg.)
    \IncludeGraphicsAux{#1}%
  }%
  \newcommand*{\IncludeGraphicsAux}[2]{%
    \XeTeXLinkBox{%
      \SavedIncludeGraphics#1{#2}%
    }%
  }%
\fi

\makeatletter
\newenvironment{kframe}{%
\medskip{}
\setlength{\fboxsep}{.8em}
 \def\at@end@of@kframe{}%
 \ifinner\ifhmode%
  \def\at@end@of@kframe{\end{minipage}}%
  \begin{minipage}{\columnwidth}%
 \fi\fi%
 \def\FrameCommand##1{\hskip\@totalleftmargin \hskip-\fboxsep
 \colorbox{shadecolor}{##1}\hskip-\fboxsep
     % There is no \\@totalrightmargin, so:
     \hskip-\linewidth \hskip-\@totalleftmargin \hskip\columnwidth}%
 \MakeFramed {\advance\hsize-\width
   \@totalleftmargin\z@ \linewidth\hsize
   \@setminipage}}%
 {\par\unskip\endMakeFramed%
 \at@end@of@kframe}
\makeatother

\makeatletter
\@ifundefined{Shaded}{
}{\renewenvironment{Shaded}{\begin{kframe}}{\end{kframe}}}
\makeatother

\newenvironment{rmdblock}[1]
  {
  \begin{itemize}
  \renewcommand{\labelitemi}{
    \raisebox{-.7\height}[0pt][0pt]{
      {\setkeys{Gin}{width=3em,keepaspectratio}\includegraphics{images/#1}}
    }
  }
  \setlength{\fboxsep}{1em}
  \begin{kframe}
  \item
  }
  {
  \end{kframe}
  \end{itemize}
  }
\newenvironment{rmdnote}
  {\begin{rmdblock}{note}}
  {\end{rmdblock}}
\newenvironment{rmdcaution}
  {\begin{rmdblock}{caution}}
  {\end{rmdblock}}
\newenvironment{rmdimportant}
  {\begin{rmdblock}{important}}
  {\end{rmdblock}}
\newenvironment{rmdtip}
  {\begin{rmdblock}{tip}}
  {\end{rmdblock}}
\newenvironment{rmdwarning}
  {\begin{rmdblock}{warning}}
  {\end{rmdblock}}

\usepackage{makeidx}
\makeindex

\urlstyle{tt}

\usepackage{amsthm}
\makeatletter
\def\thm@space@setup{%
  \thm@preskip=8pt plus 2pt minus 4pt
  \thm@postskip=\thm@preskip
}
\makeatother

\frontmatter
\usepackage[]{natbib}
\bibliographystyle{apalike}

\title{Computational Genomics With R}
\author{Altuna Akalin}
\date{2019-05-10}

\let\BeginKnitrBlock\begin \let\EndKnitrBlock\end
\begin{document}
\maketitle

%\cleardoublepage\newpage\thispagestyle{empty}\null
%\cleardoublepage\newpage\thispagestyle{empty}\null
%\cleardoublepage\newpage
\thispagestyle{empty}
\begin{center}
\includegraphics{images/dedication.pdf}
\end{center}

\setlength{\abovedisplayskip}{-5pt}
\setlength{\abovedisplayshortskip}{-5pt}

{
\hypersetup{linkcolor=}
\setcounter{tocdepth}{2}
\tableofcontents
}
\listoftables
\listoffigures
\hypertarget{preface}{%
\chapter*{Preface}\label{preface}}


The aim of this book is to provide the fundamentals for data analysis for genomics. We developed this book based on the computational genomics courses we are giving every year. We have had invariably an interdisicplinary audience with backgrounds from physics, biology, medicine, math, computer science or other quantitative fields. We want this book to be a starting point for computational genomics students and a guide for further data analysis in more specific topics in genomics. This is why we tried to cover a large variety of topics from programming to basic genome biology. As the field is interdisciplinary, it requires different starting points for people with different backgrounds. A biologist might skip sections on basic genome biology and start with R programming whereas a computer scientist might want to start with genome biology. In the same manner, a more experienced person might want to refer to this book when s/he needs to do a certain type of analysis where s/he does not have prior experience.

\includegraphics{images/by-nc-sa.png}\\
The online version of this book is licensed under the \href{http://creativecommons.org/licenses/by-nc-sa/4.0/}{Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License}.

\hypertarget{who-is-this-book-for}{%
\section*{Who is this book for?}\label{who-is-this-book-for}}


The book contains practical and theoretical aspects for computational genomics. Biology and
medicine generate more data than ever before. Therefore, we need to educate more people with data
analysis skills and understanding of computational genomics.
Since computational genomics is interdisciplinary, this book aims to be accessible for
biologists, medical scientists, computer scientists and people from other quantitative backgrounds. We wrote this book for the following audiences:

\begin{itemize}
\tightlist
\item
  Biologists and medical scientists who generate the data and are keen on analyzing it themselves.
\item
  Students and researchers who are formally starting to do research on or using computational genomics but do not have extensive domain specific knowledge but have at least a beginner level understanding in a quantitative field: math, stats
\item
  Experienced researchers looking for recipes or quick how-tos to get started in specific data analysis tasks related to computational genomics.
\end{itemize}

\hypertarget{what-will-you-get-out-of-this}{%
\subsection*{What will you get out of this?}\label{what-will-you-get-out-of-this}}


This resource describes the skills and provides how-tos that will help readers
analyze their own genomics data.

After reading:

\begin{itemize}
\tightlist
\item
  If you are not familiar with R, you will get the basics of R and dive right in to specialized uses of R for computational genomics.
\item
  You will understand genomic intervals and operations on them, such as overlap
\item
  You will be able to use R and its vast package library to do sequence analysis: Such as calculating GC content for given segments of a genome or find transcription factor binding sites
\item
  You will be familiar with visualization techniques used in genomics, such as heatmaps, meta-gene plots and genomic track visualization
\item
  You will be familiar with supervised and unsupervised learning techniques which are important in data modelling and exploratory analysis of high-dimensional data
\item
  You will be familiar with analysis of different high-throughput sequencing data
  sets mostly using R based tools.
\end{itemize}

\hypertarget{structure-of-the-book}{%
\section*{Structure of the book}\label{structure-of-the-book}}


The book is designed with the idea that practical and conceptual
understanding of data analysis methods is as important, if not more important, than the theoretical understanding, such as detailed derivation of equations in statistics or machine learning. That is why we first try to give a conceptual explanation of the concepts then we try to give essential parts of the mathematical formulas for more detailed understanding. In this sprit, we always show the code and
explain the code for a particular data analysis task. In addition, we give additional references such as books, websites , video lectures and scientific papers for readers who desire to gain deeper theoretical understanding of data analysis related methods or concepts.

``Introduction to Genomics'' chapter introduces the basic concepts in genome biology and genomics. Understanding these concepts are important for computational genomics.

``Introduction to R for Genomic Data Analysis'' chapter provides basic R skills necessary to follow the book in addition to common data analysis paradigms we observe in genomic data analysis. The ``Statistics for Genomics'', ``Exploratory Data Analysis with Unsupervised Machine Learning'' and ``Predictive Modeling with Supervised Machine Learning'' chapters introduces the necessary quantitative skills that one might need when analyzing high-dimensional genomics data.

The ``Operations on Genomic Intervals and Genome Arithmetic'' chapter introduces the fundamental tools for dealing with genomic intervals and their relationship to each other over the genome. In addition, the chapter introduces a variety of ways used in genomic data visualization. The skills introduced in this chapter are key skills that are needed to work with processed genomic data which are available through public databases such as Ensembl and UCSC browser.

The next chapters deals with specific analysis of high-throughput sequencing data and integrating different kinds of data sets. The ``Quality Check, Processing and Alignment of High-throughput Sequencing Reads'' chapter introduces quality checks that need to be done on sequencing reads and different ways to process them further. The chapters ``RNA-seq Analysis'', ``ChIP-seq Analysis'' and ``BS-seq Analysis'' deals
with the analysis techniques for popular high-throughput sequencing techniques. The last chapter ``Multi-omics Analysis'' deals with methods for integrating multiple omics data sets.

To sum it up, this book is a comprehensive guide for computational genomics. Some sections are there for the sake of the wide interdisciplinary audience and completeness, and not all sections will be equally useful to all readers of this broad audience.

\hypertarget{software-information-and-conventions}{%
\section*{Software information and conventions}\label{software-information-and-conventions}}


Package names are in bold text (e.g., \textbf{methylKit}), and inline code and filenames are formatted in a typewriter font. Function names are followed by parentheses (e.g., \texttt{genomation::ScoreMatrix()}). The double-colon operator \texttt{::} means accessing an object from a package.

\hypertarget{assignment-operator-convention}{%
\subsection*{Assignment operator convention}\label{assignment-operator-convention}}


Traditionally, \texttt{\textless{}-} is the preffered assignment operator. However, throughout the book we use \texttt{=} and \texttt{\textless{}-} as the assignment operator interchangeably.

\hypertarget{packages-needed-to-run-the-book-code}{%
\subsection*{Packages needed to run the book code}\label{packages-needed-to-run-the-book-code}}


This book is primarily about using R packages to analyze genomics data, therefore if you want to reproduce the analysis in this book you need to install the relevant packages in each chapter using \texttt{install.packages} or \texttt{BiocManager::install} functions. In each chapter, we load the necessary packages with the \texttt{library()} or \texttt{require()} function when we use the needed functions from respective packages. By looking at that calls, you can see which packages are needed for that code chunk or chapter. If you need to install all the package dependencies for the book, you can run the following command and have a cup of tea while waiting.

\begin{Shaded}
\begin{Highlighting}[]
\ControlFlowTok{if}\NormalTok{ (}\OperatorTok{!}\KeywordTok{requireNamespace}\NormalTok{(}\StringTok{"BiocManager"}\NormalTok{, }\DataTypeTok{quietly =} \OtherTok{TRUE}\NormalTok{))}
    \KeywordTok{install.packages}\NormalTok{(}\StringTok{"BiocManager"}\NormalTok{)}
\NormalTok{BiocManager}\OperatorTok{::}\KeywordTok{install}\NormalTok{(}\KeywordTok{c}\NormalTok{(}\StringTok{'qvalue'}\NormalTok{,}\StringTok{'plot3D'}\NormalTok{,}\StringTok{'ggplot2'}\NormalTok{,}\StringTok{'pheatmap'}\NormalTok{,}
                      \StringTok{'cluster'}\NormalTok{, }\StringTok{'NbClust'}\NormalTok{, }\StringTok{'fastICA'}\NormalTok{, }\StringTok{'NMF'}\NormalTok{,}
                      \StringTok{'Rtsne'}\NormalTok{, }\StringTok{'mosaic'}\NormalTok{, }\StringTok{'knitr'}\NormalTok{, }\StringTok{'genomation'}\NormalTok{,}
                      \StringTok{'ggbio'}\NormalTok{, }\StringTok{'Gviz'}\NormalTok{, }\StringTok{'DESeq2'}\NormalTok{, }\StringTok{'RUVSeq'}\NormalTok{,}
                      \StringTok{'gProfileR'}\NormalTok{, }\StringTok{'ggfortify'}\NormalTok{, }\StringTok{'corrplot'}\NormalTok{,}
                      \StringTok{'gage'}\NormalTok{, }\StringTok{'EDASeq'}\NormalTok{, }\StringTok{'citr'}\NormalTok{, }\StringTok{'formatR'}\NormalTok{,}
                      \StringTok{'svglite'}\NormalTok{, }\StringTok{'Rqc'}\NormalTok{, }\StringTok{'ShortRead'}\NormalTok{, }\StringTok{'QuasR'}\NormalTok{,}
                      \StringTok{'methylKit'}\NormalTok{,}\StringTok{'FactoMineR'}\NormalTok{, }\StringTok{'iClusterPlus'}\NormalTok{,}
                      \StringTok{'enrichR'}\NormalTok{,}\StringTok{'caret'}\NormalTok{,}\StringTok{'xgboost'}\NormalTok{,}\StringTok{'glmnet'}\NormalTok{,}
                      \StringTok{'DALEX'}\NormalTok{,}\StringTok{'kernlab'}\NormalTok{,}\StringTok{'pROC'}\NormalTok{,}\StringTok{'nnet'}\NormalTok{,}\StringTok{'RANN'}\NormalTok{,}
                      \StringTok{'ranger'}
\NormalTok{                     )}
\end{Highlighting}
\end{Shaded}

\hypertarget{data-for-the-book}{%
\section*{Data for the book}\label{data-for-the-book}}


We rely on data from different R and Bioconductor packages through out the book. For the datasets that do not ship with those packages, we created our own package \href{https://github.com/compgenomr/compGenomRData}{\textbf{compGenomRData}}. You can install this package via \texttt{devtools::install\_github("compgenomr/compGenomRData")}. We use \texttt{system.file()} function to get the path to the files. We noticed many inexperienced users are confused about this function. This function just outputs full path to the file that is installed with the data package.

\hypertarget{acknowledgements}{%
\section*{Acknowledgements}\label{acknowledgements}}


We wish to thank R and Bioconductor community for developing and maintaining libraries for genomic data analysis. Without their constant work and dedication, writing such a book will not be possible.

The following people kindly contributed fixes for typos: Thomas Schalch, Alex Gosdschan and Sarvesh Nikumbh.

\hypertarget{about-the-authors}{%
\chapter*{About the Authors}\label{about-the-authors}}


The authors have decades of combined experience in data analysis for genomics. They are developers of Bioconductor packages such as \href{https://bioconductor.org/packages/release/bioc/html/methylKit.html}{\textbf{methylKit}}, \href{https://bioconductor.org/packages/release/bioc/html/genomation.html}{\textbf{genomation}}, \href{https://bioconductor.org/packages/release/bioc/html/RCAS.html}{\textbf{RCAS}} and \href{https://bioconductor.org/packages/release/bioc/html/netSmooth.html}{\textbf{netSmooth}}. In addition, they have played key roles in developing end-to-end genomics data analysis pipelines for RNA-seq, ChIP-seq, Bisulfite-seq, and single cell RNA-seq called \href{http://bioinformatics.mdc-berlin.de/pigx/}{PiGx}.

\href{https://github.com/al2na}{\textbf{Altuna Akalin}} wrote most of the book and edited the rest.

\href{https://github.com/borauyar}{\textbf{Bora Uyar}} contributed ``RNA-seq Analysis'' chapter

\href{https://github.com/frenkiboy}{\textbf{Vedran Franke}} contributed ``ChIP-seq Analysis'' chapter

\href{https://github.com/jonathanronen}{\textbf{Jonathan Ronen}} contributed ``Multi-omics Analysis'' chapter

\mainmatter

\hypertarget{intro}{%
\chapter{Introduction to Genomics}\label{intro}}

The aim of this chapter is to provide the reader with some of the fundamentals
required for
understanding genome biology. By no means, this is a complete overview of the
subject but just a summary that will help the non-biologist reader understand
the recurring biological concepts in computational genomics. Readers that are
well-versed in genome biology and modern genome-wide quantitative assays should
feel
free to skip this chapter or skim it through.

\hypertarget{genes-dna-and-central-dogma}{%
\section{Genes, DNA and central dogma}\label{genes-dna-and-central-dogma}}

A central concept that will come up again and again is ``the gene''.
Before we can explain that we need to
introduce a few other concepts that are important to understand the gene concept.
Human body is made up of billions of cells. These cells specialize in different
tasks. For example, in the liver there are cells that help produce enzymes
to break toxins. In the heart, there are specialized muscle cells that make
the heart beat. Yet, all these different kinds of cells come from a single celled
embryo. All the instructions to make different kinds of cells are contained
within that single cell and with every division of that cell,
those instructions
are transmitted to new cells. These instructions can be coded into a string - a
molecule of DNA, a polymer made of recurring units called nucleotides. The four
nucleotides in DNA molecules, Adenine, Guanine, Cytosine and Thymine (coded as
four letters: A, C, G, and T) in a specific sequence, store the information for
life. DNA is organized in a double-helix form where two complementary polymers
interlace with each other and twist into the familiar helical shape.

\hypertarget{what-is-a-genome}{%
\subsection{What is a genome?}\label{what-is-a-genome}}

The full DNA sequence of an organism, which contains all the
hereditary information, is called a genome. The genome contains all the information
to build and maintain an organism. Genomes come in different sizes and structures.
Our genome is not only a naked strech of DNA.
In eukaryotic cells, DNA is wrapped around proteins (histones) forming higher-order structures like
nucleosomes which make up chromatins and chromosomes (see Figure \ref{fig:chromatinChr}).

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{images/chromatinChr} 

}

\caption{Chromosome structure in animals}\label{fig:chromatinChr}
\end{figure}

There might be several chromosomes
depending on the organism. However, in some species (such as most prokaryotes)
DNA is stored in a circular form. The size of genome between species differs too.
Human genome has 46 chromosomes and over 3 billion base-pairs, whereas wheat genome
has 42 chromosomes and 17 billion base-pairs, both genome size and chromosome numbers
are variable between different organisms. Genome sequences of organisms are
obtained using sequencing technology. With this technology, fragments of the
DNA sequence from the genome, called reads, are obtained.
Larger chunks of the genome sequence
is later obtained by stitching the
initial fragments to larger ones by using the overlapping reads. Latest,
sequencing technologies made genome sequencing cheaper and faster. These
technologies output more reads, longer reads and more accurate reads.

Estimated cost
of the first human genome is \$300 million in 1999-2000, today a high-quality human genome
can be obtained for \$1500. Since the costs are going down, researchers and clinicians
can generate more data. This drives up to costs for data storage and also drives
up the demand for qualified people to analyze genomic data. This was one of the
motivations behind writing this book.

\hypertarget{what-is-a-gene}{%
\subsection{What is a gene?}\label{what-is-a-gene}}

In the genome, there are specific regions containing the precise information that
encodes for physical products of genetic information. A region in the genome
with this information is traditionally called a ``gene''. However, the
precise definition of the gene is still developing. According to the classical
textbooks in molecular biology, a gene is a segment of a DNA sequence
corresponding to a single protein or to a single catalytic and structural RNA
molecule {[}1{]}. A modern definition is: ``A region (or regions) that includes all
of the sequence elements necessary to encode a functional transcript'' {[}2{]}. No
matter how variable the definitions are, all agree on the fact that genes are
basic units of heredity in all living organisms.

All cells use their hereditary
information in the same way most of the time; the DNA is replicated to
transfer the information to new cells. If activated, the genes are transcribed into
messenger RNAs (mRNAs) in nucleus (in eukaryotes), followed by mRNAs (if the
gene is protein coding) getting translated into proteins in the cytoplasm. This is
essentially a process of information transfer between information carrying
polymers; DNA, RNA and proteins, known as the ``central dogma'' of molecular
biology (see Figure \ref{fig:CentDog} for a summary).
Proteins are essential elements for life.
The growth and repair, functioning and structure of all living cells depends on them.
This is why the gene is a central
concept in genome biology, because a gene can encode information for proteins and other
functional molecules. How genes are controled and activated dictates
everything about an organism. From the identity of a cell to response to
an infection, how cells develop and behave against certain stimuli is governed
by activity of the genes and functional molecules they encode. The liver cell becomes a liver cell because certain
genes are activated and their functional products are produced to help liver
cell achieve its tasks.

\begin{figure}

{\centering \includegraphics[width=1\linewidth]{images/centDogma} 

}

\caption{Central Dogma: replication, transcription, translation}\label{fig:CentDog}
\end{figure}

\hypertarget{how-genes-are-controlled-the-transcriptional-and-the-post-transcriptional-regulation}{%
\subsection{How genes are controlled ? The transcriptional and the post-transcriptional regulation}\label{how-genes-are-controlled-the-transcriptional-and-the-post-transcriptional-regulation}}

In order to answer this question, we have to dig a little deeper on the
transcription concept we introduced via the central dogma.
The first step in a process of information transfer - a production of an RNA
copy of a part of the DNA sequence - is called transcription. This task is
carried out by the RNA polymerase enzyme. RNA polymerase-dependent initiation of
transcription is enabled by the existence of a specific region in the sequence
of DNA - a core promoter. Core promoters are regions of DNA that promote
transcription and are found upstream from the start site of transcription. In
eukaryotes, several proteins, called general transcription factors recognize and
bind to core promoters and form a pre-initiation complex. RNA polymerases
recognize these complexes and initiate synthesis of RNAs, the
polymerase travels along the template DNA and making an RNA
copy{[}3{]}. After mRNA is
produced it is often spliced by splicesosome. The sections called `introns' are removed and
sections called `exons' left in. Then, the remaining mRNA translated into proteins. Which exons
will be part of the final mature transcript can also be regulated and creates
diversity in protein structure and function (See Figure \ref{fig:TransSplice}`).

\begin{figure}

{\centering \includegraphics[width=0.7\linewidth]{images/TransSplice} 

}

\caption{Transcription could be followed by splicing, which creates different transcript isoforms. This will in return create different protein isoforms since the information required to produce the protein is encoded in the transcripts. Differences in transcript of the same gene can give rise to different protein isoforms}\label{fig:TransSplice}
\end{figure}

On the contrary to protein coding genes, non-coding RNA (ncRNAs)
genes are processed and assume their functional structures after transcription
and without going into translation, hence the name: non-coding RNAs. certain
ncRNAs can also be spliced but still not translated. ncRNAs and
other RNAs in general can form complementary base-pairs within the RNA molecule
which gives them additional complexity. This self-complementarity based
structure, termed RNA secondary structure, is often necessary for functions of many
ncRNA species.

In summary, the set of processes, from transcription initiation to production
of the functional product, is referred to as gene expression.
Gene expression quantification and regulation is a fundamental topic in
genome biology.

\hypertarget{what-does-a-gene-look-like}{%
\subsection{What does a gene look like?}\label{what-does-a-gene-look-like}}

Before we move forward, it will be good to discuss how we can visualize genes.
As someone interested in computational genomics, you will frequently encounter
a gene on a computer screen, and how it is represented on the computer will be
equivalent to what you imagine when you hear the word ``gene''.
In the online databases, the genes will appear as a sequence of letters
or as a series of connected boxes showing exon-intron structure which may
include the direction of transcription as well (see Figure \ref{fig:RealGene}).
direction. You will encounter more with the latter so this is likely what will
pop into your mind when you think of genes.

As we have mentioned DNA has two strands, and a gene can be located
on either of them, and direction of transcription will depend on that. In the
Figure you can see arrows on introns (lines connecting boxes) indicating the
direction of the gene.

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{images/RealGene} 

}

\caption{A) Representation of a gene at UCSC browser. Boxes indicate exons, and lines indicate introns. B) Partial sequence of FATE1 gene as shown in NCBI GenBank database.}\label{fig:RealGene}
\end{figure}

\hypertarget{elements-of-gene-regulation}{%
\section{Elements of gene regulation}\label{elements-of-gene-regulation}}

The mechanisms regulating gene expression are essential for
all living organisms as they dictate where and how much of a gene product (may
it be protein or ncRNA) should be manufactured. This regulation could occur
at the pre- and co-transcriptional level by controlling how many transcripts should be produced and/or which version of the transcript should be produced by regulating
splicing. Different versions of the same gene could encode for proteins by
regulating splicing the process can decide which parts will go into the final
mRNA that will code for the protein.
In addition, gene products can be regulated post-transcriptionally where certain
molecules bind to RNA and mark them for degradation even before they can be used
in protein production.

Gene regulation drives cellular differentiation; a
process during which different tissues and cell types are produced. It also
helps cells maintain differentiated states of cells/tissues. As a product of
this process, at the final stage of differentiation, different kinds of cells
maintain different expression profiles although they contain the same genetic
material. As mentioned above there are two main types of regulation and next we
will provide information on those.

\hypertarget{transcriptional-regulation}{%
\subsection{Transcriptional regulation}\label{transcriptional-regulation}}

The rate of transcription initiation is the primary regulatory element in gene
expression regulation. The rate is controlled by core promoter elements as well as
distant-acting regulatory elements such as enhancers. On top of that, processes
like histone modifications and/or DNA methylation have a crucial regulatory
impact on transcription. If a region is not accessible for the transcriptional
machinery, e.g.~in the case when chromatin structure is compacted due to the
presence of specific histone modifications, or if the promoter DNA is
methylated, transcription may not start at all. Last but the not least, gene
activity is also controlled post-transcriptionally by ncRNAs such as microRNAs
(miRNAs), as well as by cell signaling resulting in protein modification or
altered protein-protein interactions.

\hypertarget{regulation-by-transcription-factors-through-regulatory-regions}{%
\subsubsection{Regulation by transcription factors through regulatory regions}\label{regulation-by-transcription-factors-through-regulatory-regions}}

Transcripton factors are proteins that
recognize a specific DNA motif to bind on a regulatory region and regulate the transcription rate of
the gene associated with that regulatory region (See Figure \ref{fig:regSummary})`
for an illustration). These factors bind to a variety of regulatory regions summarized in
Figure \ref{fig:regSummary}, and their concerted action
controls the transcription rate. Apart from their binding preference, their
concentration, the availability of synergistic or competing transcription
factors will also affect the transcription rate.

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{images/regulationSummary} 

}

\caption{Representation of regulatory regions in animal genomes}\label{fig:regSummary}
\end{figure}

\hypertarget{core-and-proximal-promoters}{%
\paragraph{Core and proximal promoters}\label{core-and-proximal-promoters}}

Core promoters are the immediate neighboring regions around
the transcription start site (TSS) that serves as a docking site for the
transcriptional machinery and pre-initiation complex (PIC) assembly. The
textbook model for transcription initiation is as follows: The core promoter has
a TATA motif (referred as TATA-box) 30 bp upstream of an initiator sequence
(Inr), which also contains TSS. Firstly, transcription factor TFIID binds to the
TATA-box. Next, general transcription factors are recruited and transcription is
initiated on the initiator sequence. Apart from the
TATA-box and Inr, there are a number of sequence elements on the animal core
promoters that are associated with transcription initiation and PIC assembly,
such as downstream promoter elements (DPEs), the BRE elements and CpG islands.
DPEs are found 28-32 bp downstream of the TSS in TATA-less promoters of
Drosophila melanogaster, it generally co-occurs with the Inr element, and is
thought to have a similar function to the TATA-box. The BRE element is
recognized by TFIIB protein and lies upstream of the TATA-box. CpG islands
are CG dinucleotide-enriched segments of vertebrate genomes, despite the general
depletion of CG dinucleotides in those genomes. 50-70\% of promoters in
human genome are associated with CpG islands.

Proximal promoter elements are typically right upstream
of the core promoters and usually contain binding sites for activator
transcription factors and they
provide additional control over gene expression.

\hypertarget{enhancers}{%
\paragraph{Enhancers:}\label{enhancers}}

Proximal regulation is not the only, nor the most important mode
of gene regulation. Most of the transcription factor binding sites in
the human genome are found in intergenic regions or in introns .
This indicates the widespread usage of distal regulatory elements in animal
genomes. On a molecular
function level, enhancers are similar to proximal promoters; they contain binding
sites for the same transcriptional activators and they basically enhance the
gene expression. However, they are often highly modular and several of them
can affect the same promoter at the same time or in different time-points or
tissues. In addition, their activity is independent of their
orientation and their distance to the promoter they interact with. A
number of studies showed that enhancers can act upon their target genes over
several kilobases away. According to a popular
model, enhancers achieve this by looping the DNA and coming to contact with
their target genes.

\hypertarget{silencers}{%
\paragraph{Silencers:}\label{silencers}}

Silencers are similar to enhancers; however their effect is
opposite of enhancers on the transcription of the target gene, and results in
decreasing their level of transcription. They contain binding sites for
repressive transcription factors. Repressor transcription factors can either
block the binding of an activator , directly compete for the same binding site,
or induce a repressive chromatin state in which no activator binding is
possible. Silencer effects, similar to those of enhancers, are independent of
orientation and distance to target genes. In contradiction to this general view,
in Drosophila there are two types of silencers, long-range and short-range.
Short-range silencers are close to promoters and long-range silencers can
silence multiple promoters or enhancers over kilobases away. Like
enhancers, silencers bound by repressors may also induce changes in DNA
structure by looping and creating higher order structures. One class of
such repressor proteins, which is thought to initiate higher-order structures by
looping, is Polycomb group proteins (PcGs).

\hypertarget{insulators}{%
\paragraph{Insulators:}\label{insulators}}

Insulator regions limit the effect of other regulatory elements
to certain chromosomal boundaries; in other words, they create regulatory
domains untainted by the regulatory elements in regions outside that domain.
Insulators can block enhancer-promoter communication and/or prevent spreading of
repressive chromatin domains. In vertebrates and insects, some of the
well-studied insulators are bound by CTCF (CCCTC-binding factor).
Genome-wide studies from different mammalian tissues confirm that CTCF binding
is largely invariant of cell type, and CTCF motif locations are conserved in
vertebrates. At present, there are two models of explaining the insulator
function; the most prevalent model claims insulators create physically separate
domains by modifying chromosome structure. This is thought to be achieved by
CTCF-driven chromatin looping and recent evidence shows that CTCF can induce a
higher-order chromosome structure through creating loops of chromatins.
According to the second model, an insulator-bound activator cannot bind an
enhancer; thus enhancer-blocking activity is achieved and insulators can also
recruit active histone domain, creating an active domain for enhancers to
function.

\hypertarget{locus-control-regions}{%
\paragraph{Locus control regions:}\label{locus-control-regions}}

Locus control regions (LCRs) are clusters of
different regulatory elements that control entire set of genes on a locus. LCRs
help genes achieve their temporal and/or tissue-specific expression programs.
LCRs may be composed of multiple cis-regulatory elements, such as insulators,
enhancers and they act upon their targets even from long distances. However
LCRs function with an orientation dependent manner, for example the activity of
beta-globin LCR is lost if inverted. The mechanism of LCR function otherwise
seems similar to other long-range regulators described above. The evidence is
mounting in the direction of a model where DNA-looping creates a chromosomal
structure in which target genes are clustered together, which seems to be
essential for maintaining open chromatin domain.

\hypertarget{epigenetic-regulation}{%
\subsubsection{Epigenetic regulation}\label{epigenetic-regulation}}

Epigenetics in biology usually refers to
constructions (chromatin structure, DNA methylation etc.) other than DNA
sequence that influence gene regulation. In essence, epigenetic regulation is
the regulation of DNA packing and structure, the consequence of which is gene
expression regulation. A typical example is that DNA packing inside the nucleus can
directly influence gene expression by creating accessible regions for transcription
factors to bind.
There are two main mechanisms in
epigenetic regulation: i) DNA modifications ii) histone modifications. Below,
we will introduce these two mechanisms.

\hypertarget{dna-modifications-such-as-methylation}{%
\paragraph{DNA modifications such as methylation:}\label{dna-modifications-such-as-methylation}}

DNA methylation is usually associated with gene silencing.
DNA methyltransferase enzyme catalyzes the addition of a methyl group to cytosine of
CpG dinucleotides (while in mammals the addition of methyl group is largely
restricted to CpG dinucleotides, methylation can occur in other bases as well)
. This covalent modification either interferes with transcription factor
binding on the region, or methyl-CpG binding proteins induce the spread of
repressive chromatin domains, thus the gene is silenced if its promoter has
methylated CG dinucleotides. DNA methylation usually occurs in repeat
sequences to repress transposable elements, these elements when active can
jump around and insert them to random parts of the genome, potentially distrupting
the genomic functions.

.DNA methylation is also related to a
key core and proximal promoter element: CpG islands. CpG islands are usually
unmethylated, however for some genes CpG island methylation accompanies their
silenced expression. For example, during X-chromosome inactivation many CpG
islands are heavily methylated and the associated genes are silenced. In
addition, in embryonic stem cell differentiation pluripotency-associated genes
are silenced due to DNA methylation. Apart from methylation, there are other
kinds of DNA modifications present in mamalian genomes, such as hydroxy-methylation and
formylcytosine. These are other modifications under current research that are either
intermediate or stable modifications with distinct functional associations. There
are as many as 12 distinct DNA modifications observed when we look accross
all studied species.

\hypertarget{histone-modifications}{%
\paragraph{Histone Modifications:}\label{histone-modifications}}

Histones are proteins that constitute nucleosome. In
eukaryotes, eight histones nucleosomes are wrapped around by DNA and build
nucleosome. They help super-coiling of DNA and inducing high-order structure
called chromatin. In chromatin, DNA is either densely packed (called
heterochromatin or closed chromatin), or it is loosely packed (called
euchromatin or open chromatin) {[}60, 61{]}. Heterochromatin is thought to harbor
inactive genes since DNA is densely packed and transcriptional machinery cannot
access it. On the other hand, euchromatin is more accessible for transcriptional
machinery and might therefore harbor active genes. Histones have long and
unstructured N-terminal tails which can be covalently modified. The most studied
modifications include acetylation, methylation and phosphorylation {[}60{]}. Using
their tails, histones interact with neighboring nucleosomes and the
modifications on the tail affect the nucleosomes affinity to bind DNA and
therefore influence DNA packaging around nucleosomes. Different modifications on
histones are used in different combinations to program the activity of the genes
during differentiation. Histone modifications have a distinct nomenclature, for
example: H3K4me3 means the lysine (K) on the 4th position of histone H3 is
tri-methylated.

\begin{longtable}[]{@{}ll@{}}
\caption{Table 1 Histone modifications and their effects. If more than one histone modification has the same effect, they are separated by commas.}\tabularnewline
\toprule
Modifications & Effect\tabularnewline
\midrule
\endfirsthead
\toprule
Modifications & Effect\tabularnewline
\midrule
\endhead
H3K9ac & Active promoters and enhancers\tabularnewline
H3K14ac & Active transcription\tabularnewline
H3K4me3/me2/me1 Active p & romoters and enhancers, H3K4me1 and H3K27ac is enhancer-specific\tabularnewline
H3K27ac & H3K27ac is enhancer-specific\tabularnewline
H3K36me3 & Active transcribed regions\tabularnewline
H3K27me3/me2/me1 & Silent promoters\tabularnewline
H3K9me3/me2/me1 & Silent promoters\tabularnewline
\bottomrule
\end{longtable}

Histone modifications are associated with a number of different
transcription-related conditions; some of them are summarized in Table 1.
Histone modifications can indicate where the regulatory regions are and they can
also indicate activity of the genes. From a gene regulatory perspective, maybe
the most important modifications are the ones associated with enhancers and
promoters. Certain genes in mouse embryonic stem cells have both active H3K4me3
and inactive H3K27me3 modifications in their promoters. Surprisingly, most of
these genes have high CpG content, are important for development and are shown
to have paused RNA polymerase II {[}66, 67{]}. In addition, Heintzman et al.~showed
that H3K4me1 could predict tissue-specific active enhancers in human cells {[}68,
69{]}. The examples above demonstrate the capability of histone modifications in
predicting regulatory potential.

Furthermore, certain proteins can influence chromatin structure by interacting
with histones. Some of these proteins, like those of the Polycomb Group (PcG)
and CTCF, are discussed above in the insulators and silencer sections. In
vertebrates and insects, PcGs are responsible for maintaining the silent state
of developmental genes, and trithorax group proteins (trxG) for maintaining
their active state {[}70, 71{]}. PcGs and trxGs induce repressed or active states by
catalyzing histone modifications or DNA methylation. Both the proteins bind PREs
that can be on promoters or several kilobases away {[}30, 31, 71{]}. Another protein
that induces histone modifications is CTCF. In b-globin locus, CTCF binding is
shown to be associated with repressive H3K9/K27me2 modifications {[}72{]}.

\BeginKnitrBlock{rmdtip}
\textbf{Want to know more ?}

\begin{itemize}
\item
  Transcriptional regulatory elements in the human genome: \url{http://www.ncbi.nlm.nih.gov/pubmed/16719718}
\item
  On metazoan promoters: types and transcriptional properties:
  \url{http://www.ncbi.nlm.nih.gov/pubmed/22392219}
\item
  General principles of regulatory sequence function
  \url{http://www.nature.com/nrg/journal/v15/n7/abs/nrg3684.html}
\item
  DNA methylation: roles in mammalian development
  \url{http://www.nature.com/doifinder/10.1038/nrg3354}
\item
  Histone modifications and organization of the genome
  \url{http://www.nature.com/nrg/journal/v12/n1/full/nrg2905.html}
\item
  DNA methylation and histone modifications are linked
  \url{http://www.nature.com/nrg/journal/v10/n5/abs/nrg2540.html}
\end{itemize}
\EndKnitrBlock{rmdtip}

\hypertarget{post-transcriptional-regulation}{%
\subsection{Post-transcriptional regulation}\label{post-transcriptional-regulation}}

\hypertarget{regulation-by-non-coding-rnas}{%
\subsubsection{Regulation by non-coding RNAs}\label{regulation-by-non-coding-rnas}}

Recent years have witnessed an explosion in noncoding
RNA (ncRNA)-related research. Many publications implicated ncRNAs as important
regulatory elements. Plants and animals produce many
different types of ncRNAs such as long non-coding RNAs (lncRNAs),
small-interferring RNAs (siRNAs), microRNAs (miRNAs), promoter-associated RNAs
(PARs) and small nucleolar RNAs (snoRNAs). lncRNAs are typically \textgreater{}200 bp long,
they are involved in epigenetic regulation by interacting with chromatin
remodeling factors and they function in gene regulation. siRNAs are short
double-stranded RNAs which are involved in gene-regulation and transposon
control, they silence their target genes by cooperating with Argonaute proteins
. miRNAs are short single-stranded RNA molecules that interact with their
target genes by using their complementary sequence and mark them for quicker
degradation. PARs may regulate gene expression as well: they are \textasciitilde{}18-200bp
long ncRNAs originating from promoters of coding genes. snoRNAs also shown
to play roles in gene regulation, although they are mostly believed to guide
ribosomal RNA modifications.

\hypertarget{splicing-regulation}{%
\subsubsection{Splicing regulation}\label{splicing-regulation}}

Splicing is regulated by regulatory elements on the pre-mRNA and proteins
binding to those elements . Regulatory elements are categorized
as splicing
enhancers and repressors. They can be located either in exons
or introns. Depending of their activity and their locations there are four types of regulatory elements:
- exonic splicing enhancers (ESEs)
- exonic splicing silencers (ESSs)
- intronic splicing enhancers (ISEs)
- intronic splicing silencers (ISSs).

The majority of splicing repressors are heterogeneous nuclear ribonucleoproteins (hnRNPs). If splicing repressor protein bind
silencer elements they reduce the change of nearby site to be
used as splice junction. On the contrary, splicing enhancers are sites to which splicing activator proteins bind and binding
on that region increases the probability that a nearby site will be used as a splice junction. Most of the activator proteins that bind to splicing enhancers are members of the SR protein family. Such proteins can recognize specific RNA recognition motifs. By regulating splicing exons can be skipped or included
which creates protein diversity.

\BeginKnitrBlock{rmdtip}
\textbf{Want to know more ?}

\begin{itemize}
\item
  On miRNAs: Their genesis and modes of regulation
  \url{http://www.sciencedirect.com/science/article/pii/S0092867404000455}
\item
  Functions of small RNAs
  \url{http://www.nature.com/nrg/journal/v15/n9/abs/nrg3765.html}
\item
  Functions of non coding RNAs
  \url{http://www.nature.com/nrg/journal/v15/n6/abs/nrg3722.html}
\item
  on splicing: Wang, Zefeng; Christopher B. Burge (May 2008). ``Splicing regulation: From a parts list of regulatory elements to an integrated splicing code''. RNA 14 (5): 802--813. \url{doi:10.1261/rna.876308}. ISSN 1355-8382. PMC 2327353. PMID 18369186. Retrieved 2013-08-15.
\end{itemize}
\EndKnitrBlock{rmdtip}

\hypertarget{shaping-the-genome-dna-mutation}{%
\section{Shaping the genome: DNA mutation}\label{shaping-the-genome-dna-mutation}}

Human and chimpanzee genomes are 98.8\% similar. The 1.2\% difference is what separetes
us from chimpanzees. The further you move away from human in terms of evolutionary
distance the higher the difference gets. However, even between the members of
the same species differences in genome sequences exists. These differences are
due to a process called mutation which drives differences between individuals
but also the provides the fuel for evolution as the source of the genetic
variation. Individuals with beneficial mutations can adapt to their surroundings
better than others and in time these mutations which are beneficial for survival
spreads in the population due to a process called ``natural selection''. Selection
acts upon individuals with beneficial features which gives them an edge for
survival in a given environment. Genetic variation created by the mutations
in individuals provide the material on which selection can act upon. If the
selection process goes for a long time in a relatively isolated environment
that requires adaptation, this population can evolve into a different species
given enough time.This is the basic idea behind evolution in a nutshell,
and without mutations providing the genetic variation there will be no evolution.

Mutations in the genome occur due to multiple reasons. First, DNA replication
is not an error-free process. Before a cell division, the DNA is replicated with
1 mistake per 10\^{}8 to 10\^{}10 base-pairs. Second, mutagens such as UV light
can induce mutations on the genome. Third factor that contributes to mutation
is imperfect DNA repair. Every day any human cell suffers multiple instances of DNA damage.
DNA repair enzymes are there to cope with this damage but they are also not
error-free, depending on which DNA repair mechanism is used (there are multiple)
mistakes will be made at varying rates.

Mutations are classified by how many bases they effect, their effect on
DNA structure and gene function. By their effect on DNA structure
the mutations are classified as follows:

\begin{itemize}
\tightlist
\item
  \textbf{Base substitution}: A base is changed with another.
\item
  \textbf{Deletion}: One or more bases is deleted.
\item
  \textbf{Insertion}: New base or bases inserted into the genome.
\item
  \textbf{Microsatellite mutation}: Small insertions or deletions of small tandemly
  repeating DNA segments.
\item
  \textbf{Inversion}: A DNA fragment changes its orientation 180 degrees.
\item
  \textbf{Translocation}: A DNA fragment moves to another location in the genome.
\end{itemize}

Mutations can also be classified by their size as follows:

\begin{itemize}
\tightlist
\item
  \textbf{Point mutations}: mutations that involve one base. Substitutions, deletions and
  insertions are point mutations. They are also termed as single nucleotide
  polymorphisms (\textbf{SNPs}).
\item
  \textbf{small-scale mutations}: mutations that involve several bases.
\item
  \textbf{Large-scale mutations}: mutations which involve larger chromosomal regions.
  Transposable element insertions (where a segment of the genome
  jumps to another region in the genome) and segmental duplications ( a large
  region is copied multiple times in tandem) are typical large scale mutations.
\item
  \textbf{Aneuploidies}: Insertions or deletions of whole chromosomes.
\item
  \textbf{Whole-genome polyploidies}: duplications involving whole genome.
\end{itemize}

Mutations by their effect on gene function can be classified as follows:

\begin{itemize}
\tightlist
\item
  \textbf{gain-of-function mutations}: A type of mutation in which the altered gene
  product possesses a new molecular function or a new pattern of gene
  expression.
\item
  \textbf{loss-of-function mutations}: A mutation that results in reduced or abolished
  protein function. This is the more common type of mutation.
\end{itemize}

\hypertarget{high-throughput-experimental-methods-in-genomics}{%
\section{High-throughput experimental methods in genomics}\label{high-throughput-experimental-methods-in-genomics}}

Most of the biological phenomena described above relating to transcription
, gene regulation or DNA mutation can be measured over the entire genome using
high-throughput experimental techniques, which are quickly becoming
the standard for studying genome biology. In addition, their
applications in the clinic are also gaining momemntum: there are already diagnostic
tests that are based on these techniques.

Some of the things that can be measured by high-throughput assays are as follows:

\begin{itemize}
\tightlist
\item
  \emph{Which genes are expressed and how much ?}
\item
  \emph{Where does a transcription factor bind ?}
\item
  \emph{Which bases are methylated in the genome ?}
\item
  \emph{Which transcripts are translated ?}
\item
  \emph{Where does RNA-binding proteins bind ?}
\item
  \emph{Which microRNAs are expressed ?}
\item
  \emph{Which parts of the genome are in contact with each other ?}
\item
  \emph{Where are the mutations in the genome located ?}
\item
  \emph{Which parts of the genome are nucleosome-free ?}
\end{itemize}

There are many more questions one can answer using modern genome-wide techniques
and every other day a new variant of the existing techniques comes along
to answer a new question.
However, One has to keep in mind that these methods are at varying degrees
of maturity and they all come with technical limitations
and are not noise-free. Despite this, they are extremely useful for research and clinical
purposes. And, thanks to these methods we are able to sequence and annotate
genomes at a massive scale.

\hypertarget{the-general-idea-behind-high-throughput-techniques}{%
\subsection{The general idea behind high-throughput techniques}\label{the-general-idea-behind-high-throughput-techniques}}

High-throughput methods aim to quantify or locate all or most of the genome that harbours
the biological feature (expressed genes, binding sites, etc.) of interest.
Most of the methods rely on some sort of enrichment of the the targeted
biological feature. For example, if you want to measure expression of protein coding
genes you need to be able to extract mRNA molecules with special post-transcriptional
alterations that protein-coding genes acquire. If you are looking for transcription factor
binding, you need to enrich
for the DNA fragments that are bound by the protein if interest.This parts
depends on available molecular biology and chemistry technques, and the final
product of this part is RNA or DNA fragments.

Next, you need to be able to
tell where these fragments are coming from in the genome and how many of them
are there. Microarrays were the standard tool for the quantification step
until spread of sequencing techniques. In microarrays, one had to design complementary bases
,called ``oligos'' or ``probes'', to the genetic material enriched via the experimental protocol.
If the enriched material is complementary to the genetic material,
a light signal will
be produced and the intensity of the signal will be proportional to the amount of the
genetic material pairing with that oligo. There will be more probes available for
hybrdization (process of complementary bases forming bonds ), so the more fragments
available stronger the signal. For this to be able to work, you need to know
at least part of your genome sequence, and design probes. If you want to measure
gene expression, your probes should overlap with genes and should be unique
enough to not to bind sequences from other genes. This technology is now being
replaced with sequencing technology, where you directly sequence your genetic
material. If you have the sequence of your fragments, you can align them back
to genome, see where they are coming from and count them. This is a better
technology where the quantification is based on the real identiy of fragments
rather than based on hybridization to designed probes.

In summary HT techniques has the following steps, and this also summarized in
Figure \ref{fig:HTassays}:

\begin{itemize}
\tightlist
\item
  Extraction: This is the step where you extract the genetic
  material of interest, RNA or DNA.
  and nrichment
\item
  Enrichment: In this step, you enrich for the event you are interested
  in. For example, protein binding sites. In some cases such as whole-genome DNA
  sequencing there is no need for enrichment step. You just get fragments of
  genomic DNA and sequence them.
\item
  Quantification: This is where you quantify your enriched material. Depending
  on the protocol you may need to quantify a control set as well, where you
  should see no enrichment or only background enrichment.
\end{itemize}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{images/HTassays} 

}

\caption{Common steps of High-throughput assays in genome biology}\label{fig:HTassays}
\end{figure}

\hypertarget{high-throughput-sequencing}{%
\subsection{High-throughput sequencing}\label{high-throughput-sequencing}}

High-throughput sequencing, or massively parallel sequencing, is a
collection of methods and technologies that can sequence DNA thousands/millons
of fragments at a time. This is in contrast to older technologies that can
produce a limited
number of fragments at a time. Here, throughput refers to number
of sequenced bases per hour. The older low-throughput sequencing methods have
\textasciitilde{} 100 times less throughput compared to modern high-throughput methods. The
increased throughput gives the ability to measure biological features on a
genome-wide scale in a shorter time frame.

Similar to other high-throughput methods, sequencing based methods also require
an enrichment step. This step enriches for the features we are interested in.
The main difference of the sequencing based methods is the quantification step.
In high-throughput sequencing,
enriched fragments are put through the sequencer which outputs the sequences
for the fragments. Due to limitations in current leading technologies, only
limited number of bases can be sequenced using from the input fragments. However,
the length is usually enough to uniquely map the reads to the genome and quantify
the input fragments.

\hypertarget{high-throughput-sequencing-data}{%
\subsubsection{High-throughput sequencing data}\label{high-throughput-sequencing-data}}

If there is a genome available, the reads are aligned to the genome and based
on the library preperation protocol different strategies are applied for analysis. A sequencing library is composed of fragments of RNA or DNA ready to be sequenced. The library preparation primarily depends on the experiment of interest. There are a number of library preperation protocols aimed at quantifying different signals from the genome. Some of the potential analysis strategies for different library-prep protocols and processed output of read alignments
are depicted in Figure \ref{fig:HTseq}.
For example, we maybe interested to quantify the gene expression.
The experimental protocol, called RNA sequencing-
RNA-seq, enriches for fragments of RNA that are coming from protein coding genes.
Upon alignment, we can calculate the coverage profile which gives us a read count
per base along the genome. This information can be stored in a text file or
specialized file formats to be used in subsequent analysis or visualization.
We can also just count how many reads overlap with exons of each gene and record
read counts per gene for further analysis. This essentially produces a table
with gene names and read counts for different samples. As we will see in later
chapters, this is an essential information for statistical models that model
RNA-seq data. Furthermore, we can stack up the reads and count how many times
we see a base position in a read mismatches the base in the genome. Read aligners
allow for mismatches, and for this reason we can see reads with mismatches.
This information can be used to identify SNPs, and can be stored again in
a tabular format with the information of position and mismatch type and
number of reads supporting the mismatch. The original algorithms are a bit
more complicated than just counting mismatches but the general idea is the
same, what they are doing differently is trying to minimize false positive
rates by using filters, so that not every mismatch is recorded as SNP.

\begin{figure}

{\centering \includegraphics[width=0.7\linewidth]{images/HTseq} 

}

\caption{High-throughput sequencing summary}\label{fig:HTseq}
\end{figure}

\hypertarget{future-of-high-throughput-sequencing}{%
\subsubsection{Future of high-throughput sequencing}\label{future-of-high-throughput-sequencing}}

The sequencing technology is still evolving. Obtaining
longer single-molecule reads, and preferably, being able to call base modifications
on the fly is the next frontier.
With longer reads, the genome asssembly will be easier for the regions
that have high repeat content. With single-molecule sequencing, we will be able
tell how many transcripts are present in a given cell population without
relying on fragment amplification methods which can introduce biases.

Another recent development is single-cell sequencing. Current technologies usually
work on genetic material from thousands to millions of cells. This means that the
results you receive represents the population of cells that were used in the
experiment. However, there is a lot of variation between the same type of cells, but
this variation is not observed at all. Newer sequencing techniques can work
on single cells and give quantititave information on each cell.

\BeginKnitrBlock{rmdtip}
\textbf{Want to know more ?}

\begin{itemize}
\item
  Current and the future high-throughput sequencing technologies \url{http://www.sciencedirect.com/science/article/pii/S1097276515003408}
\item
  Illumina repository for different library preperation protocols for sequencing \url{http://www.illumina.com/techniques/sequencing/ngs-library-prep/library-prep-methods.html}
\end{itemize}
\EndKnitrBlock{rmdtip}

\hypertarget{visualization-and-data-repositories-for-genomics}{%
\section{Visualization and data repositories for genomics}\label{visualization-and-data-repositories-for-genomics}}

There are \textasciitilde{}100 animal genomes sequenced as of 2016. On top these, there are many
research projects from either individual labs or consortia that
produce petabytes of auxiliary genomics data, such as ChIP-seq, RNA-seq, etc.

There are two requirements to be able to visualize genomes and its associated
data, 1) you need to be able to work with a species
that has a sequenced genome and 2) you want to have annotation on that genome,
meaning, at the very least, you want to know where the genes are. Most
genomes after sequencing quickly annotated with gene-predictions or
know gene sequences are mapped on to it, you can also
have conservation to other species to filter functional elements. If you
are working with a model organism or human you will also have a lot of
auxiliary information to help demarcate the functional regions such
as regulatory regions, ncRNA, SNPs that are common in the population.
Or you might have disease or tissue specific data available.
The more the organism is worked on the more auxiliary data you will have.

\hypertarget{accessing-genome-sequences-and-annotations-via-genome-browsers}{%
\subsubsection{Accessing genome sequences and annotations via genome browsers}\label{accessing-genome-sequences-and-annotations-via-genome-browsers}}

As someone intends to work with genomics, you will need to visualize a
large amount of data to make biological inferences or simply check regions of
interest in the genome visually. Looking at the genome case by case with all
the additional datasets is a necessary step to develop hypothesis and understand
the data.

Many genomes and associated data is
available through genome browsers. A genome browser
is a website or an application
that helps you visualize the genome and all the available data associated
with it. Via genome browsers, you will be able to see where genes are in
relation to each other and other functional elements. You will be
able to see gene structure. You will be able to see auxiliary data such as
conservation, repeat content and SNPs. Here we review some of the popular
browsers.

\textbf{UCSC genome browser:} This is an online browser hosted by University of
California, Santa Cruz at \url{http://genome.ucsc.edu/}.
This is an interactive website that contains genomes
and annotations for many species. You can search for genes or genome coordinates
for the species of your interest. It is usually very responsive and allows
you to visualize large amounts of data. In addition, it has multiple other tools
that can be used in connection with the browser. One of the most useful tools
is \emph{UCSC Table Browser}, which lets you download the all the data you see on the
browser, including sequence data, in multiple formats.
Users can upload data or provide links to the data
to visualize user specific data.

\textbf{Ensembl:} This is another online browser maintained by\\
European Bioinformatics Institute and the Wellcome Trust Sanger Institute in
the UK, \url{http://www.ensembl.org}.
Similar to UCSC browser, users can visualize genes or genomic coordinates
from multiple species and it also comes with auxiliary data. Ensembl is
associated with \emph{Biomart} tool which is similar to UCSC Table browser, can
download genome data including all the auxiliary data set in multiple formats.

\textbf{IGV:} Integrated genomics viewer (IGV) is a desktop application developed
by Broad institute (\url{https://www.broadinstitute.org/igv/}). It is
developed to deal with large amounts of high-throughput sequencing data which
is harder to view in online browsers. IGV can integrate your local sequencing
results with online annotation on your desktop machine. This is useful when
viewing sequencing data, especially alignments. Other browsers mentioned above
have similar functionalities however you will need to make your large
sequencing data available online somewhere before it can be viewed by browsers.

\hypertarget{data-repositories-for-high-throughput-assays}{%
\subsubsection{Data repositories for high-throughput assays}\label{data-repositories-for-high-throughput-assays}}

Genome browsers contain lots of auxiliary high-throughput data. However, there are
many more public high-throughput data sets available and they are certainly
not available through genome browsers. Normally, every high-throughput dataset
associated with a publication should be deposited to public archives. There
are two major public archives we use to deposit data. One of them is
\emph{Gene expression Omnibus (GEO)} hosted at \url{http://www.ncbi.nlm.nih.gov/geo/},
and the other one is \emph{European nucleotide archive (ENA)} hosted at
\url{http://www.ebi.ac.uk/ena}. These repositories accept high-throughput datasets
and users can freely download and use these public data sets for their
own research. Many data sets in these repositories are in their raw format,
for example the format the sequencer provides mostly. Some data sets will also
have processed data but that is not a norm.

Apart from these repositories, there are multiple multi-national consortia
that are dedicated to certain genome biology or disease related problems and
they maintain their own databases and provide access to processed and raw data.
Some of these consortia are mentioned below.

\begin{longtable}[]{@{}ll@{}}
\toprule
\begin{minipage}[b]{0.32\columnwidth}\raggedright
Consortium\strut
\end{minipage} & \begin{minipage}[b]{0.62\columnwidth}\raggedright
what is it for?\strut
\end{minipage}\tabularnewline
\midrule
\endhead
\begin{minipage}[t]{0.32\columnwidth}\raggedright
\href{https://www.encodeproject.org/}{ENCODE}\strut
\end{minipage} & \begin{minipage}[t]{0.62\columnwidth}\raggedright
Transcription factor binding sites, gene expression and epigenomics data for cell lines\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.32\columnwidth}\raggedright
\href{http://www.roadmapepigenomics.org/}{Epigenomics Roadmap}\strut
\end{minipage} & \begin{minipage}[t]{0.62\columnwidth}\raggedright
Epigenomics data for multiple cell types\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.32\columnwidth}\raggedright
\href{http://cancergenome.nih.gov/}{The cancer genome atlas (TCGA)}\strut
\end{minipage} & \begin{minipage}[t]{0.62\columnwidth}\raggedright
Expression, mutation and epigenomics data for multiple cancer types\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.32\columnwidth}\raggedright
\href{http://www.1000genomes.org/}{1000 genomes project}\strut
\end{minipage} & \begin{minipage}[t]{0.62\columnwidth}\raggedright
Human genetic variation data obtained by sequencing 1000s of individuals\strut
\end{minipage}\tabularnewline
\bottomrule
\end{longtable}

\hypertarget{Rintro}{%
\chapter{Introduction to R for Genomic Data Analysis}\label{Rintro}}

The aim of computational genomics is to provide biological interpretation and insights from high
dimensional genomics data. Generally speaking, it is similar to any other kind
of data analysis endeavor but often times doing computational genomics will require domain specific knowledge and tools.

As new high-throughput experimental techniques are on the rise, data analysis
capabilities are sought-after features for researchers. The aim of this chapter is to first familiarize the readers with data analysis steps and then provide basics of R programming within the context of genomic data analysis. R is a free statistical programming language that is popular among researchers and data miners to build software and analyze data. Although
basic R programming tutorials are easily accessible, we are aiming to introduce
the subject with the genomic context in the background. The examples and
narrative will always be from real-life situations when you try to analyze
genomic data with R. We believe tailoring material to the context of genomics
makes a difference when learning this programming language for sake of analyzing
genomic data.

\hypertarget{steps-of-genomic-data-analysis}{%
\section{Steps of (genomic) data analysis}\label{steps-of-genomic-data-analysis}}

Regardless of the analysis type, the data analysis has a common pattern. We will
discuss this general pattern and how it applies to genomics problems. The data analysis steps typically include data collection, quality check and cleaning, processing, modeling, visualization and reporting. Although, one expects to go through these steps in a linear fashion, it is normal to go back and repeat the steps with different parameters or tools. In practice, data analysis requires going through the same steps over and over again in order to be able to do a combination of the following: a) answering other related questions, b) dealing with data quality issues that are later realized, and, c) including new data sets to the analysis.

We will now go through a brief explanation of the steps within the context of genomic data analysis.

\hypertarget{data-collection}{%
\subsection{Data collection}\label{data-collection}}

Data collection refers to any source, experiment or survey that provides data for the data analysis question you have. In genomics, data collection is done by high-throughput assays introduced in chapter \ref{intro}. One can also use publicly available data sets and specialized databases also mentioned in chapter \ref{intro}. How much data and what type of data you should collect depends on the question you are trying to answer and the technical and biological variability of the system you are studying.

\hypertarget{data-quality-check-and-cleaning}{%
\subsection{Data quality check and cleaning}\label{data-quality-check-and-cleaning}}

In general, data analysis almost always deals with imperfect data. It is
common to have missing values or measurements that are noisy. Data quality check
and cleaning aims to identify any data quality issue and clean it from the dataset.

High-throughput genomics data is produced by technologies that could embed
technical biases into the data. If we were to give an example from sequencing,
the sequenced reads do not have the same quality of bases called. Towards the
ends of the reads, you could have bases that might be called incorrectly. Identifying those low quality bases and removing them will improve read mapping step.

\hypertarget{data-processing}{%
\subsection{Data processing}\label{data-processing}}

This step refers to processing the data to a format that is suitable for
exploratory analysis and modeling. Often times, the data will not come in ready
to analyze format. You may need to convert it to other formats by transforming
data points (such as log transforming, normalizing etc), or subset the data set
with some arbitrary or pre-defined condition. In terms of genomics, processing
includes multiple steps. Following the sequencing analysis example above,
processing will include aligning reads to the genome and quantification over genes or regions of interest. This is simply counting how many reads are covering your regions of interest. This quantity can give you ideas about how much a gene is expressed if your experimental protocol was RNA sequencing. This can be followed by some normalization to aid the next step.

\hypertarget{exploratory-data-analysis-and-modeling}{%
\subsection{Exploratory data analysis and modeling}\label{exploratory-data-analysis-and-modeling}}

This phase usually takes in the processed or semi-processed data and applies machine-learning or statistical methods to explore the data. Typically, one needs to see relationship between variables measured, relationship between samples based on the variables measured. At this point, we might be looking to see if the samples group as expected by the experimental design, are there outliers or any other anomalies ? After this step you might want to do additional clean up or re-processing to deal with anomalies.

Another related step is modeling. This generally refers to modeling your variable of interest based on other variables you measured. In the context of genomics, it could be that you are trying to predict disease status of the patients from expression of genes you measured from their tissue samples. Then your variable of interest is the disease status and . This is generally called predictive modeling and could be solved with regression based or any other machine-learning methods. This kind of approach is generally called ``predictive modeling''.

Statistical modeling would also be a part of this modeling step, this can cover predictive modeling as well where we use statistical methods such as linear regression. Other analyses such as hypothesis testing, where we have an expectation and we are trying to confirm that expectation is also related to statistical modeling. A good example of this in genomics is the differential gene expression analysis. This can be formulated as comparing two data sets, in this case expression values from condition A and condition B, with the expectation that condition A and condition B has similar expression values. You will see more on this in chapter \ref{stats}.

\hypertarget{visualization-and-reporting}{%
\subsection{Visualization and reporting}\label{visualization-and-reporting}}

Visualization is necessary for all the previous steps more or less. But in the final phase, we need final figures, tables and text that describes the outcome of your analysis. This will be your report. In genomics, we use common data visualization methods as well as specific visualization methods developed or popularized by genomic data analysis. You will see many popular visualization methods in chapters \ref{stats} and @ref\{genomicIntervals\}

\hypertarget{why-use-r-for-genomics}{%
\subsection{Why use R for genomics ?}\label{why-use-r-for-genomics}}

R, with its statistical analysis
heritage, plotting features and rich user-contributed packages is one of the
best languages for the task of analyzing genomic data.
High-dimensional genomics datasets are usually suitable to
be analyzed with core R packages and functions. On top of that, Bioconductor and CRAN have an
array of specialized tools for doing genomics-specific analysis. Here is a list of computational genomics tasks that can be completed using R.

\hypertarget{data-cleanup-and-processing}{%
\subsubsection{Data cleanup and processing}\label{data-cleanup-and-processing}}

Most of general data clean up, such as removing incomplete columns and values, reorganizing and transforming data, these tasks can be achieved using R. In addition, with the help of packages R can connect to databases in various formats such as mySQL, mongoDB, etc., and query and get the data to R environment using database specific tools.

On top of these, genomic data specific processing and quality check can be achieved via R/Bioconductor packages. For example, sequencing read quality checks and even HT-read alignments can be achieved via R packages.

\hypertarget{general-data-anaylsis-and-exploration}{%
\subsubsection{General data anaylsis and exploration}\label{general-data-anaylsis-and-exploration}}

Most genomics data sets are suitable for application of general data analysis tools. In some cases, you may need to preprocess the data to get it to a state that is suitable for application of such tools. Here is a non-exhaustive list of what kind of things can be done via R.

\begin{itemize}
\tightlist
\item
  unsupervised data analysis: clustering (k-means, hierarchical), matrix factorization
  (PCA, ICA etc.)
\item
  supervised data analysis: generalized linear models, support vector machines, random forests
\end{itemize}

\hypertarget{genomics-specific-data-analysis-methods}{%
\subsubsection{Genomics-specific data analysis methods}\label{genomics-specific-data-analysis-methods}}

R/Bioconductor gives you access to multitude of other bioinformatics specific algorithms. Here are some of the things you can do.

\begin{itemize}
\tightlist
\item
  Sequence analysis: TF binding motifs, GC content and CpG counts of a given DNA sequence
\item
  Differential expression (or arrays and sequencing based measurements)
\item
  Gene set/Pathway analysis: What kind of genes are enriched in my gene set
\item
  Genomic Interval operations such as Overlapping CpG islands with transcription start sites, and filtering based on overlaps
\item
  Overlapping aligned reads with exons and counting aligned reads per gene
\end{itemize}

\hypertarget{visualization}{%
\subsubsection{Visualization}\label{visualization}}

Visualization is an important part of all data analysis techniques including computational genomics. Again, you can use core visualization techniques in R and also genomics specific ones with the help of specific packages. Here are some of the things you can do with R.

\begin{itemize}
\tightlist
\item
  Basic plots: Histograms, scatter plots, bar plots, box plots, heatmaps
\item
  ideograms and circos plots for genomics provides visualization of different features over the whole genome.
\item
  meta-profiles of genomic features, such as read enrichment over all promoters
\item
  Visualization of quantitative assays for given locus in the genome
\end{itemize}

\hypertarget{getting-started-with-r}{%
\section{Getting started with R}\label{getting-started-with-r}}

Download and install R \url{http://cran.r-project.org/} and RStudio \url{http://www.rstudio.com/} if you do not have them already. Rstudio is optional but it is a great tool if you are just starting to learn R.
You will need specific data sets to run the codes in this document. Download the data.zip{[}URL to come{]} and extract it to your directory of choice. The folder name should be ``data'' and your R working directory should be level above the data folder. That means in your R console, when you type ``dir(``data'')'' you should be able to see the contents of the data folder. You can change your working directory by \emph{setwd()} command and get your current working directory with \emph{getwd()} command in R. In RStudio, you can click on the top menu and change the location of your working directory via user interface.

\hypertarget{installing-packages}{%
\subsection{Installing packages}\label{installing-packages}}

R packages are add-ons to base R that help you achieve additional tasks that are not directly supported by base R. It is by the action of these extra functionality that R excels as a tool for computational genomics. Bioconductor project (\url{http://bioconductor.org/}) is a dedicated package repository for computational biology related packages. However main package repository of R, called CRAN, has also computational biology related packages. In addition, R-Forge(\url{http://r-forge.r-project.org/}), GitHub(\url{https://github}. com/), and googlecode(\url{http://code.google.com}) are other locations where R packages might be hosted.
You can install CRAN packages using install.packages(). (\# is the comment character in R)

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# install package named "randomForests" from CRAN}
\KeywordTok{install.packages}\NormalTok{(}\StringTok{"randomForests"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

You can install bioconductor packages with a specific installer script

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# get the installer package}
\KeywordTok{source}\NormalTok{(}\StringTok{"http://bioconductor.org/biocLite.R"}\NormalTok{)}
\CommentTok{# install bioconductor package "rtracklayer"}
\KeywordTok{biocLite}\NormalTok{(}\StringTok{"rtracklayer"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

You can install packages from github using \emph{install\_github()} function from \textbf{devtools}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(devtools)}
\KeywordTok{install_github}\NormalTok{(}\StringTok{"hadley/stringr"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Another way to install packages are from the source.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# download the source file}
\KeywordTok{download.file}\NormalTok{(}\StringTok{"http://goo.gl/3pvHYI"}\NormalTok{,}
               \DataTypeTok{destfile=}\StringTok{"methylKit_0.5.7.tar.gz"}\NormalTok{)}
\CommentTok{# install the package from the source file}
\KeywordTok{install.packages}\NormalTok{(}\StringTok{"methylKit_0.5.7.tar.gz"}\NormalTok{,}
                 \DataTypeTok{repos=}\OtherTok{NULL}\NormalTok{,}\DataTypeTok{type=}\StringTok{"source"}\NormalTok{)}
\CommentTok{# delete the source file}
\KeywordTok{unlink}\NormalTok{(}\StringTok{"methylKit_0.5.7.tar.gz"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

You can also update CRAN and Bioconductor packages.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# updating CRAN packages}
\KeywordTok{update.packages}\NormalTok{()}

\CommentTok{# updating bioconductor packages}
\KeywordTok{source}\NormalTok{(}\StringTok{"http://bioconductor.org/biocLite.R"}\NormalTok{)}
\KeywordTok{biocLite}\NormalTok{(}\StringTok{"BiocUpgrade"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{installing-packages-in-custom-locations}{%
\subsection{Installing packages in custom locations}\label{installing-packages-in-custom-locations}}

If you will be using R on servers or computing clusters rather than your personal computer it is unlikely that you will have administrator access to install packages. In that case, you can install packages in custom locations by telling R where to look for additional packages. This is done by setting up an .Renviron file in your home directory and add the following line:

\begin{verbatim}
R_LIBS=~/Rlibs
\end{verbatim}

This tells R that ``Rlibs'' directory at your home directory will be the first choice of locations to look for packages and install packages (The directory name and location is up to you above is just an example). You should go and create that directory now. After that, start a fresh R session and start installing packages. From now on, packages will be installed to your local directory where you have read-write access.

\hypertarget{getting-help-on-functions-and-packages}{%
\subsection{Getting help on functions and packages}\label{getting-help-on-functions-and-packages}}

You can get help on functions by help() and help.search() functions. You can list the functions in a package with ls() function

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(MASS)}
\KeywordTok{ls}\NormalTok{(}\StringTok{"package:MASS"}\NormalTok{) }\CommentTok{# functions in the package}
\KeywordTok{ls}\NormalTok{() }\CommentTok{# objects in your R enviroment}
\CommentTok{# get help on hist() function}
\NormalTok{?hist}
\KeywordTok{help}\NormalTok{(}\StringTok{"hist"}\NormalTok{)}
\CommentTok{# search the word "hist" in help pages}
\KeywordTok{help.search}\NormalTok{(}\StringTok{"hist"}\NormalTok{)}
\NormalTok{??hist}
\end{Highlighting}
\end{Shaded}

\hypertarget{more-help-needed}{%
\subsubsection{More help needed?}\label{more-help-needed}}

In addition, check package vignettes for help and practical understanding of the functions. All Bionconductor packages have vignettes that walk you through example analysis. Google search will always be helpful as well, there are many blogs and web pages that have posts about R. R-help, Stackoverflow and R-bloggers are usually source of good and reliable information.

\hypertarget{computations-in-r}{%
\section{Computations in R}\label{computations-in-r}}

R can be used as an ordinary calculator, some say it is an over-grown calculator. Here are some examples. Remember \textbf{\#} is the comment character. The comments give details about the operations in case they are not clear.

\begin{verbatim}
2 + 3 * 5       # Note the order of operations.
log(10)        # Natural logarithm with base e
5^2            # 5 raised to the second power
3/2            # Division
sqrt(16)      # Square root
abs(3-7)      # Absolute value of 3-7
pi             # The number
exp(2)        # exponential function
# This is a comment line

\end{verbatim}

\hypertarget{data-structures}{%
\section{Data structures}\label{data-structures}}

R has multiple data structures. If you are familiar with excel you can think of a single excel sheet as a table and data structures as building blocks of that table. Most of the time you will deal with tabular data sets or you will want to transform your raw data to a tabular data set, and you will try to manipulate this tabular data set in some way. For example, you may want to take sub-sections of the table or extract all the values in a column. For these and similar purposes, it is essential to know what are the common data structures in R and how they can be used. R deals with named data structures, this means you can give names to data structures and manipulate or operate on them using those names. It will be clear soon what we mean by this if ``named data structures'' does not ring a bell.

\hypertarget{vectors}{%
\subsection{Vectors}\label{vectors}}

Vectors are one the core R data structures. It is basically a list of elements of the same type (numeric,character or logical). Later you will see that every column of a table will be represented as a vector. R handles vectors easily and intuitively. You can create vectors with c() function, however that is not the only way. The operations on vectors will propagate to all the elements of the vectors.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{x<-}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{3}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{10}\NormalTok{,}\DecValTok{5}\NormalTok{)    }\CommentTok{#create a vector named x with 5 components}
\NormalTok{x =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{3}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{10}\NormalTok{,}\DecValTok{5}\NormalTok{)  }
\NormalTok{x}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]  1  3  2 10  5
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{y<-}\DecValTok{1}\OperatorTok{:}\DecValTok{5}              \CommentTok{#create a vector of consecutive integers y}
\NormalTok{y}\OperatorTok{+}\DecValTok{2}                 \CommentTok{#scalar addition}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 3 4 5 6 7
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\DecValTok{2}\OperatorTok{*}\NormalTok{y                 }\CommentTok{#scalar multiplication}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]  2  4  6  8 10
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{y}\OperatorTok{^}\DecValTok{2}                 \CommentTok{#raise each component to the second power}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]  1  4  9 16 25
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\DecValTok{2}\OperatorTok{^}\NormalTok{y                 }\CommentTok{#raise 2 to the first through fifth power}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]  2  4  8 16 32
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{y                   }\CommentTok{#y itself has not been unchanged}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 1 2 3 4 5
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{y<-y}\OperatorTok{*}\DecValTok{2}
\NormalTok{y                   }\CommentTok{#it is now changed}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]  2  4  6  8 10
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{r1<-}\KeywordTok{rep}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{3}\NormalTok{)        }\CommentTok{# create a vector of 1s, length 3}
\KeywordTok{length}\NormalTok{(r1)           }\CommentTok{#length of the vector}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 3
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{class}\NormalTok{(r1)            }\CommentTok{# class of the vector}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "numeric"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{a<-}\DecValTok{1}                \CommentTok{# this is actually a vector length one}
\end{Highlighting}
\end{Shaded}

\hypertarget{matrices}{%
\subsection{Matrices}\label{matrices}}

A matrix refers to a numeric array of rows and columns. You can think of it as a stacked version of vectors where each row or column is a vector. One of the easiest ways to create a matrix is to combine vectors of equal length using \emph{cbind()}, meaning `column bind'.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{x<-}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{3}\NormalTok{,}\DecValTok{4}\NormalTok{)}
\NormalTok{y<-}\KeywordTok{c}\NormalTok{(}\DecValTok{4}\NormalTok{,}\DecValTok{5}\NormalTok{,}\DecValTok{6}\NormalTok{,}\DecValTok{7}\NormalTok{)}
\NormalTok{m1<-}\KeywordTok{cbind}\NormalTok{(x,y);m1}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##      x y
## [1,] 1 4
## [2,] 2 5
## [3,] 3 6
## [4,] 4 7
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{t}\NormalTok{(m1)                }\CommentTok{# transpose of m1}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##   [,1] [,2] [,3] [,4]
## x    1    2    3    4
## y    4    5    6    7
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{dim}\NormalTok{(m1)              }\CommentTok{# 2 by 5 matrix}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 4 2
\end{verbatim}

You can also directly list the elements and specify the matrix:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{m2<-}\KeywordTok{matrix}\NormalTok{(}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{3}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{5}\NormalTok{,}\OperatorTok{-}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{3}\NormalTok{,}\DecValTok{9}\NormalTok{),}\DataTypeTok{nrow=}\DecValTok{3}\NormalTok{)}
\NormalTok{m2}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##      [,1] [,2] [,3]
## [1,]    1    5    2
## [2,]    3   -1    3
## [3,]    2    2    9
\end{verbatim}

Matrices and the next data structure \textbf{data frames} are tabular data structures. You can subset them using \textbf{\protect\hyperlink{section-21}{}} and providing desired rows and columns to subset. Figure \ref{fig:slicingDataFrames} shows how that works conceptually.

\begin{figure}

{\centering \includegraphics[width=0.8\linewidth]{images/slicingDataFrames} 

}

\caption{slicing/subsetting of a matrix and a data frame}\label{fig:slicingDataFrames}
\end{figure}

\hypertarget{data-frames}{%
\subsection{Data Frames}\label{data-frames}}

A data frame is more general than a matrix, in that different columns can have different modes (numeric, character, factor, etc.). A data frame can be constructed by data.frame() function. For example, we illustrate how to construct a data frame from genomic intervals or coordinates.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{chr <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"chr1"}\NormalTok{, }\StringTok{"chr1"}\NormalTok{, }\StringTok{"chr2"}\NormalTok{, }\StringTok{"chr2"}\NormalTok{)}
\NormalTok{strand <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"-"}\NormalTok{,}\StringTok{"-"}\NormalTok{,}\StringTok{"+"}\NormalTok{,}\StringTok{"+"}\NormalTok{)}
\NormalTok{start<-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\DecValTok{200}\NormalTok{,}\DecValTok{4000}\NormalTok{,}\DecValTok{100}\NormalTok{,}\DecValTok{400}\NormalTok{)}
\NormalTok{end<-}\KeywordTok{c}\NormalTok{(}\DecValTok{250}\NormalTok{,}\DecValTok{410}\NormalTok{,}\DecValTok{200}\NormalTok{,}\DecValTok{450}\NormalTok{)}
\NormalTok{mydata <-}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(chr,start,end,strand)}
\CommentTok{#change column names}
\KeywordTok{names}\NormalTok{(mydata) <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"chr"}\NormalTok{,}\StringTok{"start"}\NormalTok{,}\StringTok{"end"}\NormalTok{,}\StringTok{"strand"}\NormalTok{)}
\NormalTok{mydata }\CommentTok{# OR this will work too}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##    chr start end strand
## 1 chr1   200 250      -
## 2 chr1  4000 410      -
## 3 chr2   100 200      +
## 4 chr2   400 450      +
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mydata <-}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{chr=}\NormalTok{chr,}\DataTypeTok{start=}\NormalTok{start,}\DataTypeTok{end=}\NormalTok{end,}\DataTypeTok{strand=}\NormalTok{strand)}
\NormalTok{mydata}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##    chr start end strand
## 1 chr1   200 250      -
## 2 chr1  4000 410      -
## 3 chr2   100 200      +
## 4 chr2   400 450      +
\end{verbatim}

There are a variety of ways to extract the elements of a data frame. You can extract certain columns using column numbers or names, or you can extract certain rows by using row numbers. You can also extract data using logical arguments, such as extracting all rows that has a value in a column larger than your threshold.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mydata[,}\DecValTok{2}\OperatorTok{:}\DecValTok{4}\NormalTok{] }\CommentTok{# columns 2,3,4 of data frame}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##   start end strand
## 1   200 250      -
## 2  4000 410      -
## 3   100 200      +
## 4   400 450      +
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mydata[,}\KeywordTok{c}\NormalTok{(}\StringTok{"chr"}\NormalTok{,}\StringTok{"start"}\NormalTok{)] }\CommentTok{# columns chr and start from data frame}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##    chr start
## 1 chr1   200
## 2 chr1  4000
## 3 chr2   100
## 4 chr2   400
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mydata}\OperatorTok{$}\NormalTok{start }\CommentTok{# variable start in the data frame}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]  200 4000  100  400
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mydata[}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{3}\NormalTok{),] }\CommentTok{# get 1st and 3rd rows}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##    chr start end strand
## 1 chr1   200 250      -
## 3 chr2   100 200      +
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mydata[mydata}\OperatorTok{$}\NormalTok{start}\OperatorTok{>}\DecValTok{400}\NormalTok{,] }\CommentTok{# get all rows where start>400}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##    chr start end strand
## 2 chr1  4000 410      -
\end{verbatim}

\hypertarget{lists}{%
\subsection{Lists}\label{lists}}

An ordered collection of objects (components). A list allows you to gather a variety of (possibly unrelated) objects under one name.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# example of a list with 4 components}
\CommentTok{# a string, a numeric vector, a matrix, and a scalar}
\NormalTok{w <-}\StringTok{ }\KeywordTok{list}\NormalTok{(}\DataTypeTok{name=}\StringTok{"Fred"}\NormalTok{,}
       \DataTypeTok{mynumbers=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{3}\NormalTok{),}
       \DataTypeTok{mymatrix=}\KeywordTok{matrix}\NormalTok{(}\DecValTok{1}\OperatorTok{:}\DecValTok{4}\NormalTok{,}\DataTypeTok{ncol=}\DecValTok{2}\NormalTok{),}
       \DataTypeTok{age=}\FloatTok{5.3}\NormalTok{)}
\NormalTok{w}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## $name
## [1] "Fred"
## 
## $mynumbers
## [1] 1 2 3
## 
## $mymatrix
##      [,1] [,2]
## [1,]    1    3
## [2,]    2    4
## 
## $age
## [1] 5.3
\end{verbatim}

You can extract elements of a list using the \textbf{{[}\protect\hyperlink{section-21}{}{]}} convention using either its position in the list or its name.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{w[[}\DecValTok{3}\NormalTok{]] }\CommentTok{# 3rd component of the list}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##      [,1] [,2]
## [1,]    1    3
## [2,]    2    4
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{w[[}\StringTok{"mynumbers"}\NormalTok{]] }\CommentTok{# component named mynumbers in list}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 1 2 3
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{w}\OperatorTok{$}\NormalTok{age}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 5.3
\end{verbatim}

\hypertarget{factors}{%
\subsection{Factors}\label{factors}}

Factors are used to store categorical data. They are important for statistical modeling since categorical variables are treated differently in statistical models than continuous variables. This ensures categorical data treated accordingly in statistical models.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{features=}\KeywordTok{c}\NormalTok{(}\StringTok{"promoter"}\NormalTok{,}\StringTok{"exon"}\NormalTok{,}\StringTok{"intron"}\NormalTok{)}
\NormalTok{f.feat=}\KeywordTok{factor}\NormalTok{(features)}
\end{Highlighting}
\end{Shaded}

Important thing to note is that when you are reading a data.frame with read.table() or creating a data frame with \textbf{data.frame()} character columns are stored as factors by default, to change this behavior you need to set \textbf{stringsAsFactors=FALSE} in \textbf{read.table()} and/or \textbf{data.frame()} function arguments.

\hypertarget{data-types}{%
\section{Data types}\label{data-types}}

There are four common data types in R, they are \textbf{numeric}, \textbf{logical}, \textbf{character} and \textbf{integer}. All these data types can be used to create vectors natively.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#create a numeric vector x with 5 components}
\NormalTok{x<-}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{3}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{10}\NormalTok{,}\DecValTok{5}\NormalTok{)}
\NormalTok{x}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]  1  3  2 10  5
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#create a logical vector x}
\NormalTok{x<-}\KeywordTok{c}\NormalTok{(}\OtherTok{TRUE}\NormalTok{,}\OtherTok{FALSE}\NormalTok{,}\OtherTok{TRUE}\NormalTok{)}
\NormalTok{x}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]  TRUE FALSE  TRUE
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# create a character vector}
\NormalTok{x<-}\KeywordTok{c}\NormalTok{(}\StringTok{"sds"}\NormalTok{,}\StringTok{"sd"}\NormalTok{,}\StringTok{"as"}\NormalTok{)}
\NormalTok{x}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "sds" "sd"  "as"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{class}\NormalTok{(x)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "character"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# create an integer vector}
\NormalTok{x<-}\KeywordTok{c}\NormalTok{(1L,2L,3L)}
\NormalTok{x}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 1 2 3
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{class}\NormalTok{(x)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "integer"
\end{verbatim}

\hypertarget{reading-and-writing-data}{%
\section{Reading and writing data}\label{reading-and-writing-data}}

Most of the genomics data sets are in the form of genomic intervals associated with a score. That means mostly the data will be in table format with columns denoting chromosome, start positions, end positions, strand and score. One of the popular formats is BED format used primarily by UCSC genome browser but most other genome browsers and tools will support BED format. We have all the annotation data in BED format. In R, you can easily read tabular format data with read.table() function.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{enhancerFilePath=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
                \StringTok{"subset.enhancers.hg18.bed"}\NormalTok{,}
                \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{cpgiFilePath=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
                \StringTok{"subset.cpgi.hg18.bed"}\NormalTok{,}
                \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\CommentTok{# read enhancer marker BED file}
\NormalTok{enh.df <-}\StringTok{ }\KeywordTok{read.table}\NormalTok{(enhancerFilePath, }\DataTypeTok{header =} \OtherTok{FALSE}\NormalTok{) }

\CommentTok{# read CpG island BED file}
\NormalTok{cpgi.df <-}\StringTok{ }\KeywordTok{read.table}\NormalTok{(cpgiFilePath, }\DataTypeTok{header =} \OtherTok{FALSE}\NormalTok{) }

\CommentTok{# check first lines to see how the data looks like}
\KeywordTok{head}\NormalTok{(enh.df)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##      V1     V2     V3 V4   V5 V6    V7    V8 V9
## 1 chr20 266275 267925  . 1000  .  9.11 13.17 -1
## 2 chr20 287400 294500  . 1000  . 10.53 13.02 -1
## 3 chr20 300500 302500  . 1000  .  9.10 13.39 -1
## 4 chr20 330400 331800  . 1000  .  6.39 13.51 -1
## 5 chr20 341425 343400  . 1000  .  6.20 12.99 -1
## 6 chr20 437975 439900  . 1000  .  6.31 13.52 -1
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{head}\NormalTok{(cpgi.df)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##      V1     V2     V3       V4
## 1 chr20 195575 195851  CpG:_28
## 2 chr20 207789 208148  CpG:_32
## 3 chr20 219055 219437  CpG:_33
## 4 chr20 225831 227155 CpG:_135
## 5 chr20 252826 256323 CpG:_286
## 6 chr20 275376 276977 CpG:_116
\end{verbatim}

You can save your data by writing it to disk as a text file. A data frame or matrix can be written out by using write.table() function. Now let us write out cpgi.df, we will write it out as a tab-separated file, pay attention to the arguments.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{write.table}\NormalTok{(cpgi.df,}\DataTypeTok{file=}\StringTok{"cpgi.txt"}\NormalTok{,}\DataTypeTok{quote=}\OtherTok{FALSE}\NormalTok{,}
            \DataTypeTok{row.names=}\OtherTok{FALSE}\NormalTok{,}\DataTypeTok{col.names=}\OtherTok{FALSE}\NormalTok{,}\DataTypeTok{sep=}\StringTok{"}\CharTok{\textbackslash{}t}\StringTok{"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

You can save your R objects directly into a file using save() and saveRDS() and load them back in with load() and readRDS(). By using these functions you can save any R object whether or not they are in data frame or matrix classes.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{save}\NormalTok{(cpgi.df,enh.df,}\DataTypeTok{file=}\StringTok{"mydata.RData"}\NormalTok{)}
\KeywordTok{load}\NormalTok{(}\StringTok{"mydata.RData"}\NormalTok{)}
\CommentTok{# saveRDS() can save one object at a type}
\KeywordTok{saveRDS}\NormalTok{(cpgi.df,}\DataTypeTok{file=}\StringTok{"cpgi.rds"}\NormalTok{)}
\NormalTok{x=}\KeywordTok{readRDS}\NormalTok{(}\StringTok{"cpgi.rds"}\NormalTok{)}
\KeywordTok{head}\NormalTok{(x)}
\end{Highlighting}
\end{Shaded}

One important thing is that with \texttt{save()} you can save many objects at a time and when they are loaded into memory with \texttt{load(} they retain their variable names. For example, in the above code when you use \texttt{load("mydata.RData")} in a fresh R session, an object names ``cpg.df'' will be created. That means you have to figure out what name you gave it to the objects before saving them. On the contrary to that, when you save an object by \texttt{saveRDS()} and read by \texttt{readRDS()} the name of the object is not retained, you need to assign the output of \texttt{readRDS()} to a new variable (``x'' in the above code chunk).

\hypertarget{plotting-in-r}{%
\section{Plotting in R}\label{plotting-in-r}}

R has great support for plotting and customizing plots. We will show only a few below. Let us sample 50 values from normal distribution and plot them as a histogram.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# sample 50 values from normal distribution}
\CommentTok{# and store them in vector x}
\NormalTok{x<-}\KeywordTok{rnorm}\NormalTok{(}\DecValTok{50}\NormalTok{)}
\KeywordTok{hist}\NormalTok{(x) }\CommentTok{# plot the histogram of those values}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics[width=0.65\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-20-1} \end{center}

We can modify all the plots by providing certain arguments to the plotting function. Now let's give a title to the plot using \textbf{`main'} argument. We can also change the color of the bars using \textbf{`col'} argument. You can simply provide the name of the color. Below, we are using \textbf{`red'} for the color. See Figure below for the result this chunk.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{hist}\NormalTok{(x,}\DataTypeTok{main=}\StringTok{"Hello histogram!!!"}\NormalTok{,}\DataTypeTok{col=}\StringTok{"red"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics[width=0.65\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-21-1} \end{center}

Next, we will make a scatter plot. Scatter plots are one the most common plots you will encounter in data analysis. We will sample another set of 50 values and plotted those against the ones we sampled earlier. Scatterplot shows values of two variables for a set of data points. It is useful to visualize relationships between two variables. It is frequently used in connection with correlation and linear regression. There are other variants of scatter plots which show density of the points with different colors. We will show examples of those that in following chapters. The scatter plot from our sampling experiment is shown in the figure. Notice that, in addition to main we used \textbf{``xlab''} and \textbf{``ylab''} arguments to give labels to the plot. You can customize the plots even more than this. See \textbf{?plot} and \textbf{?par} for more arguments that can help you customize the plots.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# randomly sample 50 points from normal distribution}
\NormalTok{y<-}\KeywordTok{rnorm}\NormalTok{(}\DecValTok{50}\NormalTok{)}
\CommentTok{#plot a scatter plot}
\CommentTok{# control x-axis and y-axis labels}
\KeywordTok{plot}\NormalTok{(x,y,}\DataTypeTok{main=}\StringTok{"scatterplot of random samples"}\NormalTok{,}
        \DataTypeTok{ylab=}\StringTok{"y values"}\NormalTok{,}\DataTypeTok{xlab=}\StringTok{"x values"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics[width=0.65\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-22-1} \end{center}

we can also plot boxplots for vectors x and y. Boxplots depict groups of numerical data through their quartiles. The edges of the box denote 1st and 3rd quartile, and the line that crosses the box is the median. Whiskers usually are defined using interquantile range:

\emph{lowerWhisker=Q1-1.5{[}IQR{]} and upperWhisker=Q1+1.5{[}IQR{]}}

In addition, outliers can be depicted as dots. In this case, outliers are the values that remain outside the whiskers.

\begin{Shaded}
\begin{Highlighting}[]
 \KeywordTok{boxplot}\NormalTok{(x,y,}\DataTypeTok{main=}\StringTok{"boxplots of random samples"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics[width=0.65\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-23-1} \end{center}

Next up is bar plot which you can plot by \textbf{barplot()} function. We are going to plot four imaginary percentage values and color them with two colors, and this time we will also show how to draw a legend on the plot using \textbf{legend()} function.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{perc=}\KeywordTok{c}\NormalTok{(}\DecValTok{50}\NormalTok{,}\DecValTok{70}\NormalTok{,}\DecValTok{35}\NormalTok{,}\DecValTok{25}\NormalTok{)}
\KeywordTok{barplot}\NormalTok{(}\DataTypeTok{height=}\NormalTok{perc,}
        \DataTypeTok{names.arg=}\KeywordTok{c}\NormalTok{(}\StringTok{"CpGi"}\NormalTok{,}\StringTok{"exon"}\NormalTok{,}\StringTok{"CpGi"}\NormalTok{,}\StringTok{"exon"}\NormalTok{),}
        \DataTypeTok{ylab=}\StringTok{"percentages"}\NormalTok{,}\DataTypeTok{main=}\StringTok{"imagine %s"}\NormalTok{,}
        \DataTypeTok{col=}\KeywordTok{c}\NormalTok{(}\StringTok{"red"}\NormalTok{,}\StringTok{"red"}\NormalTok{,}\StringTok{"blue"}\NormalTok{,}\StringTok{"blue"}\NormalTok{))}
\KeywordTok{legend}\NormalTok{(}\StringTok{"topright"}\NormalTok{,}\DataTypeTok{legend=}\KeywordTok{c}\NormalTok{(}\StringTok{"test"}\NormalTok{,}\StringTok{"control"}\NormalTok{),}
       \DataTypeTok{fill=}\KeywordTok{c}\NormalTok{(}\StringTok{"red"}\NormalTok{,}\StringTok{"blue"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics[width=0.65\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-24-1} \end{center}

\hypertarget{saving-plots}{%
\section{Saving plots}\label{saving-plots}}

If you want to save your plots to an image file there are couple of ways of doing that. Normally, you will have to do the following:
1. Open a graphics device
2. Create the plot
3. Close the graphics device

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{pdf}\NormalTok{(}\StringTok{"mygraphs/myplot.pdf"}\NormalTok{,}\DataTypeTok{width=}\DecValTok{5}\NormalTok{,}\DataTypeTok{height=}\DecValTok{5}\NormalTok{)}
\KeywordTok{plot}\NormalTok{(x,y)}
\KeywordTok{dev.off}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

Alternatively, you can first create the plot then copy the plot to a graphic device.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{plot}\NormalTok{(x,y)}
\KeywordTok{dev.copy}\NormalTok{(pdf,}\StringTok{"mygraphs/myplot.pdf"}\NormalTok{,}\DataTypeTok{width=}\DecValTok{7}\NormalTok{,}\DataTypeTok{height=}\DecValTok{5}\NormalTok{)}
\KeywordTok{dev.off}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\hypertarget{functions-and-control-structures-for-ifelse-etc.}{%
\section{Functions and control structures (for, if/else etc.)}\label{functions-and-control-structures-for-ifelse-etc.}}

\hypertarget{user-defined-functions}{%
\subsection{User defined functions}\label{user-defined-functions}}

Functions are useful for transforming larger chunks of code to re-usable pieces of code. Generally, if you need to execute certain tasks with variable parameters then it is time you write a function. A function in R takes different arguments and returns a definite output, much like mathematical functions. Here is a simple function takes two arguments, x and y, and returns the sum of their squares.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sqSum<-}\ControlFlowTok{function}\NormalTok{(x,y)\{}
\NormalTok{result=x}\OperatorTok{^}\DecValTok{2}\OperatorTok{+}\NormalTok{y}\OperatorTok{^}\DecValTok{2}
\KeywordTok{return}\NormalTok{(result)}
\NormalTok{\}}
\CommentTok{# now try the function out}
\KeywordTok{sqSum}\NormalTok{(}\DecValTok{2}\NormalTok{,}\DecValTok{3}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 13
\end{verbatim}

Functions can also output plots and/or messages to the terminal. Here is a function that prints a message to the terminal:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{sqSumPrint<-}\ControlFlowTok{function}\NormalTok{(x,y)\{}
\NormalTok{result=x}\OperatorTok{^}\DecValTok{2}\OperatorTok{+}\NormalTok{y}\OperatorTok{^}\DecValTok{2}
\KeywordTok{cat}\NormalTok{(}\StringTok{"here is the result:"}\NormalTok{,result,}\StringTok{"}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{)}
\NormalTok{\}}
\CommentTok{# now try the function out}
\KeywordTok{sqSumPrint}\NormalTok{(}\DecValTok{2}\NormalTok{,}\DecValTok{3}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## here is the result: 13
\end{verbatim}

Sometimes we would want to execute a certain part of the code only if certain condition is satisfied. This condition can be anything from the type of an object (Ex: if object is a matrix execute certain code), or it can be more complicated such as if object value is between certain thresholds. Let us see how they can be used3. They can be used anywhere in your code, now we will use it in a function.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cpgi.df <-}\StringTok{ }\KeywordTok{read.table}\NormalTok{(}\StringTok{"intro2R_data/data/subset.cpgi.hg18.bed"}\NormalTok{, }\DataTypeTok{header =} \OtherTok{FALSE}\NormalTok{)}
\CommentTok{# function takes input one row}
\CommentTok{# of CpGi data frame}
\NormalTok{largeCpGi<-}\ControlFlowTok{function}\NormalTok{(bedRow)\{}
\NormalTok{ cpglen=bedRow[}\DecValTok{3}\NormalTok{]}\OperatorTok{-}\NormalTok{bedRow[}\DecValTok{2}\NormalTok{]}\OperatorTok{+}\DecValTok{1}
 \ControlFlowTok{if}\NormalTok{(cpglen}\OperatorTok{>}\DecValTok{1500}\NormalTok{)\{}
    \KeywordTok{cat}\NormalTok{(}\StringTok{"this is large}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{)}
\NormalTok{ \}}
 \ControlFlowTok{else} \ControlFlowTok{if}\NormalTok{(cpglen}\OperatorTok{<=}\DecValTok{1500} \OperatorTok{&}\StringTok{ }\NormalTok{cpglen}\OperatorTok{>}\DecValTok{700}\NormalTok{)\{}
    \KeywordTok{cat}\NormalTok{(}\StringTok{"this is normal}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{)}
\NormalTok{ \}}
 \ControlFlowTok{else}\NormalTok{\{}
    \KeywordTok{cat}\NormalTok{(}\StringTok{"this is short}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{)}
\NormalTok{ \}}
\NormalTok{\}}
\KeywordTok{largeCpGi}\NormalTok{(cpgi.df[}\DecValTok{10}\NormalTok{,])}
\KeywordTok{largeCpGi}\NormalTok{(cpgi.df[}\DecValTok{100}\NormalTok{,])}
\KeywordTok{largeCpGi}\NormalTok{(cpgi.df[}\DecValTok{1000}\NormalTok{,])}
\end{Highlighting}
\end{Shaded}

\hypertarget{loops-and-looping-structures-in-r}{%
\subsection{Loops and looping structures in R}\label{loops-and-looping-structures-in-r}}

When you need to repeat a certain task or a execute a function multiple times, you can do that with the help of loops. A loop will execute the task until a certain condition is reached. The loop below is called a ``for-loop'' and it executes the task sequentially 10 times.

\begin{Shaded}
\begin{Highlighting}[]
\ControlFlowTok{for}\NormalTok{(i }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{10}\NormalTok{)\{ }\CommentTok{# number of repetitions}
\KeywordTok{cat}\NormalTok{(}\StringTok{"This is iteration"}\NormalTok{) }\CommentTok{# the task to be repeated}
\KeywordTok{print}\NormalTok{(i)}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## This is iteration[1] 1
## This is iteration[1] 2
## This is iteration[1] 3
## This is iteration[1] 4
## This is iteration[1] 5
## This is iteration[1] 6
## This is iteration[1] 7
## This is iteration[1] 8
## This is iteration[1] 9
## This is iteration[1] 10
\end{verbatim}

The task above is a bit pointless, normally in a loop, you would want to do something meaningful. Let us calculate the length of the CpG islands we read in earlier. Although this is not the most efficient way of doing that particular task, it serves as a good example for looping. The code below will be execute hundred times, and it will calculate the length of the CpG islands for the first 100 islands in
the data frame (by subtracting the end coordinate from the start coordinate).

\textbf{Note:}If you are going to run a loop that has a lot of repetitions, it is smart to try the loop with few repetitions first and check the results. This will help you make sure the code in the loop works before executing it for thousands of times.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# this is where we will keep the lenghts}
\CommentTok{# for now it is an empty vector}
\NormalTok{result=}\KeywordTok{c}\NormalTok{()}
\CommentTok{# start the loop}
\ControlFlowTok{for}\NormalTok{(i }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{100}\NormalTok{)\{}
    \CommentTok{#calculate the length}
\NormalTok{    len=cpgi.df[i,}\DecValTok{3}\NormalTok{]}\OperatorTok{-}\NormalTok{cpgi.df[i,}\DecValTok{2}\NormalTok{]}\OperatorTok{+}\DecValTok{1}
    \CommentTok{#append the length to the result}
\NormalTok{    result=}\KeywordTok{c}\NormalTok{(result,len)}
\NormalTok{\}}
\CommentTok{# check the results}
\KeywordTok{head}\NormalTok{(result)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]  277  360  383 1325 3498 1602
\end{verbatim}

\hypertarget{apply-family-functions-instead-of-loops}{%
\subsubsection{apply family functions instead of loops}\label{apply-family-functions-instead-of-loops}}

R has other ways of repeating tasks that tend to be more efficient than using loops. They are known as the \textbf{``apply''} family of functions, which include \emph{apply, lapply, mapply and tapply} (and some other variants). All of these functions apply a given function to a set of instances and returns the result of those functions for each instance. The differences between them is that they take different type of inputs. For example apply works on data frames or matrices and applies the function on each row or column of the data structure. \emph{lapply} works on lists or vectors and applies a function which takes the list element as an argument. Next we will demonstrate how to use \textbf{apply()} on a matrix. The example applies the sum function on the rows of a matrix, it basically sums up the values on each row of the matrix, which is conceptualized in Figure \ref{fig:applyConcept}.

\begin{figure}

{\centering \includegraphics[width=0.8\linewidth]{images/apply} 

}

\caption{apply concept in R}\label{fig:applyConcept}
\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mat=}\KeywordTok{cbind}\NormalTok{(}\KeywordTok{c}\NormalTok{(}\DecValTok{3}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{3}\NormalTok{,}\DecValTok{3}\NormalTok{),}\KeywordTok{c}\NormalTok{(}\DecValTok{3}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{),}\KeywordTok{c}\NormalTok{(}\DecValTok{3}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{3}\NormalTok{),}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{),}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{,}\DecValTok{0}\NormalTok{),}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{,}\DecValTok{0}\NormalTok{))}
\NormalTok{result<-}\KeywordTok{apply}\NormalTok{(mat,}\DecValTok{1}\NormalTok{,sum)}
\NormalTok{result}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 12  3  5  6
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# OR you can define the function as an argument to apply()}
\NormalTok{result<-}\KeywordTok{apply}\NormalTok{(mat,}\DecValTok{1}\NormalTok{,}\ControlFlowTok{function}\NormalTok{(x) }\KeywordTok{sum}\NormalTok{(x))}
\NormalTok{result}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 12  3  5  6
\end{verbatim}

Notice that we used a second argument which equals to 1, that indicates that rows of the matrix/ data frame will be the input for the function. If we change the second argument to 2, this will indicate that columns should be the input for the function that will be applied. See Figure \ref{fig:applyConcept2} for the visualization of apply() on columns.

\begin{figure}

{\centering \includegraphics[width=0.8\linewidth]{images/apply2} 

}

\caption{apply function on columns}\label{fig:applyConcept2}
\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{result<-}\KeywordTok{apply}\NormalTok{(mat,}\DecValTok{2}\NormalTok{,sum)}
\NormalTok{result}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 9 3 6 2 3 3
\end{verbatim}

Next, we will use \textbf{lapply()}, which applies a function on a list or a vector. The function that will be applied is a simple function that takes the square of a given number.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{input=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{3}\NormalTok{)}
\KeywordTok{lapply}\NormalTok{(input,}\ControlFlowTok{function}\NormalTok{(x) x}\OperatorTok{^}\DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [[1]]
## [1] 1
## 
## [[2]]
## [1] 4
## 
## [[3]]
## [1] 9
\end{verbatim}

\textbf{mapply()} is another member of apply family, it can apply a function on an unlimited set of vectors/lists, it is like a version of lapply that can handle multiple vectors as arguments. In this case, the argument to the mapply() is the function to be applied and the sets of parameters to be supplied as arguments of the function. This conceptualized Figure \ref{fig:mapplyConcept}, the function to be applied is a function that takes to arguments and sums them up. The arguments to be summed up are in the format of vectors, Xs and Ys. \texttt{mapply()} applies the summation function to each pair in Xs and Ys vector. Notice that the order of the input function and extra arguments are different for mapply.

\begin{figure}

{\centering \includegraphics[width=0.8\linewidth]{images/mapply} 

}

\caption{mapply concept}\label{fig:mapplyConcept}
\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{Xs=}\DecValTok{0}\OperatorTok{:}\DecValTok{5}
\NormalTok{Ys=}\KeywordTok{c}\NormalTok{(}\DecValTok{2}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{3}\NormalTok{,}\DecValTok{3}\NormalTok{,}\DecValTok{3}\NormalTok{)}
\NormalTok{result<-}\KeywordTok{mapply}\NormalTok{(}\ControlFlowTok{function}\NormalTok{(x,y) }\KeywordTok{sum}\NormalTok{(x,y),Xs,Ys)}
\NormalTok{result}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 2 3 4 6 7 8
\end{verbatim}

\hypertarget{apply-family-functions-on-multiple-cores}{%
\subsubsection{apply family functions on multiple cores}\label{apply-family-functions-on-multiple-cores}}

If you have large data sets apply family functions can be slow (although probably still better than for loops). If that is the case, you can easily use the parallel versions of those functions from parallel package. These functions essentially divide your tasks to smaller chunks run them on separate CPUs and merge the results from those parallel operations. This concept is visualized at Figure below , \texttt{mcapply} runs the summation function on three different processors. Each processor executes the summation function on a part of the data set, and the results are merged and returned as a single vector that has the same order as the input parameters Xs and Ys.

\begin{figure}

{\centering \includegraphics[width=0.8\linewidth]{images/mcmapply} 

}

\caption{mcapplyconcept}\label{fig:mcapplyConcept}
\end{figure}

\hypertarget{vectorized-functions-in-r}{%
\subsubsection{Vectorized Functions in R}\label{vectorized-functions-in-r}}

The above examples have been put forward to illustrate functions and loops in R because functions using sum() are not complicated and easy to understand. You will probably need to use loops and looping structures with more complicated functions. In reality, most of the operations we used do not need the use of loops or looping structures because there are already vectorized functions that can achieve the same outcomes, meaning if the input arguments are R vectors the output will be a vector as well, so no need for loops or vectorization.

For example, instead of using mapply() and sum() functions we can just use + operator and sum up Xs and Ys.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{result=Xs}\OperatorTok{+}\NormalTok{Ys}
\NormalTok{result}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 2 3 4 6 7 8
\end{verbatim}

In order to get the column or row sums, we can use the vectorized functions colSums() and rowSums().

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{colSums}\NormalTok{(mat)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 9 3 6 2 3 3
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{rowSums}\NormalTok{(mat)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 12  3  5  6
\end{verbatim}

However, remember that not every function is vectorized in R, use the ones that are. But sooner or later, apply family functions will come in handy.

\hypertarget{exercises}{%
\section{Exercises}\label{exercises}}

\hypertarget{computations-in-r-1}{%
\subsection{Computations in R}\label{computations-in-r-1}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  Sum 2 and 3, use +
\item
  Take the square root of 36, use sqrt()
\item
  Take the log10 of 1000, use function \texttt{log10()}
\item
  Take the log2 of 32, use function \texttt{log2()}
\item
  Assign the sum of 2,3 and 4 to variable x
\item
  Find the absolute value of \texttt{5\ -\ 145} using \texttt{abs()} function
\item
  Calculate the square root of 625, divide it by 5 and assign it to variable \texttt{x}.
  Ex: \texttt{y=\ log10(1000)/5}, the previous statement takes log10 of 1000, divides it
  by 5 and assigns the value to variable y
\item
  Multiply the value you get from previous exercise with 10000, assign it variable x
  Ex: \texttt{y=y*5}, multiplies y with 5 and assigns the value to y.
\end{enumerate}

\textbf{KEY CONCEPT:} results of computations or arbitrary values can be stored in variables we can re-use those variables later on and over-write them with new values

\hypertarget{data-structures-in-r}{%
\subsection{Data structures in R}\label{data-structures-in-r}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{9}
\item
  Make a vector of 1,2,3,5 and 10 using \texttt{c()}, assign it to \texttt{vec} variable.
  Ex: \texttt{vec1=c(1,3,4)} makes a vector out of 1,3,4.
\item
  Check the length of your vector with length().
  Ex: \texttt{length(vec1)} should return 3
\item
  Make a vector of all numbers between 2 and 15.
  Ex: \texttt{vec=1:6} makes a vector of numbers between 1 and 6, assigns to vec variable
\item
  Make a vector of 4s repeated 10 times using \texttt{rep()} function. Ex: \texttt{rep(x=2,times=5)} makes a vector of 2s repeated 5 times
\item
  Make a logical vector with TRUE, FALSE values of length 4, use \texttt{c()}.
  Ex: \texttt{c(TRUE,FALSE)}
\item
  Make a character vector of gene names PAX6,ZIC2,OCT4 and SOX2.
  Ex: \texttt{avec=c("a","b","c")} a makes a character vector of a,b and c
\item
  Subset the vector using \texttt{{[}{]}} notation, get 5th and 6th elements.
  Ex: \texttt{vec1{[}1{]}} gets the first element. \texttt{vec1{[}c(1,3){]}} gets 1st and 3rd elements
\item
  You can also subset any vector using a logical vector in \texttt{{[}{]}}. Run the following:
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{myvec=}\DecValTok{1}\OperatorTok{:}\DecValTok{5}
\NormalTok{myvec[}\KeywordTok{c}\NormalTok{(}\OtherTok{TRUE}\NormalTok{,}\OtherTok{TRUE}\NormalTok{,}\OtherTok{FALSE}\NormalTok{,}\OtherTok{FALSE}\NormalTok{,}\OtherTok{FALSE}\NormalTok{)] }\CommentTok{# the length of the logical vector should be equal to length(myvec) }
\NormalTok{myvec[}\KeywordTok{c}\NormalTok{(}\OtherTok{TRUE}\NormalTok{,}\OtherTok{FALSE}\NormalTok{,}\OtherTok{FALSE}\NormalTok{,}\OtherTok{FALSE}\NormalTok{,}\OtherTok{TRUE}\NormalTok{)]}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{17}
\tightlist
\item
  \texttt{==,\textgreater{},\textless{},\ \textgreater{}=,\ \textless{}=} operators create logical vectors. See the results of the following operations:
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{myvec }\OperatorTok{>}\StringTok{ }\DecValTok{3}
\NormalTok{myvec }\OperatorTok{==}\StringTok{ }\DecValTok{4}
\NormalTok{myvec }\OperatorTok{<=}\StringTok{ }\DecValTok{2}
\NormalTok{myvec }\OperatorTok{!=}\StringTok{ }\DecValTok{4}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{18}
\item
  Use \texttt{\textgreater{}} operator in \texttt{myvec{[}\ {]}} to get elements larger than 2 in \texttt{myvec} whic is described above
\item
  make a 5x3 matrix (5 rows, 3 columns) using \texttt{matrix()}.
  Ex: matrix(1:6,nrow=3,ncol=2) makes a 3x2 matrix using numbers between 1 and 6
\item
  What happens when you use \texttt{byrow\ =\ TRUE} in your matrix() as an additional argument?
  Ex: \texttt{mat=matrix(1:6,nrow=3,ncol=2,byrow\ =\ TRUE)}
\item
  Extract first 3 columns and first 3 rows of your matrix using \texttt{{[}{]}} notation.
\item
  Extract last two rows of the matrix you created earlier.
\end{enumerate}

Ex: \texttt{mat{[}2:3,{]}} or \texttt{mat{[}c(2,3),{]}} extracts 2nd and 3rd rows.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{23}
\item
  Extract the first two columns and run \texttt{class()} on the result.
\item
  Extract first column and run \texttt{class()} on the result, compare with the above exercise.
\item
  Make a data frame with 3 columns and 5 rows, make sure first column is sequence
  of numbers 1:5, and second column is a character vector.
\end{enumerate}

Ex: \texttt{df=data.frame(col1=1:3,col2=c("a","b","c"),col3=3:1)\ \#\ 3x3\ data\ frame}.

Remember you need to make 3x5 data frame

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{26}
\tightlist
\item
  Extract first two columns and first two rows.
\end{enumerate}

\textbf{HINT:} Same notation as matrices

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{27}
\tightlist
\item
  Extract last two rows of the data frame you made.
\end{enumerate}

\textbf{HINT:} Same notation as matrices

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{28}
\item
  Extract last two columns using column names of the data frame you made.
\item
  Extract second column using column names.
  You can use \texttt{{[}{]}} or \texttt{\$} as in lists, use both in two different answers.
\item
  Extract rows where 1st column is larger than 3.
  \textbf{HINT:} you can get a logical vector using \texttt{\textgreater{}} operator
  ,logical vectors can be used in \texttt{{[}{]}} when subsetting.
\item
  Extract rows where 1st column is larger than or equal to 3.
\item
  Convert data frame to the matrix. \textbf{HINT:} use \texttt{as.matrix()}.
\end{enumerate}

Observe what happens to numeric values in the data frame.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{33}
\tightlist
\item
  Make a list using \texttt{list()} function, your list should have 4 elements
  the one below has 2.
\end{enumerate}

Ex: \texttt{mylist=\ list(a=c(1,2,3),b=c("apple,"orange"))}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{34}
\tightlist
\item
  Select the 1st element of the list you made using \texttt{\$} notation.
\end{enumerate}

Ex: \texttt{mylist\$a} selects first element named ``a''

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{35}
\item
  Select the 4th element of the list you made earlier using \texttt{\$} notation.
\item
  Select the 1st element of your list using \texttt{{[}\ {]}} notation.
\end{enumerate}

Ex: \texttt{mylist{[}1{]}} selects first element named ``a'', you get a list with one element.

Ex: \texttt{mylist{[}"a"{]}} selects first element named ``a'', you get a list with one element.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{37}
\item
  select the 4th element of your list using `\protect\hyperlink{section-21}{}' notation.
\item
  Make a factor using factor(), with 5 elements.
  Ex: \texttt{fa=factor(c("a","a","b"))}
\item
  Convert a character vector to factor using \texttt{as.factor()}.
  First, make a character vector using \texttt{c()} then use \texttt{as.factor()}.
\item
  Convert the factor you made above to character using \texttt{as.character()}.
\end{enumerate}

\hypertarget{reading-in-and-writing-data-out-in-r}{%
\subsection{Reading in and writing data out in R}\label{reading-in-and-writing-data-out-in-r}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{41}
\item
  Read CpG island (CpGi) data from the compGenomRData package \texttt{CpGi.table.hg18.txt}, this is a tab-separated file, store it in a variable called ``cpgi''. Use \texttt{cpgtFilePath=system.file("extdata",\ \ \ \ \ \ \ \ \ \ \ \ \ "CpGi.table.hg18.txt",\ \ \ \ \ \ \ \ \ \ \ \ \ package="compGenomRData")}
  to get the file path within the installed \texttt{compGenomRData} package.
\item
  Use \texttt{head()} on CpGi to see first few rows.
\item
  Why doesn't the following work? see `sep' argument at \texttt{help(read.table)}.
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cpgtFilePath=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
                \StringTok{"CpGi.table.hg18.txt"}\NormalTok{,}
                \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{cpgtFilePath}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "/Users/aakalin/Rlibs/compGenomRData/extdata/CpGi.table.hg18.txt"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cpgiSepComma=}\KeywordTok{read.table}\NormalTok{(cpgtFilePath,}\DataTypeTok{header=}\OtherTok{TRUE}\NormalTok{,}\DataTypeTok{sep=}\StringTok{","}\NormalTok{)}
\KeywordTok{head}\NormalTok{(cpgiSepComma)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##   chrom.chromStart.chromEnd.name.length.cpgNum.gcNum.perCpg.perGc.obsExp
## 1         chr1\t18598\t19673\tCpG: 116\t1075\t116\t787\t21.6\t73.2\t0.83
## 2          chr1\t124987\t125426\tCpG: 30\t439\t30\t295\t13.7\t67.2\t0.64
## 3          chr1\t317653\t318092\tCpG: 29\t439\t29\t295\t13.2\t67.2\t0.62
## 4         chr1\t427014\t428027\tCpG: 84\t1013\t84\t734\t16.6\t72.5\t0.64
## 5         chr1\t439136\t440407\tCpG: 99\t1271\t99\t777\t15.6\t61.1\t0.84
## 6            chr1\t523082\t523977\tCpG: 94\t895\t94\t570\t21\t63.7\t1.04
\end{verbatim}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{42}
\item
  What happens when \texttt{stringsAsFactors=FALSE} ?
  \texttt{cpgiHF=read.table("intro2R\_data/data/CpGi.table.hg18.txt",header=FALSE,sep="\textbackslash{}t",\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ stringsAsFactors=FALSE)}
\item
  Read only first 10 rows of the CpGi table.
\item
  Use \texttt{cpgtFilePath=system.file("extdata","CpGi.table.hg18.txt",package="compGenomRData")} to get the file path, then use
  \texttt{read.table()} with argument \texttt{header=FALSE}. Use \texttt{head()} to see the results.
\item
  Write CpG islands to a text file called ``my.cpgi.file.txt''. Write the file
  to your home folder, you can use \texttt{file="\textasciitilde{}/my.cpgi.file.txt"} in linux. \texttt{\textasciitilde{}/} denotes
  home folder.
\item
  Same as above but this time make sure use \texttt{quote=FALSE},\texttt{sep="\textbackslash{}t"} and \texttt{row.names=FALSE} arguments.
  Save the file to ``my.cpgi.file2.txt'' and compare it with ``my.cpgi.file.txt''
\item
  Write out the first 10 rows of `cpgi' data frame.
  \textbf{HINT:} use subsetting for data frames we learned before.
\item
  Write the first 3 columns of `cpgi' data frame.
\item
  Write CpG islands only on chr1. \textbf{HINT:} use subsetting with \texttt{{[}{]}}, feed a logical vector using \texttt{==} operator.
\item
  Read two other data sets ``rn4.refseq.bed'' and ``rn4.refseq2name.txt''
  with \texttt{header=FALSE}, assign them to df1 and df2 respectively.
  They are again included in the compGenomRData package, and you
  can use \texttt{system.file()} function to get the file paths.
\item
  Use \texttt{head()} to see what is inside of the the data frames above.
\item
  Merge data sets using \texttt{merge()} and assign the results to variable named `new.df', and use \texttt{head()} to see the results.
\end{enumerate}

\hypertarget{plotting-in-r-1}{%
\subsection{Plotting in R}\label{plotting-in-r-1}}

Please \textbf{run the following} for the rest of the exercises.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{1001}\NormalTok{)}
\NormalTok{x1=}\DecValTok{1}\OperatorTok{:}\DecValTok{100}\OperatorTok{+}\KeywordTok{rnorm}\NormalTok{(}\DecValTok{100}\NormalTok{,}\DataTypeTok{mean=}\DecValTok{0}\NormalTok{,}\DataTypeTok{sd=}\DecValTok{15}\NormalTok{)}
\NormalTok{y1=}\DecValTok{1}\OperatorTok{:}\DecValTok{100}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{43}
\item
  Make a scatter plot using \texttt{x1} and \texttt{y1} vectors generated above.
\item
  Use \texttt{main} argument to give a title to \texttt{plot()} as in \texttt{plot(x,y,main="title")}
\item
  Use \texttt{xlab} argument to set a label to x-axis.Use \texttt{ylab} argument to set a label to y-axis.
\item
  See what \texttt{mtext(side=3,text="hi\ there")} does. \textbf{HINT:} mtext stands for margin text.
\item
  See what \texttt{mtext(side=2,text="hi\ there")} does.check your plot after execution.
\item
  You can use \texttt{paste()} as `text' argument in mtext() try that, you need to re-plot.
  your plot first. \textbf{HINT:} \texttt{mtext(side=3,text=paste(...))}
  See what \texttt{paste()} is used for.
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{paste}\NormalTok{(}\StringTok{"Text"}\NormalTok{,}\StringTok{"here"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Text here"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{myText=}\KeywordTok{paste}\NormalTok{(}\StringTok{"Text"}\NormalTok{,}\StringTok{"here"}\NormalTok{)}
\NormalTok{myText}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Text here"
\end{verbatim}

Use \emph{mtext()} and \emph{paste()} to put a margin text on the plot.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{43}
\tightlist
\item
  \texttt{cor()} calculates correlation between two vectors.
  Pearson correlation is a measure of the linear correlation (dependence)
  between two variables X and Y.
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{corxy=}\KeywordTok{cor}\NormalTok{(x1,y1) }\CommentTok{# calculates pearson correlation}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{43}
\item
  Try use \texttt{mtext()},\texttt{cor()} and \texttt{paste()} to display correlation coefficient on your scatterplot ?
\item
  Change the colors of your plot using \texttt{col} argument.
  Ex: \texttt{plot(x,y,col="red")}
\item
  Use \texttt{pch=19} as an argument in your \texttt{plot()} command.
\item
  Use \texttt{pch=18} as an argument to your \texttt{plot()} command.
\item
  Make histogram of x1 with \texttt{hist()} function.Histogram is a graphical representation of the data distribution.
\item
  You can change colors with `col', add labels with `xlab', `ylab', and add a `title' with `main' arguments. Try all these in a histogram.
\item
  Make boxplot of y1 with \texttt{boxplot()}.
\item
  Make boxplots of \texttt{x1} and \texttt{y1} vectors in the same plot.
\item
  In boxplot use horizontal = TRUE argument
\item
  make multiple plots with par(mfrow=c(2,1))

  \begin{itemize}
  \tightlist
  \item
    run par(mfrow=c(2,1))
  \item
    make a boxplot
  \item
    make a histogram
  \end{itemize}
\item
  Do the same as above but this time with par(mfrow=c(1,2))
\item
  Save your plot using ``Export'' button in Rstudio
\item
  Save your plot by running :
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{dev.copy}\NormalTok{(pdf,}\DataTypeTok{file=}\StringTok{"plot.file.pdf"}\NormalTok{);}\KeywordTok{dev.off}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{43}
\tightlist
\item
  Save your plot running :
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{dev.copy}\NormalTok{(png,}\DataTypeTok{filename=}\StringTok{"plot.file.png"}\NormalTok{);}\KeywordTok{dev.off}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{43}
\item
  Another way to save the plot is the following
\item
  Open a graphics device
\item
  Create the plot
\item
  Close the graphics device
\item
  \textbf{EXTRA: Making color density scatterplot}. You can make a scatter plot showing density
  of points rather than points themselves. If you use points it looks like this:
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{x2=}\DecValTok{1}\OperatorTok{:}\DecValTok{1000}\OperatorTok{+}\KeywordTok{rnorm}\NormalTok{(}\DecValTok{1000}\NormalTok{,}\DataTypeTok{mean=}\DecValTok{0}\NormalTok{,}\DataTypeTok{sd=}\DecValTok{200}\NormalTok{)}
\NormalTok{y2=}\DecValTok{1}\OperatorTok{:}\DecValTok{1000}
\KeywordTok{plot}\NormalTok{(x2,y2,}\DataTypeTok{pch=}\DecValTok{19}\NormalTok{,}\DataTypeTok{col=}\StringTok{"blue"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/unnamed-chunk-115-1} \end{center}

If you use smoothScatter() function, you get the densities.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{smoothScatter}\NormalTok{(x2,y2,}\DataTypeTok{colramp=}\KeywordTok{colorRampPalette}\NormalTok{(}\KeywordTok{c}\NormalTok{(}\StringTok{"white"}\NormalTok{,}\StringTok{"blue"}\NormalTok{, }\StringTok{"green"}\NormalTok{,}\StringTok{"yellow"}\NormalTok{,}\StringTok{"red"}\NormalTok{))) }
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/unnamed-chunk-116-1} \end{center}

Now, plot with \texttt{colramp=heat.colors} argument and then use a custom color scale using the following argument.

\begin{verbatim}
colramp = colorRampPalette(c("white","blue", "green","yellow","red")))
\end{verbatim}

\hypertarget{functions-and-control-structures-for-ifelse-etc.-1}{%
\subsection{Functions and control structures (for, if/else etc.)}\label{functions-and-control-structures-for-ifelse-etc.-1}}

Read CpG island data as shown below for the rest of the exercises.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cpgtFilePath=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
                \StringTok{"CpGi.table.hg18.txt"}\NormalTok{,}
                \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{cpgi=}\KeywordTok{read.table}\NormalTok{(cpgtFilePath,}\DataTypeTok{header=}\OtherTok{TRUE}\NormalTok{,}\DataTypeTok{sep=}\StringTok{"}\CharTok{\textbackslash{}t}\StringTok{"}\NormalTok{)}
\KeywordTok{head}\NormalTok{(cpgi)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##   chrom chromStart chromEnd     name length cpgNum
## 1  chr1      18598    19673 CpG: 116   1075    116
## 2  chr1     124987   125426  CpG: 30    439     30
## 3  chr1     317653   318092  CpG: 29    439     29
## 4  chr1     427014   428027  CpG: 84   1013     84
## 5  chr1     439136   440407  CpG: 99   1271     99
## 6  chr1     523082   523977  CpG: 94    895     94
##   gcNum perCpg perGc obsExp
## 1   787   21.6  73.2   0.83
## 2   295   13.7  67.2   0.64
## 3   295   13.2  67.2   0.62
## 4   734   16.6  72.5   0.64
## 5   777   15.6  61.1   0.84
## 6   570   21.0  63.7   1.04
\end{verbatim}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{43}
\item
  Check values at perGc column using a histogram.
  `perGc' stands for GC percent =\textgreater{} percentage of C+G nucleotides
\item
  Make a boxplot for `perGc' column
\item
  Use if/else structure to decide if given GC percent high, low or medium.
  If it is low, high, or medium. low \textless{} 60, high\textgreater{}75, medium is between 60 and 75
  use greater or less than operators \textless{} or \textgreater{} .
  Fill in the values in the in code below, where it is written `YOU\_FILL\_IN'
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{GCper=}\DecValTok{65}

  \CommentTok{# check if GC value is lower than 60, }
  \CommentTok{# assign "low" to result}
  \ControlFlowTok{if}\NormalTok{(}\StringTok{'YOU_FILL_IN'}\NormalTok{)\{}
\NormalTok{    result=}\StringTok{"low"}
    \KeywordTok{cat}\NormalTok{(}\StringTok{"low"}\NormalTok{)}
\NormalTok{  \}}
  \ControlFlowTok{else} \ControlFlowTok{if}\NormalTok{(}\StringTok{'YOU_FILL_IN'}\NormalTok{)\{  }\CommentTok{# check if GC value is higher than 75, assign "high" to result}
\NormalTok{    result=}\StringTok{"high"}
    \KeywordTok{cat}\NormalTok{(}\StringTok{"high"}\NormalTok{)}
\NormalTok{  \}}\ControlFlowTok{else}\NormalTok{\{ }\CommentTok{# if those two conditions fail then it must be "medium"}
\NormalTok{    result=}\StringTok{"medium"}
\NormalTok{  \}}

\NormalTok{result}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{43}
\tightlist
\item
  Write a function that takes a value of GC percent and decides
  if it is low, high, or medium. low \textless{} 60, high\textgreater{}75, medium is between 60 and 75.
  Fill in the values in the in code below, where it is written `YOU\_FILL\_IN'
\end{enumerate}

\begin{verbatim}
GCclass<-function(my.gc){
  
  YOU_FILL_IN
  
  return(result)
}
GCclass(10) # should return "low"
GCclass(90) # should return "high"
GCclass(65) # should return "medium"
\end{verbatim}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{43}
\tightlist
\item
  Use a for loop to get GC percentage classes for gcValues below. Use the function
  you wrote above.
\end{enumerate}

\begin{verbatim}
gcValues=c(10,50,70,65,90)
for( i in YOU_FILL_IN){
  YOU_FILL_IN
}
\end{verbatim}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{43}
\tightlist
\item
  Use \texttt{lapply} to get to get GC percentage classes for gcValues. Example:
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{vec=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{4}\NormalTok{,}\DecValTok{5}\NormalTok{)}
\NormalTok{power2=}\ControlFlowTok{function}\NormalTok{(x)\{ }\KeywordTok{return}\NormalTok{(x}\OperatorTok{^}\DecValTok{2}\NormalTok{)  \}}
    \KeywordTok{lapply}\NormalTok{(vec,power2)}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{43}
\item
  Use sapply to get values to get GC percentage classes for gcValues
\item
  Is there a way to decide on the GC percentage class of given vector of GCpercentages
  without using if/else structure and loops ? if so, how can you do it?
  \textbf{HINT:} subsetting using \textless{} and \textgreater{} operators
\end{enumerate}

\hypertarget{stats}{%
\chapter{Statistics for Genomics}\label{stats}}

This chapter will summarize statistics methods frequently used
in computational genomics. As these fields are continuously evolving, the
techniques introduced here do not form an exhaustive list but mostly corner
stone methods
that are often and still being used. In addition, we focused on giving intuitive and
practical understanding of the methods with relevant examples from the field.\\
If you want to dig deeper into statistics and math, beyond what is described
here, we included appropriate references with annotation after each major
section.

\hypertarget{how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions}{%
\section{How to summarize collection of data points: The idea behind statistical distributions}\label{how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions}}

In biology and many other fields data is collected via experimentation.
The nature of the experiments and natural variation in biology makes
it impossible to get the same exact measurements every time you measure something.
For example, if you are measuring gene expression values for
a certain gene, say PAX6, and let's assume you are measuring expression
per sample and cell with any method( microarrays, rt-qPCR, etc.). You will not
get the same expression value even if your samples are homogeneous. Due
to technical bias in experiments or natural variation in the samples. Instead,
we would like to describe this collection of data some other way
that represents the general properties of the data. The figure shows a sample of
20 expression values from PAX6 gene.

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-128-1} 

}

\caption{Expression of PAX6 gene in 20 replicate experiments}\label{fig:unnamed-chunk-128}
\end{figure}

\hypertarget{describing-the-central-tendency-mean-and-median}{%
\subsection{Describing the central tendency: mean and median}\label{describing-the-central-tendency-mean-and-median}}

As seen in the figure above, the points from this sample are distributed around
a central value and the histogram below the dot plot shows number of points in
each bin. Another observation is that there are some bins that have more points than others. If we want to summarize what we observe, we can try
to represent the collection of data points
with an expression value that is typical to get, something that represents the
general tendency we observe on the dot plot and the histogram. This value is
sometimes called central
value or central tendency, and there are different ways to calculate such a value.
In the figure above, we see that all the values are spread around 6.13 (red line),
and that is indeed what we call mean value of this sample of expression values.
It can be calculated with the following formula \(\overline{X}=\sum_{i=1}^n x_i/n\),
where \(x_i\) is the expression value of an experiment and \(n\) is the number of
expression value obtained from the experiments. In R, \texttt{mean()} function will calculate the
mean of a provided vector of numbers. This is called a ``sample mean''. In reality
the possible values of PAX6 expression for all cells (provided each cell is of the
identical cell type and is in identical conditions) are much much more than 20.
If we had the time and the funding to sample all cells and measure PAX6 expression we would
get a collection values that would be called, in statistics, a ``population''. In
our case the population will look like the left hand side of the figure below. What we have done with
our 20 data points is that we took a sample of PAX6 expression values from this
population, and calculated the sample mean.

\begin{figure}

{\centering \includegraphics[width=0.75\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-129-1} 

}

\caption{Expression of all possible PAX6 gene expressions measures on all available biological samples (left). Expression of PAX6 gene from statistical sample, a random subset, from the population of biological samples (Right). }\label{fig:unnamed-chunk-129}
\end{figure}

The mean of the population is calculated the same way but traditionally
Greek letter \(\mu\) is used to denote the population mean. Normally, we would not
have access to the population and we will use sample mean and other quantities
derived from the sample to estimate the population properties. This is the basic
idea behind statistical inference which we will see this in action in later
sections as well. We
estimate the population parameters from the sample parameters and there is some
uncertainty associated with those estimates. We will be trying to assess those
uncertainties and make decisions in the presence of those uncertainties.

We are not yet done with measuring central tendency.
There are other ways to describe it, such as the median value.
Mean can be affected by outliers easily.
If certain values are very high or low from the
bulk of the sample this will shift mean towards those outliers. However, median
is not affected by outliers. It is simply the value in a distribution where half
of the values are above and the other half is below. In R, \texttt{median()} function
will calculate the mean of a provided vector of numbers.

Let's create a set of random numbers and calculate their mean and median using
R.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#create 10 random numbers from uniform distribution }
\NormalTok{x=}\KeywordTok{runif}\NormalTok{(}\DecValTok{10}\NormalTok{)}
\CommentTok{# calculate mean}
\KeywordTok{mean}\NormalTok{(x)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.3739
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# calculate median}
\KeywordTok{median}\NormalTok{(x)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.3278
\end{verbatim}

\hypertarget{describing-the-spread-measurements-of-variation}{%
\subsection{Describing the spread: measurements of variation}\label{describing-the-spread-measurements-of-variation}}

Another useful way to summarize a collection of data points is to measure
how variable the values are. You can simply describe the range of the values
, such as minimum and maximum values. You can easily do that in R with \texttt{range()}
function. A more common way to calculate variation is by calculating something
called ``standard deviation'' or the related quantity called ``variance''. This is a
quantity that shows how variable the values are, a value around zero indicates
there is not much variation in the values of the data points, and a high value
indicates high variation in the values. The variance is the squared distance of
data points from the mean. Population variance is again a quantity we usually
do not have access to and is simply calculate as follows \(\sigma^2=\sum_{i=1}^n \frac{(x_i-\mu)^2}{n}\), where \(\mu\) is the population mean, \(x_i\) is the ith
data point in the population and \(n\) is the population size. However, when the
we have only access to a sample this formulation is biased. It means that it
underestimates the population variance, so we make a small adjustment when we
calculate the sample variance, denoted as \(s^2\):
\[
\begin{aligned}
s^2=\sum_{i=1}^n \frac{(x_i-\overline{X})^2}{n-1} && \text{ where $x_i$ is the ith data point and
$\overline{X}$ is the sample mean.}
\end{aligned}
\]

The sample standard deviation is simply the square-root of the sample variance.
The good thing about standard deviation is that it has the same unit as the mean
so it is more intuitive.\\
\[s=\sqrt{\sum_{i=1}^n \frac{(x_i-\overline{X})^2}{n-1}}\]

We can calculate sample standard deviation and variation with \texttt{sd()} and \texttt{var()}
functions in R. These functions take vector of numeric values as input and
calculate the desired quantities. Below we use those functions on a randomly
generated vector of numbers.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{x=}\KeywordTok{rnorm}\NormalTok{(}\DecValTok{20}\NormalTok{,}\DataTypeTok{mean=}\DecValTok{6}\NormalTok{,}\DataTypeTok{sd=}\FloatTok{0.7}\NormalTok{)}
\KeywordTok{var}\NormalTok{(x)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.2531
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{sd}\NormalTok{(x)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.5031
\end{verbatim}

One potential problem with the variance is that it could be affected by
outliers. The points that are too far away from the mean will have a large
affect on the variance even though there might be few of them.
A way to measure variance that could be less affected by outliers is
looking at where bulk of the distribution is. How do we define where the bulk is?
One common way is to look at the the difference between 75th percentile and 25th
percentile, this effectively removes a lot of potential outliers which will be
towards the edges of the range of values.
This is called interquartile range , and
can be easily calculated using R via \texttt{IQR()} function and the quantiles of a vector
is calculated with \texttt{quantile()} function.

Let us plot the boxplot for a random vector and also calculate IQR using R.
In the boxplot below, 25th and 75th percentiles are the edges of the box, and
the median is marked with a thick line going through roughly middle the box.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{x=}\KeywordTok{rnorm}\NormalTok{(}\DecValTok{20}\NormalTok{,}\DataTypeTok{mean=}\DecValTok{6}\NormalTok{,}\DataTypeTok{sd=}\FloatTok{0.7}\NormalTok{)}
\KeywordTok{IQR}\NormalTok{(x)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.5011
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{quantile}\NormalTok{(x)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##    0%   25%   50%   75%  100% 
## 5.437 5.743 5.860 6.244 6.558
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{boxplot}\NormalTok{(x,}\DataTypeTok{horizontal =}\NormalTok{ T)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-134-1} 

}

\caption{Boxplot showing 25th percentile and 75th percentile and median for a set of points sample from a normal distribution with mean=6 and standard deviation=0.7}\label{fig:unnamed-chunk-134}
\end{figure}

\hypertarget{frequently-used-statistical-distributions}{%
\subsubsection{Frequently used statistical distributions}\label{frequently-used-statistical-distributions}}

The distributions have parameters (such as mean and variance) that
summarizes them but also they are functions that assigns each outcome of a
statistical experiment to its probability of occurrence.
One distribution that you
will frequently encounter is the normal distribution or Gaussian distribution.
The normal distribution has a typical ``bell-curve'' shape
and, characterized by mean and standard deviation. A set of data points
that
follow normal distribution mostly will be close to the mean
but spread around it controlled by the standard deviation parameter. That
means if we sample data points from a normal distribution we are more
likely to sample nearby the mean and sometimes away from the mean.
Probability of an event occurring is higher if it is nearby the mean.
The effect
of the parameters for normal distribution can be observed in the following
plot.

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-135-1} 

}

\caption{Different parameters for normal distribution and effect of those on the shape of the distribution}\label{fig:unnamed-chunk-135}
\end{figure}

The normal distribution is often denoted by \(\mathcal{N}(\mu,\,\sigma^2)\) When a random variable \(X\) is distributed normally with mean \(\mu\) and variance \(\sigma^2\), we write:

\[X\ \sim\ \mathcal{N}(\mu,\,\sigma^2).\]

The probability
density function of Normal distribution with mean \(\mu\) and standard deviation
\(\sigma\) is as follows

\[P(x)=\frac{1}{\sigma\sqrt{2\pi} } \; e^{ -\frac{(x-\mu)^2}{2\sigma^2} } \]

The probability density function gives the probability of observing a value
on a normal distribution defined by \(\mu\) and
\(\sigma\) parameters.

Often times, we do not need the exact probability of a value but we need the
probability of observing a value larger or smaller than a critical value or reference
point. For example, we might want to know the probability of \(X\) being smaller than or
equal to -2 for a normal distribution with mean 0 and standard deviation 2.
,\(P(X <= -2 \; | \; \mu=0,\sigma=2)\). In this case, what we want is the are under the
curve shaded in blue. To be able to that we need to integrate the probability
density function but we will usually let software do that. Traditionally,
one calculates a Z-score which is simply \((X-\mu)/\sigma=(-2-0)/2= -1\), and
corresponds to how many standard deviations you are away from the mean.
This is also called ``standardization'', the corresponding value is distributed in ``standard normal distribution'' where \(\mathcal{N}(0,\,1)\).

After calculating the Z-score,
we can go look up in a table, that contains the area under the curve for
the left and right side of the Z-score, but again we use software for that
tables are outdated.

Below we are showing the Z-score and the associated probabilities derived
from the calculation above for \(P(X <= -2 \; | \; \mu=0,\sigma=2)\).

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/zscore-1} 

}

\caption{Z-score and associated probabilities for Z= -1}\label{fig:zscore}
\end{figure}

In R, family of \texttt{*norm} functions (\texttt{rnorm},\texttt{dnorm},\texttt{qnorm} and \texttt{pnorm}) can
be used to
operate with normal distribution, such as calculating probabilities and
generating random numbers drawn from normal distribution.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# get the value of probability density function when X= -2,}
\CommentTok{# where mean=0 and sd=2}
\KeywordTok{dnorm}\NormalTok{(}\OperatorTok{-}\DecValTok{2}\NormalTok{, }\DataTypeTok{mean=}\DecValTok{0}\NormalTok{, }\DataTypeTok{sd=}\DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.121
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# get the probability of P(X =< -2) where mean=0 and sd=2}
\KeywordTok{pnorm}\NormalTok{(}\OperatorTok{-}\DecValTok{2}\NormalTok{, }\DataTypeTok{mean=}\DecValTok{0}\NormalTok{, }\DataTypeTok{sd=}\DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.1587
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# get the probability of P(X > -2) where mean=0 and sd=2}
\KeywordTok{pnorm}\NormalTok{(}\OperatorTok{-}\DecValTok{2}\NormalTok{, }\DataTypeTok{mean=}\DecValTok{0}\NormalTok{, }\DataTypeTok{sd=}\DecValTok{2}\NormalTok{,}\DataTypeTok{lower.tail =} \OtherTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.8413
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# get 5 random numbers from normal dist with  mean=0 and sd=2}
\KeywordTok{rnorm}\NormalTok{(}\DecValTok{5}\NormalTok{, }\DataTypeTok{mean=}\DecValTok{0}\NormalTok{ , }\DataTypeTok{sd=}\DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] -1.8109 -1.9221 -0.5147  0.8217 -0.7901
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# get y value corresponding to P(X > y) = 0.15 with  mean=0 and sd=2}
\KeywordTok{qnorm}\NormalTok{( }\FloatTok{0.15}\NormalTok{, }\DataTypeTok{mean=}\DecValTok{0}\NormalTok{ , }\DataTypeTok{sd=}\DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] -2.073
\end{verbatim}

There are many other distribution functions in R that can be used the same
way. You have to enter the distribution specific parameters along
with your critical value, quantiles or number of random numbers depending
on which function you are using in the family.We will list some of those functions below.

\begin{itemize}
\item
  \texttt{dbinom} is for binomial distribution. This distribution is usually used
  to model fractional data and binary data. Examples from genomics includes
  methylation data.
\item
  \texttt{dpois} is used for Poisson distribution and \texttt{dnbinom} is used for
  negative binomial distribution. These distributions are used to model count
  data such as sequencing read counts.
\item
  \texttt{df} (F distribution) and \texttt{dchisq} (Chi-Squared distribution) are used
  in relation to distribution of variation. F distribution is used to model
  ratios of variation and Chi-Squared distribution is used to model
  distribution of variations. You will frequently encounter these in linear models and generalized linear models.
\end{itemize}

\hypertarget{precision-of-estimates-confidence-intervals}{%
\subsection{Precision of estimates: Confidence intervals}\label{precision-of-estimates-confidence-intervals}}

When we take a random sample from a population and compute a statistic, such as
the mean, we are trying to approximate the mean of the population. How well this
sample statistic estimates the population value will always be a
concern. A confidence interval addresses this concern because it provides a
range of values which is plausible to contain the population parameter of interest.
Normally, we would not have access to a population. If we did, we would not have to estimate the population parameters and its precision.

When we do not have access
to the population, one way to estimate intervals is to repeatedly take samples from the
original sample with replacement, that is we take a data point from the sample
we replace, and we take another data point until we have sample size of the
original sample. Then, we calculate the parameter of interest, in this case mean, and
repeat this step a large number of times, such as 1000. At this point, we would have a distribution of re-sampled
means, we can then calculate the 2.5th and 97.5th percentiles and these will
be our so-called 95\% confidence interval. This procedure, resampling with replacement to
estimate the precision of population parameter estimates, is known as the \textbf{bootstrap}.

Let's see how we can do this in practice. We simulate a sample
coming from a normal distribution (but we pretend we don't know the
population parameters). We will try to estimate the precision
of the mean of the sample using bootstrap to build confidence intervals.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(mosaic)}
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{21}\NormalTok{)}
\NormalTok{sample1=}\StringTok{ }\KeywordTok{rnorm}\NormalTok{(}\DecValTok{50}\NormalTok{,}\DecValTok{20}\NormalTok{,}\DecValTok{5}\NormalTok{) }\CommentTok{# simulate a sample}

\CommentTok{# do bootstrap resampling, sampling with replacement}
\NormalTok{boot.means=}\KeywordTok{do}\NormalTok{(}\DecValTok{1000}\NormalTok{) }\OperatorTok{*}\StringTok{ }\KeywordTok{mean}\NormalTok{(}\KeywordTok{resample}\NormalTok{(sample1))}

\CommentTok{# get percentiles from the bootstrap means}
\NormalTok{q=}\KeywordTok{quantile}\NormalTok{(boot.means[,}\DecValTok{1}\NormalTok{],}\DataTypeTok{p=}\KeywordTok{c}\NormalTok{(}\FloatTok{0.025}\NormalTok{,}\FloatTok{0.975}\NormalTok{))}

\CommentTok{# plot the histogram}
\KeywordTok{hist}\NormalTok{(boot.means[,}\DecValTok{1}\NormalTok{],}\DataTypeTok{col=}\StringTok{"cornflowerblue"}\NormalTok{,}\DataTypeTok{border=}\StringTok{"white"}\NormalTok{,}
                    \DataTypeTok{xlab=}\StringTok{"sample means"}\NormalTok{)}
\KeywordTok{abline}\NormalTok{(}\DataTypeTok{v=}\KeywordTok{c}\NormalTok{(q[}\DecValTok{1}\NormalTok{], q[}\DecValTok{2}\NormalTok{] ),}\DataTypeTok{col=}\StringTok{"red"}\NormalTok{)}
\KeywordTok{text}\NormalTok{(}\DataTypeTok{x=}\NormalTok{q[}\DecValTok{1}\NormalTok{],}\DataTypeTok{y=}\DecValTok{200}\NormalTok{,}\KeywordTok{round}\NormalTok{(q[}\DecValTok{1}\NormalTok{],}\DecValTok{3}\NormalTok{),}\DataTypeTok{adj=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{0}\NormalTok{))}
\KeywordTok{text}\NormalTok{(}\DataTypeTok{x=}\NormalTok{q[}\DecValTok{2}\NormalTok{],}\DataTypeTok{y=}\DecValTok{200}\NormalTok{,}\KeywordTok{round}\NormalTok{(q[}\DecValTok{2}\NormalTok{],}\DecValTok{3}\NormalTok{),}\DataTypeTok{adj=}\KeywordTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.7\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-137-1} 

}

\caption{Precision estimate of the sample mean using 1000 bootstrap samples. Confidence intervals derived from the bootstrap samples are shown with red lines.}\label{fig:unnamed-chunk-137}
\end{figure}

If we had a convenient mathematical method to calculate confidence interval
we could also do without resampling methods. It turns out that if we take
repeated
samples from a population of with sample size \(n\), the distribution of means
( \(\overline{X}\)) of those samples
will be approximately normal with mean \(\mu\) and standard deviation
\(\sigma/\sqrt{n}\). This is also known as \textbf{Central Limit Theorem(CLT)} and
is one of the most important theorems in statistics. This also means that
\(\frac{\overline{X}-\mu}{\sigma\sqrt{n}}\) has a standard normal
distribution and we can calculate the Z-score and then we can get
the percentiles associated with the Z-score. Below, we are showing the
Z-score
calculation for the distribution of \(\overline{X}\), and then
we are deriving the confidence intervals starting with the fact that
probability of Z being between -1.96 and 1.96 is 0.95. We then use algebra
to show that the probability that unknown \(\mu\) is captured between
\(\overline{X}-1.96\sigma/\sqrt{n}\) and \(\overline{X}+1.96\sigma/\sqrt{n}\) is 0.95, which is commonly known as 95\% confidence interval.

\[\begin{array}{ccc}
Z=\frac{\overline{X}-\mu}{\sigma/\sqrt{n}}\\
P(-1.96 < Z < 1.96)=0.95 \\
P(-1.96 < \frac{\overline{X}-\mu}{\sigma/\sqrt{n}} < 1.96)=0.95\\
P(\mu-1.96\sigma/\sqrt{n} < \overline{X} < \mu+1.96\sigma/\sqrt{n})=0.95\\
P(\overline{X}-1.96\sigma/\sqrt{n} < \mu < \overline{X}+1.96\sigma/\sqrt{n})=0.95\\
confint=[\overline{X}-1.96\sigma/\sqrt{n},\overline{X}+1.96\sigma/\sqrt{n}]
\end{array}\]

A 95\% confidence interval for population mean is the most common
common interval to use, and would
mean that we would expect 95\% of the interval estimates to include the
population parameter, in this case the mean. However, we can pick any value
such as 99\% or 90\%. We can generalize the confidence interval for
\(100(1-\alpha)\) as follows:

\[\overline{X} \pm Z_{\alpha/2}\sigma/\sqrt{n}\]

In R, we can do this using \texttt{qnorm()} function to get Z-scores associated
with \({\alpha/2}\) and \({1-\alpha/2}\). As you can see, the confidence intervals we calculated using CLT are very
similar to the ones we got from bootstrap for the same sample. For bootstrap we got \([19.21, 21.989]\) and for the CLT based estimate we got \([19.23638, 22.00819]\).

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{alpha=}\FloatTok{0.05}
\NormalTok{sd=}\DecValTok{5}
\NormalTok{n=}\DecValTok{50}
\KeywordTok{mean}\NormalTok{(sample1)}\OperatorTok{+}\KeywordTok{qnorm}\NormalTok{(}\KeywordTok{c}\NormalTok{(alpha}\OperatorTok{/}\DecValTok{2}\NormalTok{,}\DecValTok{1}\OperatorTok{-}\NormalTok{alpha}\OperatorTok{/}\DecValTok{2}\NormalTok{))}\OperatorTok{*}\NormalTok{sd}\OperatorTok{/}\KeywordTok{sqrt}\NormalTok{(n)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 19.24 22.01
\end{verbatim}

The good thing about CLT as long as the sample size is large regardless of
the population distribution, the distribution of sample means drawn from
that population will always be normal. Here we are repeatedly
drawing samples 1000 times with sample size \(n\)=10,30, and 100 from a bimodal,
exponential and a uniform distribution and we are getting sample mean distributions
following normal distribution.

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/unnamed-chunk-139-1} 

}

\caption{Sample means are normally distributed regardless of the population distribution they are drawn from.}\label{fig:unnamed-chunk-139}
\end{figure}

However, we should note that how we constructed the confidence interval
using standard normal distribution, \(N(0,1)\), only works when the when we know the
population standard deviation. In reality, we usually have only access
to a sample and have no idea about the population standard deviation. If
this is the case we should use estimate the standard deviation using
sample standard deviation and use something called \emph{t distribution} instead
of standard normal distribution in our interval calculation. Our confidence interval becomes
\(\overline{X} \pm t_{\alpha/2}s/\sqrt{n}\), with t distribution
parameter \(d.f=n-1\), since now the following quantity is t distributed \(\frac{\overline{X}-\mu}{s/\sqrt{n}}\) instead of standard normal distribution.

The t distribution is similar to standard normal distribution has mean 0 but its spread is larger than the normal distribution
especially when sample size is small, and has one parameter \(v\) for
the degrees of freedom, which is \(n-1\) in this case. Degrees of freedom
is simply number of data points minus number of parameters estimated. Here
we are estimating the mean from the data and the distribution is for the means, therefore degrees of freedom is \(n-1\).

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-140-1} 

}

\caption{Normal distribution and t distribution with different degrees of freedom. With increasing degrees of freedom, t distribution approximates the normal distribution better.}\label{fig:unnamed-chunk-140}
\end{figure}

\hypertarget{how-to-test-for-differences-between-samples}{%
\section{How to test for differences between samples}\label{how-to-test-for-differences-between-samples}}

Often times we would want to compare sets of samples. Such comparisons include
if wild-type samples have different expression compared to mutants or if healthy
samples are different from disease samples in some measurable feature (blood count,
gene expression, methylation of certain loci). Since there is variability in our
measurements, we need to take that into account when comparing the sets of samples.
We can simply subtract the means of two samples, but given the variability
of sampling, at the very least we need to decide a cutoff value for differences
of means, small differences of means can be explained by random chance due to
sampling. That means we need to compare the difference we get to a value that
is typical to get if the difference between two group means were only due to
sampling. If you followed the logic above, here we actually introduced two core
ideas of something called ``hypothesis testing'', this is simply using
statistics to
determine the probability that a given hypothesis (if two sample sets
are from the same population or not) is true. Formally, those two core
ideas are as follows:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Decide on a hypothesis to test, often called ``null hypothesis'' (\(H_0\)). In our
  case, the hypothesis is there is no difference between sets of samples. An the ``Alternative hypothesis'' (\(H_1\)) is there is a difference between the
  samples.
\item
  Decide on a statistic to test the truth of the null hypothesis.
\item
  Calculate the statistic
\item
  Compare it to a reference value to establish significance, the P-value. Based on that either reject or not reject the null hypothesis, \(H_0\)
\end{enumerate}

\hypertarget{randomization-based-testing-for-difference-of-the-means}{%
\subsection{randomization based testing for difference of the means}\label{randomization-based-testing-for-difference-of-the-means}}

There is one intuitive way to go about this. If we believe there are no
differences between samples that means the sample labels (test-control or
healthy-disease) has no meaning. So, if we randomly assign labels to the
samples
that and calculate the difference of the mean, this creates a null
distribution for the \(H_0\) where we can compare the real difference and
measure how unlikely it is to get such a value under the expectation of the
null hypothesis. We can calculate all possible permutations to calculate
the null distribution. However, sometimes that is not very feasible and
equivalent approach would be generating the null distribution by taking a
smaller number of random samples with shuffled group membership.

Below, we are doing this process in R. We are first simulating two samples
from two different distributions.
These would be equivalent to gene expression measurements obtained under
different conditions. Then, we calculate the differences in the means
and do the randomization procedure to get a null distribution when we
assume there is no difference between samples, \(H_0\). We then calculate how
often we would get the original difference we calculated under the
assumption that \(H_0\) is true.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{100}\NormalTok{)}
\NormalTok{gene1=}\KeywordTok{rnorm}\NormalTok{(}\DecValTok{30}\NormalTok{,}\DataTypeTok{mean=}\DecValTok{4}\NormalTok{,}\DataTypeTok{sd=}\DecValTok{2}\NormalTok{)}
\NormalTok{gene2=}\KeywordTok{rnorm}\NormalTok{(}\DecValTok{30}\NormalTok{,}\DataTypeTok{mean=}\DecValTok{2}\NormalTok{,}\DataTypeTok{sd=}\DecValTok{2}\NormalTok{)}
\NormalTok{org.diff=}\KeywordTok{mean}\NormalTok{(gene1)}\OperatorTok{-}\KeywordTok{mean}\NormalTok{(gene2)}
\NormalTok{gene.df=}\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{exp=}\KeywordTok{c}\NormalTok{(gene1,gene2),}
                  \DataTypeTok{group=}\KeywordTok{c}\NormalTok{( }\KeywordTok{rep}\NormalTok{(}\StringTok{"test"}\NormalTok{,}\DecValTok{30}\NormalTok{),}\KeywordTok{rep}\NormalTok{(}\StringTok{"control"}\NormalTok{,}\DecValTok{30}\NormalTok{) ) )}


\NormalTok{exp.null <-}\StringTok{ }\KeywordTok{do}\NormalTok{(}\DecValTok{1000}\NormalTok{) }\OperatorTok{*}\StringTok{ }\KeywordTok{diff}\NormalTok{(mosaic}\OperatorTok{::}\KeywordTok{mean}\NormalTok{(exp }\OperatorTok{~}\StringTok{ }\KeywordTok{shuffle}\NormalTok{(group), }\DataTypeTok{data=}\NormalTok{gene.df))}
\KeywordTok{hist}\NormalTok{(exp.null[,}\DecValTok{1}\NormalTok{],}\DataTypeTok{xlab=}\StringTok{"null distribution | no difference in samples"}\NormalTok{,}
     \DataTypeTok{main=}\KeywordTok{expression}\NormalTok{(}\KeywordTok{paste}\NormalTok{(H[}\DecValTok{0}\NormalTok{],}\StringTok{" :no difference in means"}\NormalTok{) ),}
     \DataTypeTok{xlim=}\KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{2}\NormalTok{,}\DecValTok{2}\NormalTok{),}\DataTypeTok{col=}\StringTok{"cornflowerblue"}\NormalTok{,}\DataTypeTok{border=}\StringTok{"white"}\NormalTok{)}
\KeywordTok{abline}\NormalTok{(}\DataTypeTok{v=}\KeywordTok{quantile}\NormalTok{(exp.null[,}\DecValTok{1}\NormalTok{],}\FloatTok{0.95}\NormalTok{),}\DataTypeTok{col=}\StringTok{"red"}\NormalTok{ )}
\KeywordTok{abline}\NormalTok{(}\DataTypeTok{v=}\NormalTok{org.diff,}\DataTypeTok{col=}\StringTok{"blue"}\NormalTok{ )}
\KeywordTok{text}\NormalTok{(}\DataTypeTok{x=}\KeywordTok{quantile}\NormalTok{(exp.null[,}\DecValTok{1}\NormalTok{],}\FloatTok{0.95}\NormalTok{),}\DataTypeTok{y=}\DecValTok{200}\NormalTok{,}\StringTok{"0.05"}\NormalTok{,}\DataTypeTok{adj=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{0}\NormalTok{),}\DataTypeTok{col=}\StringTok{"red"}\NormalTok{)}
\KeywordTok{text}\NormalTok{(}\DataTypeTok{x=}\NormalTok{org.diff,}\DataTypeTok{y=}\DecValTok{200}\NormalTok{,}\StringTok{"org. diff."}\NormalTok{,}\DataTypeTok{adj=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{0}\NormalTok{),}\DataTypeTok{col=}\StringTok{"blue"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-141-1} 

}

\caption{The null distribution for differences of means obtained via randomization. The original difference is marked via blue line. The red line marks the value that corresponds to P-value of 0.05}\label{fig:unnamed-chunk-141}
\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{p.val=}\KeywordTok{sum}\NormalTok{(exp.null[,}\DecValTok{1}\NormalTok{]}\OperatorTok{>}\NormalTok{org.diff)}\OperatorTok{/}\KeywordTok{length}\NormalTok{(exp.null[,}\DecValTok{1}\NormalTok{])}
\NormalTok{p.val}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0
\end{verbatim}

After doing random permutations and getting a null distribution, it is possible to get a confidence interval for the distribution of difference in means.
This is simply the 2.5th and 97.5th percentiles of the null distribution, and
directly related to the P-value calculation above.

\hypertarget{using-t-test-for-difference-of-the-means-between-two-samples}{%
\subsection{Using t-test for difference of the means between two samples}\label{using-t-test-for-difference-of-the-means-between-two-samples}}

We can also calculate the difference between means using a t-test. Sometimes we will have too few data points in a sample to do meaningful
randomization test, also randomization takes more time than doing a t-test.
This is a test that depends on the t distribution. The line of thought follows
from the CLT and we can show differences in means are t distributed.
There are couple of variants of the t-test for this purpose. If we assume
the variances are equal we can use the following version

\[t = \frac{\bar {X}_1 - \bar{X}_2}{s_{X_1X_2} \cdot \sqrt{\frac{1}{n_1}+\frac{1}{n_2}}}\]
where
\[s_{X_1X_2} = \sqrt{\frac{(n_1-1)s_{X_1}^2+(n_2-1)s_{X_2}^2}{n_1+n_2-2}}\]
In the first equation above the quantity is t distributed with \(n_1+n_2-2\) degrees of freedom. We can calculate the quantity then use software
to look for the percentile of that value in that t distribution, which is our P-value. When we can not assume equal variances we use ``Welch's t-test''
which is the default t-test in R and also works well when variances and
the sample sizes are the same. For this test we calculate the following
quantity:

\[t = {\overline{X}_1 - \overline{X}_2 \over s_{\overline{X}_1 - \overline{X}_2}}\]
where
\[s_{\overline{X}_1 - \overline{X}_2} = \sqrt{{s_1^2 \over n_1} + {s_2^2  \over n_2}}
\]
and the degrees of freedom equals to:
\[\mathrm{d.f.} = \frac{(s_1^2/n_1 + s_2^2/n_2)^2}{(s_1^2/n_1)^2/(n_1-1) + (s_2^2/n_2)^2/(n_2-1)}
\]

Luckily, R does all those calculations for us. Below we will show the use of \texttt{t.test()} function in R. We will use it on the samples we simulated
above.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Welch's t-test}
\NormalTok{stats}\OperatorTok{::}\KeywordTok{t.test}\NormalTok{(gene1,gene2)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
##  Welch Two Sample t-test
## 
## data:  gene1 and gene2
## t = 3.8, df = 48, p-value = 5e-04
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  0.8724 2.8728
## sample estimates:
## mean of x mean of y 
##     4.058     2.185
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# t-test with equal varience assumption}
\NormalTok{stats}\OperatorTok{::}\KeywordTok{t.test}\NormalTok{(gene1,gene2,}\DataTypeTok{var.equal=}\OtherTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
##  Two Sample t-test
## 
## data:  gene1 and gene2
## t = 3.8, df = 58, p-value = 4e-04
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  0.8771 2.8681
## sample estimates:
## mean of x mean of y 
##     4.058     2.185
\end{verbatim}

A final word on t-tests: they generally assume population where samples coming
from have normal
distribution, however it is been shown t-test can tolerate deviations from
normality. Especially, when two distributions are moderately skewed in the
same direction. This is due to central limit theorem which says means of
samples will be distributed normally no matter the population distribution
if sample sizes are large.

\hypertarget{multiple-testing-correction}{%
\subsection{multiple testing correction}\label{multiple-testing-correction}}

We should think of hypothesis testing as a non-error-free method of making
decisions. There will be times when we declare something significant and accept
\(H_1\) but we will be wrong.
These decisions are also called ``false positives'' or ``false discoveries'', this
is also known as ``type I error''. Similarly, we can fail to reject a hypothesis
when we actually should. These cases are known as ``false negatives'', also known
as ``type II error''.

The ratio of true negatives to the sum of
true negatives and false positives (\(\frac{TN}{FP+TN}\)) is known as specificity.
And we usually want to decrease the FP and get higher specificity.
The ratio of true positives to the sum of
true positives and false negatives (\(\frac{TP}{TP+FN}\)) is known as sensitivity.
And, again we usually want to decrease the FN and get higher sensitivity.
Sensitivity is also known as ``power of a test'' in the context of hypothesis
testing. More powerful tests will be highly sensitive and will do less type
II errors. For the t-test the power is positively associated with sample size
and the effect size. Higher the sample size, smaller the standard error and
looking for the larger effect sizes will similarly increase the power.

The general summary of these the different combination of the decisions are
included in the table below.

\begin{longtable}[]{@{}lccl@{}}
\toprule
\begin{minipage}[b]{0.17\columnwidth}\raggedright
\strut
\end{minipage} & \begin{minipage}[b]{0.22\columnwidth}\centering
\(H_0\) is
TRUE,
{[}Gene is NOT
differentially
expressed{]}\strut
\end{minipage} & \begin{minipage}[b]{0.22\columnwidth}\centering
\(H_1\) is
TRUE,
{[}Gene is
differentially
expressed{]}\strut
\end{minipage} & \begin{minipage}[b]{0.27\columnwidth}\raggedright
\strut
\end{minipage}\tabularnewline
\midrule
\endhead
\begin{minipage}[t]{0.17\columnwidth}\raggedright
Accept \(H_0\)
(claim that
the gene is not
differentially
expressed)\strut
\end{minipage} & \begin{minipage}[t]{0.22\columnwidth}\centering
True Negatives (TN)\strut
\end{minipage} & \begin{minipage}[t]{0.22\columnwidth}\centering
False Negatives (FN)
,type II error\strut
\end{minipage} & \begin{minipage}[t]{0.27\columnwidth}\raggedright
\(m_0\): number of truly
null hypotheses\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.17\columnwidth}\raggedright
reject \(H_0\)
(claim that
the gene is
differentially
expressed)\strut
\end{minipage} & \begin{minipage}[t]{0.22\columnwidth}\centering
False Positives (FP)
,type I error\strut
\end{minipage} & \begin{minipage}[t]{0.22\columnwidth}\centering
True Positives (TP)\strut
\end{minipage} & \begin{minipage}[t]{0.27\columnwidth}\raggedright
\(m-m_0\): number of
truly alternative
hypotheses\strut
\end{minipage}\tabularnewline
\bottomrule
\end{longtable}

We expect to make more type I errors as the number of tests increase, that
means we will reject the null hypothesis by mistake. For example, if we
perform a test the 5\% significance level, there is a 5\% chance of
incorrectly rejecting the null hypothesis if the null hypothesis is true.
However, if we make 1000 tests where all null hypotheses are true for
each of them, the average number of incorrect rejections is 50. And if we
apply the rules of probability, there are is almost a 100\% chance that
we will have at least one incorrect rejection.
There are multiple statistical techniques to prevent this from happening.
These techniques generally shrink the P-values obtained from multiple
tests to higher values, if the individual P-value is low enough it survives
this process. The most simple method is just to multiply the individual,
P-value (\(p_i\)) with the number of tests (\(m\)): \(m \cdot p_i\), this is
called ``Bonferroni correction''. However, this is too harsh if you have thousands
of tests. Other methods are developed to remedy this. Those methods
rely on ranking the P-values and dividing \(m \cdot p_i\) by the
rank,\(i\), :\(\frac{m \cdot p_i }{i}\), this is derived from Benjamini--Hochberg
procedure. This procedure is developed to control for ``False Discovery Rate (FDR)''
, which is proportion of false positives among all significant tests. And in
practical terms, we get the ``FDR adjusted P-value'' from the procedure described
above. This gives us an estimate of proportion of false discoveries for a given
test. To elaborate, p-value of 0.05 implies that 5\% of all tests will be false positives. An FDR adjusted p-value of 0.05 implies that 5\% of significant tests will be false positives. The FDR adjusted P-values will result in a lower number of false positives.

One final method that is also popular is called the ``q-value''
method and related to the method above. This procedure relies on estimating the proportion of true null
hypotheses from the distribution of raw p-values and using that quantity
to come up with what is called a ``q-value'', which is also an FDR adjusted P-value . That can be practically defined
as ``the proportion of significant features that turn out to be false
leads.'' A q-value 0.01 would mean 1\% of the tests called significant at this
level will be truly null on average. Within the genomics community
q-value and FDR adjusted P-value are synonymous although they can be
calculated differently.

In R, the base function \texttt{p.adjust()} implements most of the p-value correction
methods described above. For the q-value, we can use the \texttt{qvalue} package from
Bioconductor. Below we are demonstrating how to use them on a set of simulated
p-values.The plot shows that Bonferroni correction does a terrible job. FDR(BH) and q-value
approach are better but q-value approach is more permissive than FDR(BH).

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(qvalue)}
\KeywordTok{data}\NormalTok{(hedenfalk)}

\NormalTok{qvalues <-}\StringTok{ }\KeywordTok{qvalue}\NormalTok{(hedenfalk}\OperatorTok{$}\NormalTok{p)}\OperatorTok{$}\NormalTok{q}
\NormalTok{bonf.pval=}\KeywordTok{p.adjust}\NormalTok{(hedenfalk}\OperatorTok{$}\NormalTok{p,}\DataTypeTok{method =}\StringTok{"bonferroni"}\NormalTok{)}
\NormalTok{fdr.adj.pval=}\KeywordTok{p.adjust}\NormalTok{(hedenfalk}\OperatorTok{$}\NormalTok{p,}\DataTypeTok{method =}\StringTok{"fdr"}\NormalTok{)}

\KeywordTok{plot}\NormalTok{(hedenfalk}\OperatorTok{$}\NormalTok{p,qvalues,}\DataTypeTok{pch=}\DecValTok{19}\NormalTok{,}\DataTypeTok{ylim=}\KeywordTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{),}
     \DataTypeTok{xlab=}\StringTok{"raw P-values"}\NormalTok{,}\DataTypeTok{ylab=}\StringTok{"adjusted P-values"}\NormalTok{)}
\KeywordTok{points}\NormalTok{(hedenfalk}\OperatorTok{$}\NormalTok{p,bonf.pval,}\DataTypeTok{pch=}\DecValTok{19}\NormalTok{,}\DataTypeTok{col=}\StringTok{"red"}\NormalTok{)}
\KeywordTok{points}\NormalTok{(hedenfalk}\OperatorTok{$}\NormalTok{p,fdr.adj.pval,}\DataTypeTok{pch=}\DecValTok{19}\NormalTok{,}\DataTypeTok{col=}\StringTok{"blue"}\NormalTok{)}
\KeywordTok{legend}\NormalTok{(}\StringTok{"bottomright"}\NormalTok{,}\DataTypeTok{legend=}\KeywordTok{c}\NormalTok{(}\StringTok{"q-value"}\NormalTok{,}\StringTok{"FDR (BH)"}\NormalTok{,}\StringTok{"Bonferroni"}\NormalTok{),}
       \DataTypeTok{fill=}\KeywordTok{c}\NormalTok{(}\StringTok{"black"}\NormalTok{,}\StringTok{"blue"}\NormalTok{,}\StringTok{"red"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/multtest-1} 

}

\caption{Adjusted P-values via different methods and their relationship to raw P-values}\label{fig:multtest}
\end{figure}

\hypertarget{moderated-t-tests-using-information-from-multiple-comparisons}{%
\subsection{moderated t-tests: using information from multiple comparisons}\label{moderated-t-tests-using-information-from-multiple-comparisons}}

In genomics, we usually do not do one test but many, as described above. That means we
may be able to use the information from the parameters obtained from all
comparisons to influence the individual parameters. For example, if you have many variances
calculated for thousands of genes across samples, you can force individual
variance estimates to shrunk towards the mean or the median of the distribution
of variances. This usually creates better performance in individual variance
estimates and therefore better performance in significance testing which
depends on variance estimates. How much the values be shrunk towards a common
value comes in many flavors. These tests in general are called moderated
t-tests or shrinkage t-tests. One approach popularized by Limma software is
to use so-called ``Empirical Bayesian methods''. The main formulation in these
methods is \(\hat{V_g} = aV_0 + bV_g\), where \(V_0\) is the background variability\\
and \(V_g\) is the individual variability. Then, these methods estimate \(a\) and \(b\)
in various ways to come up with shrunk version of variability, \(\hat{V_g}\). In a Bayesian viewpoint,
the prior knowledge is used to calculate the variability of an individual gene. In this
case, \(V_0\) would be the prior knowledge we have on variability of
the genes and we
use that knowledge to influence our estimate for the individual genes.

Below we are simulating a gene expression matrix with 1000 genes, and 3 test
and 3 control groups. Each row is a gene and in normal circumstances we would
like to find out differentially expressed genes. In this case, we are simulating
them from the same distribution so in reality we do not expect any differences.
We then use the adjusted standard error estimates in empirical Bayesian spirit but
in a very crude way. We just shrink the gene-wise standard error estimates towards the median with equal \(a\) and \(b\) weights. That is to say, we add individual estimate to the
median of standard error distribution from all genes and divide that quantity by 2.
So if we plug that in the to the above formula what we do is:

\[ \hat{V_g} = (V_0 + V_g)/2 \]

In the code below, we are avoiding for loops or apply family functions
by using vectorized operations.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{100}\NormalTok{)}

\CommentTok{#sample data matrix from normal distribution}

\NormalTok{gset=}\KeywordTok{rnorm}\NormalTok{(}\DecValTok{3000}\NormalTok{,}\DataTypeTok{mean=}\DecValTok{200}\NormalTok{,}\DataTypeTok{sd=}\DecValTok{70}\NormalTok{)}
\NormalTok{data=}\KeywordTok{matrix}\NormalTok{(gset,}\DataTypeTok{ncol=}\DecValTok{6}\NormalTok{)}

\CommentTok{# set groups}
\NormalTok{group1=}\DecValTok{1}\OperatorTok{:}\DecValTok{3}
\NormalTok{group2=}\DecValTok{4}\OperatorTok{:}\DecValTok{6}
\NormalTok{n1=}\DecValTok{3}
\NormalTok{n2=}\DecValTok{3}
\NormalTok{dx=}\KeywordTok{rowMeans}\NormalTok{(data[,group1])}\OperatorTok{-}\KeywordTok{rowMeans}\NormalTok{(data[,group2])}
  
\KeywordTok{require}\NormalTok{(matrixStats)}

\CommentTok{# get the esimate of pooled variance }
\NormalTok{stderr <-}\StringTok{ }\KeywordTok{sqrt}\NormalTok{( (}\KeywordTok{rowVars}\NormalTok{(data[,group1])}\OperatorTok{*}\NormalTok{(n1}\DecValTok{-1}\NormalTok{) }\OperatorTok{+}\StringTok{ }\KeywordTok{rowVars}\NormalTok{(data[,group2])}\OperatorTok{*}\NormalTok{(n2}\DecValTok{-1}\NormalTok{)) }\OperatorTok{/}\StringTok{ }\NormalTok{(n1}\OperatorTok{+}\NormalTok{n2}\DecValTok{-2}\NormalTok{) }\OperatorTok{*}\StringTok{ }\NormalTok{( }\DecValTok{1}\OperatorTok{/}\NormalTok{n1 }\OperatorTok{+}\StringTok{ }\DecValTok{1}\OperatorTok{/}\NormalTok{n2 ))}

\CommentTok{# do the shrinking towards median}
\NormalTok{mod.stderr <-}\StringTok{ }\NormalTok{(stderr }\OperatorTok{+}\StringTok{ }\KeywordTok{median}\NormalTok{(stderr)) }\OperatorTok{/}\StringTok{ }\DecValTok{2} \CommentTok{# moderation in variation}

\CommentTok{# esimate t statistic with moderated variance}
\NormalTok{t.mod =}\StringTok{ }\NormalTok{dx }\OperatorTok{/}\StringTok{ }\NormalTok{mod.stderr}

\CommentTok{# calculate P-value of rejecting null }
\NormalTok{p.mod =}\StringTok{ }\DecValTok{2}\OperatorTok{*}\KeywordTok{pt}\NormalTok{( }\OperatorTok{-}\KeywordTok{abs}\NormalTok{(t.mod), n1}\OperatorTok{+}\NormalTok{n2}\DecValTok{-2}\NormalTok{ )}

\CommentTok{# esimate t statistic without moderated variance}
\NormalTok{t =}\StringTok{ }\NormalTok{dx }\OperatorTok{/}\StringTok{ }\NormalTok{stderr}

\CommentTok{# calculate P-value of rejecting null }
\NormalTok{p =}\StringTok{ }\DecValTok{2}\OperatorTok{*}\KeywordTok{pt}\NormalTok{( }\OperatorTok{-}\KeywordTok{abs}\NormalTok{(t), n1}\OperatorTok{+}\NormalTok{n2}\DecValTok{-2}\NormalTok{ )}

\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\KeywordTok{hist}\NormalTok{(p,}\DataTypeTok{col=}\StringTok{"cornflowerblue"}\NormalTok{,}\DataTypeTok{border=}\StringTok{"white"}\NormalTok{,}\DataTypeTok{main=}\StringTok{""}\NormalTok{,}\DataTypeTok{xlab=}\StringTok{"P-values t-test"}\NormalTok{)}
\KeywordTok{mtext}\NormalTok{(}\KeywordTok{paste}\NormalTok{(}\StringTok{"signifcant tests:"}\NormalTok{,}\KeywordTok{sum}\NormalTok{(p}\OperatorTok{<}\FloatTok{0.05}\NormalTok{))  )}
\KeywordTok{hist}\NormalTok{(p.mod,}\DataTypeTok{col=}\StringTok{"cornflowerblue"}\NormalTok{,}\DataTypeTok{border=}\StringTok{"white"}\NormalTok{,}\DataTypeTok{main=}\StringTok{""}\NormalTok{,}\DataTypeTok{xlab=}\StringTok{"P-values mod. t-test"}\NormalTok{)}
\KeywordTok{mtext}\NormalTok{(}\KeywordTok{paste}\NormalTok{(}\StringTok{"signifcant tests:"}\NormalTok{,}\KeywordTok{sum}\NormalTok{(p.mod}\OperatorTok{<}\FloatTok{0.05}\NormalTok{))  )}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-142-1} 

}

\caption{The distributions of P-values obtained by t-tests and moderated t-tests}\label{fig:unnamed-chunk-142}
\end{figure}

\BeginKnitrBlock{rmdtip}
\textbf{Want to know more ?}

\begin{itemize}
\tightlist
\item
  basic statistical concepts

  \begin{itemize}
  \tightlist
  \item
    ``Cartoon guide to statistics'' by Gonick \& Smith
  \item
    ``Introduction to statistics'' by Mine Rundel, et al.~(Free e-book)
  \end{itemize}
\item
  Hands-on statistics recipes with R

  \begin{itemize}
  \tightlist
  \item
    ``The R book'' by Crawley
  \end{itemize}
\item
  moderated tests

  \begin{itemize}
  \tightlist
  \item
    comparison of moderated tests for differential expression \url{http://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-11-17}
  \item
    limma method: Smyth, G. K. (2004). Linear models and empirical Bayes methods for assessing differential expression in microarray experiments. Statistical Applications in Genetics and Molecular Biology, 3, No.~1, Article 3. \url{http://www.statsci.org/smyth/pubs/ebayes.pdf}
  \end{itemize}
\end{itemize}
\EndKnitrBlock{rmdtip}

\hypertarget{relationship-between-variables-linear-models-and-correlation}{%
\section{Relationship between variables: linear models and correlation}\label{relationship-between-variables-linear-models-and-correlation}}

In genomics, we would often need to measure or model the relationship between
variables. We might want to know about expression of a particular gene in liver
in relation to the dosage of a drug that patient receives. Or, we may want to know
DNA methylation of certain locus in the genome in relation to age of the sample
donor's. Or, we might be interested in the relationship between histone
modifications and gene expression. Is there a linear relationship, the more
histone modification the more the gene is expressed ?

In these
situations and many more, linear regression or linear models can be used to
model the relationship with a ``dependent'' or ``response'' variable (expression or
methylation
in the above examples) and one or more ``independent''" or ``explanatory'' variables (age, drug dosage or histone modification in the above examples). Our simple linear model has the
following components.

\[  Y= \beta_0+\beta_1X + \epsilon \]

In the equation above, \(Y\) is the response variable and \(X\) is the explanatory
variable. \(\epsilon\) is the mean-zero error term. Since, the line fit will not
be able to precisely predict the \(Y\) values, there will be some error associated
with each prediction when we compare it to the original \(Y\) values. This error
is captured in \(\epsilon\) term. We can alternatively write the model as
follows to emphasize that the model approximates \(Y\), in this case notice that we removed the \(\epsilon\) term: \(Y \sim \beta_0+\beta_1X\)

The graph below shows the relationship between
histone modification (trimethylated forms of histone H3 at lysine 4, aka H3K4me3)
and gene expression for 100 genes. The blue line is our model with estimated
coefficients (\(\hat{y}=\hat{\beta}_0 + \hat{\beta}_1X\), where \(\hat{\beta}_0\)
and \(\hat{\beta}_1\) the estimated values of \(\beta_0\) and
\(\beta_1\), and \(\hat{y}\) indicates the prediction). The red lines indicate the individual
errors per data point, indicated as \(\epsilon\) in the formula above.

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-143-1} 

}

\caption{Relationship between histone modification score and gene expression. Increasing histone modification, H3K4me3, seems to be associated with increasing gene expression. Each dot is a gene}\label{fig:unnamed-chunk-143}
\end{figure}

There could be more than one explanatory variable, we then simply add more \(X\)
and \(\beta\) to our model. If there are two explanatory variables our model
will look like this:

\[  Y= \beta_0+\beta_1X_1 +\beta_2X_2 + \epsilon \]

In this case, we will be fitting a plane rather than a line. However, the fitting
process which we will describe in the later sections will not change. For our
gene expression problem. We can introduce one more histone modification, H3K27me3. We will then have a linear model with 2 explanatory variables and the
fitted plane will look like the one below. The gene expression values are shown
as dots below and above the fitted plane.

\begin{figure}

{\centering \includegraphics[width=0.7\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-144-1} 

}

\caption{Association of Gene expression with H3K4me3 and H27Kme3 histone modifications.}\label{fig:unnamed-chunk-144}
\end{figure}

\hypertarget{matrix-notation-for-linear-models}{%
\subsubsection{Matrix notation for linear models}\label{matrix-notation-for-linear-models}}

We can naturally have more explanatory variables than just two.The formula
below has \(n\) explanatory variables.

\[Y= \beta_0+\beta_1X_1+\beta_2X_2 +  \beta_3X_3 + .. + \beta_nX_n +\epsilon\]

If there are many variables, it would be easier
to write the model in matrix notation. The matrix form of linear model with
two explanatory variables will look like the one
below. First matrix would be our data matrix. This contains our explanatory
variables and a column of 1s. The second term is a column vector of \(\beta\)
values. We add a vector of error terms,\(\epsilon\)s to the matrix multiplication.

\[
 \mathbf{Y} = \left[\begin{array}{rrr}
1 & X_{1,1} & X_{1,2} \\
1 & X_{2,1} & X_{2,2} \\
1 & X_{3,1} & X_{3,2} \\
1 & X_{4,1} & X_{4,2}
\end{array}\right]
%
\left[\begin{array}{rrr}
\beta_0 \\
\beta_1 \\
\beta_2 
\end{array}\right]
% 
+
\left[\begin{array}{rrr}
\epsilon_1 \\
\epsilon_2 \\ 
\epsilon_3 \\ 
\epsilon_0
\end{array}\right]
\]

The multiplication of data matrix and \(\beta\) vector and addition of the
error terms simply results in the the following set of equations per data point:

\[
\begin{aligned}
Y_1= \beta_0+\beta_1X_{1,1}+\beta_2X_{1,2} +\epsilon_1 \\
Y_2= \beta_0+\beta_1X_{2,1}+\beta_2X_{2,2} +\epsilon_2 \\
Y_3= \beta_0+\beta_1X_{3,1}+\beta_2X_{3,2} +\epsilon_3 \\
Y_4= \beta_0+\beta_1X_{4,1}+\beta_2X_{4,2} +\epsilon_4 
\end{aligned}
\]

This expression involving the multiplication of the data matrix, the
\(\beta\) vector and vector of error terms (\(\epsilon\))
could be simply written as follows.

\[Y=X\beta + \epsilon\]

In the equation above \(Y\) is the vector of response variables and \(X\) is the
data matrix and \(\beta\) is the vector of coefficients.
This notation is more concise and often used in scientific papers. However, this
also means you need some understanding of linear algebra to follow the math
laid out in such resources.

\hypertarget{how-to-fit-a-line}{%
\subsection{How to fit a line}\label{how-to-fit-a-line}}

At this point a major questions is left unanswered: How did we fit this line?
We basically need to define \(\beta\) values in a structured way.
There are multiple ways or understanding how
to do this, all of which converges to the same
end point. We will describe them one by one.

\hypertarget{the-cost-or-loss-function-approach}{%
\subsubsection{The cost or loss function approach}\label{the-cost-or-loss-function-approach}}

This is the first approach and in my opinion is easiest to understand.
We try to optimize a function, often called ``cost function'' or ``loss function''.
The cost function
is the sum of squared differences between the predicted \(\hat{Y}\) values from our model
and the original \(Y\) values. The optimization procedure tries to find \(\beta\) values
that minimizes this difference between reality and the predicted values.

\[min \sum{(y_i-(\beta_0+\beta_1x_i))^2}\]

Note that this is related to the the error term, \(\epsilon\), we already mentioned
above, we are trying to minimize the squared sum of \(\epsilon_i\) for each data
point. We can do this minimization by a bit of calculus.
The rough algorithm is as follows:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Pick a random starting point, random \(\beta\) values
\item
  Take the partial derivatives of the cost function to see which direction is
  the way to go in the cost function.
\item
  Take a step toward the direction that minimizes the cost function.

  \begin{itemize}
  \tightlist
  \item
    step size is parameter to choose, there are many variants.
  \end{itemize}
\item
  repeat step 2,3 until convergence.
\end{enumerate}

This is the basis of ``gradient descent'' algorithm. With the help of partial
derivatives we define a ``gradient'' on the cost function and follow that through
multiple iterations and until convergence, meaning until the results do not
improve defined by a margin. The algorithm usually converges to optimum \(\beta\)
values. Below, we show the cost function over various \(\beta_0\) and \(\beta_1\)
values for the histone modification and gene expression data set. The algorithm
will pick a point on this graph and traverse it incrementally based on the
derivatives and converge on the bottom of the cost function ``well''.

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/3dcostfunc-1} 

}

\caption{Cost function landscape for linear regression with changing beta values. The optimization process tries to find the lowest point in this landscape by implementing a strategy for updating beta values towards the lowest point in the landscape.}\label{fig:3dcostfunc}
\end{figure}

\hypertarget{not-cost-function-but-maximum-likelihood-function}{%
\subsubsection{Not cost function but maximum likelihood function}\label{not-cost-function-but-maximum-likelihood-function}}

We can also think of this problem from more a statistical point of view. In
essence, we are looking for best statistical parameters, in this
case \(\beta\) values, for our model that are most likely to produce such a
scatter of data points given the explanatory variables.This is called
``Maximum likelihood'' approach. The approach assumes that a given response variable \(y_i\) follows a normal distribution with mean \(\beta_0+\beta_1x_i\) and variance \(s^2\). Therefore probability of observing any given \(y_i\) value is dependent on \(\beta_0\) and \(\beta_1\) values. Since \(x_i\), the explanatory variable, is fixed within our data set, by varying \(\beta_0\) and \(\beta_1\) values we can maximize the probability of observing any given \(y_i\). The trick is to find \(\beta_0\) and \(\beta_1\) values that maximizes the probability of observing all the response variables in the dataset given the explanatory variables. The probability of observing a response variable \(y_i\) with assumptions we described above is shown below. Note that this assumes variance is constant and \(s^2=\frac{\sum{\epsilon_i}}{n-2}\) is an unbiased estimation for population variance, \(\sigma^2\).

\[P(y_{i})=\frac{1}{s\sqrt{2\pi} }e^{-\frac{1}{2}\left(\frac{y_i-(\beta_0 + \beta_1x_i)}{s}\right)^2}\]

Following from the probability equation above, the likelihood function (shown as \(L\) below) for
linear regression is multiplication of \(P(y_{i})\) for all data points.

\[L=P(y_1)P(y_2)P(y_3)..P(y_n)=\prod\limits_{i=1}^n{P_i}\]

This can be simplified to this by some algebra and taking logs (since it is
easier to add than multiply)

\[ln(L) = -nln(s\sqrt{2\pi}) - \frac{1}{2s^2} \sum\limits_{i=1}^n{(y_i-(\beta_0 + \beta_1x_i))^2} \]

As you can see, the right part of the function is the negative of the cost function
defined above. If we wanted to optimize this function we would need to take derivative of
the function with respect to \(\beta\) parameters. That means we can ignore the
first part since there is no \(\beta\) terms there. This simply reduces to the
negative of the cost function. Hence, this approach produces exactly the same
result as the cost function approach. The difference is that we defined our
problem
within the domain of statistics. This particular function has still to be optimized. This can be done with some calculus without the need for an
iterative approach.

\hypertarget{linear-algebra-and-closed-form-solution-to-linear-regression}{%
\subsubsection{Linear algebra and closed-form solution to linear regression}\label{linear-algebra-and-closed-form-solution-to-linear-regression}}

The last approach we will describe is the minimization process using linear
algebra. If you find this concept challenging, feel free to skip it but scientific publications and other books frequently use matrix notation and linear algebra to define and solve regression problems. In this case, we do not use an iterative approach. Instead, we will
minimize cost function by explicitly taking its derivatives with respect to
\(\beta\)'s and setting them to zero. This is doable by employing linear algebra
and matrix calculus. This approach is also called ``ordinary least squares''. We
will not
show the whole derivation here but the following expression
is what we are trying to minimize in matrix notation, this is basically a
different notation of the same minimization problem defined above. Remember
\(\epsilon_i=Y_i-(\beta_0+\beta_1x_i)\)

\[
\begin{aligned}
\sum\epsilon_{i}^2=\epsilon^T\epsilon=(Y-{\beta}{X})^T(Y-{\beta}{X}) \\
=Y^T{Y}-2{\beta}^T{Y}+{\beta}^TX^TX{\beta}
\end{aligned}
\]
After rearranging the terms, we take the derivative of \(\epsilon^T\epsilon\)
with respect to \(\beta\), and equalize that to zero. We then arrive at
the following for estimated \(\beta\) values, \(\hat{\beta}\):

\[\hat{\beta}=(X^TX)^{-1}X^TY\]

This requires for you to calculate the inverse of the \(X^TX\) term, which could
be slow for large matrices. Iterative approach over the cost function
derivatives will be faster for larger problems.
The linear algebra notation is something you will see in the papers
or other resources often. If you input the data matrix X and solve the \((X^TX)^{-1}\)
,
you get the following values for \(\beta_0\) and \(\beta_1\) for simple regression
. However, we should note that this simple linear regression case can easily
be solved algebraically without the need for matrix operations. This can be done
by taking the derivative of \(\sum{(y_i-(\beta_0+\beta_1x_i))^2}\) with respect to
\(\beta_1\), rearranging the terms and equalizing the derivative to zero.

\[\hat{\beta_1}=\frac{\sum{(x_i-\overline{X})(y_i-\overline{Y})}}{ \sum{(x_i-\overline{X})^2} }\]
\[\hat{\beta_0}=\overline{Y}-\hat{\beta_1}\overline{X}\]

\hypertarget{fitting-lines-in-r}{%
\subsubsection{Fitting lines in R}\label{fitting-lines-in-r}}

After all this theory, you will be surprised how easy it is to fit lines in R.
This is achieved just by \texttt{lm()} command, stands for linear models. Let's do this
for a simulated data set and plot the fit. First step is to simulate the
data, we will decide on \(\beta_0\) and \(\beta_1\) values. The we will decide
on the variance parameter,\(\sigma\) to be used in simulation of error terms,
\(\epsilon\). We will first find \(Y\) values, just using the linear equation
\(Y=\beta0+\beta_1X\), for
a set of \(X\) values. Then, we will add the error terms get our simulated values.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# set random number seed, so that the random numbers from the text}
\CommentTok{# is the same when you run the code.}
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{32}\NormalTok{)}

\CommentTok{# get 50 X values between 1 and 100}
\NormalTok{x =}\StringTok{ }\KeywordTok{runif}\NormalTok{(}\DecValTok{50}\NormalTok{,}\DecValTok{1}\NormalTok{,}\DecValTok{100}\NormalTok{)}

\CommentTok{# set b0,b1 and varience (sigma)}
\NormalTok{b0 =}\StringTok{ }\DecValTok{10}
\NormalTok{b1 =}\StringTok{ }\DecValTok{2}
\NormalTok{sigma =}\StringTok{ }\DecValTok{20}
\CommentTok{# simulate error terms from normal distribution}
\NormalTok{eps =}\StringTok{ }\KeywordTok{rnorm}\NormalTok{(}\DecValTok{50}\NormalTok{,}\DecValTok{0}\NormalTok{,sigma)}
\CommentTok{# get y values from the linear equation and addition of error terms}
\NormalTok{y =}\StringTok{ }\NormalTok{b0 }\OperatorTok{+}\StringTok{ }\NormalTok{b1}\OperatorTok{*}\NormalTok{x}\OperatorTok{+}\StringTok{ }\NormalTok{eps}
\end{Highlighting}
\end{Shaded}

Now let us fit a line using lm() function. The function requires a formula, and
optionally a data frame. We need the pass the following expression within the
lm function, \texttt{y\textasciitilde{}x}, where \texttt{y} is the simulated \(Y\) values and \texttt{x} is the explanatory variables \(X\). We will then use \texttt{abline()} function to draw the fit.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mod1=}\KeywordTok{lm}\NormalTok{(y}\OperatorTok{~}\NormalTok{x)}

\CommentTok{# plot the data points}
\KeywordTok{plot}\NormalTok{(x,y,}\DataTypeTok{pch=}\DecValTok{20}\NormalTok{,}
     \DataTypeTok{ylab=}\StringTok{"Gene Expression"}\NormalTok{,}\DataTypeTok{xlab=}\StringTok{"Histone modification score"}\NormalTok{)}
\CommentTok{# plot the linear fit}
\KeywordTok{abline}\NormalTok{(mod1,}\DataTypeTok{col=}\StringTok{"blue"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-146-1} 

}

\caption{Gene expression and histone modification score modelled by linear regression}\label{fig:unnamed-chunk-146}
\end{figure}

\hypertarget{how-to-estimate-the-error-of-the-coefficients}{%
\subsection{How to estimate the error of the coefficients}\label{how-to-estimate-the-error-of-the-coefficients}}

Since we are using a sample to estimate the coefficients they are
not exact, with every random sample they will vary. Below, we
are taking multiple samples from the population and fitting lines to each
sample, with each sample the lines slightly change.We are overlaying the
points and the lines for each sample on top of the other samples
.When we take 200 samples and fit lines for each of them,the lines fit are
variable. And,
we get a normal-like distribution of \(\beta\) values with a defined mean
and standard deviation a, which is called standard error of the
coefficients.

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/unnamed-chunk-147-1} 

}

\caption{Regression coefficients vary with every random sample. The figure illustrates the variability of regression coefficients when regression is done using a sample of data points. Histograms depict this variability for $b_0$ and $b_1$ coefficients.}\label{fig:unnamed-chunk-147}
\end{figure}

As usually we will not have access to the population to do repeated sampling,
model fitting and estimation of the standard error for the coefficients. But
there is statistical theory that helps us infer the population properties from
the sample. When we assume that error terms have constant variance and mean zero
, we can model the uncertainty in the regression coefficients, \(\beta\)s.
The estimates for standard errors of \(\beta\)s for simple regression are as
follows and shown without derivation.

\[
\begin{aligned}
s=RSE=\sqrt{\frac{\sum{(y_i-(\beta_0+\beta_1x_i))^2}}{n-2}  } =\sqrt{\frac{\sum{\epsilon^2}}{n-2}  } \\
SE(\hat{\beta_1})=\frac{s}{\sqrt{\sum{(x_i-\overline{X})^2}}} \\
SE(\hat{\beta_0})=s\sqrt{ \frac{1}{n} + \frac{\overline{X}^2}{\sum{(x_i-\overline{X})^2} }  }
\end{aligned}
\]

Notice that that \(SE(\beta_1)\) depends on the estimate of variance of
residuals shown as \(s\) or \textbf{Residual Standard Error (RSE)}.
Notice alsos standard error depends on the spread of \(X\). If \(X\) values have more
variation, the standard error will be lower. This intuitively makes sense since if the
spread of the \(X\) is low, the regression line will be able to wiggle more
compared to a regression line that is fit to the same number of points but
covers a greater range on the X-axis.

The standard error estimates can also be used to calculate confidence intervals and test
hypotheses, since the following quantity called t-score approximately follows a
t-distribution with \(n-p\) degrees of freedom, where \(n\) is the number
of data points and \(p\) is the number of coefficients estimated.

\[ \frac{\hat{\beta_i}-\beta_test}{SE(\hat{\beta_i})}\]

Often, we would like to test the null hypothesis if a coefficient is equal to
zero or not. For simple regression this could mean if there is a relationship
between explanatory variable and response variable. We would calculate the
t-score as follows \(\frac{\hat{\beta_i}-0}{SE(\hat{\beta_i})}\), and compare it
t-distribution with \(d.f.=n-p\) to get the p-value.

We can also
calculate the uncertainty of the regression coefficients using confidence
intervals, the range of values that are likely to contain \(\beta_i\). The 95\%
confidence interval for \(\hat{\beta_i}\) is
\(\hat{\beta_i}\)  \(t_{0.975}SE(\hat{\beta_i})\).
\(t_{0.975}\) is the 97.5\% percentile of
the t-distribution with \(d.f. = n  p\).

In R, \texttt{summary()} function will test all the coefficients for the null hypothesis
\(\beta_i=0\). The function takes the model output obtained from the \texttt{lm()}
function. To demonstrate this, let us first get some data. The procedure below
simulates data to be used in a regression setting and it is useful to examine
what the linear model expect to model the data.

Since we have the data, we can build our model and call the \texttt{summary} function.
We will then use \texttt{confint()} function to get the confidence intervals on the
coefficients and \texttt{coef()} function to pull out the estimated coefficients from
the model.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mod1=}\KeywordTok{lm}\NormalTok{(y}\OperatorTok{~}\NormalTok{x)}
\KeywordTok{summary}\NormalTok{(mod1)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## Call:
## lm(formula = y ~ x)
## 
## Residuals:
##    Min     1Q Median     3Q    Max 
## -77.11 -18.44   0.33  16.06  57.23 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(>|t|)    
## (Intercept)  13.2454     6.2887    2.11    0.038 *  
## x             0.4995     0.0513    9.74  4.5e-16 ***
## ---
## Signif. codes:  
## 0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Residual standard error: 28.8 on 98 degrees of freedom
## Multiple R-squared:  0.492,  Adjusted R-squared:  0.486 
## F-statistic: 94.8 on 1 and 98 DF,  p-value: 4.54e-16
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# get confidence intervals }
\KeywordTok{confint}\NormalTok{(mod1)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##              2.5 %  97.5 %
## (Intercept) 0.7657 25.7251
## x           0.3977  0.6014
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# pull out coefficients from the model}
\KeywordTok{coef}\NormalTok{(mod1)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## (Intercept)           x 
##     13.2454      0.4995
\end{verbatim}

The \texttt{summary()} function prints out an extensive list of values.
The ``Coefficients'' section has the estimates, their standard error, t score
and the p-value from the hypothesis test \(H_0:\beta_i=0\). As you can see, the
estimate we get for the coefficients and their standard errors are close to
the ones we get from the repeatedly sampling and getting a distribution of
coefficients. This is statistical inference at work, we can estimate the
population properties within a certain error using just a sample.

\hypertarget{accuracy-of-the-model}{%
\subsection{Accuracy of the model}\label{accuracy-of-the-model}}

If you have observed the table output by \texttt{summary()} function, you must have noticed there are some other outputs, such as ``Residual standard error'',
``Multiple R-squared'' and ``F-statistic''. These are metrics that are useful
for assessing the accuracy of the model. We will explain them one by one.

\_ (RSE)\_ simply is the square-root of the
the sum of squared error terms, divided by degrees of freedom, \(n-p\), for simple
linear regression case, \(n-2\). Sum of of the squares of the error terms is also
called \textbf{``Residual sum of squares''}, RSS. So RSE is
calculated as follows:

\[ s=RSE=\sqrt{\frac{\sum{(y_i-\hat{Y_i})^2 }}{n-p}}=\sqrt{\frac{RSS}{n-p}}\]

RSE is a way of assessing the model fit. The larger the RSE the worse the
model is. However, this is an absolute measure in the units of \(Y\) and we have nothing to
compare against. One idea is that we divide it by RSS of a simpler model
for comparative purposes. That simpler model is in this case is the model
with the intercept,\(\beta_0\). A very bad model will have close zero
coefficients for explanatory variables, and the RSS of that model
will be close to the RSS of the model with only the intercept. In such
a model intercept will be equal to \(\overline{Y}\). As it turns out, RSS of
the the model with
just the intercept is called \emph{``Total Sum of Squares'' or TSS}. A good model will have a low \(RSS/TSS\). The metric \(R^2\) uses these quantities to calculate a score between 0 and 1, and closer to 1 the better the model. Here is how
it is calculated:

\[R^2=1-\frac{RSS}{TSS}=\frac{TSS-RSS}{TSS}=1-\frac{RSS}{TSS}\]

\(TSS-RSS\) part of the formula often referred to as ``explained variability'' in
the model. The bottom part is for ``total variability''. With this interpretation, higher
the ``explained variability'' better the model. For simple linear regression
with one explanatory variable, the square root of \(R^2\) is a quantity known
as absolute value of the correlation coefficient, which can be calculated for any pair of variables, not only
the
response and the explanatory variables. \emph{Correlation} is a general measure of
linear
relationship between two variables. One
of the most popular flavors of correlation is the Pearson correlation coefficient. Formally, It is the
\emph{covariance} of X and Y divided by multiplication of standard deviations of
X and Y. In R, it can be calculated with \texttt{cor()} function.

\[ 
r_{xy}=\frac{cov(X,Y)}{\sigma_x\sigma_y}
      =\frac{\sum\limits_{i=1}^n (x_i-\bar{x})(y_i-\bar{y})}
            {\sqrt{\sum\limits_{i=1}^n (x_i-\bar{x})^2 \sum\limits_{i=1}^n (y_i-\bar{y})^2}}
\]
In the equation above, cov is the covariance, this is again a measure of
how much two variables change together, like correlation. If two variables
show similar behavior they will usually have positive covariance value, if they have opposite behavior, the covariance will have negative value.
However, these values are boundless. A normalized way of looking at
covariance is to divide covariance by the multiplication of standard
errors of X and Y. This bounds the values to -1 and 1, and as mentioned
above called Pearson correlation coefficient. The values that change in a similar manner will have a positive coefficient, the values that change in
opposite manner will have negative coefficient, and pairs do not have
a linear relationship will have 0 or near 0 correlation. In
the figure below, we are showing \(R^2\), correlation
coefficient and covariance for different scatter plots.

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/unnamed-chunk-150-1} 

}

\caption{Correlation and covariance for different scatter plots}\label{fig:unnamed-chunk-150}
\end{figure}

For simple linear regression, correlation can be used to asses the model. However, this becomes useless as a measure of general accuracy
if the there are more than one explanatory
variable as in multiple linear regression. In that case, \(R^2\) is a measure
of accuracy for the model. Interestingly, square of the
correlation of predicted values
and original response variables (\((cor(Y,\hat{Y}))^2\) ) equals to \(R^2\) for
multiple linear regression.

The last accuracy measure or the model fit in general we are going to explain is \emph{F-statistic}. This is a quantity that depends on RSS and TSS again. It can also answer one important question that other metrics can
not easily answer. That question is whether or not any of the explanatory
variables have predictive value or in other words if all the explanatory variables are zero. We can write the null hypothesis as follows:

\[H_0: \beta_1=\beta_2=\beta_3=...=\beta_p=0 \]

where the alternative is:

\[H_1: \text{at least one } \beta_i \neq 0 \]

Remember \(TSS-RSS\) is analogous to ``explained variability'' and the RSS is
analogous to ``unexplained variability''. For the F-statistic, we divide explained variance to
unexplained variance. Explained variance is just the \(TSS-RSS\) divided
by degrees of freedom, and unexplained variance is the RSE.
The ratio will follow the F-distribution
with two parameters, the degrees of freedom for the explained variance and
the degrees of freedom for the the unexplained variance.F-statistic for a linear model is calculated as follows.

\[F=\frac{(TSS-RSS)/(p-1)}{RSS/(n-p)}=\frac{(TSS-RSS)/(p-1)}{RSE} \sim F(p-1,n-p)\]

If the variances are the same, the ratio will be 1, and when \(H_0\) is true, then
it can be shown that expected value of \((TSS-RSS)/(p-1)\) will be \(\sigma^2\)
which is estimated by RSE. So, if the variances are significantly different,
the ratio will need to be significantly bigger than 1.
If the ratio is large enough we can reject the null hypothesis. To asses that
we need to use software or look up the tables for F statistics with calculated
parameters. In R, function \texttt{qf()} can be used to calculate critical value of the
ratio. Benefit of the F-test over
looking at significance of coefficients one by one is that we circumvent
multiple testing problem. If there are lots of explanatory variables
at least 5\% of the time (assuming we use 0.05 as P-value significance
cutoff), p-values from coefficient t-tests will be wrong. In summary,
F-test is a better choice for testing if there is any association
between the explanatory variables and the response variable.

\hypertarget{regression-with-categorical-variables}{%
\subsection{Regression with categorical variables}\label{regression-with-categorical-variables}}

An important feature of linear regression is that categorical variables can
be used as explanatory variables, this feature is very useful in genomics
where explanatory variables often could be categorical. To put it in
context, in our histone modification example we can also include if
promoters have CpG islands or not as a variable. In addition, in
differential gene expression, we usually test the difference between
different condition which can be encoded as categorical variables in
a linear regression. We can sure use t-test for that as well if there
are only 2 conditions, but if there are more conditions and other variables
to control for such as Age or sex of the samples, we need to take those
into account for our statistics, and t-test alone can not handle such
complexity. In addition, when we have categorical variables we can also
have numeric variables in the model and we certainly do not have to include
only one type of variable in a model.

The simplest model with categorical variables include two levels that
can be encoded in 0 and 1.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{100}\NormalTok{)}
\NormalTok{gene1=}\KeywordTok{rnorm}\NormalTok{(}\DecValTok{30}\NormalTok{,}\DataTypeTok{mean=}\DecValTok{4}\NormalTok{,}\DataTypeTok{sd=}\DecValTok{2}\NormalTok{)}
\NormalTok{gene2=}\KeywordTok{rnorm}\NormalTok{(}\DecValTok{30}\NormalTok{,}\DataTypeTok{mean=}\DecValTok{2}\NormalTok{,}\DataTypeTok{sd=}\DecValTok{2}\NormalTok{)}
\NormalTok{gene.df=}\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{exp=}\KeywordTok{c}\NormalTok{(gene1,gene2),}
                  \DataTypeTok{group=}\KeywordTok{c}\NormalTok{( }\KeywordTok{rep}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{30}\NormalTok{),}\KeywordTok{rep}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{30}\NormalTok{) ) )}

\NormalTok{mod2=}\KeywordTok{lm}\NormalTok{(exp}\OperatorTok{~}\NormalTok{group,}\DataTypeTok{data=}\NormalTok{gene.df)}
\KeywordTok{summary}\NormalTok{(mod2)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## Call:
## lm(formula = exp ~ group, data = gene.df)
## 
## Residuals:
##    Min     1Q Median     3Q    Max 
## -4.729 -1.066  0.012  1.384  4.563 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(>|t|)    
## (Intercept)    2.185      0.352    6.21    6e-08 ***
## group          1.873      0.497    3.77  0.00039 ***
## ---
## Signif. codes:  
## 0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Residual standard error: 1.93 on 58 degrees of freedom
## Multiple R-squared:  0.196,  Adjusted R-squared:  0.183 
## F-statistic: 14.2 on 1 and 58 DF,  p-value: 0.000391
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{require}\NormalTok{(mosaic)}
\KeywordTok{plotModel}\NormalTok{(mod2)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/unnamed-chunk-151-1} 

}

\caption{Linear model with a categorical variable coded as 0 and 1}\label{fig:unnamed-chunk-151}
\end{figure}

we can even compare more levels, we do not even have to encode them
ourselves. We can pass categorical variables to \texttt{lm()} function.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gene.df=}\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{exp=}\KeywordTok{c}\NormalTok{(gene1,gene2,gene2),}
                  \DataTypeTok{group=}\KeywordTok{c}\NormalTok{( }\KeywordTok{rep}\NormalTok{(}\StringTok{"A"}\NormalTok{,}\DecValTok{30}\NormalTok{),}\KeywordTok{rep}\NormalTok{(}\StringTok{"B"}\NormalTok{,}\DecValTok{30}\NormalTok{),}\KeywordTok{rep}\NormalTok{(}\StringTok{"C"}\NormalTok{,}\DecValTok{30}\NormalTok{) ) }
\NormalTok{                  )}

\NormalTok{mod3=}\KeywordTok{lm}\NormalTok{(exp}\OperatorTok{~}\NormalTok{group,}\DataTypeTok{data=}\NormalTok{gene.df)}
\KeywordTok{summary}\NormalTok{(mod3)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## Call:
## lm(formula = exp ~ group, data = gene.df)
## 
## Residuals:
##    Min     1Q Median     3Q    Max 
## -4.729 -1.079 -0.098  1.484  4.563 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(>|t|)    
## (Intercept)    4.058      0.378    10.7  < 2e-16 ***
## groupB        -1.873      0.535    -3.5  0.00073 ***
## groupC        -1.873      0.535    -3.5  0.00073 ***
## ---
## Signif. codes:  
## 0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Residual standard error: 2.07 on 87 degrees of freedom
## Multiple R-squared:  0.158,  Adjusted R-squared:  0.139 
## F-statistic: 8.17 on 2 and 87 DF,  p-value: 0.000558
\end{verbatim}

\hypertarget{regression-pitfalls}{%
\subsection{Regression pitfalls}\label{regression-pitfalls}}

In most cases one should look at the error terms (residuals) vs fitted
values plot. Any structure in this plot indicates problems such as
non-linearity, correlation of error terms, non-constant variance or
unusual values driving the fit. Below we briefly explain the potential
issues with the linear regression.

\hypertarget{non-linearity}{%
\paragraph{non-linearity}\label{non-linearity}}

If the true relationship is far from linearity, prediction accuracy
is reduced and all the other conclusions are questionable. In some cases,
transforming the data with \(logX\), \(\sqrt{X}\) and \(X^2\) could resolve
the issue.

\hypertarget{correlation-of-explanatory-variables}{%
\paragraph{correlation of explanatory variables}\label{correlation-of-explanatory-variables}}

If the explanatory variables are correlated that could lead to something
known as multicolinearity. When this happens SE estimates of the coefficients will be too large. This is usually observed in time-course
data.

\hypertarget{correlation-of-error-terms}{%
\paragraph{correlation of error terms}\label{correlation-of-error-terms}}

This assumes that the errors of the response variables are uncorrelated with each other. If they are confidence intervals in the coefficients
might too narrow.

\hypertarget{non-constant-variance-of-error-terms}{%
\paragraph{Non-constant variance of error terms}\label{non-constant-variance-of-error-terms}}

This means that different response variables have the same variance in their errors, regardless of the values of the predictor variables. If
the errors are not constant, if for the errors grow as X grows this
will result in unreliable estimates in standard errors as the model
assumes constant variance. Transformation of data, such as
\(logX\) and \(\sqrt{X}\) could help in some cases.

\hypertarget{outliers-and-high-leverage-points}{%
\paragraph{outliers and high leverage points}\label{outliers-and-high-leverage-points}}

Outliers are extreme values for Y and high leverage points are unusual
X values. Both of these extremes have power to affect the fitted line
and the standard errors. In some cases (measurement error), they can be
removed from the data for a better fit.

\BeginKnitrBlock{rmdtip}
\textbf{Want to know more ?}

\begin{itemize}
\tightlist
\item
  linear models and derivations of equations including matrix notation

  \begin{itemize}
  \tightlist
  \item
    Applied Linear Statistical Models by Kutner, Nachtsheim, et al.
  \item
    Elements of statistical learning by Hastie \& Tibshirani
  \item
    An Introduction to statistical learning by James, Witten, et al.
  \end{itemize}
\end{itemize}
\EndKnitrBlock{rmdtip}

\hypertarget{exercises-1}{%
\section{Exercises}\label{exercises-1}}

\hypertarget{how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions-1}{%
\subsection{How to summarize collection of data points: The idea behind statistical distributions}\label{how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions-1}}

\hypertarget{section}{%
\subsubsection{}\label{section}}

Calculate the means and variances
of the rows of the following simulated data set, plot the distributions
of means and variances using \texttt{hist()} and \texttt{boxplot()} functions.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{100}\NormalTok{)}

\CommentTok{#sample data matrix from normal distribution}
\NormalTok{gset=}\KeywordTok{rnorm}\NormalTok{(}\DecValTok{600}\NormalTok{,}\DataTypeTok{mean=}\DecValTok{200}\NormalTok{,}\DataTypeTok{sd=}\DecValTok{70}\NormalTok{)}
\NormalTok{data=}\KeywordTok{matrix}\NormalTok{(gset,}\DataTypeTok{ncol=}\DecValTok{6}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{section-1}{%
\subsubsection{}\label{section-1}}

Using the data generated above, calculate the standard deviation of the
distribution of the means using \texttt{sd()} function. Compare that to the expected
standard error obtained from central limit theorem keeping in mind the
population parameters were \(\sigma=70\) and \(n=6\). How does the estimate
from the random samples change if we simulate more data with
\texttt{data=matrix(rnorm(6000,mean=200,sd=70),ncol=6)}

\hypertarget{section-2}{%
\subsubsection{}\label{section-2}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{-1}
\tightlist
\item
  simulate 30 random variables using \texttt{rpois()} function, do this 1000 times and calculate means of sample. Plot the sampling distributions of the means
  using a histogram. Get the 2.5th and 97.5th percentiles of the
  distribution.
\item
  Use \texttt{t.test} function to calculate confidence intervals
  of the first random sample \texttt{pois1} simulated from\texttt{rpois()} function below.
\item
  Use bootstrap confidence interval for the mean on \texttt{pois1}
\item
  compare all the estimates
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{100}\NormalTok{)}

\CommentTok{#sample 30 values from poisson dist with lamda paramater =30}
\NormalTok{pois1=}\KeywordTok{rpois}\NormalTok{(}\DecValTok{30}\NormalTok{,}\DataTypeTok{lambda=}\DecValTok{5}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{section-3}{%
\subsubsection{}\label{section-3}}

Optional exercise:
Try to recreate the following figure, which demonstrates the CLT concept.

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/unnamed-chunk-157-1} \end{center}

\hypertarget{how-to-test-for-differences-in-samples}{%
\subsection{How to test for differences in samples}\label{how-to-test-for-differences-in-samples}}

\hypertarget{section-4}{%
\subsubsection{}\label{section-4}}

Test the difference of means of the following simulated genes
using the randomization, t-test and \texttt{wilcox.test()} functions.
Plot the distributions using histograms and boxplots.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{101}\NormalTok{)}
\NormalTok{gene1=}\KeywordTok{rnorm}\NormalTok{(}\DecValTok{30}\NormalTok{,}\DataTypeTok{mean=}\DecValTok{4}\NormalTok{,}\DataTypeTok{sd=}\DecValTok{3}\NormalTok{)}
\NormalTok{gene2=}\KeywordTok{rnorm}\NormalTok{(}\DecValTok{30}\NormalTok{,}\DataTypeTok{mean=}\DecValTok{3}\NormalTok{,}\DataTypeTok{sd=}\DecValTok{3}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{section-5}{%
\subsubsection{}\label{section-5}}

Test the difference of means of the following simulated genes
using the randomization, t-test and \texttt{wilcox.test()} functions.
Plot the distributions using histograms and boxplots.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{100}\NormalTok{)}
\NormalTok{gene1=}\KeywordTok{rnorm}\NormalTok{(}\DecValTok{30}\NormalTok{,}\DataTypeTok{mean=}\DecValTok{4}\NormalTok{,}\DataTypeTok{sd=}\DecValTok{2}\NormalTok{)}
\NormalTok{gene2=}\KeywordTok{rnorm}\NormalTok{(}\DecValTok{30}\NormalTok{,}\DataTypeTok{mean=}\DecValTok{2}\NormalTok{,}\DataTypeTok{sd=}\DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{section-6}{%
\subsubsection{}\label{section-6}}

read the gene expression data set with \texttt{data=readRDS("StatisticsForGenomics/geneExpMat.rds")}.
The data has 100 differentially expressed genes.First 3 columns
are the test samples, and the last 3 are the control samples. Do
a t-test for each gene (each row is a gene), record the p-values.
Then, do a moderated t-test, as shown in the lecture notes and record
the p-values. Do a p-value histogram and compare two approaches in terms
of the number of significant tests with 0.05 threshold.
On the p-values use FDR (BH), bonferroni and q-value adjustment methods.
Calculate how many adjusted p-values are below 0.05 for each approach.

\hypertarget{relationship-between-variables-linear-models-and-correlation-1}{%
\subsection{Relationship between variables: linear models and correlation}\label{relationship-between-variables-linear-models-and-correlation-1}}

\hypertarget{section-7}{%
\subsubsection{}\label{section-7}}

Below we are going to simulate X and Y values.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Run the code then fit a line to predict Y based on X.
\item
  Plot the scatter plot and the fitted line.
\item
  Calculate correlation and R\^{}2.
\item
  Run the \texttt{summary()} function and
  try to extract P-values for the model from the object
  returned by \texttt{summary}. see \texttt{?summary.lm}
\item
  Plot the residuals vs fitted values plot, by calling \texttt{plot}
  function with \texttt{which=1} as the second argument. First argument
  is the model returned by \texttt{lm}.
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# set random number seed, so that the random numbers from the text}
\CommentTok{# is the same when you run the code.}
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{32}\NormalTok{)}

\CommentTok{# get 50 X values between 1 and 100}
\NormalTok{x =}\StringTok{ }\KeywordTok{runif}\NormalTok{(}\DecValTok{50}\NormalTok{,}\DecValTok{1}\NormalTok{,}\DecValTok{100}\NormalTok{)}

\CommentTok{# set b0,b1 and varience (sigma)}
\NormalTok{b0 =}\StringTok{ }\DecValTok{10}
\NormalTok{b1 =}\StringTok{ }\DecValTok{2}
\NormalTok{sigma =}\StringTok{ }\DecValTok{20}
\CommentTok{# simulate error terms from normal distribution}
\NormalTok{eps =}\StringTok{ }\KeywordTok{rnorm}\NormalTok{(}\DecValTok{50}\NormalTok{,}\DecValTok{0}\NormalTok{,sigma)}
\CommentTok{# get y values from the linear equation and addition of error terms}
\NormalTok{y =}\StringTok{ }\NormalTok{b0 }\OperatorTok{+}\StringTok{ }\NormalTok{b1}\OperatorTok{*}\NormalTok{x}\OperatorTok{+}\StringTok{ }\NormalTok{eps}
\end{Highlighting}
\end{Shaded}

\hypertarget{section-8}{%
\subsubsection{}\label{section-8}}

Read the data set histone modification data set with using a variation of:
\texttt{df=readRDS("StatisticsForGenomics\_data/HistoneModeVSgeneExp.rds")}. There
are 3 columns in the data set these are measured levels of H3K4me3,
H3K27me3 and gene expression per gene.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  plot the scatter plot for H3K4me3 vs expression
\item
  plot the scatter plot for H3K27me3 vs expression
\item
  fit the model model for prediction of expression data using:

  \begin{itemize}
  \tightlist
  \item
    only H3K4me3 as explanatory variable
  \item
    only H3K27me3 as explanatory variable
  \item
    using both H3K4me3 and H3K27me3 as explanatory variables
  \end{itemize}
\item
  inspect summary() function output in each case, which terms are significant
\item
  Is using H3K4me3 and H3K27me3 better than the model with only H3K4me3.
\item
  Plot H3k4me3 vs H3k27me3. Inspect the points that does not
  follow a linear trend. Are they clustered at certain segments
  of the plot. Bonus: Is there any biological or technical interpretation
  for those points ?
\end{enumerate}

\hypertarget{unsupervisedLearning}{%
\chapter{Exploratory Data Analysis with Unsupervised Machine Learning}\label{unsupervisedLearning}}

In this chapter, we will focus on using some of the machine learning techniques to explore genomics data. The goals of data exploration is usually many. Generally, we want to understand how the variables in our data set relate to each other and how the samples defined by those variables relate to each other. These points of information can be used to generate hypothesis, find outliers in the samples or identify sample groups that need more data points. In this chapter, we will focus on two main classes of techniques: ``clustering'' and ``dimension reduction''. We will show how to use these techniques and how to visualize them using R.

\hypertarget{clustering-grouping-samples-based-on-their-similarity}{%
\section{Clustering: grouping samples based on their similarity}\label{clustering-grouping-samples-based-on-their-similarity}}

In genomics, we would very frequently want to assess how our samples relate to each other. Are our replicates similar to each other? Do the samples from the same treatment group have the similar genome-wide signals ? Do the patients with similar diseases have similar gene expression profiles ?
Take the last question for example. We need to define a distance or similarity metric between patients' expression profiles and use that metric to find groups of patients that are more similar to each other than the rest of the patients. This, in essence, is the general idea behind clustering. We need a distance metric and a method to utilize that distance metric to find self-similar groups. Clustering is a ubiquitous procedure in bioinformatics as well as any field that deals with high-dimensional data. It is very likely every genomics paper containing multiple samples have some sort of clustering. Due to this ubiquity and general usefulness, it is an essential technique to learn.

\hypertarget{distance-metrics}{%
\subsection{Distance metrics}\label{distance-metrics}}

The first required step for clustering is the distance metric. This is simply a measurement of how similar gene expressions to each other are. There are many options for distance metrics and the choice of the metric is quite important for clustering. Consider a simple example where we have four patients and expression of three genes measured. Which patients look similar to each other based on their gene expression profiles ?

\begin{table}

\caption{\label{tab:expTable}Gene expressions from patients}
\centering
\begin{tabular}[t]{l|r|r|r}
\hline
  & IRX4 & OCT4 & PAX6\\
\hline
patient1 & 11 & 10 & 1\\
\hline
patient2 & 13 & 13 & 3\\
\hline
patient3 & 2 & 4 & 10\\
\hline
patient4 & 1 & 3 & 9\\
\hline
\end{tabular}
\end{table}

It may not be obvious from the table at first sight but if we plot the gene expression profile for each patient, we will see that expression profiles of patient 1 and patient 2 is more similar to each other than patient 3 or patient 4.

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/expPlot-1} 

}

\caption{Gene expression values for different patients. Certain patients have similar gene expression values to each other.}\label{fig:expPlot}
\end{figure}

But how can we quantify what see by eye ? A simple metric for distance between gene expression vectors between a given patient pair is the sum of absolute difference between gene expression values This can be formulated as follows: \(d_{AB}={\sum _{i=1}^{n}|e_{Ai}-e_{Bi}|}\), where \(d_{AB}\) is the distance between patient A and B, and \(e_{Ai}\) and \(e_{Bi}\) expression value of the \(i\)th gene for patient A and B. This distance metric is called \textbf{``Manhattan distance''} or \textbf{``L1 norm''}.

Another distance metric using sum of squared distances and taking a square root of resulting value, that can be formulaized as: \(d_{AB}={{\sqrt {\sum _{i=1}^{n}(e_{Ai}-e_{Bi})^{2}}}}\). This distance is called \textbf{``Euclidean Distance''} or \textbf{``L2 norm''}. This is usually the default distance metric for many clustering algorithms. due to squaring operation values that are very different get higher contribution to the distance. Due to this, compared to Manhattan distance it can be more affected by outliers but generally if the outliers are rare this distance metric works well.

The last metric we will introduce is the \textbf{``correlation distance''}. This is simply \(d_{AB}=1-\rho\), where \(\rho\) is the pearson correlation coefficient between two vectors, in our case those vectors are gene expression profiles of patients. Using this distance the gene expression vectors that have a similar pattern will have a small distance whereas when the vectors have different patterns they will have a large distance. In this case, the linear correlation between vectors matters, the the scale of the vectors might be different.

Now let's see how we can calculate these distance in R. First, we have our gene expression per patient table.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{df}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##          IRX4 OCT4 PAX6
## patient1   11   10    1
## patient2   13   13    3
## patient3    2    4   10
## patient4    1    3    9
\end{verbatim}

Next, we calculate the distance metrics using \texttt{dist} function and \texttt{1-cor()}.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{dist}\NormalTok{(df,}\DataTypeTok{method=}\StringTok{"manhattan"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##          patient1 patient2 patient3
## patient2        7                  
## patient3       24       27         
## patient4       25       28        3
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{dist}\NormalTok{(df,}\DataTypeTok{method=}\StringTok{"euclidean"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##          patient1 patient2 patient3
## patient2    4.123                  
## patient3   14.071   15.843         
## patient4   14.595   16.733    1.732
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{as.dist}\NormalTok{(}\DecValTok{1}\OperatorTok{-}\KeywordTok{cor}\NormalTok{(}\KeywordTok{t}\NormalTok{(df)))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##          patient1 patient2 patient3
## patient2 0.004129                  
## patient3 1.988522 1.970725         
## patient4 1.988522 1.970725 0.000000
\end{verbatim}

\hypertarget{scaling-before-calculating-the-distance}{%
\subsubsection{Scaling before calculating the distance}\label{scaling-before-calculating-the-distance}}

Before we proceed to the clustering, one more thing we need to take care. Should we normalize our data ? Scale of the vectors in our expression matrix can affect the distance calculation. Gene expression tables are usually have some sort of normalization, so the values are in comparable scales. But somehow if a gene's expression values were on much higher scale than the other genes, that gene will effect the distance more than other when using Euclidean or Manhattan distance. If that is the case we can scale the variables.The traditional way of scaling variables is to subtract their mean, and divide by their standard deviation, this operation is also called ``standardization''. If this is done on all genes, each gene will have the same affect on distance measures. The decision to apply scaling ultimately depends on our data and what you want to achieve. If the gene expression values are previously normalized between patients, having genes that dominate the distance metric could have a biological meaning and therefore it may not be desireable to further scale variables. In R, the standardization is done via \texttt{scale()} function. Here we scale the gene expression values.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{df}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##          IRX4 OCT4 PAX6
## patient1   11   10    1
## patient2   13   13    3
## patient3    2    4   10
## patient4    1    3    9
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{scale}\NormalTok{(df)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##             IRX4    OCT4    PAX6
## patient1  0.6933  0.5213 -1.0734
## patient2  1.0195  1.1468 -0.6214
## patient3 -0.7748 -0.7298  0.9604
## patient4 -0.9379 -0.9383  0.7344
## attr(,"scaled:center")
## IRX4 OCT4 PAX6 
## 6.75 7.50 5.75 
## attr(,"scaled:scale")
##  IRX4  OCT4  PAX6 
## 6.131 4.796 4.425
\end{verbatim}

\hypertarget{hiearchical-clustering}{%
\subsection{Hiearchical clustering}\label{hiearchical-clustering}}

This is one of the most ubiqutous clustering algorithms. Using this algorithm you can see the relationship of individual data points and relationships of clusters. This is achieved successively joining small clusters to each other based on the intercluster distance. Eventually, you get a tree structure or a dendrogram that shows the relationship between the individual data points and clusters. The height of the dendrogram is the distance between clusters. Here we can show how to use this on our toy data set from four patients. The base function in R to do hierarchical clustering in \texttt{hclust()}. Below, we apply that function on Euclidean distances between patients.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{d=}\KeywordTok{dist}\NormalTok{(df)}
\NormalTok{hc=}\KeywordTok{hclust}\NormalTok{(d,}\DataTypeTok{method=}\StringTok{"complete"}\NormalTok{)}
\KeywordTok{plot}\NormalTok{(hc)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/toyClust-1} 

}

\caption{Dendrogram of distance matrix}\label{fig:toyClust}
\end{figure}

In the above code snippet, we have used \texttt{method="complete"} argument without explaining it. The \texttt{method} argument defines the criteria that directs how the sub-clusters are merged. During clustering starting with single-member clusters, the clusters are merged based on the distance between them. There are many different ways to define distance between clusters and based on which definition you use the hierarchical clustering results change. So the \texttt{method} argument controls that. There are a couple of values this argument can take, we list them and their description below:

\begin{itemize}
\tightlist
\item
  \textbf{``complete''} stands for ``Complete Linkage'' and the distance between two clusters is defined as largest distance between any members of the two clusters.
\item
  \textbf{``single''} stands for ``Single Linkage'' and the distance between two clusters is defined as smallest distance between any members of the two clusters.
\item
  \textbf{``average''} stands for ``Average Linkage'' or more precisely UPGMA (Unweighted Pair Group Method with Arithmetic Mean) method. In this case, the distance between two clusters is defined as average distance between any members of the two clusters.
\item
  \textbf{``ward.D2''} and \textbf{``ward.D''} stands for different implementations of Ward's minimum variance method. This method aims to find compact, spherical clusters by selecting clusters to merge based on the change in the cluster variances. The clusters are merged if the increase in the combined variance over the sum of the cluster specific variances is minimum compared to alternative merging operations.
\end{itemize}

In real life, we would get expression profiles from thousands of genes and we will typically have many more patients than our toy example. One such data set is gene expression values from 60 bone marrow samples of patients with one of the four main types of leukemia (ALL, AML, CLL, CML) or no-leukemia controls. We trimmed that data set down to top 1000 most variable genes to be able to work with it easier and in addition genes that are not very variable do not contribute much to the distances between patients. We will now use this data set to cluster the patients and display the values as a heatmap and a dendrogram. The heatmap shows the expression values of genes across patients in a color coded manner. The heatmap function, \texttt{pheatmap()}, we will use performs the clustering as well. The matrix that contains gene expressions has the genes in the rows and the patients in the columns. Therefore, we will also use a column-side color code to mark the patients based on their leukemia type. For the hierarchical clustering, we will use Ward's method designated by \texttt{clustering\_method} argument to \texttt{pheatmap()} function.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(pheatmap)}
\NormalTok{expFile=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}\StringTok{"leukemiaExpressionSubset.rds"}\NormalTok{,}\DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{mat=}\KeywordTok{readRDS}\NormalTok{(expFile)}

\CommentTok{# set the leukemia type annotation for each sample}
\NormalTok{annotation_col =}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}
                    \DataTypeTok{LeukemiaType =}\KeywordTok{substr}\NormalTok{(}\KeywordTok{colnames}\NormalTok{(mat),}\DecValTok{1}\NormalTok{,}\DecValTok{3}\NormalTok{))}
\KeywordTok{rownames}\NormalTok{(annotation_col)=}\KeywordTok{colnames}\NormalTok{(mat)}
  

\KeywordTok{pheatmap}\NormalTok{(mat,}\DataTypeTok{show_rownames=}\OtherTok{FALSE}\NormalTok{,}\DataTypeTok{show_colnames=}\OtherTok{FALSE}\NormalTok{,}\DataTypeTok{annotation_col=}\NormalTok{annotation_col,}\DataTypeTok{scale =} \StringTok{"none"}\NormalTok{,}\DataTypeTok{clustering_method=}\StringTok{"ward.D2"}\NormalTok{,}\DataTypeTok{clustering_distance_cols=}\StringTok{"euclidean"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/heatmap1-1} \end{center}

As we can observe in the heatmap each cluster has a distinct set of expression values. The main clusters almost perfectly distinguish the leukemia types. Only one CML patient is clustered as a non-leukemia sample. This could mean that gene expression profiles are enough to classify leukemia type. More detailed analysis and experiments are needed to verify that but by looking at this exploratory analysis we can decide where to focus our efforts next.

\hypertarget{where-to-cut-the-tree}{%
\subsubsection{where to cut the tree ?}\label{where-to-cut-the-tree}}

The example above seems like a clear cut example where we can pick by eye clusters from the dendrogram. This is mostly due to the Ward's method where compact clusters are preffered. However, as it is usually the case we do not have patient labels and it would be difficult to tell which leaves (patients) in the dendrogram we should consider as part of the same cluster. In other words, how deep we should cut the dendrogram so that every patient sample still connected via the remaining sub-dendrograms constitute clusters. The \texttt{cutree()} function provides the functionality to output either desired number of clusters or clusters obtained from cutting the dendrogram at a certain height. Below, we will cluster the patients with hierarchical clustering using the default method ``complete linkage'' and cut the dendrogram at a certain height. In this case, you will also observe that, changing from Ward's distance to complete linkage had an effect on clustering. Now two clusters that are defined by Ward's distance are closer to each other and harder to separate from each other.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{hcl=}\KeywordTok{hclust}\NormalTok{(}\KeywordTok{dist}\NormalTok{(}\KeywordTok{t}\NormalTok{(mat)))}
\KeywordTok{plot}\NormalTok{(hcl,}\DataTypeTok{labels =} \OtherTok{FALSE}\NormalTok{, }\DataTypeTok{hang=} \DecValTok{-1}\NormalTok{)}
\KeywordTok{rect.hclust}\NormalTok{(hcl, }\DataTypeTok{h =} \DecValTok{80}\NormalTok{, }\DataTypeTok{border =} \StringTok{"red"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/hclustNcut-1} \end{center}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{clu.k5=}\KeywordTok{cutree}\NormalTok{(hcl,}\DataTypeTok{k=}\DecValTok{5}\NormalTok{) }\CommentTok{# cut tree so that there are 4 clusters}

\NormalTok{clu.h80=}\KeywordTok{cutree}\NormalTok{(hcl,}\DataTypeTok{h=}\DecValTok{80}\NormalTok{) }\CommentTok{# cut tree/dendrogram from height 80}
\KeywordTok{table}\NormalTok{(clu.k5) }\CommentTok{# number of samples for each cluster}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## clu.k5
##  1  2  3  4  5 
## 12  3  9 12 24
\end{verbatim}

Apart from the arbitrary values for the height or the number of the clusters, how can we define clusters more systematically? As this is a general question, we will show later how to decide the optimal number of clusters later in this chapter.

\hypertarget{k-means-clustering}{%
\subsection{K-means clustering}\label{k-means-clustering}}

Another, very common clustering algorithm is k-means.This method divides or partitions the data points, our working example patients, into a pre-determined, ``k'' number of clusters. Hence, this type of methods are generally called ``partioning'' methods. The algorithm is initialized with randomly choosen \(k\) centers or centroids. In a sense, a centroid is a data point with multiple values. In our working example, it is a hypothetical patient with gene expression values. But in the initialization phase, those gene expression values are choosen randomly within the boundaries of the gene expression distributions from real patients. As the next step in the algorithm, each patient is assigned to the closest centroid and in the next iteration centroids are set to the mean of values of the genes in the cluster. This process of setting centroids and assigning patients to the clusters repeats itself until sum of squared distances to cluster centroids is minimized.

As you might see, the cluster algorithm starts with random initial centroids. This feature might yield different results for each run of the algorithm. We will know show how to use k-means method on the gene expression data set. We will use \texttt{set.seed()} for reproducbility. In the wild, you might want to run this algorithm multiple times to see if your clustering results are stable.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{101}\NormalTok{)}

\CommentTok{# we have to transpore the matrix t()}
\CommentTok{# so that we calculate distances between patients}
\NormalTok{kclu=}\KeywordTok{kmeans}\NormalTok{(}\KeywordTok{t}\NormalTok{(mat),}\DataTypeTok{centers=}\DecValTok{5}\NormalTok{)  }

\CommentTok{# number of data points in each cluster}
\KeywordTok{table}\NormalTok{(kclu}\OperatorTok{$}\NormalTok{cluster)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
##  1  2  3  4  5 
## 12 12 14 11 11
\end{verbatim}

Now let us check the percentage of each leukemia type in each cluster. We can visualize this as a table. Looking at the table below, we see that each of the 5 clusters are predominantly representing one of the 4 leukemia types or the control patients without leukemia.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{type2kclu =}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}
                    \DataTypeTok{LeukemiaType =}\KeywordTok{substr}\NormalTok{(}\KeywordTok{colnames}\NormalTok{(mat),}\DecValTok{1}\NormalTok{,}\DecValTok{3}\NormalTok{),}
                    \DataTypeTok{cluster=}\NormalTok{kclu}\OperatorTok{$}\NormalTok{cluster)}

\KeywordTok{table}\NormalTok{(type2kclu)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##             cluster
## LeukemiaType  1  2  3  4  5
##          ALL 12  0  0  0  0
##          AML  0  0  1  0 11
##          CLL  0 12  0  0  0
##          CML  0  0  1 11  0
##          NoL  0  0 12  0  0
\end{verbatim}

Another related and maybe more robust algorithm is called \textbf{``k-medoids''} clustering. The procedure is almost identical to k-means clustering with a couple of differences. In this case, centroids choosen are real data points in our case patients, and the metric we are trying to optimize in each iteration is based on manhattan distance to the centroid. In k-means this was based on sum of squared distances so euclidean distance. Below we are showing how to use k-medoids clustering function \texttt{pam()} from the \texttt{cluster} package.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{kmclu=cluster}\OperatorTok{::}\KeywordTok{pam}\NormalTok{(}\KeywordTok{t}\NormalTok{(mat),}\DataTypeTok{k=}\DecValTok{5}\NormalTok{) }\CommentTok{#  cluster using k-medoids}

\CommentTok{# make a data frame with Leukemia type and cluster id}
\NormalTok{type2kmclu =}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}
                    \DataTypeTok{LeukemiaType =}\KeywordTok{substr}\NormalTok{(}\KeywordTok{colnames}\NormalTok{(mat),}\DecValTok{1}\NormalTok{,}\DecValTok{3}\NormalTok{),}
                    \DataTypeTok{cluster=}\NormalTok{kmclu}\OperatorTok{$}\NormalTok{cluster)}

\KeywordTok{table}\NormalTok{(type2kmclu)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##             cluster
## LeukemiaType  1  2  3  4  5
##          ALL 12  0  0  0  0
##          AML  0 10  1  1  0
##          CLL  0  0  0  0 12
##          CML  0  0  0 12  0
##          NoL  0  0 12  0  0
\end{verbatim}

We can not visualize the clustering from partioning methods with a tree like we did for hierarchical clustering. Even if we can get the distances between patients the algorithm does not return the distances between clusters out of the box. However, if we had a way to visualize the distances between patients in 2 dimensions we could see the how patients and clusters relate each other. It turns out, that there is a way to compress between patient distances to a 2-dimensional plot. There are many ways to do this and we introduce these dimension reduction methods including the one we will use now later in this chapter. For now, we are going to use a method called ``multi-dimensional scaling'' and plot the patients in a 2D plot color coded by their cluster assignments.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Calculate distances}
\NormalTok{dists=}\KeywordTok{dist}\NormalTok{(}\KeywordTok{t}\NormalTok{(mat))}

\CommentTok{# calculate MDS}
\NormalTok{mds=}\KeywordTok{cmdscale}\NormalTok{(dists)}

\CommentTok{# plot the patients in the 2D space}
\KeywordTok{plot}\NormalTok{(mds,}\DataTypeTok{pch=}\DecValTok{19}\NormalTok{,}\DataTypeTok{col=}\KeywordTok{rainbow}\NormalTok{(}\DecValTok{5}\NormalTok{)[kclu}\OperatorTok{$}\NormalTok{cluster])}

\CommentTok{# set the legend for cluster colors}
\KeywordTok{legend}\NormalTok{(}\StringTok{"bottomright"}\NormalTok{,}
       \DataTypeTok{legend=}\KeywordTok{paste}\NormalTok{(}\StringTok{"clu"}\NormalTok{,}\KeywordTok{unique}\NormalTok{(kclu}\OperatorTok{$}\NormalTok{cluster)),}
       \DataTypeTok{fill=}\KeywordTok{rainbow}\NormalTok{(}\DecValTok{5}\NormalTok{)[}\KeywordTok{unique}\NormalTok{(kclu}\OperatorTok{$}\NormalTok{cluster)],}
       \DataTypeTok{border=}\OtherTok{NA}\NormalTok{,}\DataTypeTok{box.col=}\OtherTok{NA}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/kmeansmds-1} \end{center}

The plot we obtained shows the separetion between clusters. However, it does not do a great job showing the separation between cluster 3 and 4, which represent CML and ``no leukemia'' patients. We might need another dimension to properly visualize that separation. In addition, those two clusters were closely related in the hierarhical clustering as well.

\hypertarget{how-to-choose-k-the-number-of-clusters}{%
\subsection{how to choose ``k'', the number of clusters}\label{how-to-choose-k-the-number-of-clusters}}

Up to this point, we have avoided the question of selecting optimal number clusters. How do we know where to cut our dendrogram or which k to choose ?
First of all, this is a difficult question. Usually, clusters have different granuality. Some clusters are tight and compact and some are wide,and both these types of clusters can be in the same data set. When visualized, some large clusters may look like they may have sub-clusters. So should we consider the large cluster as one cluster or should we consider the sub-clusters as individual clusters ? There are some metrics to help but there is no definite answer. We will show a couple of them below.

\hypertarget{silhouhette}{%
\subsubsection{Silhouhette}\label{silhouhette}}

One way to determine how well the clustering is to measure the expected self-similar nature of the points in a set of clusters. The silhouette value does just that and it is a measure of how similar a data point is to its own cluster compared to other clusters. The silhouette value ranges from -1 to +1, where values that are positive indicates that the data point is well matched to its own cluster, if the value is zero it is a borderline case and if the value is minus it means that the data point might be mis-clustered because it is more simialr to a neighboring cluster. If most data points have a high value, then the clustering is appropriate. Ideally, one can create many different clusterings with different parameters such as \(k\),number of clusters and assess their appropriateness using the average
silhouette values. In R, silhouette values are refered to as silhouette widths in the documentation.

A silhouette value is calculated for each data point. In our working example, each patient will get silhouette values showing how well they are matched to their assigned clusters. Formally this calculated as follows. For each data point \(i\), we calculate \({\displaystyle a(i)}\), which denotes the average distance between \(i\) and all other data points within the same cluster. This shows how well the point fits into that cluster. For the same data point, we also calculate \({\displaystyle b(i)}\) b(i) denotes the lowest average distance of \({\displaystyle i}\) to all points in any other cluster, of which \({\displaystyle i}\) is not a member. The cluster with this lowest average \(b(i)\) is the ``neighbouring cluster'' of data point \({\displaystyle i}\) since it is the next best fit cluster for that data point. Then, the silhouette value for a given data point is:

\(s(i) = \frac{b(i) - a(i)}{\max\{a(i),b(i)\}}\)

As described, this quantity is positive when \(b(i)\) is high and \(a(i)\) is low, meaning that the data point \(i\) is self-similar to its cluster. And the silhouette value, \(s(i)\), is negative if it is more similar to its neighbours than its assigned cluster.

In R, we can calculate silhouette values using \texttt{cluster::silhouette()} function. Below, we calculate the silhouette values for k-medoids clustering with \texttt{pam()} function with \texttt{k=5}.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(cluster)}
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{101}\NormalTok{)}
\NormalTok{pamclu=cluster}\OperatorTok{::}\KeywordTok{pam}\NormalTok{(}\KeywordTok{t}\NormalTok{(mat),}\DataTypeTok{k=}\DecValTok{5}\NormalTok{)}
\KeywordTok{plot}\NormalTok{(}\KeywordTok{silhouette}\NormalTok{(pamclu),}\DataTypeTok{main=}\OtherTok{NULL}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/sill-1} \end{center}

Now, let us calculate average silhouette value different \(k\) values and compare. We will use \texttt{sapply()} function to get average silhouette values accross \(k\) values between 2 and 7. Within \texttt{sapply()} there is an anonymous function that that does the clustering and calculates average silhouette values for each \(k\).

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{Ks=}\KeywordTok{sapply}\NormalTok{(}\DecValTok{2}\OperatorTok{:}\DecValTok{7}\NormalTok{,}
    \ControlFlowTok{function}\NormalTok{(i) }
      \KeywordTok{summary}\NormalTok{(}\KeywordTok{silhouette}\NormalTok{(}\KeywordTok{pam}\NormalTok{(}\KeywordTok{t}\NormalTok{(mat),}\DataTypeTok{k=}\NormalTok{i)))}\OperatorTok{$}\NormalTok{avg.width)}
\KeywordTok{plot}\NormalTok{(}\DecValTok{2}\OperatorTok{:}\DecValTok{7}\NormalTok{,Ks,}\DataTypeTok{xlab=}\StringTok{"k"}\NormalTok{,}\DataTypeTok{ylab=}\StringTok{"av. silhouette"}\NormalTok{,}\DataTypeTok{type=}\StringTok{"b"}\NormalTok{,}
     \DataTypeTok{pch=}\DecValTok{19}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/sillav-1} \end{center}

In this case, it seems the best value for \(k\) is 4. The k-medoids function \texttt{pam()} will usually cluster CML and noLeukemia cases together when \texttt{k=4}, which are also related clusters according to hierarchical clustering we did earlier.

\hypertarget{gap-statistic}{%
\subsubsection{Gap statistic}\label{gap-statistic}}

As clustering aims to find self-similar data points, it would be reasonable to expect with the correct number of clusters the total within-cluster variation is minimized. Within-cluster variation for a single cluster can simply be defined as sum of squares from the cluster mean, which in this case is the centroid we defined in k-means algorithm. The total within-cluster variation is then sum of within-cluster variations for each cluster. This can be formally defined as follows:

\(\displaystyle W_k = \sum_{k=1}^K \sum_{\mathrm{x}_i \in C_k} (\mathrm{x}_i - \mu_k )^2\)

Where \(\mathrm{x}_i\) is data point in cluster \(k\), and \(\mu_k\) is the cluster mean, and \(W_k\) is the total within-cluster variation quantity we described. However, the problem is that the variation quantity decreases with number of clusters. The more centroids we have, the smaller the distances to the centroids get. A more reliable approach would be somehow calculating the expected variation from a reference null distribution and compare that to the observed variation for each \(k\). In gap statistic approach, the expected distribution is calculated via sampling points from the boundaries of the original data and calculating within-cluster variation quantity for multiple rounds of sampling. This way we have an expectation how about the variability when there is no expected clustering, and then compare that expected variation to the observed within-cluster variation. The expected variation should also go down with increasing number of clusters, but for the optimal number of clusters the expected variation will be furthest away from observed variation. This distance is called the \textbf{``gap statistic''} and defined as follows:
\(\displaystyle \mathrm{Gap}_n(k) = E_n^*\{\log W_k\} - \log W_k\), where \(E_n^*\{\log W_k\}\) is the expected variation in log-scale under a sample size \(n\) from the reference distribution and \(\log W_k\) is the observed variation. Our aim is choose the \(k\), number of clusters, that maximizes \(\mathrm{Gap}_n(k)\).

We can easily calculate the gap statistic with \texttt{cluster::clusGap()} function. We will now use that function to calculate the gap statistic for our patient gene expression data.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(cluster)}
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{101}\NormalTok{)}
\CommentTok{# define the clustering function}
\NormalTok{pam1 <-}\StringTok{ }\ControlFlowTok{function}\NormalTok{(x,k) }
  \KeywordTok{list}\NormalTok{(}\DataTypeTok{cluster =} \KeywordTok{pam}\NormalTok{(x,k, }\DataTypeTok{cluster.only=}\OtherTok{TRUE}\NormalTok{))}

\CommentTok{# calculate the gap statistic}
\NormalTok{pam.gap=}\StringTok{ }\KeywordTok{clusGap}\NormalTok{(}\KeywordTok{t}\NormalTok{(mat), }\DataTypeTok{FUN =}\NormalTok{ pam1, }\DataTypeTok{K.max =} \DecValTok{8}\NormalTok{,}\DataTypeTok{B=}\DecValTok{50}\NormalTok{)}

\CommentTok{# plot the gap statistic accross k values}
\KeywordTok{plot}\NormalTok{(pam.gap, }\DataTypeTok{main =} \StringTok{"Gap statistic for the 'Leukemia' data"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/clusGap-1} \end{center}

In this case, gap statistic shows \(k=7\) is the best. However, after \(K=6\) the statistic has more or less a stable curve. In this case, we know that there are 5 main patient categories but this does not mean there is no sub-categories or sub-types for the cancers we are looking at.

\url{https://statweb.stanford.edu/~gwalther/gap}

\hypertarget{other-methods}{%
\subsubsection{Other methods}\label{other-methods}}

There are several other methods that provide insight into how many clusters. In fact, the package \texttt{NbClust} provides 30 different ways to determine the number of optimal clusters and can offer a voting mechanism to pick the best number. Below, we are showing how to use this function for some of the optimal number of cluster detection methods.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(NbClust)}
\NormalTok{nb =}\StringTok{ }\KeywordTok{NbClust}\NormalTok{(}\DataTypeTok{data=}\KeywordTok{t}\NormalTok{(mat), }
             \DataTypeTok{distance =} \StringTok{"euclidean"}\NormalTok{, }\DataTypeTok{min.nc =} \DecValTok{2}\NormalTok{,}
        \DataTypeTok{max.nc =} \DecValTok{7}\NormalTok{, }\DataTypeTok{method =} \StringTok{"kmeans"}\NormalTok{,}
        \DataTypeTok{index=}\KeywordTok{c}\NormalTok{(}\StringTok{"kl"}\NormalTok{,}\StringTok{"ch"}\NormalTok{,}\StringTok{"cindex"}\NormalTok{,}\StringTok{"db"}\NormalTok{,}\StringTok{"silhouette"}\NormalTok{,}
                \StringTok{"duda"}\NormalTok{,}\StringTok{"pseudot2"}\NormalTok{,}\StringTok{"beale"}\NormalTok{,}\StringTok{"ratkowsky"}\NormalTok{,}
                \StringTok{"gap"}\NormalTok{,}\StringTok{"gamma"}\NormalTok{,}\StringTok{"mcclain"}\NormalTok{,}\StringTok{"gplus"}\NormalTok{,}
                \StringTok{"tau"}\NormalTok{,}\StringTok{"sdindex"}\NormalTok{,}\StringTok{"sdbw"}\NormalTok{))}

\KeywordTok{table}\NormalTok{(nb}\OperatorTok{$}\NormalTok{Best.nc[}\DecValTok{1}\NormalTok{,]) }\CommentTok{# consensus seems to be 3 clusters }
\end{Highlighting}
\end{Shaded}

However, the readers should keep in mind that clustering is an exploratory technique. If you have solid labels for your data points maybe clustering is just a sanity check, and you should just do predictive modeling instead. However, in biology there are rarely solid labels and things have different granularity. Take the leukemia patients case we have been using for example, it is know that leukemia types have subtypes and those sub-types that have different mutation profiles and consequently have different molecular signatures. Because of this, it is not surprising that some optimal cluster number techniques will find more clusters to be appropriate. On the other hand, CML (Chronic myeloid leukemia ) is a slow progressing disease and maybe as molecular signatures goes could be the closest to no leukemia patients, clustering algorithms may confuse the two depending on what granuality they are operating with. It is always good to look at the heatmaps after clustering, if you have meaningful self-similar data points even if the labels you have do not agree that there can be different clusters you can perform downstream analysis to understand the sub-clusters better. As we have seen, we can estimate optimal number of clusters but we can not take that estimation as the absolute truth, given more data points or different set of expression signatures you may have different optimal clusterings, or the supposed optimal clustering might overlook previously known sub-groups of your data.

\hypertarget{dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d}{%
\section{Dimensionality reduction techniques: visualizing complex data sets in 2D}\label{dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d}}

In statistics, dimension reduction techniques are a set of processes for reducing the number of random variables by obtaining a set of principal variables. For example, in the context of a gene expression matrix accross different patient samples, this might mean getting a set of new variables that cover the variation in sets of genes. This way samples can be represented by a couple of principal variables instead of thousands of genes. This is useful for visualization, clustering and predictive modeling.

\hypertarget{principal-component-analysis}{%
\subsection{Principal component analysis}\label{principal-component-analysis}}

Principal component analysis (PCA) is maybe the most popular technique to examine high-dimensional data. There are multiple interpretations of how PCA reduces dimensionality. We will first focus on geometrical interpretation, where this operation can be interpreted as rotating the orignal dimensions of the data. For this, we go back to our example gene expression data set. In this example, we will represent our patients with expression profiles of just two genes, CD33 (ENSG00000105383) and PYGL (ENSG00000100504) genes. This way we can visualize them in a scatterplot.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{plot}\NormalTok{(mat[}\KeywordTok{rownames}\NormalTok{(mat)}\OperatorTok{==}\StringTok{"ENSG00000100504"}\NormalTok{,],}
\NormalTok{     mat[}\KeywordTok{rownames}\NormalTok{(mat)}\OperatorTok{==}\StringTok{"ENSG00000105383"}\NormalTok{,],}\DataTypeTok{pch=}\DecValTok{19}\NormalTok{,}
     \DataTypeTok{ylab=}\StringTok{"CD33 (ENSG00000105383)"}\NormalTok{,}
     \DataTypeTok{xlab=}\StringTok{"PYGL (ENSG00000100504)"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-163-1} \end{center}

PCA rotates the original data space such that the axes of the new coordinate system point into the directions of highest variance of the data. The axes or new variables are termed principal components (PCs) and are ordered by variance: The first component, PC 1, represents the direction of the highest variance of the data. The direction of the second component, PC 2, represents the highest of the remaining variance orthogonal to the first component. This can be naturally extended to obtain the required number of components which together span a component space covering the desired amount of variance. In our toy example with only two genes, the principal componets are drawn over the original scatter plot and in the next plot we show the new coordinate system based on the pricinpal components. We will calculate the PCA with the \texttt{princomp()} function, this function returns the new coordinates as well. These new coordinates are simply a projection of data over the new coordinates. We will decorate the scatter plots with eigenvectors showing the direction of greatest variation. Then, we will plot the new coordinates. These are automatically calculated by \texttt{princomp()} function. Notice that we are using the \texttt{scale()} function when plotting coordinates and also before calculating PCA. This function centers the data, meaning substracts the mean of the each column vector from the elements in the vector. This essentially gives the columns a zero mean. It also divides the data by the standard deviation of the centered columns. These two operations helps bring the data to a common scale which is important for PCA not to be affected by different scales in the data.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}

\CommentTok{# create the subset of the data with two genes only}
\CommentTok{# notice that we transpose the matrix so samples are }
\CommentTok{# on the columns}
\NormalTok{sub.mat=}\KeywordTok{t}\NormalTok{(mat[}\KeywordTok{rownames}\NormalTok{(mat) }\OperatorTok{%in%}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"ENSG00000100504"}\NormalTok{,}\StringTok{"ENSG00000105383"}\NormalTok{),])}

\CommentTok{# ploting our genes of interest as scatter plots}
\KeywordTok{plot}\NormalTok{(}\KeywordTok{scale}\NormalTok{(mat[}\KeywordTok{rownames}\NormalTok{(mat)}\OperatorTok{==}\StringTok{"ENSG00000100504"}\NormalTok{,]),}
     \KeywordTok{scale}\NormalTok{(mat[}\KeywordTok{rownames}\NormalTok{(mat)}\OperatorTok{==}\StringTok{"ENSG00000105383"}\NormalTok{,]),}
     \DataTypeTok{pch=}\DecValTok{19}\NormalTok{,}
     \DataTypeTok{ylab=}\StringTok{"CD33 (ENSG00000105383)"}\NormalTok{,}
     \DataTypeTok{xlab=}\StringTok{"PYGL (ENSG00000100504)"}\NormalTok{,}
     \DataTypeTok{col=}\NormalTok{annotation_col}\OperatorTok{$}\NormalTok{LeukemiaType,}
     \DataTypeTok{xlim=}\KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{2}\NormalTok{,}\DecValTok{2}\NormalTok{),}\DataTypeTok{ylim=}\KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{2}\NormalTok{,}\DecValTok{2}\NormalTok{))}

\CommentTok{# create the legend for the Leukemia types}
\KeywordTok{legend}\NormalTok{(}\StringTok{"bottomright"}\NormalTok{,}
       \DataTypeTok{legend=}\KeywordTok{unique}\NormalTok{(annotation_col}\OperatorTok{$}\NormalTok{LeukemiaType),}
       \DataTypeTok{fill =}\KeywordTok{palette}\NormalTok{(}\StringTok{"default"}\NormalTok{),}
       \DataTypeTok{border=}\OtherTok{NA}\NormalTok{,}\DataTypeTok{box.col=}\OtherTok{NA}\NormalTok{)}

\CommentTok{# calculate the PCA only for our genes and all the samples}
\NormalTok{pr=}\KeywordTok{princomp}\NormalTok{(}\KeywordTok{scale}\NormalTok{(sub.mat))}


\CommentTok{# plot the direction of eigenvectors}
\CommentTok{# pr$loadings returned by princomp has the eigenvectors}
\KeywordTok{arrows}\NormalTok{(}\DataTypeTok{x0=}\DecValTok{0}\NormalTok{, }\DataTypeTok{y0=}\DecValTok{0}\NormalTok{, }\DataTypeTok{x1 =}\NormalTok{ pr}\OperatorTok{$}\NormalTok{loadings[}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{], }
         \DataTypeTok{y1 =}\NormalTok{ pr}\OperatorTok{$}\NormalTok{loadings[}\DecValTok{2}\NormalTok{,}\DecValTok{1}\NormalTok{],}\DataTypeTok{col=}\StringTok{"pink"}\NormalTok{,}\DataTypeTok{lwd=}\DecValTok{3}\NormalTok{)}
\KeywordTok{arrows}\NormalTok{(}\DataTypeTok{x0=}\DecValTok{0}\NormalTok{, }\DataTypeTok{y0=}\DecValTok{0}\NormalTok{, }\DataTypeTok{x1 =}\NormalTok{ pr}\OperatorTok{$}\NormalTok{loadings[}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{], }
         \DataTypeTok{y1 =}\NormalTok{ pr}\OperatorTok{$}\NormalTok{loadings[}\DecValTok{2}\NormalTok{,}\DecValTok{2}\NormalTok{],}\DataTypeTok{col=}\StringTok{"gray"}\NormalTok{,}\DataTypeTok{lwd=}\DecValTok{3}\NormalTok{)}


\CommentTok{# plot the samples in the new coordinate system}
\KeywordTok{plot}\NormalTok{(}\OperatorTok{-}\NormalTok{pr}\OperatorTok{$}\NormalTok{scores,}\DataTypeTok{pch=}\DecValTok{19}\NormalTok{,}
     \DataTypeTok{col=}\NormalTok{annotation_col}\OperatorTok{$}\NormalTok{LeukemiaType,}
     \DataTypeTok{ylim=}\KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{2}\NormalTok{,}\DecValTok{2}\NormalTok{),}\DataTypeTok{xlim=}\KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{4}\NormalTok{,}\DecValTok{4}\NormalTok{))}

\CommentTok{# plot the new coordinate basis vectors}
\KeywordTok{arrows}\NormalTok{(}\DataTypeTok{x0=}\DecValTok{0}\NormalTok{, }\DataTypeTok{y0=}\DecValTok{0}\NormalTok{, }\DataTypeTok{x1 =}\OperatorTok{-}\DecValTok{2}\NormalTok{, }
         \DataTypeTok{y1 =} \DecValTok{0}\NormalTok{,}\DataTypeTok{col=}\StringTok{"pink"}\NormalTok{,}\DataTypeTok{lwd=}\DecValTok{3}\NormalTok{)}
\KeywordTok{arrows}\NormalTok{(}\DataTypeTok{x0=}\DecValTok{0}\NormalTok{, }\DataTypeTok{y0=}\DecValTok{0}\NormalTok{, }\DataTypeTok{x1 =} \DecValTok{0}\NormalTok{, }
         \DataTypeTok{y1 =} \DecValTok{-1}\NormalTok{,}\DataTypeTok{col=}\StringTok{"gray"}\NormalTok{,}\DataTypeTok{lwd=}\DecValTok{3}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/pcaRot-1} 

}

\caption{Geometric interpretation of PCA finding eigenvectors that point to direction of highest variance. Eigenvectors can be used as a new coordinate system.}\label{fig:pcaRot}
\end{figure}

As you can see, the new coordinate system is useful by itself.The X-axis which represents the first component separates the data along the lympoblastic and myeloid leukemias.

PCA in this case is obtained by calculating eigenvectors of the covariance matrix via an operation called eigen decomposition. Covariance matrix is obtained by covariance of pairwise variables of our expression matrix, which is simply \({ \operatorname{cov} (X,Y)={\frac {1}{n}}\sum _{i=1}^{n}(x_{i}-\mu_X)(y_{i}-\mu_Y)}\), where \(X\) and \(Y\) expression values of genes in a sample in our example. This is a measure of how things vary together, if high expressed genes in sample A are also highly expressed in sample B and lowly expressed in sample A are also lowly expressed in sample B, then sample A and B will have positive covariance. If the opposite is true then they will have negative covariance. This quantity is related to correlation and in fact correlation is standardized covariance. Covariance of variables can be obtained with \texttt{cov()} function, and eigen decomposition of such a matrix will produce a set of ortahogonal vectors that span the directions of highest variation. In 2D, you can think of this operation as rotating two perpendicular lines together until they point to the directions where most of the variation in the data lies on, similar to the figure \ref{fig:pcaRot}. An important intuition is that, after the rotation prescribed by eigenvectors is complete the covariance between variables in this rotated dataset will be zero. There is a proper mathematical relationship between covariances of the rotated dataset and the original dataset. That's why operating on covariance matrix is related to the rotation of the original dataset.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cov.mat=}\KeywordTok{cov}\NormalTok{(sub.mat) }\CommentTok{# calculate covariance matrix}
\NormalTok{cov.mat}
\KeywordTok{eigen}\NormalTok{(cov.mat) }\CommentTok{# obtain eigen decomposition for eigen values and vectors}
\end{Highlighting}
\end{Shaded}

Eigenvectors and eigenvalues of the covariance matrix indicates the direction and the magnitute of variation of the data. In our visual example the eigenvectors are so-called principal components. The eigenvector indicates the direction and the eigen values indicate the variation in that direction. Eigenvectors and values exist in pairs: every eigenvector has a corresponding eigenvalue and the eigenvectors are linearly independent from each other, this means they are orthogonal or uncorrelated in the our working example above. The eigenvectors are ranked by their corresponding eigen value, the higher the eigen value the more important the eigenvector is, because it explains more of the variation compared to the other eigenvectors. This feature of PCA makes the dimension reduction possible. We can sometimes display data sets that have many variables only in 2D or 3D because the these top eigenvectors are sometimes enough to capture most of variation in the data.

\hypertarget{singular-value-decomposition-and-principal-component-analysis}{%
\subsubsection{Singular value decomposition and principal component analysis}\label{singular-value-decomposition-and-principal-component-analysis}}

A more common way to calculate PCA is through something called singular value decomposition (SVD). This results in another interpretation of PCA, which is called ``latent factor'' or ``latent component'' interpretation. In a moment, it will be more clear what we mean by ``latent factors''. SVD is a matrix factorization or decomposition algorithm that decomposes an input matrix,\(X\), to three matrices as follows: \(\displaystyle \mathrm{X} = USV^T\). In essence many matrices can be decomposed as a product of multiple matrices and we will come to other techniques later in this chapter. Singular Value Decomposition is shown in figure \ref{fig:SVDcartoon}. \(U\) is the matrix with eigenarrays on the columns and this has the same dimensions as the input matrix, you might see elsewhere the columns are named as eigenassays. \(S\) is the matrix that contain the singular values on the diagonal. The singular values are also known as eigenvalues and their square is proportional to explained variation by each eigenvector. Finally, the matrix \(V^T\) contains the eigenvectors on its rows. It is interpretation is still the same. Geometrically, eigenvectors point to the direction of highest variance in the data. They are uncorrolated or geometrically orthogonal to each other. These interpretations are identical to the ones we made before. The slight difference is that the decomposition seem to output \(V^T\) which is just the transpose of the matrix \(V\). However, the SVD algorithms in R usually return the matrix \(V\). If you want the eigenvectors, you either simply use the columns of matrix \(V\) or rows of \(V^T\).

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{images/SVDcartoon} 

}

\caption{Singular Value Decomposition (SVD) explained in a diagram. }\label{fig:SVDcartoon}
\end{figure}

One thing that is new in the figure \ref{fig:SVDcartoon} is the concept of eigenarrays. The eigenarrays or sometimes called eigenassays reprensent the sample space and can be used to plot the relationship between samples rather than genes. In this way, SVD offers additional information than the PCA using the covariance matrix. It offers us a way to summarize both genes and samples. As we can project the gene expression profiles over the top two eigengenes and get a 2D representation of genes, but with SVD we can also project the samples over the the top two eigenarrays and get a representation of samples in 2D scatterplot. Eigenvector could represent independent expression programs across samples, such as cell-cycle if we had time-based expression profiles. However, there is no guarantee that each eigenvector will be biologically meaningful. Similarly each eigenarray represent samples with specific expression characteristics. For example, the samples that have a particular pathway activated might be corrolated to an eigenarray returned by SVD.

Previously, in order to map samples to the reduced 2D space we had to transpose the genes-by-samples matrix when using \texttt{princomp()} function. We will now first use SVD on genes-by-samples matrix to get eigenarrays and use that to plot samples on the reduced dimensions. We will project the columns in our original expression data on eigenarrays and use the first two dimensions in the scatter plot. If you look at the code you will see that for the projection we use \(U^T X\) operation, which is just \(V^T\) if you follow the linear algebra. We will also perform the PCA this time with \texttt{prcomp()} function on the transposed genes-by-samples matrix to get a similar information, and plot the samples on the reduced coordinates.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\NormalTok{d=}\KeywordTok{svd}\NormalTok{(}\KeywordTok{scale}\NormalTok{(mat)) }\CommentTok{# apply SVD}
\NormalTok{assays=}\KeywordTok{t}\NormalTok{(d}\OperatorTok{$}\NormalTok{u) }\OperatorTok{%*%}\StringTok{ }\KeywordTok{scale}\NormalTok{(mat) }\CommentTok{# projection on eigenassays}
\KeywordTok{plot}\NormalTok{(assays[}\DecValTok{1}\NormalTok{,],assays[}\DecValTok{2}\NormalTok{,],}\DataTypeTok{pch=}\DecValTok{19}\NormalTok{,}
     \DataTypeTok{col=}\NormalTok{annotation_col}\OperatorTok{$}\NormalTok{LeukemiaType)}
\CommentTok{#plot(d$v[,1],d$v[,2],pch=19,}
\CommentTok{#     col=annotation_col$LeukemiaType)}
\NormalTok{pr=}\KeywordTok{prcomp}\NormalTok{(}\KeywordTok{t}\NormalTok{(mat),}\DataTypeTok{center=}\OtherTok{TRUE}\NormalTok{,}\DataTypeTok{scale=}\OtherTok{TRUE}\NormalTok{) }\CommentTok{# apply PCA on transposed matrix}

\CommentTok{# plot new coordinates from PCA, projections on eigenvectors}
\CommentTok{# since the matrix is transposed eigenvectors represent }
\KeywordTok{plot}\NormalTok{(pr}\OperatorTok{$}\NormalTok{x[,}\DecValTok{1}\NormalTok{],pr}\OperatorTok{$}\NormalTok{x[,}\DecValTok{2}\NormalTok{],}\DataTypeTok{col=}\NormalTok{annotation_col}\OperatorTok{$}\NormalTok{LeukemiaType)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.65\linewidth]{compgenomrReloaded_files/figure-latex/svd-1} 

}

\caption{SVD on matrix and its transpose}\label{fig:svd}
\end{figure}

As you can see in the figure \ref{fig:svd}, the two approaches yield separation of samples, although they are slightly different. The difference comes from the centering and scaling. In the first case, we scale and center columns and the second case we scale and center rows since the matrix is transposed. If we do not do any scaling or centering we would get identical plots.

\hypertarget{eigenvectors-as-latent-factorsvariables}{%
\paragraph{Eigenvectors as latent factors/variables}\label{eigenvectors-as-latent-factorsvariables}}

Finally, we can introduce the latent factor interpretation of PCA via SVD. As we have already mentioned eigenvectors can also be interpreted as expression programs that are shared by several genes such as cell cycle expression program when measuring gene expression accross samples taken in different time points. In this intrepretation, linear combination of expression programs makes up the expression profile of the genes. Linear combination simply means multiplying the expression program with a weight and adding them up. Our \(USV^T\) matrix multiplication can be rearranged to yield such an understanding, we can multiply eigenarrays \(U\) with the diagonal eigenvalues \(S\), to produce a m-by-n weights matrix called \(W\), so \(W=US\) and we can re-write the equation as just weights by eigenvectors matrix, \(X=WV^T\) as shown in figure \ref{fig:SVDasWeigths}.

\begin{figure}

{\centering \includegraphics[width=0.7\linewidth]{images/SVDasWeights} 

}

\caption{Singular Value Decomposition (SVD) reorgonized as multiplication of m-by-n weights matrix and eigenvectors }\label{fig:SVDasWeigths}
\end{figure}

This simple transformation now makes it clear that indeed if eigenvectors are representing expression programs, their linear combination is making up individual gene expression profiles. As an example, we can show the liner combination of the first two eigenvectors can approximate the expression profile of an hypothetical gene in the gene expression matrix. The figure \ref{fig:SVDlatentExample} shows eigenvector 1 and eigenvector 2 combined with certain weights in \(W\) matrix can approximate gene expression pattern our example gene.

\begin{figure}

{\centering \includegraphics[width=19.68in]{images/SVDlatentExample} 

}

\caption{Gene expression of a gene can be thought as linear combination of eigenvectors. }\label{fig:SVDlatentExample}
\end{figure}

However, SVD does not care about biology. The eigenvectors are just obtained from the data with constraints of ortagonality and the direction of variation. There are examples of eigenvectors representing
real expression programs but that does not mean eigenvectors will always be biologically meaningful. Sometimes combination of them might make more sense in biology than single eigenvectors. This is also the same for the other matrix factorization techniques we describe below.

\hypertarget{other-dimension-reduction-techniques-using-other-matrix-factorization-methods}{%
\subsection{Other dimension reduction techniques using other matrix factorization methods}\label{other-dimension-reduction-techniques-using-other-matrix-factorization-methods}}

We must mention a few other techniques that are similar to SVD in spirit. Remember we mentioned that every matrix can be decomposed to other matrices where matrix multiplication operations reconstruct the original matrix. In the case of SVD/PCA, the constraint is that eigenvectors/arrays are ortogonal, however there are other decomposition algorithms with other constraints.

\hypertarget{independent-component-analysis-ica}{%
\subsubsection{Independent component analysis (ICA)}\label{independent-component-analysis-ica}}

We will first start with independent component analysis (ICA) which is an extension of PCA. ICA algorithm decomposes a given matrix \(X\) as follows: \(X=SA\). The rows of \(A\) could be interpreted similar to the eigengenes and columns of \(S\) could be interpreted as eigenarrays, these components are sometimes called metagenes and metasamples in the literature. Traditionally, \(S\) is called source matrix and \(A\) is called mixing matrix. ICA is developed for a problem called ``blind-source separation''. In this problem, multiple microphones record sound from multiple instruments, and the task is disentagle sounds from original instruments since each microphone is recording a combination of sounds. In this respect, the matrix \(S\) contains the original signals (sounds from different instruments) and their linear combinations identified by the weights in \(A\), and the product of \(A\) and \(S\) makes up the matrix \(X\), which is the observed signal from different microphones. With this interpretation in mind, if the interest is strictly expression patterns similar that represent the hidden expression programs we see that genes-by-samples matrix is transposed to a samples-by-genes matrix, so that the columns of \(S\) represent these expression patterns , here refered to as ``metagenes'', hopefully representing distinct expression programs (Figure \ref{fig:ICAcartoon} ).

\begin{figure}

{\centering \includegraphics[width=26.24in]{images/ICAcartoon} 

}

\caption{Independent Component Analysis (ICA)}\label{fig:ICAcartoon}
\end{figure}

ICA requires that the columns of \(S\) matrix, the ``metagenes'' in our example above to be statistical independent. This is a stronger constraint than uncorrelatedness. In this case, there should be no relationship between non-linear transformation of the data either. There are different ways of ensuring this statistical indepedence and this is the main constraint when finding the optimal \(A\) and \(S\) matrices. The various ICA algorithms use different proxies for statistical independence, and the definition of that proxy is the main difference between many ICA algorithms. The algorithm we are going to use requires that metagenes or sources in the \(S\) matrix are non-gaussian as possible. Non-gaussianity is shown to be related to statistical independence {[}REF{]}. Below, we are using \texttt{fastICA::fastICA()} function to extract 2 components and plot the rows of matrix \(A\) which represents metagenes. This way, we can visualize samples in a 2D plot. If we wanted to plot the relationship between genes we would use the the columns of matrix \(S\).

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(fastICA)}
\NormalTok{ica.res=}\KeywordTok{fastICA}\NormalTok{(}\KeywordTok{t}\NormalTok{(mat),}\DataTypeTok{n.comp=}\DecValTok{2}\NormalTok{) }\CommentTok{# apply ICA}

\CommentTok{# plot reduced dimensions}
\KeywordTok{plot}\NormalTok{(ica.res}\OperatorTok{$}\NormalTok{S[,}\DecValTok{1}\NormalTok{],ica.res}\OperatorTok{$}\NormalTok{S[,}\DecValTok{2}\NormalTok{],}\DataTypeTok{col=}\NormalTok{annotation_col}\OperatorTok{$}\NormalTok{LeukemiaType)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-165-1} \end{center}

\hypertarget{non-negative-matrix-factorization-nmf}{%
\subsubsection{Non-negative matrix factorization (NMF)}\label{non-negative-matrix-factorization-nmf}}

Non-negative matrix factorization algorithms are series of algorithms that aim to decompose the matrix \(X\) into the product or matrices \(W\) and \(H\), \(X=WH\) (Figure \ref{fig:NMFcartoon}). The constraint is that \(W\) and \(H\) must contain non-negative values, so must \(X\). This is well suited for data sets that can not contain negative values such as gene expression. This also implies addivity of components, in our example expression of a gene across samples are addition of multiple metagenes. Unlike ICA and SVD/PCA, the metagenes can never be combined in subtractive way. In this sense, expression programs potentially captured by metagenes are combined additively.

\begin{figure}

{\centering \includegraphics[width=21.32in]{images/NMFcartoon} 

}

\caption{Non-negative matrix factorization}\label{fig:NMFcartoon}
\end{figure}

The algorithms that compute NMF tries to minimize the cost function \(D(X,WH)\), which is the distance between \(X\) and \(WH\). The early algorithms just use the euclidean distance which translates to \(\sum(X-WH)^2\), this is also known as Frobenious norm and you will see in the literature it is written as :\(\||V-WH||_{F}\)
However this is not the only distance metric, other distance metrics are also used in NMF algorithms. In addition, there could be other parameters to optimize that relates to sparseness of the \(W\) and \(H\) matrices. With sparse \(W\) and \(H\), each entry in the \(X\) matrix is expressed as the sum of a small number of components. This makes the interpretation easier, if the weights are 0 than there is not contribution from the corresponding factors.

Below, we are plotting the values of metagenes (rows of \(H\)) for component 1 and 3. In this context, these values can also be interpreted as relationship between samples. If we wanted to plot the relationship between genes we would plot the columns of \(W\) matrix.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(NMF)}
\NormalTok{res=}\KeywordTok{nmf}\NormalTok{(mat,}\DataTypeTok{rank=}\DecValTok{3}\NormalTok{,}\DataTypeTok{seed=}\DecValTok{123}\NormalTok{) }\CommentTok{# nmf with 3 components/factors}
\NormalTok{w <-}\StringTok{ }\KeywordTok{basis}\NormalTok{(res) }\CommentTok{# get W}
\NormalTok{h <-}\StringTok{ }\KeywordTok{coef}\NormalTok{(res)  }\CommentTok{# get H}

\CommentTok{# plot 1st factor against 3rd factor}
\KeywordTok{plot}\NormalTok{(h[}\DecValTok{1}\NormalTok{,],h[}\DecValTok{3}\NormalTok{,],}\DataTypeTok{col=}\NormalTok{annotation_col}\OperatorTok{$}\NormalTok{LeukemiaType,}\DataTypeTok{pch=}\DecValTok{19}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/nmfCode-1} \end{center}

We should add the note that due to random starting points of the optimization algorithm, NMF is usually run multiple times and a consensus clustering approach is used when clustering samples. This simply means that samples are clustered together if they cluster together in multiple runs of the NMF. The NMF package we used above has built-in ways to achieve this. In addition, NMF is a family of algorithms the choice of cost function to optimize the difference between \(X\) and \(WH\) and the methods used for optimization creates multiple variants of NMF. The ``method'' parameter in the above \texttt{nmf()} function controls the which algorithm for NMF.

\hypertarget{chosing-the-number-of-components-and-ranking-components-in-importance}{%
\subsubsection{chosing the number of components and ranking components in importance}\label{chosing-the-number-of-components-and-ranking-components-in-importance}}

In both ICA and NMF, there is no well-defined way to rank components or to select the number of components. There are couple of approaches that might suit to both ICA and NMF for ranking components. One can use the norms of columns/rows in mixing matrices. This could simply mean take the sum of absolute values in mixing matrices. In our examples above, For our ICA example above, ICA we would take the sum of the absolute values of the rows of \(A\) since we transposed the input matrix \(X\) before ICA. And for the NMF, we would use the columns of \(W\). These ideas assume that the larger coefficients in the weight or mixing matrices indicate more important components.

For selecting the optimal number of components, NMF package provides different strategies. One way is to calculate RSS for each \(k\), number of components, and take the \(k\) where the RSS curve starts to stabilize.However, these strategies require that you run the algorithm with multiple possible component numbers. \texttt{nmf} function will run these automatically when the \texttt{rank} argument is a vector of numbers. For ICA there is no straightforward way to choose the right number of components, a common strategy is to start with as many components as variables and try to rank them by their usefullness.

\BeginKnitrBlock{rmdtip}
\textbf{Want to know more ?}

NMF package vignette has extensive information on how to run NMF to get stable resuts and getting an estimate of components \url{https://cran.r-project.org/web/packages/NMF/vignettes/NMF-vignette.pdf}
\EndKnitrBlock{rmdtip}

\hypertarget{multi-dimensional-scaling}{%
\subsection{Multi-dimensional scaling}\label{multi-dimensional-scaling}}

MDS is a set of data analysis techniques that display the structure of distance data in a high dimensional space into a lower dimensional space without much loss of information. The overall goal of MDS is to faithfully represent these distances with the lowest possible dimensions. So called ``classical multi-dimensional scaling'' algorithm, tries to minimize the following function:

\({\displaystyle Stress_{D}(z_{1},z_{2},...,z_{N})={\Biggl (}{\frac {\sum _{i,j}{\bigl (}d_{ij}-\|z_{i}-z_{j}\|{\bigr )}^{2}}{\sum _{i,j}d_{ij}^{2}}}{\Biggr )}^{1/2}}\)

Here the function compares the new data points on lower dimension \((z_{1},z_{2},...,z_{N})\) to the input distances between data points or distance between samples in our gene expression example. It turns out, this problem can be efficiently solved with SVD/PCA on the scaled distance matrix, the projection on eigenvectors will be the most optimal solution for the equation above. Therefore, classical MDS is sometimes called Principal Coordinates Analysis in the litereuature. However, later variants improve on classical MDS this by using this as a starting point and optimize a slightly different cost function that again measures how well the low-dimensional distances correspond to high-dimensional distances. This variant is called non-metric MDS and due to the nature of the cost function, it assumes a less stringent relationship between the low-dimensional distances \$\textbar{}z\_\{i\}-z\_\{j\}\textbar{} and input distances \(d_{ij}\). Formally, this procedure tries to optimize the following function.

\({\displaystyle Stress_{D}(z_{1},z_{2},...,z_{N})={\Biggl (}{\frac {\sum _{i,j}{\bigl (}\|z_{i}-z_{j}\|-\theta(d_{ij}){\bigr )}^{2}}{\sum _{i,j}\|z_{i}-z_{j}\|^{2}}}{\Biggr )}^{1/2}}\)

The core of a non-metric MDS algorithm is a twofold optimization process. First the optimal monotonic transformation of the distances has to be found, this is shown in the above formula as \(\theta(d_{ij})\). Secondly, the points on a low dimension configuration have to be optimally arranged, so that their distances match the scaled distances as closely as possible. This two steps are repeated until some convergence criteria is reached. This usually means that the cost function does not improve much after certain number of iterations. The basic steps in a non-metric MDS algorithm are:
1. Find a random low dimensional configuration of points, or in the variant we will be using below we start with the configuration returned by classical MDS
2. Calculate the distances between the points in the low dimension \$\textbar{}z\_\{i\}-z\_\{j\}\textbar{}, \(z_{i}\) and \(z_{j}\) are vector of positions for sample \(i\) and \(j\).
3. Find the optimal monotonic transformation of the input distance, \({\textstyle \theta(d_{ij})}\), to approximate input distances to low-dimensional distances. This is achieved by isotonic regression, where a monotonically increasing free-form function is fit. This step practically ensures that ranking of low-dimensional distances are similar to rankings of input distances.
4. Minimize the stress function by re-configuring low-dimensional space and keeping \(\theta\) function constant.
5. repeat from step 2 until convergence.

We will now demonstrate both classical MDS and Kruskal's isometric MDS.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mds=}\KeywordTok{cmdscale}\NormalTok{(}\KeywordTok{dist}\NormalTok{(}\KeywordTok{t}\NormalTok{(mat)))}
\NormalTok{isomds=MASS}\OperatorTok{::}\KeywordTok{isoMDS}\NormalTok{(}\KeywordTok{dist}\NormalTok{(}\KeywordTok{t}\NormalTok{(mat)))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## initial  value 15.907414 
## final  value 13.462986 
## converged
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# plot the patients in the 2D space}
\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\KeywordTok{plot}\NormalTok{(mds,}\DataTypeTok{pch=}\DecValTok{19}\NormalTok{,}\DataTypeTok{col=}\NormalTok{annotation_col}\OperatorTok{$}\NormalTok{LeukemiaType,}
     \DataTypeTok{main=}\StringTok{"classical MDS"}\NormalTok{)}
\KeywordTok{plot}\NormalTok{(isomds}\OperatorTok{$}\NormalTok{points,}\DataTypeTok{pch=}\DecValTok{19}\NormalTok{,}\DataTypeTok{col=}\NormalTok{annotation_col}\OperatorTok{$}\NormalTok{LeukemiaType,}
     \DataTypeTok{main=}\StringTok{"isotonic MDS"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/mds2-1} \end{center}

In this example, there is not much difference between isotonic MDS and classical MDS. However, there might be cases where different MDS methods provides visiable changes in the scatter plots.

\hypertarget{t-distributed-stochastic-neighbor-embedding-t-sne}{%
\subsection{t-Distributed Stochastic Neighbor Embedding (t-SNE)}\label{t-distributed-stochastic-neighbor-embedding-t-sne}}

t-SNE maps the distances in high-dimensional space to lower dimensions and it is similar to MDS method in this respect. But the benefit of this particular method is that it tries to preserve the local structure of the data so the distances and grouping of the points we observe in a lower dimensions such as a 2D scatter plot is as close as possible to the distances we observe in the high-dimensional space. As with other dimension reduction methods, you can choose how many lower dimensions you need. The main difference of t-SNE is that it tries to preserve the local structure of the data. This kind of local structure embedding is missing in the MDS algorithm which also has a similar goal. MDS tries to optimize the distances as a whole, whereas t-SNE optimizes the distances with the local structure in mind. This is defined by the ``perplexity'' parameter in the arguments. This parameter controls how much the local structure influences the distance calculation. The lower the value the more the local structure is take into account. Similar to MDS, the process is an optimization algorithm. Here, we also try to minimize the divergence between observed distances and lower dimension distances. However, in the case of t-SNE, the observed distances and lower dimensional distances are transformed using a probabilistic framework with their local variance in mind.

From here on, we will provide a bit more detail on how the algorithm works in case conceptual description above is too shallow. In t-SNE the euclidiean distances between data points are transformed into a conditional similarity between points. This is done by assuming a normal distribution on each data point with a variance calculated ultimately by the use of ``perplexity'' parameter. The perplexity paramater is, in a sense, a guess about the number of the closest neighbors each point has. Setting it to higher values gives more weight to global structure. Given \(d_{ij}\) is the euclidean distance between point \(i\) and \(j\), the similarity score \(p_{ij}\) is calculated as shown below.

\(p_{j | i} = \frac{\exp(-\|d_{ij}\|^2 / 2 _i^2)}{_{k \neq i} \exp(-\|d_{ik}\|^2 / 2 _i^2)}\)

This distance is symmetrized by incorparating \$p\_\{i \textbar{} j\} as shown below.

\(p_{i j}=\frac{p_{j|i} + p_{i|j}}{2n}\)

For the distances in the reduced dimension, we use t-distribution with one degree of freedom. In the formula below, \(| y_i-y_j\|^2\) is euclidean distance between points \(i\) and \(j\) in the reduced dimensions.

\[
q_{i j} = \frac{(1+ \| y_i-y_j\|^2)^{-1}}{(_{k \neq l} 1+ \| y_k-y_l\|^2)^{-1} }
\]

As most of the algorithms we have seen in this section, t-SNE is an optimization process in essence. In every iteration the points along lower dimensions are re-arranged to minimize the formulated difference between the the observed joint probabilities (\(p_{i j}\)) and low-dimensional joint probabilities (\(q_{i j}\)). Here we are trying to compare probability distributions. In this case, this is done using a method called Kullback-Leibler divergence, or KL-divergence. In the formula below, since the \(p_{i j}\) is pre-defined using original distances, only way to optimize is to play with \(q_{i j}\)) because it depends on the configuration of points in the lower dimensional space. This configuration is optimized to minimize the KL-divergence between \(p_{i j}\) and \(q_{i j}\).

\[
KL(P||Q) = \sum_{i, j} p_{ij} \, \log \frac{p_{ij}}{q_{ij}}.
\]
Strictly speaking, KL-divergence measures how well the distribution \(P\) which is observed using the original data points can be aproximated by distribution \(Q\), which is modeled using points on the lower dimension. If the distributions are identical KL-divergence would be 0. Naturally, the more divergent the distributions are the higher the KL-divergence will be.

We will now show how to use t-SNE on our gene expression data set. We are setting the random seed because again t-SNE optimization algorithm have random starting points and this might create non-identical results in every run. After calculating the t-SNE lower dimension embeddings we will plot the points in a 2D scatter plot.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(}\StringTok{"Rtsne"}\NormalTok{)}
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{42}\NormalTok{) }\CommentTok{# Set a seed if you want reproducible results}
\NormalTok{tsne_out <-}\StringTok{ }\KeywordTok{Rtsne}\NormalTok{(}\KeywordTok{t}\NormalTok{(mat),}\DataTypeTok{perplexity =} \DecValTok{10}\NormalTok{) }\CommentTok{# Run TSNE}
 \CommentTok{#image(t(as.matrix(dist(tsne_out$Y))))}
\CommentTok{# Show the objects in the 2D tsne representation}
\KeywordTok{plot}\NormalTok{(tsne_out}\OperatorTok{$}\NormalTok{Y,}\DataTypeTok{col=}\NormalTok{annotation_col}\OperatorTok{$}\NormalTok{LeukemiaType,}
     \DataTypeTok{pch=}\DecValTok{19}\NormalTok{)}

\CommentTok{# create the legend for the Leukemia types}
\KeywordTok{legend}\NormalTok{(}\StringTok{"bottomleft"}\NormalTok{,}
       \DataTypeTok{legend=}\KeywordTok{unique}\NormalTok{(annotation_col}\OperatorTok{$}\NormalTok{LeukemiaType),}
       \DataTypeTok{fill =}\KeywordTok{palette}\NormalTok{(}\StringTok{"default"}\NormalTok{),}
       \DataTypeTok{border=}\OtherTok{NA}\NormalTok{,}\DataTypeTok{box.col=}\OtherTok{NA}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/tsne-1} \end{center}

As you might have noticed, we set again a random seed with \texttt{set.seed()} function. The optimization algorithm starts with random configuration of points in the lower dimension space, and each iteration it tries to improve on the previous lower dimension confugration, that is why starting points can result in different final outcomes.

\BeginKnitrBlock{rmdtip}
\textbf{Want to know more ?}

\begin{itemize}
\tightlist
\item
  How perplexity effects t-sne, interactive examples \url{https://distill.pub/2016/misread-tsne/}
\item
  more on perplexity: \url{https://blog.paperspace.com/dimension-reduction-with-t-sne/}
\item
  Intro to t-SNE \url{https://www.oreilly.com/learning/an-illustrated-introduction-to-the-t-sne-algorithm}
\end{itemize}
\EndKnitrBlock{rmdtip}

\hypertarget{exercises-2}{%
\section{Exercises}\label{exercises-2}}

TBD

\hypertarget{supervisedLearning}{%
\chapter{Predictive Modeling with Supervised Machine Learning}\label{supervisedLearning}}

In this chapter we will introduce supervised machine learning applications for predictive modeling. In genomics, we are often faced a couple biological questions to answer using lots of data. Some of those questions can easily fit in the domain of machine learning, where algorithms will learn a mathematical model of the input data in order to make decisions about similar data, previously unseen by the model. Often we are trying to predict a medical or biological variable of interest using molecular signatures obtained via genomics methods. To give you a better idea, we listed some of the machine learning applications in genomics:

\begin{itemize}
\tightlist
\item
  predicting gene expression from epigenetic modifications (\citet{pmid22950368})
\item
  predicting gene locations (\citet{pmid12364589})
\item
  predicting enhancer or other regulatory regions (\citet{pmid22328731})
\item
  predicting drug response based on genomics (\citet{pmid21428770})
\item
  predicting healthy/disease status or disease subtypes based on genomics (\citet{pmid25750696})
\item
  predicting the effect of SNPs on gene regulation (\citet{pmid26301843})
\item
  calling SNPs (\citet{pmid30247488})
\end{itemize}

Apart from prediction of an outcome, machine learning can be used to understand which predictor variables are the most important for the prediction performance. This often gives insights into the biology as well. Many machine learning algorithms have either built-in variable importance assessment or can be wrapped into a model-agnostic variable importance method. For example, we may want to find which epigenetic modifications are most important for gene expression prediction. Although decades of molecular biology gives a pretty good idea for this, we could arrive at similar conclusions by building a machine learning model to predict gene expression using histone modifications H3K27ac, H3K27me, H3K4me1, H3K4me3 and DNA methylation. We can then check which of these are most important for gene expression prediction using variable importance metrics.

In this chapter, we will show how to use supervised machine learning models to solve problems in genomics. We will go over general steps in machine learning applications. In addition, we will introduce how to use some of the most popular supervised machine learning models in practice.

\hypertarget{how-machine-learning-models-are-fit}{%
\section{How machine learning models are fit?}\label{how-machine-learning-models-are-fit}}

We have already have quite an insight on how machine learning models are fit. We have previously seen clustering methods, which are unsupervised machine learning models, and we have seen linear regression which is a simple machine learning model if we disregard its objectives for statistical inference.

Machine learning models are optimization methods in their core. They all depend on defining a ``cost'' or ``loss'' function to minimize. For example, in linear regression the difference between predicted and the original values are being minimized. When we have a data set with the correct answer such as original values or class labels, this is called supervised learning. We use the structure in the data the predict a value and optimization methods help us use the right structure or patterns in the data. The supervised machine learning methods use predictor variables such as gene expression values or other genomic scores to build a mathematical function, or a mapping method if you will. This function maps a predictor variable vector or matrix from a given sample to the response variable: labels/classes or numeric values. The response variable is also called ``dependent variable''. Then, the predictions are simply output of mathematical functions, \(f(X)\). These functions take predictor variables, \(X\), as input. The variables in \(X\) are also called ``independent variables'',``explanatory variables'' or ``features''. The functions also have internal parameters that help map \(X\) to the predicted values. The optimization works on the parameters of \(f(X)\) and tries to minimize the difference between the function output and original response variables (\(Y\)): \(\sum(Y-f(X))^2\). Now, this is just a simplification of the actual ``cost'' or ``loss'' function. Especially, in classification problems cost functions can take different forms but the idea behind is the same. You have a mathematical expression you can minimize by searching for the optimal parameter values. The core ingredients of a machine learning algorithm are the same and they are listed as follows:

\begin{enumerate}
\def\labelenumi{\arabic{enumi})}
\tightlist
\item
  Define a prediction function or method \(f(X)\)
\item
  Devise a function (called loss or cost function) to optimize the difference between your predictions and observed values, such as \(\sum (Y-f(X))^2\)
\item
  Apply mathematical optimization methods to find best parameter values for \(f(X)\) in relation to the cost/loss function.
\end{enumerate}

Similarly, clustering and dimension reduction techniques can use optimization methods but they do so without having a correct answer to predict or train with. In this case, they find patterns or structure in the data without trying to estimate a correct answer. These patterns are groupings of samples or variables, such as common gene expression patterns that can be obtained from dimension reduction techniques such as PCA. In general, dimension reduction algorithms can be thought as optimization procedures that are trying to minimize \(X-WH\). Here, \(X\) is our original data set and \(WH\) is the product of potentially two lower dimension matrices, \(W\) and \(H\). In this case, the optimization procedure hopefully gives us the lower dimensional space we can represent our data without loosing too much information.

\hypertarget{machine-learning-vs-statistics}{%
\subsection{Machine learning vs Statistics}\label{machine-learning-vs-statistics}}

Machine learning and statistics are related and sometimes overlapping fields. Statistical inference is the main purpose of statistics. The aim of inference is to find statistical properties of the underlying data and also estimate the uncertainty about those properties. However, while doing so field of statistics developed dimension reduction and regression techniques that are the corner stone of machine learning applications.

Both machine learning and statistics share the same overarching goal, which is learning from the data.
The difference between the two is that machine learning emphasizes optimization and performance over statistical inference. Statistics is also concerned about performance but would like to calculate the uncertainty associated with parameters of the model. It will try to model the population statistics from the sample data points to assess that uncertainty. Having said that, many machine learning algorithms, including a couple we will introduce below, are developed by scientists who will define themselves as statisticians, and work at statistics departments of universities.

\hypertarget{steps-in-supervised-machine-learning}{%
\section{Steps in supervised machine learning}\label{steps-in-supervised-machine-learning}}

There are many methods to use for supervised learning problems. However, there are similar steps that you will need to follow whatever machine learning method you choose to train. These steps are briefly described below and we will get back to these in detail later in the chapter:

\begin{itemize}
\tightlist
\item
  pre-processing data: Normalization, data transformation
\item
  training and test data split: Decide which strategy you want to use for evaluation purposes. You need to use a test set to evaluate your model later on.
\item
  training the model: This is where your choice of supervised learning algorithm becomes relevant.``Training'' generally means your data set is used in optimization of the loss function to find parameters for \(f(x)\).
\item
  Estimating performance of the model: This is about which metrics to use to evaluate performance and how to calculate those metrics
\item
  Model tuning and selection: We try different parameters and select the best model
\end{itemize}

Many of these steps are identical for different supervised learning methods. Therefore, we will use \href{http://topepo.github.io/caret/index.html}{\textbf{caret}} package to perform these steps, which streamlines the steps and provides a similar interface for different supervised learning methods. There are other similar packages such as \href{https://mlr.mlr-org.com/}{\textbf{mlr}} that can provide similar functionality. For now, we will focus on classification models which is a subset of supervised learning models. In these types of models, we try to predict a categorical response variable, such as if a patient has the disease or not, or what type of disease the patient has based on predictor variables.

\hypertarget{use-case-disease-subtype-from-genomics-data}{%
\section{Use case: Disease subtype from genomics data}\label{use-case-disease-subtype-from-genomics-data}}

We will start our illustration of machine learning using a real dataset from tumor biopsies. We will use the gene expression data of glioblastoma tumor samples from ``The Cancer Genome Atlas'' project. We will try to predict the subtype of this disease using molecular markers. This subtype is characterized by large scale epigenetic alterations called ``CpG island methylator phenotype'' or ``CIMP'' ( \citet{pmid20399149}), half of the patients in our data set have this subtype and the rest do not, and we will try to predict which ones have CIMP subtype. There two data objects we need for this exercise, one for gene expression values per tumor sample and the other one is subtype annotation per patient. In the expression data set, every row is a patient and every column is a gene expression value. There are 184 tumor samples. This data set might a bit small for real world applications, however it is very relevant for the genomics focus of this book and the small datasets takes less time to train, which is useful for reproduciblity purposes. We will read these data sets from \textbf{compGenomRData} package now with \texttt{readRDS()} function.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# get file paths}
\NormalTok{fileLGGexp=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
                      \StringTok{"LGGrnaseq.rds"}\NormalTok{,}
                      \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{fileLGGann=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
                      \StringTok{"patient2LGGsubtypes.rds"}\NormalTok{,}
                      \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\CommentTok{# gene expression values}
\NormalTok{gexp=}\KeywordTok{readRDS}\NormalTok{(fileLGGexp)}
\KeywordTok{head}\NormalTok{(gexp[,}\DecValTok{1}\OperatorTok{:}\DecValTok{5}\NormalTok{])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##       TCGA-CS-4941 TCGA-CS-4944 TCGA-CS-5393
## A1BG         72.23        24.71        46.38
## A1CF          0.00         0.00         0.00
## A2BP1       524.50       105.41       323.58
## A2LD1       144.09        18.02        29.09
## A2ML1       521.39       159.37       164.62
## A2M       17944.72     10894.96     16480.11
##       TCGA-CS-5394 TCGA-CS-5395
## A1BG        37.966        19.52
## A1CF         0.000         0.00
## A2BP1       19.739       299.54
## A2LD1        7.595       202.12
## A2ML1       63.566       953.41
## A2M       9217.792     10801.85
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{dim}\NormalTok{(gexp)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 20501   184
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# patient annotation}
\NormalTok{patient=}\KeywordTok{readRDS}\NormalTok{(fileLGGann)}
\KeywordTok{head}\NormalTok{(patient)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##              subtype
## TCGA-FG-8185    CIMP
## TCGA-DB-5276    CIMP
## TCGA-P5-A77X    CIMP
## TCGA-IK-8125    CIMP
## TCGA-DU-A5TR    CIMP
## TCGA-E1-5311    CIMP
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{dim}\NormalTok{(patient)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 184   1
\end{verbatim}

\hypertarget{data-preprocessing}{%
\section{Data preprocessing}\label{data-preprocessing}}

We will have to preprocess the data before we start training. This might include exploratory data analysis to see how variables and samples relate to each other. For example, we might want to check correlation between predictor variables and keep only one variable from that group. In addition, some training algorithms might be sensitive to data scales or outliers. We should deal with those issues in this step. In some cases, the data might have missing values. We can choose the remove the samples that have missing values or try to impute them. Many machine learning algorithms will not be able to deal with missing values.

We will show how to do this in practice using the \texttt{caret::preProcess()} function and base R functions. Please note that there are more preprocessing options available than what we will show both via \texttt{caret::preProcess()} and base R functions, we are just going to cover a few basics.

\hypertarget{data-transformation}{%
\subsection{data transformation}\label{data-transformation}}

First thing we will do is data normalization and transformation. We have to take care of data scale issues that might come from how the experiments are performed and the potential problems that might occur during data collection. Ideally, each tumor sample has a similar distribution of gene expression values. Systematic differences between tumor samples must be corrected. We check if there are such differences using box plots.
We will only plot the first 50 tumor samples so that the figure is not too
squished.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{boxplot}\NormalTok{(gexp[,}\DecValTok{1}\OperatorTok{:}\DecValTok{50}\NormalTok{],}\DataTypeTok{outline=}\OtherTok{FALSE}\NormalTok{,}\DataTypeTok{col=}\StringTok{"cornflowerblue"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/boxML-1} 

}

\caption{boxplots for gene expression values}\label{fig:boxML}
\end{figure}

It seems there was some normalization done on this data. Gene expression values per sample looks to have the same scale. However, it looks like they have long tailed distributions, a log transformation may fix that. These long tailed distributions have outliers and this might adversely affect the models. We show how to distributions looks like for one of the patients samples without and with log transformation. We add a pseudo count of 1 to avoid \texttt{log(0)}.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\KeywordTok{hist}\NormalTok{(gexp[,}\DecValTok{5}\NormalTok{],}\DataTypeTok{xlab=}\StringTok{"gene expression"}\NormalTok{,}\DataTypeTok{main=}\StringTok{""}\NormalTok{,}\DataTypeTok{border=}\StringTok{"blue4"}\NormalTok{,}
     \DataTypeTok{col=}\StringTok{"cornflowerblue"}\NormalTok{)}
\KeywordTok{hist}\NormalTok{(}\KeywordTok{log10}\NormalTok{(gexp}\OperatorTok{+}\DecValTok{1}\NormalTok{)[,}\DecValTok{5}\NormalTok{], }\DataTypeTok{xlab=}\StringTok{"gene expression log scale"}\NormalTok{,}\DataTypeTok{main=}\StringTok{""}\NormalTok{,}
     \DataTypeTok{border=}\StringTok{"blue4"}\NormalTok{,}\DataTypeTok{col=}\StringTok{"cornflowerblue"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/logTransform-1} 

}

\caption{Gene expression distribution for the 5th patient (left). log transformed Gene expression distribution for the same patient (right)}\label{fig:logTransform}
\end{figure}

Since taking a log seems to work to tame the extreme values we do that below and also add 1 pseudo-count to be able to deal with 0 values.:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gexp=}\KeywordTok{log10}\NormalTok{(gexp}\OperatorTok{+}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Other things we can do in combination with this is to winsorize the data which caps extreme values to 1st and 99th percentiles. But before we go forward, we should transpose our data. In this case, the predictor variables are gene expression values and they should be on the column side. It was OK to leave them on the row side, to check systematic errors with box plots, but machine learning algorithms require predictor variables are on the column side.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# transpose the data set}
\NormalTok{tgexp <-}\StringTok{ }\KeywordTok{t}\NormalTok{(gexp)}
\end{Highlighting}
\end{Shaded}

\hypertarget{filtering-data-and-scaling}{%
\subsection{Filtering data and scaling}\label{filtering-data-and-scaling}}

We can filter predictor variables which has low variation. They are not likely to have any predictive importance since there is not much variation and they will just slow our algorithms. The more variables the slower the algorithms will be generally. \texttt{caret::preProcess()} function can help filter the predictor variables with near zero variance.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# remove near zero variation for the columns at least}
\CommentTok{# 85% of the values are the same}
\CommentTok{# this function creates the filter but doesn't apply it yet}
\NormalTok{nzv=}\KeywordTok{preProcess}\NormalTok{(tgexp,}\DataTypeTok{method=}\StringTok{"nzv"}\NormalTok{,}\DataTypeTok{uniqueCut =} \DecValTok{15}\NormalTok{)}

\CommentTok{# apply the filter using "predict" function}
\CommentTok{# return the filtered dataset and assign it to nzv_tgexp}
\CommentTok{# variable}
\NormalTok{nzv_tgexp=}\KeywordTok{predict}\NormalTok{(nzv,tgexp)}
\end{Highlighting}
\end{Shaded}

In addition, we can also choose arbitrary cutoffs for variability. For example, we can choose to take the top 1000 variable predictors.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{SDs=}\KeywordTok{apply}\NormalTok{(tgexp,}\DecValTok{2}\NormalTok{,sd )}
\NormalTok{topPreds=}\KeywordTok{order}\NormalTok{(SDs,}\DataTypeTok{decreasing =} \OtherTok{TRUE}\NormalTok{)[}\DecValTok{1}\OperatorTok{:}\DecValTok{1000}\NormalTok{]}
\NormalTok{tgexp=tgexp[,topPreds]}
\end{Highlighting}
\end{Shaded}

We can also center the data which is as we have seen in chapter 4 is subtracting the mean. Following this, the predictor variables will have zero means. In addition, we can scale the data. When we scale, each value of the predictor
variable is divided by its standard deviation. Therefore predictor variables will have the same standard deviation. These manipulations are generally used to improve the numerical stability of some calculations. In distance based metrics, it could be beneficial to at least center the data. We will now center the data. We are centering using \texttt{preProcess}. This is more practical than \texttt{scale()} function because when we get a new data point we can use \texttt{predict()} function and \texttt{processCenter} object to process it just like we did for the training samples.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(caret)}
\NormalTok{processCenter=}\KeywordTok{preProcess}\NormalTok{(tgexp, }\DataTypeTok{method =} \KeywordTok{c}\NormalTok{(}\StringTok{"center"}\NormalTok{))}
\NormalTok{tgexp=}\KeywordTok{predict}\NormalTok{(processCenter,tgexp)}
\end{Highlighting}
\end{Shaded}

We will next filter the predictor variables that are highly correlated. You may choose not to do this as some methods can handle correlation between predictor variables. However, the less predictor variables we have the faster the model fitting can be done.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# create a filter for removing higly correlated variables}
\CommentTok{# if two variables are highly correlated only one of them}
\CommentTok{# is removed}
\NormalTok{corrFilt=}\KeywordTok{preProcess}\NormalTok{(tgexp, }\DataTypeTok{method =} \StringTok{"corr"}\NormalTok{,}\DataTypeTok{cutoff =} \FloatTok{0.9}\NormalTok{)}
\NormalTok{tgexp=}\KeywordTok{predict}\NormalTok{(corrFilt,tgexp)}
\end{Highlighting}
\end{Shaded}

\hypertarget{dealing-with-missing-values}{%
\subsection{Dealing with missing values}\label{dealing-with-missing-values}}

In real life situations, there will be missing values in our data. In genomics, we might not have values for certain genes or genomic locations due to technical problems during experiments. We have to be able to deal with these missing values. For demonstration purposes, we will now introduce NA values in our data, the ``NA'' values is normally used to encode missing values in R. We then show how to check and deal with those. One way is to impute them, here we use again a machine learning algorithm to guess the missing values. Another option is to discard the samples with missing values or discard the predictor variables with missing values. First, we replace one of the values as NA and check if it is there.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{missing_tgexp=tgexp}
\NormalTok{missing_tgexp[}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{]=}\OtherTok{NA}
\KeywordTok{anyNA}\NormalTok{(missing_tgexp) }\CommentTok{# check if there are NA values}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] TRUE
\end{verbatim}

Next, we will try to remove that gene from the set. Removing genes or samples have both downsides. You might be removing a predictor variable that could be important for the prediction. Removing the samples with missing values will decrease the number of samples in the training set. The code below checks which values are NA in the matrix, then runs a column sum and keeps everything where column sum is equal to 0. The column sums where there are NA values will be higher than 0 depending on how many NA values there are in a column.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gexpnoNA=missing_tgexp[ , }\KeywordTok{colSums}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(missing_tgexp)) }\OperatorTok{==}\StringTok{ }\DecValTok{0}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

We will next try to impute the missing value(s). Imputation can be as simple as assigning missing values to the mean or median value of the variable, or assigning the mean/median of values from nearest neighbors of the sample having the missing value. We will show both using the \texttt{caret::preProcess()} function. First, let us run the median imputation.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(caret)}
\NormalTok{mImpute=}\KeywordTok{preProcess}\NormalTok{(missing_tgexp,}\DataTypeTok{method=}\StringTok{"medianImpute"}\NormalTok{)}
\NormalTok{imputedGexp=}\KeywordTok{predict}\NormalTok{(mImpute,missing_tgexp)}
\end{Highlighting}
\end{Shaded}

Another imputation methods that is more precise than the median imputation is to impute the missing values based on the nearest neighbors of the samples. In this case, the algorithm finds most similar samples to the sample vector with NA values. Next, the algorithm averages the non-missing values from those neighbors and replaces the missing value with that value.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(RANN)}
\NormalTok{knnImpute=}\KeywordTok{preProcess}\NormalTok{(missing_tgexp,}\DataTypeTok{method=}\StringTok{"knnImpute"}\NormalTok{)}
\NormalTok{knnimputedGexp=}\KeywordTok{predict}\NormalTok{(knnImpute,missing_tgexp) }
\end{Highlighting}
\end{Shaded}

\hypertarget{splitting-the-data}{%
\section{Splitting the data}\label{splitting-the-data}}

At this point we might choose to split the data to the test and the training partitions. The reason for this is that we need an independent test we did not train on. This will become more clear in the following sections but without having a separate test set we can not assess the performance of our model or tune it properly.

\hypertarget{holdout-test-dataset}{%
\subsection{Holdout test dataset}\label{holdout-test-dataset}}

There are multiple data split strategies. For starters, we will split the 30\% of the data as test. This method is the gold standard for testing performance of our model. By doing this, we have a separate data set that the model has never seen. First, we create a single data frame with predictors and response variables

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tgexp=}\KeywordTok{merge}\NormalTok{(patient,tgexp,}\DataTypeTok{by=}\StringTok{"row.names"}\NormalTok{)}

\CommentTok{# push sample ids back to the row names}
\KeywordTok{rownames}\NormalTok{(tgexp)=tgexp[,}\DecValTok{1}\NormalTok{]}
\NormalTok{tgexp=tgexp[,}\OperatorTok{-}\DecValTok{1}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

Now the response variable or the class label is merged with our dataset we can split it to test and training set with \texttt{caret::createPartition()}.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{3031}\NormalTok{) }\CommentTok{# set the random number seed for reproducibility }

\CommentTok{# get indices for 70% of the data set}
\NormalTok{intrain <-}\StringTok{ }\KeywordTok{createDataPartition}\NormalTok{(}\DataTypeTok{y =}\NormalTok{ tgexp[,}\DecValTok{1}\NormalTok{], }\DataTypeTok{p=} \FloatTok{0.7}\NormalTok{)[[}\DecValTok{1}\NormalTok{]]}

\CommentTok{# seperate test and training sets}
\NormalTok{training <-}\StringTok{ }\NormalTok{tgexp[intrain,]}
\NormalTok{testing <-}\StringTok{ }\NormalTok{tgexp[}\OperatorTok{-}\NormalTok{intrain,]}
\end{Highlighting}
\end{Shaded}

\hypertarget{cross-validation}{%
\subsection{Cross-validation}\label{cross-validation}}

In some cases, we might have too few data points and it might be to costly to set aside a significant portion of the data set as a holdout test set. In these cases a resampling based techniques such as cross-validation may be useful.

Cross-validation works by splitting the data into randomly sampled \(k\) subsets, called k-folds. So, for example, in the case of 5-fold cross-validation with 100 data points, we would create 5 folds each containing 20 data points. We would then build models and estimate errors 5 times. Each time four of the groups are combined (resulting in 80 data points) and used to train your model. Then the 5th group of 20 points that was not used to construct the model is used to estimate the test error. In the case of 5-fold cross-validation, we would have 5 error estimates that could be averaged to obtain a more robust estimate of the test error.

An extreme case of k-fold cross-validation, is to equalize the \(k\) to the number of data points or in our case number of tumor samples. This is called leave-one-out cross-validation (LOOCV). This could be better than k-fold cross-validation but it takes too much time to train that many models if number of data points are large.

The \textbf{caret} package has built-in cross-validation functionality for all the machine learning methods and we will be using that in the later sections.

\hypertarget{bootstrap-resampling}{%
\subsection{Bootstrap resampling}\label{bootstrap-resampling}}

Another method to be used to estimate the prediction error is to use bootstrap resampling. This is a general method we have already introduced in Chapter \ref{stats}. It can be used the estimate variability of any statistical parameter. In this case, that parameter is the test error or test accuracy.

The training set is drawn from the original set with replacement (same size as the original set), then we build a model with this bootstrap resampled set. Next, we take the data points that are not selected for the random sample and predict labels for them. These data points are called ``out-of-the-bag (OOB) sample''. We repeat this process many times and record the error for the OOB samples. We can take the average of OOB error to estimate the real test error. This is a powerful method that is not only used to estimate test error but incorporated into the training part of some machine learning methods such as random forests. Normally, we should repeat the process hundreds or up to a thousand times to get good estimates. However, limiting factor would be the time it takes construct and test that many models. 20-30 repetitions might be enough if the time cost of training is too high. Again, \textbf{caret} package provides the bootstrap interface for many machine learning models for sampling before training and estimating the error on OOB samples.

\hypertarget{predicting-the-subtype-with-k-nearest-neighbors}{%
\section{Predicting the subtype with k-nearest neighbors}\label{predicting-the-subtype-with-k-nearest-neighbors}}

One of the easiest things to wrap our heads around when we are trying to predict a label such as disease subtype is to look for similar samples and assign the labels of the those similar samples to our sample.

Conceptually, k-nearest neighbors (k-NN) is very similar to clustering algorithms we have seen earlier. If we have a measure of distance between the samples, we can find the nearest \(k\) samples to our new sample and use a voting method to decide on the label of our new sample.

Let us run the k-NN algorithm with our cancer data. For illustrative purposes, we provide the same data set for training and test data. Providing the training data as the test data shows us the training error or accuracy, which is how the model is doing on the data it is trained with. Below we are running k-NN with \texttt{caret:knn3()} function. The most important argument is \texttt{k} which is the number of nearest neighbors to consider. In this case, we set it to 5. We will later discuss how to find the best \texttt{k}.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(caret)}
\NormalTok{knnFit=}\KeywordTok{knn3}\NormalTok{(}\DataTypeTok{x=}\NormalTok{training[,}\OperatorTok{-}\DecValTok{1}\NormalTok{], }\CommentTok{# training set}
            \DataTypeTok{y=}\NormalTok{training[,}\DecValTok{1}\NormalTok{], }\CommentTok{# training set class labels}
            \DataTypeTok{k=}\DecValTok{5}\NormalTok{)}
\CommentTok{# predictions on the test set}
\NormalTok{trainPred=}\KeywordTok{predict}\NormalTok{(knnFit,training[,}\OperatorTok{-}\DecValTok{1}\NormalTok{])}
\end{Highlighting}
\end{Shaded}

\hypertarget{assessing-the-performance-of-our-model}{%
\section{Assessing the performance of our model}\label{assessing-the-performance-of-our-model}}

We have to define some metrics to see if our model worked. The algorithm is trying to reduce the classification error, or in other words it is trying to increase the training accuracy. For the assessment of performance, there are other different metrics to consider. All the metrics for 2-class classification depend on the table below. Which shows the number of true positives (TP), false positives (FP), true negatives (TN) and false negatives (FN), similar to a table we used in hypothesis testing section in the statistics chapter previously.

\begin{longtable}[]{@{}lcc@{}}
\toprule
\begin{minipage}[b]{0.20\columnwidth}\raggedright
\strut
\end{minipage} & \begin{minipage}[b]{0.27\columnwidth}\centering
Actual CIMP\strut
\end{minipage} & \begin{minipage}[b]{0.28\columnwidth}\centering
Actual noCIMP\strut
\end{minipage}\tabularnewline
\midrule
\endhead
\begin{minipage}[t]{0.20\columnwidth}\raggedright
Predicted as
CIMP\strut
\end{minipage} & \begin{minipage}[t]{0.27\columnwidth}\centering
True Positives (TP)\strut
\end{minipage} & \begin{minipage}[t]{0.28\columnwidth}\centering
False Positive (FP)\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.20\columnwidth}\raggedright
Predicted as
noCIMP\strut
\end{minipage} & \begin{minipage}[t]{0.27\columnwidth}\centering
False Positives (FN)\strut
\end{minipage} & \begin{minipage}[t]{0.28\columnwidth}\centering
True negatives (TN)\strut
\end{minipage}\tabularnewline
\bottomrule
\end{longtable}

Accuracy is the first metric to look at. This metric is is simply
\((TP+TN)/(TP+TN+FP+FN)\) and shows the proportion of times we were right. There are other accuracy metrics that are important and output by \textbf{caret} functions. We will go over some of them here.

Precision, \(TP/(TP+FP)\), is about the confidence we have on our CIMP calls. If our method is very precise, we will have low false positives. That means every time we call a CIMP event, we would relatively certain it is not a false positive.

Sensitivity, \(TP/(TP+FN)\), is how often we miss CIMP cases and call them as noCIMP. Making less mistakes in noCIMP cases will increase our sensitivity. You can think of sensitivity also as in sick/healthy context. A highly sensitive method will be good at classifying sick people when they are indeed sick.

Specificity,\(TN/(TN+FP)\), is about how sure we are when we call something as noCIMP. If our method is not very specific, we would call many patients as CIMP while in fact they did not have the subtype. In the sick/healthy context, a highly specific method, will be good at not calling healthy people sick.

An alternative to accuracy we showed earlier is ``balanced accuracy''. Accuracy does not perform well when classes have very different number of samples (class imbalance). For example, if you have 90 CIMP cases and 10 noCIMP cases, classifying all the samples as CIMP gives 0.9 accuracy score by default. Using the ``balanced accuracy'' metric can help in such situations. This is simply \((Precision+Sensitivity)/2\). In this case above with the class imbalance scenario, the ``balanced accuracy'' would be 0.5. Another metric that takes into account accuracy that could be generated by chance is ``Kappa statistic'' or ``Cohen's Kappa''. This metric includes expected accuracy which is affected by class imbalance in the training set and provides a metric corrected by that.

In the k-NN example above, we trained and tested on the same data. The model returned the predicted labels for our training. We can calculate the accuracy metrics using \texttt{caret::confusionMatrix()} function. This is sometimes what it is called training accuracy. If you take \(1-accuracy\), it will be the ``training error''.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# gte k-NN prediction on the training data itself, with k=5}
\NormalTok{knnFit=}\KeywordTok{knn3}\NormalTok{(}\DataTypeTok{x=}\NormalTok{training[,}\OperatorTok{-}\DecValTok{1}\NormalTok{], }\CommentTok{# training set}
            \DataTypeTok{y=}\NormalTok{training[,}\DecValTok{1}\NormalTok{], }\CommentTok{# training set class labels}
            \DataTypeTok{k=}\DecValTok{5}\NormalTok{)}

\CommentTok{# predictions on the training set}
\NormalTok{trainPred=}\KeywordTok{predict}\NormalTok{(knnFit,training[,}\OperatorTok{-}\DecValTok{1}\NormalTok{],}\DataTypeTok{type=}\StringTok{"class"}\NormalTok{)}

\CommentTok{# compare the predicted labels to real labels}
\CommentTok{# get different performance metrics}
\KeywordTok{confusionMatrix}\NormalTok{(}\DataTypeTok{data=}\NormalTok{training[,}\DecValTok{1}\NormalTok{],}\DataTypeTok{reference=}\NormalTok{trainPred)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Confusion Matrix and Statistics
## 
##           Reference
## Prediction CIMP noCIMP
##     CIMP     65      0
##     noCIMP    5     60
##                                         
##                Accuracy : 0.962         
##                  95% CI : (0.913, 0.987)
##     No Information Rate : 0.538         
##     P-Value [Acc > NIR] : <2e-16        
##                                         
##                   Kappa : 0.923         
##  Mcnemar's Test P-Value : 0.0736        
##                                         
##             Sensitivity : 0.929         
##             Specificity : 1.000         
##          Pos Pred Value : 1.000         
##          Neg Pred Value : 0.923         
##              Prevalence : 0.538         
##          Detection Rate : 0.500         
##    Detection Prevalence : 0.500         
##       Balanced Accuracy : 0.964         
##                                         
##        'Positive' Class : CIMP          
## 
\end{verbatim}

Now, let us see what our test set accuracy looks like again using the \texttt{knn} function and the \texttt{confusionMatrix()} function on the predicted and real classes.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# predictions on the test set, return class labels}
\NormalTok{testPred=}\KeywordTok{predict}\NormalTok{(knnFit,testing[,}\OperatorTok{-}\DecValTok{1}\NormalTok{],}\DataTypeTok{type=}\StringTok{"class"}\NormalTok{)}

\CommentTok{# compare the predicted labels to real labels}
\CommentTok{# get different performance metrics}
\KeywordTok{confusionMatrix}\NormalTok{(}\DataTypeTok{data=}\NormalTok{testing[,}\DecValTok{1}\NormalTok{],}\DataTypeTok{reference=}\NormalTok{testPred)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Confusion Matrix and Statistics
## 
##           Reference
## Prediction CIMP noCIMP
##     CIMP     27      0
##     noCIMP    1     26
##                                     
##                Accuracy : 0.981     
##                  95% CI : (0.901, 1)
##     No Information Rate : 0.519     
##     P-Value [Acc > NIR] : 2.02e-14  
##                                     
##                   Kappa : 0.963     
##  Mcnemar's Test P-Value : 1         
##                                     
##             Sensitivity : 0.964     
##             Specificity : 1.000     
##          Pos Pred Value : 1.000     
##          Neg Pred Value : 0.963     
##              Prevalence : 0.519     
##          Detection Rate : 0.500     
##    Detection Prevalence : 0.500     
##       Balanced Accuracy : 0.982     
##                                     
##        'Positive' Class : CIMP      
## 
\end{verbatim}

Test set accuracy is not as good as the training accuracy, which is usually the case. That is why the best way to evaluate performance is to use test data that is not used by the model for training. That gives you an idea about real world performance where the model will be used to predict data that is not previously seen.

\hypertarget{receiver-operating-characteristic-roc-curves}{%
\subsection{Receiver Operating Characteristic (ROC) Curves}\label{receiver-operating-characteristic-roc-curves}}

One important and popular metric when evaluating performance is looking at receiver operating characteristic (ROC) curves. The ROC curve is created by evaluating the class probabilities for the model across a continuum of thresholds. Typically, in the case of two class classification the methods return a probability for one of the classes. If that probability is higher than 0.5, you call the label as, for example, class A. If less than 0.5, we call the label class B. However, we can move that threshold and change what we call class A or B. For each candidate threshold, the resulting the sensitivity and 1-specificity are plotted against each other. The best possible prediction would result a point in the upper left corner, representing 100\% sensitivity (no false negatives) and 100\% specificity (no false positives). For the best model, the curve will be almost like a square. Since this is an important information, area under the curve (AUC) is calculated. This is a quantity between 0 and 1, and closer to 1 better the performance of your classifier in terms of sensitivity and specificity. For an uninformative classification model AUC will be 0.5. Although, ROC curves are initially designed fro two class problems, later extensions made it possible to use ROC curves for multi-class problems.

ROC curves can also be used to determine alternate cutoffs for class probabilities for two class problems. However, this will always result in a trade-off between sensitivity and specificity. Sometimes it might be desirable to limit number of false positives as making such mistakes would be too costly for the individual cases. For example, if predicted with a certain disease, you might be recommended to go under surgery. However, if your classifier has a relatively high false positive rate, low specificity, you might go under the surgery for no reason. Typically, you want your classification model to have high specificity and sensitivity which may not be always possible in the real world. You might have to choose what is more important for a specific problem and try to increase that.

Next, we will show how to use ROC curves for our k-NN application. The method requires classification probabilities in the format where 0 probability denotes class ``noCIMP'' and probability 1 denotes class ``CIMP''. This way ROC curve can be drawn by varying the probability cutoff for calling class ``noCIMP'' or ``CIMP''. Below we are getting a similar probability from k-NN but we have to transform it to the format we described above. Then, we feed those class probabilities to \texttt{pROC::roc()} function to calculate ROC curve and area-under-the-curve.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(pROC)}

\CommentTok{# get k-NN class probabilities}
\CommentTok{# prediction probabilities on the test set}
\NormalTok{testProbs=}\KeywordTok{predict}\NormalTok{(knnFit,testing[,}\OperatorTok{-}\DecValTok{1}\NormalTok{])}

\CommentTok{# get the roc curve}
\NormalTok{rocCurve <-}\StringTok{ }\NormalTok{pROC}\OperatorTok{::}\KeywordTok{roc}\NormalTok{(}\DataTypeTok{response =}\NormalTok{ testing[,}\DecValTok{1}\NormalTok{],}
                \DataTypeTok{predictor =}\NormalTok{ testProbs[,}\DecValTok{1}\NormalTok{],}
              \CommentTok{## This function assumes that the second}
              \CommentTok{## class is the class of interest, so we}
              \CommentTok{## reverse the labels.}
              \DataTypeTok{levels =} \KeywordTok{rev}\NormalTok{(}\KeywordTok{levels}\NormalTok{(testing[,}\DecValTok{1}\NormalTok{])))}
\CommentTok{# plot the curve}
\KeywordTok{plot}\NormalTok{(rocCurve, }\DataTypeTok{legacy.axes =} \OtherTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/ROC-1} 

}

\caption{ROC curve for k-NN}\label{fig:ROC}
\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# return area under the curve}
\NormalTok{pROC}\OperatorTok{::}\KeywordTok{auc}\NormalTok{(rocCurve)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Area under the curve: 0.999
\end{verbatim}

\hypertarget{model-tuning-and-avoiding-overfitting}{%
\section{Model tuning and avoiding overfitting}\label{model-tuning-and-avoiding-overfitting}}

How can we know we pick the best \(k\)? One straightforward way is that we can try many different \(k\) values and check accuracy of our model. We will first check the effect of different \(k\) on training accuracy. Below, we will go through many \(k\) values and calculate the training accuracy for each.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{101}\NormalTok{)}
\NormalTok{k=}\DecValTok{1}\OperatorTok{:}\DecValTok{12} \CommentTok{# set k values}
\NormalTok{trainErr=}\KeywordTok{c}\NormalTok{() }\CommentTok{# set vector for training errors}
\ControlFlowTok{for}\NormalTok{( i }\ControlFlowTok{in}\NormalTok{ k)\{}
  
\NormalTok{  knnFit=}\KeywordTok{knn3}\NormalTok{(}\DataTypeTok{x=}\NormalTok{training[,}\OperatorTok{-}\DecValTok{1}\NormalTok{], }\CommentTok{# training set}
              \DataTypeTok{y=}\NormalTok{training[,}\DecValTok{1}\NormalTok{], }\CommentTok{# training set class labels}
              \DataTypeTok{k=}\NormalTok{i)}

  \CommentTok{# predictions on the training set}
\NormalTok{  class.res=}\KeywordTok{predict}\NormalTok{(knnFit,training[,}\OperatorTok{-}\DecValTok{1}\NormalTok{],}\DataTypeTok{type=}\StringTok{"class"}\NormalTok{)}

  \CommentTok{# training error}
\NormalTok{  err=}\DecValTok{1}\OperatorTok{-}\KeywordTok{confusionMatrix}\NormalTok{(training[,}\DecValTok{1}\NormalTok{],class.res)}\OperatorTok{$}\NormalTok{overall[}\DecValTok{1}\NormalTok{]}
\NormalTok{  trainErr[i]=err}
\NormalTok{\}}

\CommentTok{# plot training error vs k}
\KeywordTok{plot}\NormalTok{(k,trainErr,}\DataTypeTok{type=}\StringTok{"p"}\NormalTok{,}\DataTypeTok{col=}\StringTok{"#CC0000"}\NormalTok{,}\DataTypeTok{pch=}\DecValTok{20}\NormalTok{)}

\CommentTok{# add a smooth line for the trend}
\KeywordTok{lines}\NormalTok{(}\KeywordTok{loess.smooth}\NormalTok{(}\DataTypeTok{x=}\NormalTok{k, trainErr,}\DataTypeTok{degree=}\DecValTok{2}\NormalTok{),}\DataTypeTok{col=}\StringTok{"#CC0000"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/trainingErrork-1} 

}

\caption{Training error for k-NN classification of glioma tumor samples}\label{fig:trainingErrork}
\end{figure}

We can see the effect of \(k\) in training error, as \(k\) increases the model tends to a bit worse on training. This makes sense because with large \(k\) we take into account more and more neighbors, at some point we start considering data points from the other classes as well and that decreases our accuracy.

However, looking at the training accuracy is not the right way to test the model as we have mentioned. The models are generally tested on the datasets that are not used when building model. There are different strategies to do this. We have already split part of our data set as test, so let us see how we do it on the test data.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{31}\NormalTok{)}
\NormalTok{k=}\DecValTok{1}\OperatorTok{:}\DecValTok{12}
\NormalTok{testErr=}\KeywordTok{c}\NormalTok{()}
\ControlFlowTok{for}\NormalTok{( i }\ControlFlowTok{in}\NormalTok{ k)\{}

\NormalTok{  knnFit=}\KeywordTok{knn3}\NormalTok{(}\DataTypeTok{x=}\NormalTok{training[,}\OperatorTok{-}\DecValTok{1}\NormalTok{], }\CommentTok{# training set}
              \DataTypeTok{y=}\NormalTok{training[,}\DecValTok{1}\NormalTok{], }\CommentTok{# training set class labels}
              \DataTypeTok{k=}\NormalTok{i)}

  \CommentTok{# predictions on the training set}
\NormalTok{  class.res=}\KeywordTok{predict}\NormalTok{(knnFit,testing[,}\OperatorTok{-}\DecValTok{1}\NormalTok{],}\DataTypeTok{type=}\StringTok{"class"}\NormalTok{)}
\NormalTok{  testErr[i]=}\DecValTok{1}\OperatorTok{-}\KeywordTok{confusionMatrix}\NormalTok{(testing[,}\DecValTok{1}\NormalTok{],}
\NormalTok{                                 class.res)}\OperatorTok{$}\NormalTok{overall[}\DecValTok{1}\NormalTok{]}
 
\NormalTok{\}}

\CommentTok{# plot training error}
\KeywordTok{plot}\NormalTok{(k,trainErr,}\DataTypeTok{type=}\StringTok{"p"}\NormalTok{,}\DataTypeTok{col=}\StringTok{"#CC0000"}\NormalTok{,}
     \DataTypeTok{ylim=}\KeywordTok{c}\NormalTok{(}\FloatTok{0.000}\NormalTok{,}\FloatTok{0.08}\NormalTok{),}
     \DataTypeTok{ylab=}\StringTok{"prediction error (1-accuracy)"}\NormalTok{,}\DataTypeTok{pch=}\DecValTok{19}\NormalTok{)}
\CommentTok{# add a smooth line for the trend}
\KeywordTok{lines}\NormalTok{(}\KeywordTok{loess.smooth}\NormalTok{(}\DataTypeTok{x=}\NormalTok{k, trainErr,}\DataTypeTok{degree=}\DecValTok{2}\NormalTok{), }\DataTypeTok{col=}\StringTok{"#CC0000"}\NormalTok{)}

\CommentTok{# plot test error}
\KeywordTok{points}\NormalTok{(k,testErr,}\DataTypeTok{col=}\StringTok{"#00CC66"}\NormalTok{,}\DataTypeTok{pch=}\DecValTok{19}\NormalTok{) }
\KeywordTok{lines}\NormalTok{(}\KeywordTok{loess.smooth}\NormalTok{(}\DataTypeTok{x=}\NormalTok{k,testErr,}\DataTypeTok{degree=}\DecValTok{2}\NormalTok{), }\DataTypeTok{col=}\StringTok{"#00CC66"}\NormalTok{)}
\CommentTok{# add legend}
\KeywordTok{legend}\NormalTok{(}\StringTok{"bottomright"}\NormalTok{,}\DataTypeTok{fill=}\KeywordTok{c}\NormalTok{(}\StringTok{"#CC0000"}\NormalTok{,}\StringTok{"#00CC66"}\NormalTok{),}
       \DataTypeTok{legend=}\KeywordTok{c}\NormalTok{(}\StringTok{"training"}\NormalTok{,}\StringTok{"test"}\NormalTok{),}\DataTypeTok{bty=}\StringTok{"n"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/testTrainErr-1} 

}

\caption{Training and test error for k-NN classification of glioma tumor samples}\label{fig:testTrainErr}
\end{figure}

The test data shows a different thing of course. It is not the best strategy to increase the \(k\) indefinitely. The test error rate increases after a while. Increasing \(k\) results in too many data points influencing the decision about the class of the new sample, this may not be desirable since it this strategy might include points from other classes eventually. On the other hand, if we set \(k\) too low, we are restricting the model to only look for few neighbors.

In addition, \(k\) values that gives the best performance for the training set is not the best \(k\) for the test set.In fact, if we stick with \(k=1\) as the best \(k\) obtained from the training set, we would obtain a worse performance on the test set. In this case, we can talk about the concept of overfitting. This happens when our models are fitting the data in the training set extremely well but can not perform well in the test data, in other words they can not generalize. Similarly, underfitting could occur when our models do not learn well from the training data and they are overly simplistic. Ideally, we should use methods that helps us estimate the real test error when tuning the models such as cross-validation, bootstrap or holdout test set.

\hypertarget{model-complexity-and-bias-variance-trade-off}{%
\subsection{Model complexity and bias variance trade-off}\label{model-complexity-and-bias-variance-trade-off}}

The case of over- and underfitting is closely related to the model complexity and the related bias-variance trade-off. We will introduce these concepts now. First, let us point out that prediction error depends on the real value of the class label of the test case and predicted value. The test case label or value is not dependent on the prediction, the only thing that is variable here is the model. Therefore, if we could train multiple models with different data sets for the same problem, our predictions for the test set would vary. That means, our prediction error would also vary. Now, with this setting we can talk about expected prediction error for a given machine learning model. This is the average error you would get for a test set if you were able to train multiple models. This expected prediction error can largely be decomposed into the variability of the predictions due to the model variability (Variance) and the difference between the expected prediction values and the correct value of the response (Bias). Formally, the expected prediction error,\(E[Error]\) is decomposed as follows:
\[
 E[Error]=Bias^2 + Variance + \sigma_e^2
\]
Note that in the above equation \(\sigma_e^2\) is the irreducible error. This is the noise term that cannot fundamentally be accounted by any model. The bias is formally the difference between the expected prediction value and the correct response value, \(Y\): \(Bias=(Y-E[PredictedValue])\). The variance is simply the variability of the prediction values when we construct models multiple times with different training sets for the same problem: \(Variance=E[(PredictedValue-E[PredictedValue])^2]\). Note that this the value of the variance does not depend of the correct value of the test cases.

The models that have high variance are generally more complex models that have many knobs or parameters than can fit the training data well. These models, due to their flexibility, can fit training data too much that it creates poor prediction performance in a new data set. On the other hand, simple, less complex models do not have the flexibility to fit every data set that well, so they can avoid overfitting. However, they can underfit if they are not flexible enough to model or at least approximate the true relationship between predictors and the response variable. The bias term is mostly about the general model performance (expected or average value of predictions ) that can be attributed to approximating a real life problem with simpler models. These simple models can have less variability in their predictions, then the prediction error will be mostly composed of the bias term.

In reality, there is always a trade-off between bias and variance (See Figure \ref{fig:varBias}). Increasing the variance with complex models will decrease the bias but that might overfit. Conversely, simple models will increase the bias in the expense of the model variance and that might underfit. There is an optimal point for model complexity, a balance between overfitting and underfitting. In practice, there is no analytical way to find this optimal complexity. Instead we must use an accurate measure of prediction error and explore different levels of model complexity and choose the complexity level that minimizes the overall error. Another approach to this is to use ``The one standard error rule''. Instead of choosing the parameter that minimizes the error estimate, we can choose the simplest model whose error estimate is within one standard error of the best model (see chapter 7 of \citep{friedman2001elements}). The rationale behind that is to choose a simple model with the hope that it would perform better in the unseen data since its performance is not different from the best model in a statistically significant way. You might see the option to choose ``one-standard-error'' model in some machine learning packages.

\begin{figure}

{\centering \includegraphics[width=0.8\linewidth]{images/Variance-bias} 

}

\caption{Variance-Bias trade-off visualized as components of total prediction error in relation to model complexity}\label{fig:varBias}
\end{figure}

In our k-NN example, lower \(k\) values creates a more flexible model. This might be counter intuitive but as we have explained before having small \(k\) values will fit the data in very data specific manner. It will probably not generalize well. Therefore in this respect, lower \(k\) values will result in more complex models with high variance. On the other hand, higher \(k\) values will result in less variance but higher bias. Figure \ref{fig:kNNboundary} shows the decision boundary for two different k-NN models with \(k=2\) and \(k=12\), to be able to plot this in 2D we ran the model on principal component 1 and 2 of the training data set, and predicted the class label of many points in this 2D space. As you can see, \(k=2\) creates a more variable model which tries aggressively to include all training samples in the correct class. This creates a high variance model because the model could change drastically from data set to the data set. On the other hand, setting \(k=12\) creates a model with a smoother decision boundary. This model will have less variance since it considers many points for a decision, therefore the decision boundary is smoother.

\begin{figure}

{\centering \includegraphics[width=0.7\linewidth]{images/knnDecisionBoundPCA} 

}

\caption{Decision boundary for different k values in k-NN models. k=12 creates a smooth decision boundary and ignores certain data points on either side of the boundary. k=2 is less smooth and more variable}\label{fig:kNNboundary}
\end{figure}

\hypertarget{data-split-strategies-for-model-tuning-and-testing}{%
\subsection{Data split strategies for model tuning and testing}\label{data-split-strategies-for-model-tuning-and-testing}}

The data split strategy is essential for accurate prediction of the test error. As we have seen in the model complexity/bias-variance discussion, estimating the prediction error is central for model tuning in order to find the model with the right complexity. Therefore, we will revisit this and show how to build and test models, and measure their prediction error in practice.

\hypertarget{training-validation-test}{%
\subsubsection{training-validation-test}\label{training-validation-test}}

This data split strategy is creates three partitions of the data set, training, validation and test sets. In this strategy, training set is used to train the data and the validation set is used to tune the model to the best possible model. The final partition, ``test'', is only used for the final test and should not be used to tune the model, this is regarded as the real world prediction error for your model. This strategy works when you a lot of data to do a three way split. The test set we used above is most likely too small to measure the prediction error with just using a test set. In such cases, bootstrap or cross-validation should yield more stable results.

\hypertarget{cross-validation-1}{%
\subsubsection{cross-validation}\label{cross-validation-1}}

A more realistic approach when you do not have a lot of data to do the three way split is cross-validation. You can use cross-validation in the model tuning phase as well, instead of going a single train-validation split. As with the three-way split, the final prediction error could be estimated with the test set. In other words, we can separate 80\% of the data for model building with cross-validation, and the final model performance will be measured on the test set.

We have already split our glioma data set into training and test set. Now, we will show how to use run a k-NN model with cross-validation using \texttt{caret::train()} function. This function will use cross-validation to train models for different \(k\) values. Every \(k\) value will be trained and tested with cross-validation to estimate prediction performance for each \(k\).

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{17}\NormalTok{)}
\CommentTok{# this method controls everything about training}
\CommentTok{# we will just set up 10-fold cross validation}
\NormalTok{trctrl <-}\StringTok{ }\KeywordTok{trainControl}\NormalTok{(}\DataTypeTok{method =} \StringTok{"cv"}\NormalTok{,}\DataTypeTok{number=}\DecValTok{10}\NormalTok{)}

\CommentTok{# we will now train k-NN model}
\NormalTok{knn_fit <-}\StringTok{ }\KeywordTok{train}\NormalTok{(subtype}\OperatorTok{~}\NormalTok{., }\DataTypeTok{data =}\NormalTok{ training, }
                 \DataTypeTok{method =} \StringTok{"knn"}\NormalTok{,}
                 \DataTypeTok{trControl=}\NormalTok{trctrl,}
                 \DataTypeTok{tuneGrid =} \KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{k=}\DecValTok{1}\OperatorTok{:}\DecValTok{12}\NormalTok{))}

\CommentTok{# best k value by cross-validation accuracy}
\NormalTok{knn_fit}\OperatorTok{$}\NormalTok{bestTune}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##   k
## 6 6
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# plot k vs prediction error}
\KeywordTok{plot}\NormalTok{(}\DataTypeTok{x=}\DecValTok{1}\OperatorTok{:}\DecValTok{12}\NormalTok{,}\DecValTok{1}\OperatorTok{-}\NormalTok{knn_fit}\OperatorTok{$}\NormalTok{results[,}\DecValTok{2}\NormalTok{],}\DataTypeTok{pch=}\DecValTok{19}\NormalTok{,}
     \DataTypeTok{ylab=}\StringTok{"prediction error"}\NormalTok{,}\DataTypeTok{xlab=}\StringTok{"k"}\NormalTok{)}
\KeywordTok{lines}\NormalTok{(}\KeywordTok{loess.smooth}\NormalTok{(}\DataTypeTok{x=}\DecValTok{1}\OperatorTok{:}\DecValTok{12}\NormalTok{,}\DecValTok{1}\OperatorTok{-}\NormalTok{knn_fit}\OperatorTok{$}\NormalTok{results[,}\DecValTok{2}\NormalTok{],}\DataTypeTok{degree=}\DecValTok{2}\NormalTok{),}
      \DataTypeTok{col=}\StringTok{"#CC0000"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/kknCv-1} 

}

\caption{Cross-validated estimate of Prediction error of k in k-NN models}\label{fig:kknCv}
\end{figure}

Based on figure \ref{fig:kknCv} the cross-validation accuracy reveals that \(k=5\) is the best \(k\) value. On the other hand, we can also try bootstrap resampling and check the prediction error that way. We will again use \texttt{caret::trainControl()} function to do the bootstrap sampling and estimate OOB based error. However, for small number of samples like we have in our example the difference between the estimated and the true value of the prediction error can be large.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{17}\NormalTok{)}
\CommentTok{# this method controls everything about training}
\CommentTok{# we will just set up 100 bootstrap samples and for each }
\CommentTok{# bootstrap OOB samples to test the error}
\NormalTok{trctrl <-}\StringTok{ }\KeywordTok{trainControl}\NormalTok{(}\DataTypeTok{method =} \StringTok{"boot"}\NormalTok{,}\DataTypeTok{number=}\DecValTok{20}\NormalTok{,}
                       \DataTypeTok{returnResamp=}\StringTok{"all"}\NormalTok{)}

\CommentTok{# we will now train k-NN model}
\NormalTok{knn_fit <-}\StringTok{ }\KeywordTok{train}\NormalTok{(subtype}\OperatorTok{~}\NormalTok{., }\DataTypeTok{data =}\NormalTok{ training, }
                 \DataTypeTok{method =} \StringTok{"knn"}\NormalTok{,}
                 \DataTypeTok{trControl=}\NormalTok{trctrl,}
                 \DataTypeTok{tuneGrid =} \KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{k=}\DecValTok{1}\OperatorTok{:}\DecValTok{12}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\hypertarget{variable-importance}{%
\section{Variable importance}\label{variable-importance}}

Another important purpose of machine learning models could be to learn which variables are more important for the prediction. This information could lead to potential biological insights or could help design better data collection methods or experiments.

Variable importance metrics can be separated into two groups: those that are model dependent and those are not. Many machine-learning methods comes with built-in variable importance measures. These may be able to incorporate the correlation structure between the predictors into the importance calculation. Model independent methods are not able to use any internal model data. We will go over some model independent strategies below. The model dependent importance measures will be mentioned when we introduce machine learning methods that have built-in variable importance measures.

One simple method for variable importance is to correlate or apply statistical tests to test the association with the predictor variable with the response variable. Variables can be ranked based on the strength of those associations. For classification problems, ROC curves can be computed by thresholding the predictor variable, and for each variable an AUC can be computed. The variables can be ranked based on these values. However, these methods completely ignores how variables would behave in the presence of other variables. \texttt{caret::filterVarImp()} function implements some of these strategies.

If a variable important for prediction removing that variable before model training will cause a drop in performance. With this understanding, we can remove the variables one by one and train models without them and rank them by the loss of performance. The most important variables must cause the largest loss of performance. This strategy requires training and testing models as many times as the number of predictor variables. This will consume a lot of time. A related but more practical approach have been put forward to measure variable importance in a model independent manner but without re-training \citep{dalex, mcr}. In this case, instead of removing the variables at training, variables are permuted at the test phase. The loss in prediction performance is calculated by comparing the labels/values from the original response variable to the labels/values obtained by running the permuted test data through the model. This is called ``variable dropout loss''. In this case, we are not really dropping out variables but by permuting them we destroy their relationship to the response variable. The dropout loss is compared to the ``worst case'' scenario where response variable is permuted and compared against the original response variables, this is called ``baseline loss''. The algorithm ranks the variables by their variable dropout loss or by their ratio of variable dropout to baseline loss. Both quantities are proportional but the second one contains information about the baseline loss.

Below, we are running \texttt{DALEX::explain()} function to do the permutation drop-out strategy for the variables. The function needs the machine learning model, and new data and its labels to do the permutation-based dropout strategy. In this case, we are feeding the function with the data we used for training. For visualization we use \texttt{DALEX::variable\_importance()} function which plots the loss

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(DALEX)}
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{102}\NormalTok{)}
\NormalTok{explainer_knn<-}\StringTok{ }\NormalTok{DALEX}\OperatorTok{::}\KeywordTok{explain}\NormalTok{(knn_fit, }
                               \DataTypeTok{label=}\StringTok{"knn"}\NormalTok{, }
                               \DataTypeTok{data =}\NormalTok{training[,}\OperatorTok{-}\DecValTok{1}\NormalTok{], }
                               \DataTypeTok{y =} \KeywordTok{as.numeric}\NormalTok{(training[,}\DecValTok{1}\NormalTok{]))}

\NormalTok{viknn=}\KeywordTok{variable_importance}\NormalTok{(explainer_knn,}\DataTypeTok{n_sample=}\OperatorTok{-}\DecValTok{1}\NormalTok{)}
\KeywordTok{plot}\NormalTok{(viknn)}
\end{Highlighting}
\end{Shaded}

Although the variable drop-out strategy will still be slow if you have a lot of variables, the upside is that you can use any black-box model as long as you have access to the model to run new predictions. Later sections in this chapter will show methods with built-in variable importance metrics, since these are calculated during training it comes with less of an additional compute cost.

\hypertarget{how-to-deal-with-class-imbalance}{%
\section{How to deal with class imbalance}\label{how-to-deal-with-class-imbalance}}

A common hurdle in many applications of machine learning on genomic data is the large class imbalance. The imbalance refers to relative difference in the sizes of the groups being classified. For example, if we had class imbalance in our example data set we could have much more CIMP samples in the training than noCIMP samples, or the other way around. Another example with severe class imbalance would be enhancer prediction \citep{enhancerImbalance}. Depending on which training data set you use you can have a couple of hundred to thousands of positive examples for enhancer locations in the human genome. In either case, the negative set, ``not enhancer'', set will overwhelm the training, because the human genome is 3 billion base-pairs long and most of that do not overlap with an enhancer annotation. In whatever strategy you pick to build a negative set it will contain much more data points than the positive set. As we have mentioned in the model performance section above, if we have a severe imbalance in the class sizes, the training algorithm may get better accuracy just by calling everything one class. This will be evident in specificity and sensitivity metrics, and the related balanced accuracy metric. Below, we will discuss a couple of techniques that might help when the training set has class imbalance.

\hypertarget{sampling-for-class-balance}{%
\subsection{Sampling for class balance}\label{sampling-for-class-balance}}

If we think class imbalance is a problem based on looking at the relative sizes of the classes, and relevant accuracy metrics of a model there are a couple of things that might help. First, we can try sampling or ``stratified'' sampling when we are constructing our training set. This simply means that before training we can we build the classification model with samples of the data so we have the same size classes. This could be down-sampling the classes with too many data points. For this purpose, you can simply use \texttt{sample()} or \texttt{caret::downSample()} function and create your training set prior to modeling. In addition, minority class could be up-sampled for the missing number of data points using sampling with replacement similar to bootstrap sampling with \texttt{caret::upSample()} function. There are more advanced up-sampling methods such as synthetic up-sampling method SMOTE \citep{smote}. In this method, each data point from the minority class is up-sampled synthetically by adding variability to the predictor variable vector from the one of the k-nearest neighbors of the data point. Specifically, one neighbor is randomly chosen and the difference between predictor variables of the neighbor and the original data point is added to the original predictor variables after multiplying the difference values with a random number between 0 and 1. This creates synthetic data points data are similar to original data points but not identical. This method and other similar methods of synthetic sampling is available at \href{https://cran.r-project.org/web/packages/smotefamily/index.html}{\textbf{smotefamily}} package in CRAN.

In addition to the strategies above, some methods can do sampling during training to cope with the effects of class imbalance. For example, random forests has a sampling step during training, and this step can be altered to do stratified sampling. We will be introducing random forests later in the chapter.

However, even if we are doing the sampling on the training set to avoid problems. The test set proportions should have original class label proportions to evaluate the performance in a real world situation.

\hypertarget{altering-case-weights}{%
\subsection{Altering case weights}\label{altering-case-weights}}

For some methods, we can use different case weights proportional to the imbalance suffered by the minority class. This means cases from minority class will have higher case weights, this causes an effect as if we are up-sampling the minority class. Logistic regression based methods and boosting methods are examples of algorithms that can utilize case weights. Both of which will be introduced later.

\hypertarget{selecting-different-classification-score-cutoffs}{%
\subsection{selecting different classification score cutoffs}\label{selecting-different-classification-score-cutoffs}}

Another simple approach for dealing with class imbalance is to select a prediction score cutoff that minimizes the excess true positives or false positives depending on the direction of the class imbalance. This can simply be done using ROC curves. For example, classical prediction cutoff for 2-class classification problems is 0.5. We can alter this cutoff to optimize sensitivity and specificity.

\hypertarget{dealing-with-correlated-predictors}{%
\section{Dealing with correlated predictors}\label{dealing-with-correlated-predictors}}

Highly correlated predictors can lead to collinearity issues and this can greatly increase the model variance especially in the context of regression. In some cases, there could be relationships between multiple predictor variables and this is called multicollinearity. Having correlated variables will result in unnecessarily complex models with more than necessary predictor variables. From a data collection point of view, spending time and money of collecting correlated variables could be a waste of effort. In terms of linear regression or the models that are based on regression, the collinearity problem is more severe because it creates unstable models where statistical inference becomes difficult or unreliable. On the other hand, correlation between variables may not be a problem for the predictive performance if the correlation structure in the training and the future tests data sets are the same. However, more often correlated structures within the training set might lead to overfitting.

Here are couple of things to do if collinearity is a problem:

\begin{itemize}
\item
  We can do PCA on the training data which creates new variables removing the collinearity between them. We can then training models on these new dimensions. Downside is it is harder to interpret these variables. They are now linear combinations of original variables. The variable importance would be harder to interpret.
\item
  As we have already shown in the data preprocessing section, we can try variable filtering and reduce the number of correlated variables. However, this will may not address the multicollinearity issue where linear combinations of variables might be correlated while themselves are not directly correlated.
\item
  Method specific techniques such as regularization can decrease the effects of collinearity. Regularization as we will see in the later chapter is a technique that is used to prevent overfitting and it can also dampen the effects of collinearity. In addition, decision tree based methods could suffer less from the effects of collinearity.
\end{itemize}

\hypertarget{trees-and-forests-random-forests-in-action}{%
\section{Trees and forests: Random forests in action}\label{trees-and-forests-random-forests-in-action}}

\hypertarget{decision-trees}{%
\subsection{decision trees}\label{decision-trees}}

Decision trees are a popular method for various machine learning tasks mostly because their interpretability is very high. A decision tree is a series of filters on the predictor variables. The series of filters end up in a class prediction. Each filter is binary a yes/no question, this creates bifurcations in the series of filters thus leading to a tree like structure. The filters are dependent on the type of the predictor variables. If the variables are categorical such as gender then the filters could be ``is gender female'' type of questions. If the variables are continuous such as gene expression, the filter could be ``is PIGX expression larger than 210?''. Every point where we filter samples based on these questions are called ``decision nodes''. The tree fitting algorithm finds the best variables at decision nodes depending on how well they split the samples into classes after the application of the decision node. Decision trees handle both categorical and numeric predictor variables, they are easy to interpret and they can deal with missing variables. Despite their advantages, decision trees tend to overfit if they are grown very deep and can learn irregular patterns.

There are many variants of tree based machine learning algorithms. However, most algorithms construct decision nodes in a top down manner. They select the best variables to use in decision nodes based on how homogeneous the sample sets are after the split. One measure of homogeneity is ``Gini impurity''. This measure is calculated for each subset after the split and later summed up as weighted average. For a decision node that splits the data perfectly in a two class problem, the gini impurity will be 0, and for a node that splits the data to a subset that that has 50\% class 2 and 50\% class two the impurity will be 0.5. Formally, gini impurity,\({I}_{G}(p)\), of a set of samples with known class labels for \(K\) classes is the following, where \(p_{i}\) is the probability of observing class \(i\) in the subset:

\[
{\displaystyle {I}_{G}(p)=\sum _{i=1}^{K}p_{i}(1-p_{i})=\sum _{i=1}^{K}p_{i}-\sum _{i=1}^{K}{p_{i}}^{2}=1-\sum _{i=1}^{K}{p_{i}}^{2}}
\]

For example, if a subset of data after split has 75\% class A and 25\% class B for that subset the impurity would be \(1-(0.75^2+0.25^2)=0.375\). If the other subset had 5\% class A and 95\% class B its impurity would be \(1-(0.95^2+0.05^2)=0.095\). If the subset sizes after the split were equal total weighted impurity would be \(0.5*0.375+0.5*0.095= 0.235\). These calculations will be done for each potential variable and the split, and every node will be constructed based on gini impurity decrease. If the variable is continuous the cutoff value will be decided based on the best impurity. For example, gene expression values will have splits such as ``PIGX expression \textless{} 2.1''. Here 2.1 is the cutoff value is the one that produces the best impurity. There are other homogeneity measures, however gini impurity is the one that is used for the random forests which we will introduce next.

\hypertarget{trees-to-forests}{%
\subsection{Trees to forests}\label{trees-to-forests}}

Random forests are devised to counter the short comings of decision trees. They are simply ensembles of decision trees. Each tree is trained with a different randomly selected part of the data with randomly selected predictor variables. The goal of introducing randomness is to reduce the variance of the model so it does not overfit, with the expense of a small increase in the bias and some loss of interpretability. This strategy generally boosts the performance of the final model.

The random forests algorithm tries to decorrelate the trees so that they learn different things about the data. It does this by selecting a random subset of variables. If one or a few predictor variables are very strong predictors for the response variable, these features will be selected in many of the trees, causing them to become correlated. By random subsampling of predictor variables it ensures that not always the best predictors overall will be selected for every tree and the model has a chance to learn other features of the data.

Another sampling method introduce when building random forest models is the bootstrap resampling before constructing each tree. This brings the advantage of out-of-the-bag (OOB) error prediction. In this case, the prediction error can be estimated for training samples that were OOB, meaning they were not used in the training, for some percentage of the trees. The prediction error for each sample can be estimated from the trees where that sample was OOB. OOB estimates claimed to be a good alternative to cross-validation estimated errors \citep{breiman2001random}.

\begin{figure}

{\centering \includegraphics[width=0.7\linewidth]{images/ml-random-forest-features} 

}

\caption{Random forest concept. Individual decision trees are built with sampling strategies. Votes from each tree defines the final class}\label{fig:RFcartoon}
\end{figure}

For the demonstration purposes, we will use \textbf{caret} package interface to \textbf{ranger} random forest package. This is a fast implementation of the original random forest algorithm. For random forests, we have two critical arguments. One is the most critical argument for random forest is the predictor variables to sample in each split of the tree. As this parameter controls the independence between the trees, and as explain before this limits the overfitting. Below, we are going to fit a random forest model to our tumor subtype problem. We will set \texttt{mtry=100} and do not let select the training procedure to find the best \texttt{mtry} value for simplicity. However, it is good practice
to run the model with cross-validation let it pick best parameters based on the cross-validation performance. It defaults to the square root of number of predictor variables. Another variable we can tune is the minimum node size of terminal nodes in the trees (min.node.size). This controls the depth of the trees grown. Setting this to larger numbers might cost a small loss in accuracy but the algorithm will run faster.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{17}\NormalTok{)}

\CommentTok{# we will do no resampling based prediction error}
\CommentTok{# although it is advised to do so even for random forests}
\NormalTok{trctrl <-}\StringTok{ }\KeywordTok{trainControl}\NormalTok{(}\DataTypeTok{method =} \StringTok{"none"}\NormalTok{)}

\CommentTok{# we will now train random forest model}
\NormalTok{rfFit <-}\StringTok{ }\KeywordTok{train}\NormalTok{(subtype}\OperatorTok{~}\NormalTok{., }
               \DataTypeTok{data =}\NormalTok{ training, }
               \DataTypeTok{method =} \StringTok{"ranger"}\NormalTok{,}
               \DataTypeTok{trControl=}\NormalTok{trctrl,}
               \DataTypeTok{importance=}\StringTok{"permutation"}\NormalTok{, }\CommentTok{# calculate importance}
               \DataTypeTok{tuneGrid =} \KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{mtry=}\DecValTok{100}\NormalTok{,}
                                     \DataTypeTok{min.node.size =} \DecValTok{1}\NormalTok{,}
                                     \DataTypeTok{splitrule=}\StringTok{"gini"}\NormalTok{)}
\NormalTok{               )}
\CommentTok{# print OOB error}
\NormalTok{rfFit}\OperatorTok{$}\NormalTok{finalModel}\OperatorTok{$}\NormalTok{prediction.error}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.03846
\end{verbatim}

\hypertarget{variable-importance-1}{%
\subsection{Variable importance}\label{variable-importance-1}}

Random forests comes with built-in variable importance metrics. One of the metrics is similar to the ``variable dropout metric'' where the predictor variables are permuted. In this case, OOB samples are used and the variables are permuted one at a time. Every time, the samples with the permuted variables are fed to the network and decrease in accuracy with the permuted variable is measured. Using this quantity, the variables can be ranked.

A less costly method with similar performance is to use gini impurity. Every time a variable is used in a tree to make a split, the gini impurity is less than the parent node. This method adds up these gini impurity decreases for each individual variable across the trees and divided by number of the trees in the forest. This metric is often consistent with the permutation importance measure \citep{breiman2001random}. Below, we are going to plot the ``permutation'' based importance metric. This metric has been calculated during the run of the model above. We will use the \texttt{caret::varImp()} function to access the importance values and plot them using the \texttt{plot()} function.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{plot}\NormalTok{(}\KeywordTok{varImp}\NormalTok{(rfFit),}\DataTypeTok{top=}\DecValTok{10}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.5\linewidth]{compgenomrReloaded_files/figure-latex/RFvarImp-1} 

}

\caption{Top 10 important variables based on permutation-based method for the random forest classification}\label{fig:RFvarImp}
\end{figure}

\hypertarget{logistic-regression-and-regularization}{%
\section{Logistic regression and regularization}\label{logistic-regression-and-regularization}}

Logistic regression is a statistical method that is used to model a binary response variable based on predictor variables. Although initially devised for two-class or binary response problems, this method can be generalized to multiclass problems. However, our example tumor sample data is a binary response or two-class problem, therefore we will not go into multiclass case in this chapter.

Logistic regression is very similar to the linear regression as a concept and it can be thought of as a ``maximum likelihood estimation'' problem where we are trying to find statistical parameters that maximizes the likelihood of the observed data being sampled from the statistical distribution of interest. Which is also very related to the general cost/loss function approach we see in supervised machine learning algorithms. In the case of binary response variables, simple linear regression model, such as \(y_i \sim \beta _{0}+\beta _{1}x_i\), would be a poor choice because it can easily generate values outside of 0 to 1 boundary. What we need is a
model that restricts the lower bound of the prediction to zero and an upper
bound to 1. First thing towards this requirement is to formulate the problem differently. If \(y_i\) can only be 0 or 1, we can formulate \(y_i\) as a realization of a random variable that can take the values one and zero with probabilities \(p_i\) and \(1-{p_i}\), respectively. This random variable follows the Bernoulli distribution, and instead of predicting the binary variable we can formulate the problem as \(p_i \sim \beta _{0}+\beta _{1}x_i\). However, our initial problem still stands, simple linear regression will still result in values that are beyond 0 and 1 boundary. A model that satisfies the boundary requirement is the logistic equation shown below.
\[
{\displaystyle p_i={\frac {e^{(\beta _{0}+\beta _{1}x_i)}}{1+e^{(\beta _{0}+\beta_{1}x_i)}}}}
\]

This equation can be linearized by the following transformation

\[
{\displaystyle \operatorname{logit} (p_i)=\ln \left({\frac {p_i}{1-p_i}}\right)=\beta _{0}+\beta _{1}x_i}
\]
The left-hand side is termed the logit, which stands for ``logistic unit.'' It is also known as the log odds. In this case, our model will produce values on the log scale and with the logistic equation above, we can transform the values to 0-1 range. Now, the question remains: ``What are the best parameter estimates for our training set''. Within the maximum likelihood framework, the best parameter estimates are the ones that maximizes the likelihood of the statistical the model actually producing the observed data. You can think of this fitting a probability distribution to an observed data set. The parameters of the probability distribution should maximize the likelihood that the observed data came from the distribution of in question. If we were using a Gaussian distribution we would change the mean and variance parameters until the observed data is more plausible to be drawn from that specific Gaussian distribution.

In logistic regression, the response variable is modeled with a binomial distribution or its special case Bernoulli distribution. The value of each response variable, \(y_i\), is 0 or 1, and we need to figure out parameter \(p_i\) values that could generate such a distribution of 0s and 1s. If we can find the best \(p_i\) values for each tumor sample \(i\), we would be maximizing the log-likelihood function of the model over the observed data. The maximum log-likelihood function for our binary response variable case is shown as equation \eqref{eq:logLik}.

\begin{equation}
 \operatorname{\ln} (L)=\sum_{i=1}^N\bigg[{\ln(1-p_i)+y_i\ln \left({\frac {p_i}{1-p_i}}\right)\bigg]}
 \label{eq:logLik}
\end{equation}

In order to maximize this equation we have to find optimum \(p_i\) values which are dependent on parameters \(\beta_0\) and \(\beta_1\), and also dependent on the values of predictor variables \(x_i\). We can rearrange the equation replacing \(p_i\) with the logistic equation. In addition, many optimization functions minimize rather than maximize. Therefore, we will be using negative log likelihood, this also called ``log loss'' or ``logistic loss'' function. The function below is the ``log loss'' function. We substituted \(p_i\) with the logistic equation and simplified the expression.

\begin{equation}
\operatorname L_{log}=-{\ln}(L)=-\sum_{i=1}^N\bigg[-{\ln(1+e^{(\beta _{0}+\beta _{1}x_i)})+y_i \left(\beta _{0}+\beta _{1}x_i\right)\bigg]}
 \label{eq:llog}
\end{equation}

Now, let us see how this works in practice. First, as in the example above we will use one predictor variable, the expression of one gene to classify tumor samples to ``CIMP'' and ``noCIMP'' subtypes. We will be using PDPN gene expression, which was one of the most important variables in our random forest model. We will use the formula interface in \emph{caret}, where we will supply the names of response and and predictor variables in a formula. In this case, the we will be using a core R function \texttt{glm()} from \emph{stats} package. ``glm'' stands for generalized linear models, and it is the main interface for different types of regression
in R.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# fit logistic regression model}
\CommentTok{# method and family defines the type of regression}
\CommentTok{# in this case these arguments mean that we are doing logistic}
\CommentTok{# regression}
\NormalTok{lrFit =}\StringTok{ }\KeywordTok{train}\NormalTok{(subtype }\OperatorTok{~}\StringTok{ }\NormalTok{PDPN,  }
               \DataTypeTok{data=}\NormalTok{training, }\DataTypeTok{trControl=}\KeywordTok{trainControl}\NormalTok{(}\StringTok{"none"}\NormalTok{),}
               \DataTypeTok{method=}\StringTok{"glm"}\NormalTok{, }\DataTypeTok{family=}\StringTok{"binomial"}\NormalTok{)}

\CommentTok{# create data to plot the sigmoid curve}
\NormalTok{newdat <-}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{PDPN=}\KeywordTok{seq}\NormalTok{(}\KeywordTok{min}\NormalTok{(training}\OperatorTok{$}\NormalTok{PDPN),}
                               \KeywordTok{max}\NormalTok{(training}\OperatorTok{$}\NormalTok{PDPN),}\DataTypeTok{len=}\DecValTok{100}\NormalTok{))}

\CommentTok{# predict probabilities for the simulated data}
\NormalTok{newdat}\OperatorTok{$}\NormalTok{subtype =}\StringTok{ }\KeywordTok{predict}\NormalTok{(lrFit, }\DataTypeTok{newdata=}\NormalTok{newdat, }\DataTypeTok{type=}\StringTok{"prob"}\NormalTok{)[,}\DecValTok{1}\NormalTok{]}

\CommentTok{# plot the sigmoid curve and the training data}
\KeywordTok{plot}\NormalTok{(}\KeywordTok{ifelse}\NormalTok{(subtype}\OperatorTok{==}\StringTok{"CIMP"}\NormalTok{,}\DecValTok{1}\NormalTok{,}\DecValTok{0}\NormalTok{) }\OperatorTok{~}\StringTok{ }\NormalTok{PDPN, }
     \DataTypeTok{data=}\NormalTok{training, }\DataTypeTok{col=}\StringTok{"red4"}\NormalTok{,}
     \DataTypeTok{ylab=}\StringTok{"subtype as 0 or 1"}\NormalTok{, }\DataTypeTok{xlab=}\StringTok{"PDPN expression"}\NormalTok{)}
\KeywordTok{lines}\NormalTok{(subtype }\OperatorTok{~}\StringTok{ }\NormalTok{PDPN, newdat, }\DataTypeTok{col=}\StringTok{"green4"}\NormalTok{, }\DataTypeTok{lwd=}\DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/logReg1-1} 

}

\caption{Sigmoid curve for prediction of subtype based on one predictor variable}\label{fig:logReg1}
\end{figure}

The figure \ref{fig:logReg1} shows the sigmoidal curve that is fitted by the logistic regression. ``noCIMP'' subtype has higher expression of PDPN gene than the ``CIMP'' subtype. In other words, the higher the values of PDPN, the more likely that the tumor sample will be classified as ``noCIMP''. We can also assess the performance of our model with the test set and the training set. Let us try to do that with again \texttt{caret::predict()} and \texttt{caret::confusionMatrix()} functions.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# training accuracy }
\NormalTok{class.res=}\KeywordTok{predict}\NormalTok{(lrFit,training[,}\OperatorTok{-}\DecValTok{1}\NormalTok{])}
\KeywordTok{confusionMatrix}\NormalTok{(training[,}\DecValTok{1}\NormalTok{],class.res)}\OperatorTok{$}\NormalTok{overall[}\DecValTok{1}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Accuracy 
##   0.9231
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# test accuracy }
\NormalTok{class.res=}\KeywordTok{predict}\NormalTok{(lrFit,testing[,}\OperatorTok{-}\DecValTok{1}\NormalTok{])}
\KeywordTok{confusionMatrix}\NormalTok{(testing[,}\DecValTok{1}\NormalTok{],class.res)}\OperatorTok{$}\NormalTok{overall[}\DecValTok{1}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Accuracy 
##   0.8519
\end{verbatim}

The test accuracy is slightly worse than the training accuracy. Overall this is not as good as k-NN, but remember we used only one predictor variable. We have thousands of genes as predictor variables. Now we will try to use all of them in the classification problem. After fitting model we will check training and test accuracy. We fit the model again with \texttt{caret::train()} function.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{lrFit2 =}\StringTok{ }\KeywordTok{train}\NormalTok{(subtype }\OperatorTok{~}\StringTok{ }\NormalTok{.,  }
                \DataTypeTok{data=}\NormalTok{training, }
                \CommentTok{# no model tuning with sampling}
                \DataTypeTok{trControl=}\KeywordTok{trainControl}\NormalTok{(}\StringTok{"none"}\NormalTok{),}
                \DataTypeTok{method=}\StringTok{"glm"}\NormalTok{, }\DataTypeTok{family=}\StringTok{"binomial"}\NormalTok{)}

\CommentTok{# training accuracy }
\NormalTok{class.res=}\KeywordTok{predict}\NormalTok{(lrFit2,training[,}\OperatorTok{-}\DecValTok{1}\NormalTok{])}
\KeywordTok{confusionMatrix}\NormalTok{(training[,}\DecValTok{1}\NormalTok{],class.res)}\OperatorTok{$}\NormalTok{overall[}\DecValTok{1}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Accuracy 
##        1
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# test accuracy }
\NormalTok{class.res=}\KeywordTok{predict}\NormalTok{(lrFit2,testing[,}\OperatorTok{-}\DecValTok{1}\NormalTok{])}
\KeywordTok{confusionMatrix}\NormalTok{(testing[,}\DecValTok{1}\NormalTok{],class.res)}\OperatorTok{$}\NormalTok{overall[}\DecValTok{1}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Accuracy 
##   0.5556
\end{verbatim}

Training accuracy is 1 so training error is 0, nothing is missclassified in the training set. However, test accuracy/error is close to terrible. It does only little better than a random guess. If we randomly assigned class labels we would get 0.5 accuracy. The test set accuracy is 0.55 despite the 100\% training accuracy. This is because the model overfits to the training data. There are too many variables in the model. The number of predictor variables is \textasciitilde{}6.5 times more than the number of samples. The excess of predictor variables makes the model very flexible (high variance), and this leads to overfitting.

\hypertarget{regularization-in-order-to-avoid-overfitting}{%
\subsection{regularization in order to avoid overfitting}\label{regularization-in-order-to-avoid-overfitting}}

If we can limit the flexibility of the model, this might help with performance on the unseen, new data sets. Generally, any modification of the learning method to improve performance on the unseen datasets is called regularization. We need regularization to introduce bias to the model and to decrease the variance. This can be achieved by modifying the loss function with a penalty term which effectively shrinks the estimates of the coefficients. Therefore these types of methods within the framework of regression are also called ``shrinkage'' methods or ``penalized regression'' methods.

One way to ensure shrinkage is to add the penalty term, \(\lambda\sum{\beta_j}^2\), to the loss function. This penalty term is also known as The L2 norm or L2 penalty. It is calculated as the square root of the sum of the squared vector values. This term will help shrink the coefficients in the regression towards zero. The new loss function is as follows, where \(j\) is the number of parameters/coefficients in the model and \(L_{log}\) is the log loss function in Eq. \eqref{eq:llog}.

\begin{equation}
L_{log}+\lambda\sum_{j=1}^p{\beta_j}^2
\label{eq:L2norm}
\end{equation}

This penalized loss function is called ``ridge regression'' \citep{hoerl1970ridge}. When we add the penalty, the only way the optimization procedure keeps the overall loss function minimum is to assign smaller values to the coefficients. The \(\lambda\) parameter controls how much emphasis is given to the penalty term. Higher the \(\lambda\) value, the more coefficients in the regression will be pushed towards zero. However, they will never be exactly zero. This is not desirable if we want the model to select important variables. A small modification to the penalty is to use the absolute values of \(B_j\) instead of squared values. This penalty is called ``L1 norm'' or ``L1 penalty''. The regression method that uses the L1 penalty is known as ``Lasso regression'' \citep{tibshirani1996regression}.

\[
L_{log}+\lambda\sum_{j=1}^p{|\beta_j}|
\]
However, the L1 penalty tends to pick one variable at random when predictor variables are correlated. In this case, it looks like one of the variables are not important although it might still have predictive power. The Ridge regression on the other hand shrinks coefficients of correlated variables towards each other, keeping all of them. It has been shown that both Lasso and Ridge regression has their drawbacks and advantages {[}friedman2010regularization{]}. More recently, a method called ``elastic net'' is proposed to include best of the both worlds \citep{zou2005regularization}. This method uses both L1 and L2 penalties. The equation below shows the modified loss function by this penalty. As you can see the \(\lambda\) parameter still controls the weight that is given to the penalty. This time the additional parameter \(\alpha\) controls the weight given to L1 or L2 penalty and it is a value between 0 and 1.
\[
L_{log}+\lambda\sum_{j=1}^p{(\alpha\beta_j^2+(1-\alpha)|\beta_j}|)
\]

We have now got the concept behind regularization and we can see how it works in practice. We are going to use elastic net on our tumor subtype prediction problem. We will let cross-validation select the best \(\lambda\) and we will fix the \(\alpha\) parameter at 0.5.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{17}\NormalTok{)}
\KeywordTok{library}\NormalTok{(glmnet)}

\CommentTok{# this method controls everything about training}
\CommentTok{# we will just set up 10-fold cross validation}
\NormalTok{trctrl <-}\StringTok{ }\KeywordTok{trainControl}\NormalTok{(}\DataTypeTok{method =} \StringTok{"cv"}\NormalTok{,}\DataTypeTok{number=}\DecValTok{10}\NormalTok{)}

\CommentTok{# we will now train elastic net model}
\CommentTok{# it will try}
\NormalTok{enetFit <-}\StringTok{ }\KeywordTok{train}\NormalTok{(subtype}\OperatorTok{~}\NormalTok{., }\DataTypeTok{data =}\NormalTok{ training, }
                 \DataTypeTok{method =} \StringTok{"glmnet"}\NormalTok{,}
                 \DataTypeTok{trControl=}\NormalTok{trctrl,}
                 \CommentTok{# alpha and lambda paramters to try}
                 \DataTypeTok{tuneGrid =} \KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{alpha=}\FloatTok{0.5}\NormalTok{,}
                                       \DataTypeTok{lambda=}\KeywordTok{seq}\NormalTok{(}\FloatTok{0.1}\NormalTok{,}\FloatTok{0.7}\NormalTok{,}\FloatTok{0.05}\NormalTok{)))}

\CommentTok{# best alpha and lambda values by cross-validation accuracy}
\NormalTok{enetFit}\OperatorTok{$}\NormalTok{bestTune}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##   alpha lambda
## 1   0.5    0.1
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# test accuracy }
\NormalTok{class.res=}\KeywordTok{predict}\NormalTok{(enetFit,testing[,}\OperatorTok{-}\DecValTok{1}\NormalTok{])}
\KeywordTok{confusionMatrix}\NormalTok{(testing[,}\DecValTok{1}\NormalTok{],class.res)}\OperatorTok{$}\NormalTok{overall[}\DecValTok{1}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Accuracy 
##        1
\end{verbatim}

As you can see regularization worked, the tuning step selected \(\lambda=1\) and we were able to get a satisfactory test set accuracy with the best model.

\hypertarget{variable-importance-2}{%
\subsection{variable importance}\label{variable-importance-2}}

The variable importance for the penalized regression especially for lasso and elastic net is more or less out of the box. As discussed, these methods will set regression coefficients for irrelevant variables to zero. This provides a system for selecting important variables but it does not necessarily provide a way to rank them. Using the size of the regression coefficients is a way to rank predictor variables, however if the data is not normalized you will get different scales for different variables. In our case, we normalized the data and we know that the variables have the same scale before they went into the training. We can use this fact and rank them based on the regression coefficients. The \texttt{caret::varImp()} function uses the coefficients to rank the variables from the elastic net model. Below, were going to plot top 10 important variables which are normalized to the importance of the most important variable.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{plot}\NormalTok{(}\KeywordTok{varImp}\NormalTok{(enetFit),}\DataTypeTok{top=}\DecValTok{10}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/varImpEnet-1} 

}

\caption{Variable importance metric for elastic net. This metric is using regression coefficients as importance}\label{fig:varImpEnet}
\end{figure}

\hypertarget{want-to-know-more}{%
\subsubsection{Want to know more}\label{want-to-know-more}}

\begin{itemize}
\tightlist
\item
  Lecture by Trevor Hastie on regularized regression. You probably need to understand basics of regression and its terminology to follow this. However, the lecture is not very heavy on math. \url{https://youtu.be/BU2gjoLPfDc}
\end{itemize}

\hypertarget{other-supervised-algorithms}{%
\section{Other supervised algorithms}\label{other-supervised-algorithms}}

We will next introduce a couple of other supervised algorithms for completeness but in less detail. These algorithms are also as popular as the others we introduced above and people who are interested in computational genomics see them used in the field for different problems. These algorithms also fit to the general framework of optimization of a cost/loss function. However, the approaches to the construction of the cost function and the cost function itself is different in each case.

\hypertarget{gradient-boosting}{%
\subsection{Gradient boosting}\label{gradient-boosting}}

Gradient boosting is a prediction model that uses an ensemble of decision trees similar to random forest. However, the decision trees are added sequentially. These models is why this models are also called ``Multiple Additive Regression Trees (MART)'' \citep{friedman2003mart}. Apart from this, you will see similar methods named as ``Gradient boosting machines (GBM)''\citep{friedman2001gbm} or ``Boosted regression trees (BRT)'' \citep{elith2008brt} in the literature.

Generally, ``boosting'' refers to an iterative learning approach where each new model tries to focus on data points where previous ensemble of simple models did not predict well. Gradient boosting is an improvement over that, where each new model tries to focus on the residual errors (prediction error for the current ensemble of models) of the previous model. Specifically in gradient boosting, the simple models are trees. As in random forests, many trees are grown but in this case, trees are sequentially grown and each tree is focusing on fixing the shortcomings of the previous trees. One of the most widely used algorithms for gradient boosting is XGboost which stands for ``extreme gradient boosting''\citep{chen2016xgboost}. Below we will demonstrate how to use this on our problem. XGboost as well as other gradient boosting methods has many parameters to regularize and optimize the complexity of the model. Finding the best parameters for your problem might take some time. However, this flexibility comes with benefits, methods depending on XGboost have won many machine learning competitions boosting \citep{chen2016xgboost}.

The most important parameters are number of trees (nrounds), tree depth (max\_depth) and learning rate or shrinkage (eta). Generally, the more trees we have the better the algorithm will learn because each tree tries to fix misclassification errors that previous tree ensemble could not perform. Having too many trees might cause overfitting. However, learning rate parameter, eta, combats that by shrinking the contribution of each new tree. This can be set to lower values if you have many trees. You can either set a large number of trees and then tune the model with the learning rate parameter or set the learning rate low, say to 0.01 or 0.1 and tune the number of trees. Similarly, tree depth is also controls for overfitting. The deeper the tree, the more usually it will overfit. This has to be tuned as well, the default is at 6. You can try to explore a range around the default. Apart from these, as in random forests, you can subsample the training data and/or the predictive variables. These strategies also can help you counter overfitting.

We are now going to use XGboost with caret package on our cancer subtype classification problem. We are going to try different learning rate parameters. In this instance, we also subsample the dataset before we train each tree. The ``subsample'' parameter controls this and we set this to be 0.5, which means that before we train a tree we will sample 50\% of the data and use only that portion to train the tree.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(xgboost)}
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{17}\NormalTok{)}

\CommentTok{# we will just set up 5-fold cross validation}
\NormalTok{trctrl <-}\StringTok{ }\KeywordTok{trainControl}\NormalTok{(}\DataTypeTok{method =} \StringTok{"cv"}\NormalTok{,}\DataTypeTok{number=}\DecValTok{5}\NormalTok{)}

\CommentTok{# we will now train elastic net model}
\CommentTok{# it will try}
\NormalTok{gbFit <-}\StringTok{ }\KeywordTok{train}\NormalTok{(subtype}\OperatorTok{~}\NormalTok{., }\DataTypeTok{data =}\NormalTok{ training, }
                 \DataTypeTok{method =} \StringTok{"xgbTree"}\NormalTok{,}
                 \DataTypeTok{trControl=}\NormalTok{trctrl,}
                 \CommentTok{# paramters to try}
                 \DataTypeTok{tuneGrid =} \KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{nrounds=}\DecValTok{200}\NormalTok{,}
                                       \DataTypeTok{eta=}\KeywordTok{c}\NormalTok{(}\FloatTok{0.05}\NormalTok{,}\FloatTok{0.1}\NormalTok{,}\FloatTok{0.3}\NormalTok{),}
                                       \DataTypeTok{max_depth=}\DecValTok{4}\NormalTok{,}
                                       \DataTypeTok{gamma=}\DecValTok{0}\NormalTok{,}
                                       \DataTypeTok{colsample_bytree=}\DecValTok{1}\NormalTok{,}
                                       \DataTypeTok{subsample=}\FloatTok{0.5}\NormalTok{,}
                                       \DataTypeTok{min_child_weight=}\DecValTok{1}\NormalTok{))}
                                       

\CommentTok{# best parameters by cross-validation accuracy}
\NormalTok{gbFit}\OperatorTok{$}\NormalTok{bestTune}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##   nrounds max_depth  eta gamma colsample_bytree
## 1     200         4 0.05     0                1
##   min_child_weight subsample
## 1                1       0.5
\end{verbatim}

Similar to random forests, we can estimate the variable importance for gradient boosting using the improvement in gini impurity or other performance relates metrics every time a variable is selected in a tree. Again, \texttt{caret::varImp} function can be used to plot the importance metrics.

\hypertarget{want-to-know-more-1}{%
\subsubsection{want to know more\ldots{}}\label{want-to-know-more-1}}

\begin{itemize}
\tightlist
\item
  \href{https://xgboost.readthedocs.io/en/latest/tutorials/model.html}{More background on gradient boosting and XGboost}. This explains the cost/loss function and regularization in more detail.
\item
  \href{https://youtu.be/wPqtzj5VZus}{Lecture on Gradient boosting and random forests by Trevor Hastie}
\end{itemize}

\hypertarget{support-vector-machines-svm}{%
\subsection{Support Vector Machines (SVM)}\label{support-vector-machines-svm}}

Support vector machines (SVM) are popularized in the 90s due the efficiency and the performance of the algorithm \citep{boser1992svm}. The algorithm works by identifying the optimal decision boundary that separates the data points to different groups (or classes), and then predicts the class of new observations based on this separation boundary. Depending on the situation, the different groups might be separable by a linear straight line or by a non-linear boundary line or plane. If you review k-NN decision boundaries in figure \ref{fig:kNNboundary}, you can see that the decision boundary is not linear. SVM can deal with the linear or non-linear decision boundaries.

First, SVM can map the data to higher dimensions where the decision boundary can be linear. This is achieved by applying certain mathematical functions, called ``kernel functions'', to the predictor variable space. For example, a second degree polynomial can be applied to predictor variables which creates new variables and in this new space the problem is linearly separable.

Second, SVM not only tries to find a decision boundary but tries to find the boundary with largest buffer zone on the sides of the boundary. Having a boundary with a large buffer or ``margin'' ,as it is formally called, will perform better for the new data points not used in the model training. In addition, SVM calculates the decision boundary with some error toleration. As we have seen it may not be always possible to find a linear boundary that perfectly separates the classes. SVM tolerates some degree of error, as in data points on the wrong side of the decision boundary.

Another important feature of the algorithm is that SVM decides on the decision boundary by only relying on the ``landmark'' data points, formally known as ``support vectors''. These are points that are closest to the decision boundary and harder to classify. By keeping track of such points only for decision boundary creation, the computational complexity of the algorithm is reduced. However, this depends on the margin or the buffer zone. If we have large margin than there are many landmark points. The extent of the margin is also related to the variance-bias trade off, the small margin, the classification will try to find a boundary that makes less errors in the training set therefore might overfit. If the margin is larger, it will tolerate more errors in the training set and might generalize better. Practically, this is controlled by the ``C'' or ``Cost'' parameter in SVM example we will show below. Another important choice we will make is the kernel function. Below we are using the radial basis kernel function. This function provides extra predictor dimension where the problem is linearly separable. The model we will use has only one parameter, which is ``C''. It is recommended that \(C\) is in the form of \(2^k\) where \(k\) is in the range of -5 and 15 \citep{hsu2003practical}. Another parameter that can be tuned is related to the radial basis function called ``sigma''. Smaller sigma means less bias and more variance while larger sigma means less variance and more bias. Again, exponential sequences are recommended for tuning that \citep{hsu2003practical}. We will set it to 1 for demonstration purposes below.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#svm code here}
\KeywordTok{library}\NormalTok{(kernlab)}
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{17}\NormalTok{)}

\CommentTok{# we will just set up 5-fold cross validation}
\NormalTok{trctrl <-}\StringTok{ }\KeywordTok{trainControl}\NormalTok{(}\DataTypeTok{method =} \StringTok{"cv"}\NormalTok{,}\DataTypeTok{number=}\DecValTok{5}\NormalTok{)}

\CommentTok{# we will now train elastic net model}
\CommentTok{# it will try}
\NormalTok{svmFit <-}\StringTok{ }\KeywordTok{train}\NormalTok{(subtype}\OperatorTok{~}\NormalTok{., }\DataTypeTok{data =}\NormalTok{ training, }
                \CommentTok{# this SVM used radial basis function}
                 \DataTypeTok{method =} \StringTok{"svmRadial"}\NormalTok{, }
                 \DataTypeTok{trControl=}\NormalTok{trctrl,}
                \DataTypeTok{tuneGrid=}\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{C=}\KeywordTok{c}\NormalTok{(}\FloatTok{0.25}\NormalTok{,}\FloatTok{0.5}\NormalTok{,}\DecValTok{1}\NormalTok{),}
                                    \DataTypeTok{sigma=}\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\hypertarget{want-to-know-more-2}{%
\subsubsection{Want to know more\ldots{}}\label{want-to-know-more-2}}

\begin{itemize}
\tightlist
\item
  \href{https://youtu.be/_PwhiWxHK8o}{MIT lecture by Patrick Winston on SVM}. This lecture explains the concept with some mathematical background. It is not hard to follow. You should be able to follow this if you know what vectors are and if you have some knowledge on derivatives and basic algebra.
\item
  \href{https://cs.stanford.edu/people/karpathy/svmjs/demo/}{Online demo for SVM}. You can play with sigma and C parameters for radial basis SVM and see how they affect the decision boundary.
\end{itemize}

\hypertarget{neural-networks-and-deep-versions-of-it}{%
\subsection{Neural networks and deep versions of it}\label{neural-networks-and-deep-versions-of-it}}

Neural networks are another popular machine learning method which is recently regaining popularity. The earlier versions of the algorithm were popularized in the 80s and 90s. The advantage of neural networks is like SVM, they can model non-linear decision boundaries. The basic idea of neural networks is to combine the predictor variables in order to model the response variable as a non-linear function. In a neural network, input variables pass through several layers that combine the variables and transform those combinations and recombine outputs depending on how many layers the network has. In the conceptual example in Figure @\ref(fig:neuralNetDiagram) the input nodes receive predictor variables and make linear combinations of them in the form of \(\sum ( w_ixi +b)\). Simply put, the variables are multiplied with weights and summed up. This is what we by ``linear combination''. These quantities are further fed into another layer called hidden layer where an activation function is applied on the sums. And these results are further fed into an output node which outputs class probabilities assuming we are working on a classification algorithm. There could be many more hidden layers that will even further combine the output from hidden layers before them. The algorithm in the end also has a cost function similar to the logistic regression cost function but it now has to estimate all the weight parameters: \(w_i\). This is a more complicated problem than logistic regression because of number of parameters to be estimated but neural networks are able to fit complex functions due their parameter space flexibility as well.

\begin{figure}

{\centering \includegraphics[width=0.8\linewidth]{images/neuralNetDiagram} 

}

\caption{Diagram for a simple neural network, their combinations pass through hidden layers and combined again for the output. Predictor variables are fed to the network and weights are adjusted to optimize the cost function}\label{fig:neuralNetDiagram}
\end{figure}

In practical sense, the number of nodes in the hidden layer (size) and some regularization on the weights can be applied to control for overfitting. This is called the calculated (decay) parameter controls for overfitting.

We will train a simple neural network on our cancer data set. In this simple example, the network architecture is somewhat fixed. We can only choose number of nodes (denoted by ``size'') in the hidden layer and a regularization parameter ( denoted by ``decay''). Increasing the number of nodes in hidden layer or in other implementations increasing the number of hidden layers will help model non-linear relationships but can overfit. One way to combat that is to limit the number of nodes in the hidden layer, another way is to regularize the weights. The decay parameter does just that, it penalizes the loss function by \(decay(weigths^2)\). In the example below, we are trying 1 or 2 nodes in the hidden layer in the interest of simplicity and run-time. In addition, we are setting \texttt{decay=0}, which will correspond to not doing any regularization.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#svm code here}
\KeywordTok{library}\NormalTok{(nnet)}
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{17}\NormalTok{)}

\CommentTok{# we will just set up 5-fold cross validation}
\NormalTok{trctrl <-}\StringTok{ }\KeywordTok{trainControl}\NormalTok{(}\DataTypeTok{method =} \StringTok{"cv"}\NormalTok{,}\DataTypeTok{number=}\DecValTok{5}\NormalTok{)}

\CommentTok{# we will now train neural net model}
\CommentTok{# it will try}
\NormalTok{nnetFit <-}\StringTok{ }\KeywordTok{train}\NormalTok{(subtype}\OperatorTok{~}\NormalTok{., }\DataTypeTok{data =}\NormalTok{ training, }
                 \DataTypeTok{method =} \StringTok{"nnet"}\NormalTok{,}
                 \DataTypeTok{trControl=}\NormalTok{trctrl,}
                 \DataTypeTok{tuneGrid=}\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{size=}\DecValTok{1}\OperatorTok{:}\DecValTok{2}\NormalTok{,}\DataTypeTok{decay=}\DecValTok{0}
\NormalTok{                                      ),}
                 \CommentTok{# this is maximum number of weights}
                 \CommentTok{# needed for the nnet method}
                 \DataTypeTok{MaxNWts=}\DecValTok{2000}\NormalTok{) }
\end{Highlighting}
\end{Shaded}

The example we used above is a bit outdated. The modern ``deep'' neural networks provides much more flexibility in the number of nodes, number of layers and regularization options. In many areas especially computer vision deep neural networks are the state-of-the-art \citep{lecun2015deep}. These modern implementations of neural networks are available in R via \textbf{keras} package and also can be trained via the \textbf{caret} package with the similar interface we have shown until now.

\hypertarget{want-to-know-more-3}{%
\subsubsection{want to know more\ldots{}}\label{want-to-know-more-3}}

\begin{itemize}
\tightlist
\item
  \href{https://keras.rstudio.com/}{Deep neural networks in R}. There are examples and background information on deep neural networks.
\item
  \href{https://cs.stanford.edu/~karpathy/svmjs/demo/demonn.html}{Online demo for neural networks}. You can see the affect of number of hidden layers and number of nodes on the decision boundary.
\end{itemize}

\hypertarget{ensemble-learning}{%
\subsection{Ensemble learning}\label{ensemble-learning}}

Ensemble learning models are simply combinations of different machine learning models. By now, we already introduced the concept of ensemble learning in random forests and gradient boosting. However, this concept can be generalized to combining all kinds of different models. ``random forests'' is an ensemble of the same type of models, decision trees. We can also have ensembles of different types of models. For example, we can combine random Forest, k-NN and elastic net models, and make class predictions based on the votes from those different models. Below, we are showing how to do this. We are going to get predictions for three different models on the test set, use majority voting to decide on the class label and then check performance using \texttt{caret::confusionMatrix()}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# predict with k-NN model}
\NormalTok{knnPred=}\KeywordTok{as.character}\NormalTok{(}\KeywordTok{predict}\NormalTok{(knnFit,testing[,}\OperatorTok{-}\DecValTok{1}\NormalTok{],}\DataTypeTok{type=}\StringTok{"class"}\NormalTok{))}
\CommentTok{# predict with elastic Net model}
\NormalTok{enetPred=}\KeywordTok{as.character}\NormalTok{(}\KeywordTok{predict}\NormalTok{(enetFit,testing[,}\OperatorTok{-}\DecValTok{1}\NormalTok{]))}
\CommentTok{# predict with random forest model}
\NormalTok{rfPred=}\KeywordTok{as.character}\NormalTok{(}\KeywordTok{predict}\NormalTok{(rfFit,testing[,}\OperatorTok{-}\DecValTok{1}\NormalTok{]))}

\CommentTok{# do voting for class labels}
\CommentTok{# code finds the most frequent class label per row}
\NormalTok{votingPred=}\KeywordTok{apply}\NormalTok{(}\KeywordTok{cbind}\NormalTok{(knnPred,enetPred,rfPred),}\DecValTok{1}\NormalTok{,}\ControlFlowTok{function}\NormalTok{(x) }\KeywordTok{names}\NormalTok{(}\KeywordTok{which.max}\NormalTok{(}\KeywordTok{table}\NormalTok{(x))))}

\CommentTok{# check accuracy}
\KeywordTok{confusionMatrix}\NormalTok{(}\DataTypeTok{data=}\NormalTok{testing[,}\DecValTok{1}\NormalTok{],}
                \DataTypeTok{reference=}\KeywordTok{as.factor}\NormalTok{(votingPred))}\OperatorTok{$}\NormalTok{overall[}\DecValTok{1}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Accuracy 
##        1
\end{verbatim}

In the test set, we were able to obtain perfect accuracy after voting. More complicated and accurate ways to build ensembles exist. We could also use mean of class probabilities instead of voting for final class predictions. We can even combine models in a regression based scheme to assign weights to the votes or to the predicted class probabilities of each model. In these cases, the prediction performance of the ensembles can also be tested with sampling techniques such as cross-validation. You can think of this as another layer of optimization or modeling for combining results from different models. We will not pursue this further in this chapter but packages such as \href{https://cran.r-project.org/web/packages/caretEnsemble/}{\textbf{caretEnsemble}}, \href{https://cran.r-project.org/web/packages/SuperLearner/index.html}{\textbf{SuperLearner}} or \href{https://mlr.mlr-org.com/}{\textbf{mlr}} can combine models in various ways described above.

\hypertarget{predicting-continuous-variables-regression-with-machine-learning}{%
\section{Predicting continuous variables: regression with machine learning}\label{predicting-continuous-variables-regression-with-machine-learning}}

Until now, we only considered methods that can help us predict class labels. However, all the methods we have shown can also be used to predict continuous variables. In this case, the methods will try to optimize the prediction in error which is usually in the form of sum of squared errors (SSE): \(SSE=\sum (Y-f(X))^2\), where \(Y\) is the continuous response variable and \(f(X)\) is the outcome of the machine learning model.

In this section, we are going to show how to use a supervised learning method for regression. All the methods we have introduced previously in the context of classification can also do regression. Technically, this is just a simple change in the cost function format and optimization step still tries to optimize the parameters of the cost function. In many cases, if your response variable is numeric methods in the \textbf{caret} package will automatically apply regression.

\hypertarget{use-case-predicting-age-from-dna-methylation}{%
\subsection{Use case: Predicting age from DNA methylation}\label{use-case-predicting-age-from-dna-methylation}}

We will demonstrate random forest regression using a different data set which has a continuous response variable. This time we are going to try to predict age of individuals from their DNA methylation levels. Methylation is a DNA modification which has implications in gene regulation and cell state. We are going to introduce methylation in depth in chapter \ref{bsseq}, however for now what we need to know is that there are 24 million CpG dinucleotides in the human genome. Their methylation status can be measured with quantitative assays and the value is between 0 and 1. If it is 0, the CpG is not methylated in any of the cells in the sample, and if it is 1, the CpG is methylated in all the cells of the sample. It has been shown that methylation is predictive of the age of the individual that the sample is taken from \citep{numata2012dna, horvath2013dna}. Now, we will try to test that with a data set containing hundreds of individuals, their age and methylation values for \textasciitilde{}27000 CpGs. We first read in the files and construct a training set.

\hypertarget{reading-and-processing-the-data}{%
\subsection{reading and processing the data}\label{reading-and-processing-the-data}}

Let us first read in the data. When we run the summary and histogram we see that the methylation values are between 0 and 1 and there are 108 samples (See Figure @\ref(fig:readMethAge)). Typically, methylation values have bimodal distribution. In this case many of them have values around 0 and second most frequent value bracket is around 0.9.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# file path for CpG methylation and age}
\NormalTok{fileMethAge=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
                      \StringTok{"CpGmeth2Age.rds"}\NormalTok{,}
                      \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}

\CommentTok{# read methylation-age table}
\NormalTok{ameth=}\KeywordTok{readRDS}\NormalTok{(fileMethAge)}
\KeywordTok{dim}\NormalTok{(ameth)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]   108 27579
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{summary}\NormalTok{(ameth[,}\DecValTok{1}\OperatorTok{:}\DecValTok{3}\NormalTok{])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##       Age         cg26211698       cg03790787    
##  Min.   :-0.5   Min.   :0.0122   Min.   :0.0500  
##  1st Qu.:-0.4   1st Qu.:0.0188   1st Qu.:0.0782  
##  Median :18.9   Median :0.0227   Median :0.0896  
##  Mean   :25.9   Mean   :0.0248   Mean   :0.0930  
##  3rd Qu.:49.6   3rd Qu.:0.0289   3rd Qu.:0.1042  
##  Max.   :83.6   Max.   :0.0488   Max.   :0.1627
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# plot histogram of methylation values}
\KeywordTok{hist}\NormalTok{(}\KeywordTok{unlist}\NormalTok{(ameth[,}\OperatorTok{-}\DecValTok{1}\NormalTok{]),}\DataTypeTok{border=}\StringTok{"white"}\NormalTok{,}
      \DataTypeTok{col=}\StringTok{"cornflowerblue"}\NormalTok{,}\DataTypeTok{main=}\StringTok{""}\NormalTok{,}\DataTypeTok{xlab=}\StringTok{"methylation values"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/readMethAge-1} 

}

\caption{Histogram of methylation values in the training set for age prediction.}\label{fig:readMethAge}
\end{figure}

There are \textasciitilde{}27000 predictor variables. We should remove the ones that have low variation across samples. In this case, the methylation values are between 0 and 1, the CpGs that have low variation are not likely to have any association with age, they could simply be technical variation of the experiment. We will remove CpGs that have less than 0.1 standard deviation.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ameth=ameth[,}\KeywordTok{c}\NormalTok{(}\OtherTok{TRUE}\NormalTok{,matrixStats}\OperatorTok{::}\KeywordTok{colSds}\NormalTok{(}\KeywordTok{as.matrix}\NormalTok{(ameth[,}\OperatorTok{-}\DecValTok{1}\NormalTok{]))}\OperatorTok{>}\FloatTok{0.1}\NormalTok{)]}
\KeywordTok{dim}\NormalTok{(ameth)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]  108 2290
\end{verbatim}

\hypertarget{running-random-forest-regression}{%
\subsection{Running random forest regression}\label{running-random-forest-regression}}

Now we can use random forest regression to predict the age from methylation values. We are then going to plot predicted vs observed ages and see how well our predictions are.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{18}\NormalTok{)}

\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}

\CommentTok{# we are not going to do any cross-validatin}
\CommentTok{# and rely on OOB error}
\NormalTok{trctrl <-}\StringTok{ }\KeywordTok{trainControl}\NormalTok{(}\DataTypeTok{method =} \StringTok{"none"}\NormalTok{)}

\CommentTok{# we will now train random forest model}
\NormalTok{rfregFit <-}\StringTok{ }\KeywordTok{train}\NormalTok{(Age}\OperatorTok{~}\NormalTok{., }
               \DataTypeTok{data =}\NormalTok{ ameth, }
               \DataTypeTok{method =} \StringTok{"ranger"}\NormalTok{,}
               \DataTypeTok{trControl=}\NormalTok{trctrl,}
               \CommentTok{# calculate importance}
               \DataTypeTok{importance=}\StringTok{"permutation"}\NormalTok{, }
               \DataTypeTok{tuneGrid =} \KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{mtry=}\DecValTok{50}\NormalTok{,}
                                     \DataTypeTok{min.node.size =} \DecValTok{5}\NormalTok{,}
                                     \DataTypeTok{splitrule=}\StringTok{"variance"}\NormalTok{)}
\NormalTok{               )}
\CommentTok{# plot Observed vs OOB predicted values from the model}
\KeywordTok{plot}\NormalTok{(ameth}\OperatorTok{$}\NormalTok{Age,rfregFit}\OperatorTok{$}\NormalTok{finalModel}\OperatorTok{$}\NormalTok{predictions,}
     \DataTypeTok{pch=}\DecValTok{19}\NormalTok{,}\DataTypeTok{xlab=}\StringTok{"observed Age"}\NormalTok{,}
     \DataTypeTok{ylab=}\StringTok{"OOB predicted Age"}\NormalTok{)}
\KeywordTok{mtext}\NormalTok{(}\KeywordTok{paste}\NormalTok{(}\StringTok{"R-squared"}\NormalTok{,}
            \KeywordTok{format}\NormalTok{(rfregFit}\OperatorTok{$}\NormalTok{finalModel}\OperatorTok{$}\NormalTok{r.squared,}\DataTypeTok{digits=}\DecValTok{2}\NormalTok{)))}

\CommentTok{# plot residuals}
\KeywordTok{plot}\NormalTok{(ameth}\OperatorTok{$}\NormalTok{Age,(rfregFit}\OperatorTok{$}\NormalTok{finalModel}\OperatorTok{$}\NormalTok{predictions}\OperatorTok{-}\NormalTok{ameth}\OperatorTok{$}\NormalTok{Age),}
     \DataTypeTok{pch=}\DecValTok{18}\NormalTok{,}\DataTypeTok{ylab=}\StringTok{"residuals (predicted-observed)"}\NormalTok{,}
     \DataTypeTok{xlab=}\StringTok{"observed Age"}\NormalTok{,}\DataTypeTok{col=}\StringTok{"blue3"}\NormalTok{)}
\KeywordTok{abline}\NormalTok{(}\DataTypeTok{h=}\DecValTok{0}\NormalTok{,}\DataTypeTok{col=}\StringTok{"red4"}\NormalTok{,}\DataTypeTok{lty=}\DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.7\linewidth]{compgenomrReloaded_files/figure-latex/predictAge-1} 

}

\caption{Observed vs predicted age (Left). Residual plot showing for older people the error increases (Right)}\label{fig:predictAge}
\end{figure}

In this instance, we are using OOB errors and \(R^2\) value which shows how the model performs on OOB samples. The model can capture the general trend and it has acceptable OOB performance. It is not perfect as it makes errors on average close to 10 years when predicting the age, and the errors are more severe for older people (Figure \ref{fig:predictAge}). This could be due to having fewer older people to model or missing/inadequate predictor variables. However, everything we discussed in classification applies here. We had even fewer data points than the classification problem we did not do a split for a test data set. This should also be done for regression problems especially when we are going to compare the performance of different models or want to have a better idea of the real world performance of our model. We might be also interested in which variables are most important as in the classification problem, we can use \texttt{caret:varImp()} function to get access to random forest specific variable importance metrics.

\hypertarget{exercises-3}{%
\section{Exercises}\label{exercises-3}}

\hypertarget{classification}{%
\subsection{classification}\label{classification}}

TBD

\hypertarget{regression}{%
\subsection{Regression}\label{regression}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Split 20\% of the methylation-age data as test data and run elastic net regression on the training portion to tune parameters and test it on the test portion.(intermediate difficulty)
\item
  Run an ensemble model for regression using \textbf{caretEnsemble} or \textbf{mlr} package and compare the results with the elastic net and random forest model.
\end{enumerate}

\hypertarget{genomicIntervals}{%
\chapter{Operations on Genomic Intervals and Genome Arithmetic}\label{genomicIntervals}}

A considerable time in computational genomics is spent on overlapping different
features of the genome. Each feature can be represented with a genomic interval
within the chromosomal coordinate system. In addition, each interval can carry
different sorts of information. An interval may for instance represent exon coordinates or a transcription factor binding site. On the other hand,
you can have base-pair resolution, continuous scores over the genome such as read coverage or
scores that could be associated with only certain bases such as in the case of CpG
methylation (See Figure \ref{fig:gintsum} ).
Typically, you will need to overlap intervals on interest with other features of
the genome, again represented as intervals. For example, you may want to overlap
transcription factor binding sites with CpG islands or promoters to quantify what percentage of binding sites overlap with your regions of interest. Overlapping mapped reads from high-throughput sequencing experiments with genomic features such as exons, promoters, enhancers can also be classified as operations on genomic intervals. You can think of a million other ways that involves overlapping two sets of different features on the genome. This chapter aims to show how to do analysis involving operations on genomic intervals.

\begin{figure}

{\centering \includegraphics[width=0.75\linewidth]{images/genomeIntervalSummary} 

}

\caption{Summary of genomic intervals with different kinds of information}\label{fig:gintsum}
\end{figure}

\hypertarget{operations-on-genomic-intervals-with-genomicranges-package}{%
\section{Operations on Genomic Intervals with GenomicRanges package}\label{operations-on-genomic-intervals-with-genomicranges-package}}

The \href{http://bioconductor.org}{Bioconductor} project has a dedicated package called \href{http://www.bioconductor.org/packages/release/bioc/html/GenomicRanges.html}{\textbf{GenomicRanges}} to deal with genomic intervals. In this section, we will provide use cases involving operations on genomic intervals. The main reason we will stick to this package is that it provides tools to do overlap operations. However package requires that users operate on specific data types that are conceptually similar to a tabular data structure implemented in a way that makes overlapping and related operations easier. The main object we will be using is called \texttt{GRanges} object and we will also see some other related objects from the \texttt{GenomicRanges} package.

\hypertarget{how-to-create-and-manipulate-a-granges-object}{%
\subsection{How to create and manipulate a GRanges object}\label{how-to-create-and-manipulate-a-granges-object}}

\texttt{GRanges} (from GenomicRanges package) is the main object that holds the genomic intervals and extra information about those intervals. Here we will show how to create one. Conceptually, it is similar to a data frame and some operations such as using \textbf{\protect\hyperlink{section-21}{}} notation to subset the table will work also on GRanges, but keep in mind that not everything that works for data frames will work on GRanges objects.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(GenomicRanges)}
\NormalTok{gr=}\KeywordTok{GRanges}\NormalTok{(}\DataTypeTok{seqnames=}\KeywordTok{c}\NormalTok{(}\StringTok{"chr1"}\NormalTok{,}\StringTok{"chr2"}\NormalTok{,}\StringTok{"chr2"}\NormalTok{),}
           \DataTypeTok{ranges=}\KeywordTok{IRanges}\NormalTok{(}\DataTypeTok{start=}\KeywordTok{c}\NormalTok{(}\DecValTok{50}\NormalTok{,}\DecValTok{150}\NormalTok{,}\DecValTok{200}\NormalTok{),}
                          \DataTypeTok{end=}\KeywordTok{c}\NormalTok{(}\DecValTok{100}\NormalTok{,}\DecValTok{200}\NormalTok{,}\DecValTok{300}\NormalTok{)),}
           \DataTypeTok{strand=}\KeywordTok{c}\NormalTok{(}\StringTok{"+"}\NormalTok{,}\StringTok{"-"}\NormalTok{,}\StringTok{"-"}\NormalTok{)}
\NormalTok{)}
\NormalTok{gr}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          <Rle> <IRanges>  <Rle>
##   [1]     chr1    50-100      +
##   [2]     chr2   150-200      -
##   [3]     chr2   200-300      -
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# subset like a data frame}
\NormalTok{gr[}\DecValTok{1}\OperatorTok{:}\DecValTok{2}\NormalTok{,]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## GRanges object with 2 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          <Rle> <IRanges>  <Rle>
##   [1]     chr1    50-100      +
##   [2]     chr2   150-200      -
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths
\end{verbatim}

As you can see it looks a bit like a data frame. Also, note that the peculiar second argument ``ranges'' which basically contains start and end positions of the genomic intervals. However, you can not just give start and end positions you actually have to provide another object of IRanges. Do not let this confuse you, GRanges actually depends on another object that is very similar to itself called \texttt{IRanges} and you have to provide the ``ranges'' argument as an IRanges object. In its simplest for, an IRanges object can be constructed by providing start and end positions to IRanges() function. Think of it as something you just have to provide in order to construct the GRanges object.

GRanges can also contain other information about the genomic interval such as scores, names, etc. You can provide extra information at the time of the construction or you can add it later. Here is how you can do those:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{gr=}\KeywordTok{GRanges}\NormalTok{(}\DataTypeTok{seqnames=}\KeywordTok{c}\NormalTok{(}\StringTok{"chr1"}\NormalTok{,}\StringTok{"chr2"}\NormalTok{,}\StringTok{"chr2"}\NormalTok{),}
           \DataTypeTok{ranges=}\KeywordTok{IRanges}\NormalTok{(}\DataTypeTok{start=}\KeywordTok{c}\NormalTok{(}\DecValTok{50}\NormalTok{,}\DecValTok{150}\NormalTok{,}\DecValTok{200}\NormalTok{),}
                          \DataTypeTok{end=}\KeywordTok{c}\NormalTok{(}\DecValTok{100}\NormalTok{,}\DecValTok{200}\NormalTok{,}\DecValTok{300}\NormalTok{)),}
           \DataTypeTok{names=}\KeywordTok{c}\NormalTok{(}\StringTok{"id1"}\NormalTok{,}\StringTok{"id3"}\NormalTok{,}\StringTok{"id2"}\NormalTok{),}
           \DataTypeTok{scores=}\KeywordTok{c}\NormalTok{(}\DecValTok{100}\NormalTok{,}\DecValTok{90}\NormalTok{,}\DecValTok{50}\NormalTok{)}
\NormalTok{)}
\CommentTok{# or add it later (replaces the existing meta data)}
\KeywordTok{mcols}\NormalTok{(gr)=}\KeywordTok{DataFrame}\NormalTok{(}\DataTypeTok{name2=}\KeywordTok{c}\NormalTok{(}\StringTok{"pax6"}\NormalTok{,}\StringTok{"meis1"}\NormalTok{,}\StringTok{"zic4"}\NormalTok{),}
                    \DataTypeTok{score2=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{3}\NormalTok{))}

\NormalTok{gr=}\KeywordTok{GRanges}\NormalTok{(}\DataTypeTok{seqnames=}\KeywordTok{c}\NormalTok{(}\StringTok{"chr1"}\NormalTok{,}\StringTok{"chr2"}\NormalTok{,}\StringTok{"chr2"}\NormalTok{),}
           \DataTypeTok{ranges=}\KeywordTok{IRanges}\NormalTok{(}\DataTypeTok{start=}\KeywordTok{c}\NormalTok{(}\DecValTok{50}\NormalTok{,}\DecValTok{150}\NormalTok{,}\DecValTok{200}\NormalTok{),}
                          \DataTypeTok{end=}\KeywordTok{c}\NormalTok{(}\DecValTok{100}\NormalTok{,}\DecValTok{200}\NormalTok{,}\DecValTok{300}\NormalTok{)),}
           \DataTypeTok{names=}\KeywordTok{c}\NormalTok{(}\StringTok{"id1"}\NormalTok{,}\StringTok{"id3"}\NormalTok{,}\StringTok{"id2"}\NormalTok{),}
           \DataTypeTok{scores=}\KeywordTok{c}\NormalTok{(}\DecValTok{100}\NormalTok{,}\DecValTok{90}\NormalTok{,}\DecValTok{50}\NormalTok{)}
\NormalTok{)}

\CommentTok{# or appends to existing meta data}
\KeywordTok{mcols}\NormalTok{(gr)=}\KeywordTok{cbind}\NormalTok{(}\KeywordTok{mcols}\NormalTok{(gr),}
                          \KeywordTok{DataFrame}\NormalTok{(}\DataTypeTok{name2=}\KeywordTok{c}\NormalTok{(}\StringTok{"pax6"}\NormalTok{,}\StringTok{"meis1"}\NormalTok{,}\StringTok{"zic4"}\NormalTok{)) )}
\NormalTok{gr}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## GRanges object with 3 ranges and 3 metadata columns:
##       seqnames    ranges strand |       names
##          <Rle> <IRanges>  <Rle> | <character>
##   [1]     chr1    50-100      * |         id1
##   [2]     chr2   150-200      * |         id3
##   [3]     chr2   200-300      * |         id2
##          scores       name2
##       <numeric> <character>
##   [1]       100        pax6
##   [2]        90       meis1
##   [3]        50        zic4
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# elementMetadata() and values() do the same things}
\KeywordTok{elementMetadata}\NormalTok{(gr)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## DataFrame with 3 rows and 3 columns
##         names    scores       name2
##   <character> <numeric> <character>
## 1         id1       100        pax6
## 2         id3        90       meis1
## 3         id2        50        zic4
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{values}\NormalTok{(gr)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## DataFrame with 3 rows and 3 columns
##         names    scores       name2
##   <character> <numeric> <character>
## 1         id1       100        pax6
## 2         id3        90       meis1
## 3         id2        50        zic4
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# you may also add metadata using the $ operator, as for data frames}
\NormalTok{gr}\OperatorTok{$}\NormalTok{name3 =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"A"}\NormalTok{,}\StringTok{"C"}\NormalTok{, }\StringTok{"B"}\NormalTok{)}
\NormalTok{gr}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## GRanges object with 3 ranges and 4 metadata columns:
##       seqnames    ranges strand |       names
##          <Rle> <IRanges>  <Rle> | <character>
##   [1]     chr1    50-100      * |         id1
##   [2]     chr2   150-200      * |         id3
##   [3]     chr2   200-300      * |         id2
##          scores       name2       name3
##       <numeric> <character> <character>
##   [1]       100        pax6           A
##   [2]        90       meis1           C
##   [3]        50        zic4           B
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths
\end{verbatim}

\hypertarget{getting-genomic-regions-into-r-as-granges-objects}{%
\subsection{Getting genomic regions into R as GRanges objects}\label{getting-genomic-regions-into-r-as-granges-objects}}

There are multiple ways you can read in your genomic features into R and create a GRanges object. Most genomic interval data comes as a tabular format that has the basic information about the location of the interval and some other information. We already showed how to read BED files as data frame. Now we will show how to convert it to GRanges object. This is one way of doing it, but there are more convenient ways described further in the text.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# read CpGi data set}
\NormalTok{filePath=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
                      \StringTok{"cpgi.hg19.chr21.bed"}\NormalTok{,}
                      \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{cpgi.df =}\StringTok{ }\KeywordTok{read.table}\NormalTok{(filePath, }\DataTypeTok{header =} \OtherTok{FALSE}\NormalTok{,}
                     \DataTypeTok{stringsAsFactors=}\OtherTok{FALSE}\NormalTok{) }
\CommentTok{# remove chr names with "_"}
\NormalTok{cpgi.df =cpgi.df [}\KeywordTok{grep}\NormalTok{(}\StringTok{"_"}\NormalTok{,cpgi.df[,}\DecValTok{1}\NormalTok{],}\DataTypeTok{invert=}\OtherTok{TRUE}\NormalTok{),]}

\NormalTok{cpgi.gr=}\KeywordTok{GRanges}\NormalTok{(}\DataTypeTok{seqnames=}\NormalTok{cpgi.df[,}\DecValTok{1}\NormalTok{],}
                \DataTypeTok{ranges=}\KeywordTok{IRanges}\NormalTok{(}\DataTypeTok{start=}\NormalTok{cpgi.df[,}\DecValTok{2}\NormalTok{],}
                              \DataTypeTok{end=}\NormalTok{cpgi.df[,}\DecValTok{3}\NormalTok{]))}
\end{Highlighting}
\end{Shaded}

You may need to do some pre-processing before/after reading in the BED file. Below is an example of getting transcription start sites from BED files containing refseq transcript locations.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# read refseq file}
\NormalTok{filePathRefseq=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
                      \StringTok{"refseq.hg19.chr21.bed"}\NormalTok{,}
                      \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}


\NormalTok{ref.df =}\StringTok{ }\KeywordTok{read.table}\NormalTok{(filePathRefseq, }\DataTypeTok{header =} \OtherTok{FALSE}\NormalTok{,}
                     \DataTypeTok{stringsAsFactors=}\OtherTok{FALSE}\NormalTok{) }
\NormalTok{ref.gr=}\KeywordTok{GRanges}\NormalTok{(}\DataTypeTok{seqnames=}\NormalTok{ref.df[,}\DecValTok{1}\NormalTok{],}
               \DataTypeTok{ranges=}\KeywordTok{IRanges}\NormalTok{(}\DataTypeTok{start=}\NormalTok{ref.df[,}\DecValTok{2}\NormalTok{],}
                              \DataTypeTok{end=}\NormalTok{ref.df[,}\DecValTok{3}\NormalTok{]),}
               \DataTypeTok{strand=}\NormalTok{ref.df[,}\DecValTok{6}\NormalTok{],}\DataTypeTok{name=}\NormalTok{ref.df[,}\DecValTok{4}\NormalTok{])}
\CommentTok{# get TSS}
\NormalTok{tss.gr=ref.gr}
\CommentTok{# end of the + strand genes must be equalized to start pos}
\KeywordTok{end}\NormalTok{(tss.gr[}\KeywordTok{strand}\NormalTok{(tss.gr)}\OperatorTok{==}\StringTok{"+"}\NormalTok{,])  =}\KeywordTok{start}\NormalTok{(tss.gr[}\KeywordTok{strand}\NormalTok{(tss.gr)}\OperatorTok{==}\StringTok{"+"}\NormalTok{,])}
\CommentTok{# startof the - strand genes must be equalized to end pos}
\KeywordTok{start}\NormalTok{(tss.gr[}\KeywordTok{strand}\NormalTok{(tss.gr)}\OperatorTok{==}\StringTok{"-"}\NormalTok{,])=}\KeywordTok{end}\NormalTok{(tss.gr[}\KeywordTok{strand}\NormalTok{(tss.gr)}\OperatorTok{==}\StringTok{"-"}\NormalTok{,])}
\CommentTok{# remove duplicated TSSes ie alternative transcripts}
\CommentTok{# this keeps the first instance and removes duplicates}
\NormalTok{tss.gr=tss.gr[}\OperatorTok{!}\KeywordTok{duplicated}\NormalTok{(tss.gr),]}
\end{Highlighting}
\end{Shaded}

Another way of doing this is from a BED file is to use \texttt{readTranscriptfeatures()}
function from the \texttt{genomation} package. This function takes care of the steps described in the code chunk above.

Reading the genomic features as text files and converting to GRanges is not the only way to create GRanges object. With the help of the \href{http://www.bioconductor.org/packages/release/bioc/html/rtracklayer.html}{rtracklayer} package we can directly import BED files.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{require}\NormalTok{(rtracklayer)}

\CommentTok{# we are reading a BED file, the path to the file}
\CommentTok{# is stored in filePathRefseq variable}
\KeywordTok{import.bed}\NormalTok{(filePathRefseq)}
\end{Highlighting}
\end{Shaded}

Next, we will show how to use other methods to automatically obtain the data in GRanges format from online databases. But you will not be able to use these methods for every data set so it is good to now how to read data from flat files as well. We will use rtracklayer package to download data from \href{http://genome-euro.ucsc.edu/cgi-bin/hgGateway?hgsid=197855328_moRAG0Dqe1qlcOEtNDAt09E1e6ab}{UCSC browser}. We will download CpG islands as GRanges objects. The \texttt{rtracklayer} workflow we show below works like using the UCSC table browser. You need to select which species you are working with, then you need to select which dataset you need to download and lastly you download the UCSC dataset or track as \texttt{GRanges} object.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{require}\NormalTok{(rtracklayer)}
\NormalTok{session <-}\StringTok{ }\KeywordTok{browserSession}\NormalTok{(}\StringTok{"UCSC"}\NormalTok{,}\DataTypeTok{url =} \StringTok{'http://genome-euro.ucsc.edu/cgi-bin/'}\NormalTok{)}
\KeywordTok{genome}\NormalTok{(session) <-}\StringTok{ "mm9"}
\CommentTok{## choose CpG island track on chr12}
\NormalTok{query <-}\StringTok{ }\KeywordTok{ucscTableQuery}\NormalTok{(session, }\DataTypeTok{track=}\StringTok{"CpG Islands"}\NormalTok{,}\DataTypeTok{table=}\StringTok{"cpgIslandExt"}\NormalTok{,}
        \DataTypeTok{range=}\KeywordTok{GRangesForUCSCGenome}\NormalTok{(}\StringTok{"mm9"}\NormalTok{, }\StringTok{"chr12"}\NormalTok{))}
\CommentTok{## get the GRanges object for the track}
\KeywordTok{track}\NormalTok{(query)}
\end{Highlighting}
\end{Shaded}

There is also an interface to Ensembl database called \href{https://bioconductor.org/packages/release/bioc/html/biomaRt.html}{biomaRt}.
This package will enable you to access and import all of the datasets included
in Ensembl. Another similar package is \href{https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html}{AnnotationHub}.
. This package is an aggregator for different datasets from various sources.
Using AnnotationHub one can access data sets from UCSC browser, Ensembl browser
and data sets from genomics consortiums such as ENCODE and RoadmapEpigenomics.
Although, we do not provide examples, we invite curious readers to read the
vignettes of the aforementioned annotation packages.

\hypertarget{frequently-used-file-formats-and-how-to-read-them-into-r-as-a-table}{%
\subsubsection{Frequently used file formats and how to read them into R as a table}\label{frequently-used-file-formats-and-how-to-read-them-into-r-as-a-table}}

There are multiple file formats in genomics but some of them you will see more
frequently than others. We already mentioned some of them. Here is a list of files
and functions to read them into R as \texttt{GRanges} objects or something coercible to
GRanges objects.

\textbf{BED}: These are used and popularized by UCSC browser, and can hold a variety of
information including exon/intron structure of transcripts in a single line.

\begin{itemize}
\tightlist
\item
  \texttt{genomation::readBed()}
\item
  \texttt{genomation::readTranscriptFeatures()} good for getting intron/exon/promoters from BED12 files
\item
  \texttt{rtracklayer::import.bed()}
\end{itemize}

\textbf{GFF}: GFF format is a tabular text format for genomic features similar to BED. However,
it is a more flexible format than BED, which makes it harder to parse at times. Many gene annotation files are in this format.

\begin{itemize}
\tightlist
\item
  \texttt{genomation::gffToGranges()}
\item
  \texttt{rtracklayer::impot.gff()}
\end{itemize}

\textbf{BAM}: BAM format is compressed and indexed tabular file format designed for sequencing
reads.

\begin{itemize}
\tightlist
\item
  \texttt{GenomicAlignments::readGAlignments}
\item
  \texttt{Rsamtools::scanBam} returns a data frame with columns from SAM/BAM file.
\end{itemize}

\textbf{BigWig}: This is used to for storing scores associated with genomic intervals. It is an indexed format. Similar to BAM, this makes it easier to query and only necessary portions
of the file could be loaded into memory.

\begin{itemize}
\tightlist
\item
  \texttt{rtracklayer::import.bw()}
\end{itemize}

\textbf{Generic Text files}: This represents any text file with the minimal information of chromosome, start and end coordinates.

\begin{itemize}
\tightlist
\item
  \texttt{genomation::readGeneric()}
\end{itemize}

\textbf{Tabix/Bcf}: These are tabular file formats indexed and compressed similar to
BAM. The following functions return lists rather than tabular data structures. These
formats are mostly used to store genomic variation data such as SNPs and indels.

\begin{itemize}
\tightlist
\item
  \texttt{Rsamtools::scanTabix}
\item
  \texttt{Rsamtools::scanBcf}
\end{itemize}

\hypertarget{finding-regions-that-dodo-not-overlap-with-another-set-of-regions}{%
\subsection{Finding regions that do/do not overlap with another set of regions}\label{finding-regions-that-dodo-not-overlap-with-another-set-of-regions}}

This is one of the most common tasks in genomics. Usually, you have a set of regions that you are interested in and you want to see if they overlap with another set of regions or see how many of them overlap. A good example is transcription factor binding sites determined by \href{http://en.wikipedia.org/wiki/ChIP-sequencing}{ChIP-seq} experiments. In these types of experiments and followed analysis, one usually ends up with genomic regions that are bound by transcription factors. One of the standard next questions would be to annotate binding sites with genomic annotations such as promoter,exon,intron and/or CpG islands. Below is a demonstration of how transcription factor binding sites can be annotated using CpG islands. First, we will get the subset of binding sites that overlap with the CpG islands. In this case, binding sites are ChIP-seq peaks.

We can find the subset of peaks that overlap with the CpG islands using the \textbf{subsetByoverlaps()} function. You will also see another way of converting \textbf{data frames} to \textbf{GRanges}.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(genomation)}
\NormalTok{filePathPeaks=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,             }\StringTok{"wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep1.broadPeak.gz"}\NormalTok{,}
                      \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}

\CommentTok{# read the peaks from a bed file}
\NormalTok{pk1.gr=}\KeywordTok{readBroadPeak}\NormalTok{(filePathPeaks)}

\CommentTok{# get the peaks that overlap with CpG islands}
\KeywordTok{subsetByOverlaps}\NormalTok{(pk1.gr,cpgi.gr)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## GRanges object with 44 ranges and 5 metadata columns:
##        seqnames            ranges strand |        name
##           <Rle>         <IRanges>  <Rle> | <character>
##    [1]    chr21   9825360-9826582      * |   peak14562
##    [2]    chr21   9968469-9968984      * |   peak14593
##    [3]    chr21 15755368-15755956      * |   peak14828
##    [4]    chr21 19191579-19192525      * |   peak14840
##    [5]    chr21 26979619-26980048      * |   peak14854
##    ...      ...               ...    ... .         ...
##   [40]    chr21 46237464-46237809      * |   peak15034
##   [41]    chr21 46707702-46708084      * |   peak15037
##   [42]    chr21 46961552-46961875      * |   peak15039
##   [43]    chr21 47743587-47744125      * |   peak15050
##   [44]    chr21 47878412-47878891      * |   peak15052
##            score signalValue    pvalue    qvalue
##        <integer>   <numeric> <integer> <integer>
##    [1]        56      183.11        -1        -1
##    [2]       947     3064.92        -1        -1
##    [3]        90       291.9        -1        -1
##    [4]       290      940.03        -1        -1
##    [5]        32      104.67        -1        -1
##    ...       ...         ...       ...       ...
##   [40]        32      106.36        -1        -1
##   [41]        67      217.02        -1        -1
##   [42]        38      124.31        -1        -1
##   [43]       353     1141.58        -1        -1
##   [44]       104      338.78        -1        -1
##   -------
##   seqinfo: 23 sequences from an unspecified genome; no seqlengths
\end{verbatim}

For each CpG island, we can count the number of peaks that overlap with a given CpG island with countOverlaps().

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{counts=}\KeywordTok{countOverlaps}\NormalTok{(pk1.gr,cpgi.gr)}
\KeywordTok{head}\NormalTok{(counts)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0 0 0 0 0 0
\end{verbatim}

The \texttt{findOverlaps()} function can be used to see one-to-one overlaps between peaks and CpG islands. It returns a matrix showing which peak overlaps with which CpGi island.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{findOverlaps}\NormalTok{(pk1.gr,cpgi.gr)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Hits object with 45 hits and 0 metadata columns:
##        queryHits subjectHits
##        <integer>   <integer>
##    [1]     14562           1
##    [2]     14593           3
##    [3]     14828           8
##    [4]     14840          13
##    [5]     14854          16
##    ...       ...         ...
##   [41]     15034         155
##   [42]     15037         166
##   [43]     15039         176
##   [44]     15050         192
##   [45]     15052         200
##   -------
##   queryLength: 26121 / subjectLength: 205
\end{verbatim}

Another interesting thing would be to look at the distances to nearest CpG islands for each peak. In addition, just finding the nearest CpG island could also be interesting. Often times, you will need to find nearest TSS or gene to your regions of interest, and the code below is handy for doing that.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# find nearest CpGi to each TSS}
\NormalTok{n.ind=}\KeywordTok{nearest}\NormalTok{(pk1.gr,cpgi.gr)}
\CommentTok{# get distance to nearest}
\NormalTok{dists=}\KeywordTok{distanceToNearest}\NormalTok{(pk1.gr,cpgi.gr,}\DataTypeTok{select=}\StringTok{"arbitrary"}\NormalTok{)}
\NormalTok{dists}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Hits object with 620 hits and 1 metadata column:
##         queryHits subjectHits |  distance
##         <integer>   <integer> | <integer>
##     [1]     14440           1 |    384188
##     [2]     14441           1 |    382968
##     [3]     14442           1 |    381052
##     [4]     14443           1 |    379311
##     [5]     14444           1 |    376978
##     ...       ...         ... .       ...
##   [616]     15055         205 |     26212
##   [617]     15056         205 |     27402
##   [618]     15057         205 |     30468
##   [619]     15058         205 |     31611
##   [620]     15059         205 |     34090
##   -------
##   queryLength: 26121 / subjectLength: 205
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# histogram of the distances to nearest TSS}
\NormalTok{dist2plot=}\KeywordTok{mcols}\NormalTok{(dists)[,}\DecValTok{1}\NormalTok{]}
\KeywordTok{hist}\NormalTok{(}\KeywordTok{log10}\NormalTok{(dist2plot),}\DataTypeTok{xlab=}\StringTok{"log10(dist to nearest TSS)"}\NormalTok{,}
     \DataTypeTok{main=}\StringTok{"Distances"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/findNearest-1} 

}

\caption{histogram of distances}\label{fig:findNearest}
\end{figure}

\hypertarget{dealing-with-mapped-high-throughput-sequencing-reads}{%
\section{Dealing with mapped high-throughput sequencing reads}\label{dealing-with-mapped-high-throughput-sequencing-reads}}

The reads from sequencing machines are usually pre-proccessed and aligned to the genome with the help of specific bioinformatics tools. We have introduced the read processing , quality check and alignment methods in chapter \ref{processingReads}. In this section we will deal with mapped reads. Since each mapped read has a start and end position the genome, mapped reads can be thought as genomic intervals stored in a file. After mapping, the next task is to quantify the enrichment of those aligned reads in the regions of interest. You may want to count how many reads overlapping with your promoter set of interest or you may want to quantify RNA-seq reads overlapping with exons. This is similar to operations on genomic intervals which are described previously. If you can read all your alignments into the memory and create a \texttt{GRanges} object, you can apply the previously described operations. However, most of the time we can not read all mapped reads into the memory, so we have to use specialized tools to query and quantify alignments on a given set of regions. One of the most common alignment formats is SAM/BAM format, most aligners will produce SAM/BAM output or you will be able to convert your specific alignment format to SAM/BAM format. The BAM format is a binary version of the human readable SAM format. The SAM format has specific columns that contain different kind of information about the alignment such as mismatches, qualities etc. (see {[}\url{http://samtools.sourceforge.net/SAM1.pdf}{]} for SAM format specification).

\hypertarget{counting-mapped-reads-for-a-set-of-regions}{%
\subsection{Counting mapped reads for a set of regions}\label{counting-mapped-reads-for-a-set-of-regions}}

Rsamtools package has functions to query BAM files. The function we will use in the first example is countBam which takes input of the BAM file and param argument. ``param'' argument takes a ScanBamParam object. The object is instantiated using \textbf{ScanBamParam()} and contains parameters for scanning the BAM file. The example below is a simple example where ScanBamParam() only includes regions of interest, promoters on chr21.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{promoter.gr=tss.gr}
\KeywordTok{start}\NormalTok{(promoter.gr)=}\KeywordTok{start}\NormalTok{(promoter.gr)}\OperatorTok{-}\DecValTok{1000}
\KeywordTok{end}\NormalTok{(promoter.gr)  =}\KeywordTok{end}\NormalTok{(promoter.gr)}\OperatorTok{+}\DecValTok{1000}
\NormalTok{promoter.gr=promoter.gr[}\KeywordTok{seqnames}\NormalTok{(promoter.gr)}\OperatorTok{==}\StringTok{"chr21"}\NormalTok{]}

\KeywordTok{library}\NormalTok{(Rsamtools)}
\NormalTok{bamfilePath=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
            \StringTok{"wgEncodeHaibTfbsGm12878Sp1Pcr1xAlnRep1.chr21.bam"}\NormalTok{,}
                      \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}

\CommentTok{# get reads for regions of interest from the bam file}
\NormalTok{param <-}\StringTok{ }\KeywordTok{ScanBamParam}\NormalTok{(}\DataTypeTok{which=}\NormalTok{promoter.gr)}
\NormalTok{counts=}\KeywordTok{countBam}\NormalTok{(bamfilePath, }\DataTypeTok{param=}\NormalTok{param)}
\end{Highlighting}
\end{Shaded}

Alternatively, aligned reads can be read in using GenomicAlignments package (which on this occasion relies on RSamtools package).

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(GenomicAlignments)}
\NormalTok{alns <-}\StringTok{ }\KeywordTok{readGAlignments}\NormalTok{(bamfilePath, }\DataTypeTok{param=}\NormalTok{param)}
\end{Highlighting}
\end{Shaded}

\hypertarget{dealing-with-continuous-scores-over-the-genome}{%
\section{Dealing with continuous scores over the genome}\label{dealing-with-continuous-scores-over-the-genome}}

Most high-throughput data can be viewed as a continuous score over the bases of the genome. In case of RNA-seq or ChIP-seq experiments the data can be represented as read coverage values per genomic base position. In addition, other information (not necessarily from high-throughput experiments) can be represented this way. The GC content and conservation scores per base are prime examples of other data sets that can be represented as scores. This sort of data can be stored as a generic text file or can have special formats such as Wig (stands for wiggle) from UCSC, or the bigWig format is which is indexed binary format of the wig files. The bigWig format is great for data that covers large fraction of the genome with varying scores, because the file is much smaller than regular text files that have the same information and it can be queried easier since it is indexed.

In R/Bioconductor, the continuous data can also be represented in a compressed format, in a format called Rle vector, which stands for run-length encoded vector. This gives superior memory performance over regular vectors because repeating consecutive values are represented as one value in the Rle vector (See Figure \ref{fig:Rle} ).

\begin{figure}

{\centering \includegraphics[width=1\linewidth]{images/Rle_demo} 

}

\caption{Rle encoding explained}\label{fig:Rle}
\end{figure}

Typically, for genome-wide data you will have a RleList object which is a list of Rle vectors per chromosome. You can obtain such vectors by reading the reads in and calling \emph{coverage()} function from \emph{GenomicRanges} package. Let's try that on the above data set.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{covs=}\KeywordTok{coverage}\NormalTok{(alns) }\CommentTok{# get coverage vectors}
\NormalTok{covs}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## RleList of length 24
## $chr1
## integer-Rle of length 249250621 with 1 run
##   Lengths: 249250621
##   Values :         0
## 
## $chr2
## integer-Rle of length 243199373 with 1 run
##   Lengths: 243199373
##   Values :         0
## 
## $chr3
## integer-Rle of length 198022430 with 1 run
##   Lengths: 198022430
##   Values :         0
## 
## $chr4
## integer-Rle of length 191154276 with 1 run
##   Lengths: 191154276
##   Values :         0
## 
## $chr5
## integer-Rle of length 180915260 with 1 run
##   Lengths: 180915260
##   Values :         0
## 
## ...
## <19 more elements>
\end{verbatim}

Alternatively, you can get the coverage from the Bam file directly. Below, we are getting the coverage directly from the Bam file for our previously defined promoters.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{covs=}\KeywordTok{coverage}\NormalTok{(bamfilePath, }\DataTypeTok{param=}\NormalTok{param) }\CommentTok{# get coverage vectors}
\end{Highlighting}
\end{Shaded}

One of the most common ways of storing score data is, as mentioned, wig or bigWig format. Most of the ENCODE project data can be downloaded in bigWig format. In addition, conservation scores can also be downloaded as wig/bigWig format. You can import bigWig files into R using \emph{import()} function from \emph{rtracklayer} package. However, it is generally not advisable to read the whole bigWig file in memory as it was the case with BAM files. Usually, you will be interested in only a fraction of the genome, such as promoters, exons etc. So it is best you extract the data for those regions and read those into memory rather than the whole file. Below we read the a bigWig file only for promoters. The operation returns an \emph{GRanges} object with score column which indicates the scores in the BigWig file per genomic region.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(rtracklayer)}

\CommentTok{# File from ENCODE ChIP-seq tracks}
\NormalTok{bwFile=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
                      \StringTok{"wgEncodeHaibTfbsA549.chr21.bw"}\NormalTok{,}
                      \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{bw.gr=}\KeywordTok{import}\NormalTok{(bwFile, }\DataTypeTok{which=}\NormalTok{promoter.gr) }\CommentTok{# get coverage vectors}
\NormalTok{bw.gr}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## GRanges object with 9205 ranges and 1 metadata column:
##          seqnames            ranges strand |     score
##             <Rle>         <IRanges>  <Rle> | <numeric>
##      [1]    chr21   9825456-9825457      * |         1
##      [2]    chr21   9825458-9825464      * |         2
##      [3]    chr21   9825465-9825466      * |         4
##      [4]    chr21   9825467-9825470      * |         5
##      [5]    chr21           9825471      * |         6
##      ...      ...               ...    ... .       ...
##   [9201]    chr21 48055809-48055856      * |         2
##   [9202]    chr21 48055857-48055858      * |         1
##   [9203]    chr21 48055872-48055921      * |         1
##   [9204]    chr21 48055944-48055993      * |         1
##   [9205]    chr21 48056069-48056118      * |         1
##   -------
##   seqinfo: 1 sequence from an unspecified genome
\end{verbatim}

Following this we can create an \texttt{RleList} object from the GRanges with \texttt{coverage}
function.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cov.bw=}\KeywordTok{coverage}\NormalTok{(bw.gr,}\DataTypeTok{weight =} \StringTok{"score"}\NormalTok{)}

\CommentTok{# or get this directly from}
\NormalTok{cov.bw=}\KeywordTok{import}\NormalTok{(bwFile, }\DataTypeTok{which=}\NormalTok{promoter.gr,}\DataTypeTok{as =} \StringTok{"RleList"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{extracting-subsections-of-rle-and-rlelist-objects}{%
\subsection{Extracting subsections of Rle and RleList objects}\label{extracting-subsections-of-rle-and-rlelist-objects}}

Frequently, we will need to extract subsections of the Rle vectors or RleList objects.
We will need to do this to visualize that subsection or get some statistics out
of those sections. For example, we could be interested in average coverage per
base for the regions we are interested in. We have to extract those regions
from RleList object and apply summary statistics. Below, we show how to extract
subsections of RleList object. We are extracting promoter regions from ChIP-seq
read coverage RleList. Following that, we will plot the one of the promoters associated coverage values.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{myViews=}\KeywordTok{Views}\NormalTok{(cov.bw,}\KeywordTok{as}\NormalTok{(promoter.gr,}\StringTok{"IRangesList"}\NormalTok{)) }\CommentTok{# get subsets of coverage}
\CommentTok{# there is a views object for each chromosome}
\NormalTok{myViews}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## RleViewsList of length 1
## $chr21
## Views on a 48129895-length Rle subject
## 
## views:
##          start      end width
##   [1] 42218039 42220039  2001 [2 0 0 0 0 0 0 0 0 ...]
##   [2] 17441841 17443841  2001 [0 0 0 0 0 0 0 0 0 ...]
##   [3] 17565698 17567698  2001 [0 0 0 0 0 0 0 0 0 ...]
##   [4] 30395937 30397937  2001 [0 0 0 0 0 0 0 0 0 ...]
##   [5] 27542138 27544138  2001 [1 1 1 1 1 1 1 1 1 ...]
##   [6] 27511708 27513708  2001 [0 0 0 0 0 0 0 0 0 ...]
##   [7] 32930290 32932290  2001 [0 0 0 0 0 0 0 0 0 ...]
##   [8] 27542446 27544446  2001 [0 0 0 0 0 0 0 0 0 ...]
##   [9] 28338439 28340439  2001 [0 0 0 0 0 0 0 0 0 ...]
##   ...      ...      ...   ... ...
## [370] 47517032 47519032  2001 [1 1 1 1 1 1 1 1 1 ...]
## [371] 47648157 47650157  2001 [1 1 1 1 1 1 1 1 1 ...]
## [372] 47603373 47605373  2001 [0 0 0 0 0 0 0 0 0 ...]
## [373] 47647738 47649738  2001 [2 2 2 2 2 2 2 2 2 ...]
## [374] 47704236 47706236  2001 [0 0 0 0 0 0 0 0 0 ...]
## [375] 47742785 47744785  2001 [0 0 0 0 0 0 0 0 0 ...]
## [376] 47881383 47883383  2001 [1 1 1 1 1 1 1 1 1 ...]
## [377] 48054506 48056506  2001 [0 0 0 0 0 0 0 0 0 ...]
## [378] 48024035 48026035  2001 [1 1 1 1 1 1 1 1 1 ...]
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{myViews[[}\DecValTok{1}\NormalTok{]]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Views on a 48129895-length Rle subject
## 
## views:
##          start      end width
##   [1] 42218039 42220039  2001 [2 0 0 0 0 0 0 0 0 ...]
##   [2] 17441841 17443841  2001 [0 0 0 0 0 0 0 0 0 ...]
##   [3] 17565698 17567698  2001 [0 0 0 0 0 0 0 0 0 ...]
##   [4] 30395937 30397937  2001 [0 0 0 0 0 0 0 0 0 ...]
##   [5] 27542138 27544138  2001 [1 1 1 1 1 1 1 1 1 ...]
##   [6] 27511708 27513708  2001 [0 0 0 0 0 0 0 0 0 ...]
##   [7] 32930290 32932290  2001 [0 0 0 0 0 0 0 0 0 ...]
##   [8] 27542446 27544446  2001 [0 0 0 0 0 0 0 0 0 ...]
##   [9] 28338439 28340439  2001 [0 0 0 0 0 0 0 0 0 ...]
##   ...      ...      ...   ... ...
## [370] 47517032 47519032  2001 [1 1 1 1 1 1 1 1 1 ...]
## [371] 47648157 47650157  2001 [1 1 1 1 1 1 1 1 1 ...]
## [372] 47603373 47605373  2001 [0 0 0 0 0 0 0 0 0 ...]
## [373] 47647738 47649738  2001 [2 2 2 2 2 2 2 2 2 ...]
## [374] 47704236 47706236  2001 [0 0 0 0 0 0 0 0 0 ...]
## [375] 47742785 47744785  2001 [0 0 0 0 0 0 0 0 0 ...]
## [376] 47881383 47883383  2001 [1 1 1 1 1 1 1 1 1 ...]
## [377] 48054506 48056506  2001 [0 0 0 0 0 0 0 0 0 ...]
## [378] 48024035 48026035  2001 [1 1 1 1 1 1 1 1 1 ...]
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# get the coverage vector from the 5th view and plot}
\KeywordTok{plot}\NormalTok{(myViews[[}\DecValTok{1}\NormalTok{]][[}\DecValTok{5}\NormalTok{]],}\DataTypeTok{type=}\StringTok{"l"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/getViews-1} 

}

\caption{Coverage vector extracted from RleList via Views() function is plotted as a line plot.}\label{fig:getViews}
\end{figure}

Next, we are interested in average coverage per base for the promoters using summary
functions that works on Views object.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# get the mean of the views}
\KeywordTok{head}\NormalTok{(}
  \KeywordTok{viewMeans}\NormalTok{(myViews[[}\DecValTok{1}\NormalTok{]])}
\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.2259 0.3498 1.2244 0.4998 2.0905 0.6997
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# get the max of the views}
\KeywordTok{head}\NormalTok{(}
  \KeywordTok{viewMaxs}\NormalTok{(myViews[[}\DecValTok{1}\NormalTok{]])}
\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]  2  4 12  4 21  6
\end{verbatim}

\hypertarget{genomic-intervals-with-more-information-summarizedexperiment-class}{%
\section{Genomic intervals with more information: SummarizedExperiment class}\label{genomic-intervals-with-more-information-summarizedexperiment-class}}

As we have seen, genomic intervals can be mainly contained in a \texttt{GRanges} object.
It can also contain additional columns associated with each interval, here
you can save information such as read counts or other scores associated with the
interval. However,
genomics data is often have many layers. With \texttt{GRanges} you can have a table
associated with the intervals, but what happens if you have many tables and each
table has some metadata associated with it. In addition, rows and columns might
have additional annotation that can not be contained by row or column names.
For these cases, \texttt{SummarizedExperiment} class is ideal. It can hold multi-layered
tabular data associated with each genomic interval and the meta-data associated with
rows and columns, or associated with each table. For example, genomic intervals
associated with the SummarizedExperiment object can be gene locations, and
each tabular data structure can be RNA-seq read counts in a time course experiment.
Each table could represent different conditions in which experiments performed.
The SummarizedExperiment class is outlined in the figure below (Figure \ref{fig:SumExpOv} ).

\begin{figure}

{\centering \includegraphics[width=1\linewidth]{images/Summarized.Experiment} 

}

\caption{Overview of SummarizedExperiment class and functions. Adapted from SummerizedExperiment package vignette}\label{fig:SumExpOv}
\end{figure}

\hypertarget{create-a-summarizedexperiment-object}{%
\subsection{Create a SummarizedExperiment object}\label{create-a-summarizedexperiment-object}}

Here we show how to create a basic SummarizedExperiment object. We will first
create a matrix of read counts. This matrix will represent read counts from
a series RNA-seq experiments from different timepoints. Following that,
we create \texttt{GRanges} object to represent the locations of the genes, and a table
for column annotations. This will include the names for the columns and any
other value we want to represent. Finally, we will create a \texttt{SummarizedExperiment}
object by combining all those pieces.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# simulate an RNA-seq read counts table}
\NormalTok{nrows <-}\StringTok{ }\DecValTok{200}
\NormalTok{ncols <-}\StringTok{ }\DecValTok{6}
\NormalTok{counts <-}\StringTok{ }\KeywordTok{matrix}\NormalTok{(}\KeywordTok{runif}\NormalTok{(nrows }\OperatorTok{*}\StringTok{ }\NormalTok{ncols, }\DecValTok{1}\NormalTok{, }\FloatTok{1e4}\NormalTok{), nrows)}

\CommentTok{# create gene locations}
\NormalTok{rowRanges <-}\StringTok{ }\KeywordTok{GRanges}\NormalTok{(}\KeywordTok{rep}\NormalTok{(}\KeywordTok{c}\NormalTok{(}\StringTok{"chr1"}\NormalTok{, }\StringTok{"chr2"}\NormalTok{), }\KeywordTok{c}\NormalTok{(}\DecValTok{50}\NormalTok{, }\DecValTok{150}\NormalTok{)),}
                     \KeywordTok{IRanges}\NormalTok{(}\KeywordTok{floor}\NormalTok{(}\KeywordTok{runif}\NormalTok{(}\DecValTok{200}\NormalTok{, }\FloatTok{1e5}\NormalTok{, }\FloatTok{1e6}\NormalTok{)), }\DataTypeTok{width=}\DecValTok{100}\NormalTok{),}
                     \DataTypeTok{strand=}\KeywordTok{sample}\NormalTok{(}\KeywordTok{c}\NormalTok{(}\StringTok{"+"}\NormalTok{, }\StringTok{"-"}\NormalTok{), }\DecValTok{200}\NormalTok{, }\OtherTok{TRUE}\NormalTok{),}
                     \DataTypeTok{feature_id=}\KeywordTok{paste0}\NormalTok{(}\StringTok{"gene"}\NormalTok{, }\DecValTok{1}\OperatorTok{:}\DecValTok{200}\NormalTok{))}

\CommentTok{# create table for the columns}
\NormalTok{colData <-}\StringTok{ }\KeywordTok{DataFrame}\NormalTok{(}\DataTypeTok{timepoint=}\DecValTok{1}\OperatorTok{:}\DecValTok{6}\NormalTok{,}
                     \DataTypeTok{row.names=}\NormalTok{LETTERS[}\DecValTok{1}\OperatorTok{:}\DecValTok{6}\NormalTok{])}


\CommentTok{# create SummarizedExperiment object}
\NormalTok{se=}\KeywordTok{SummarizedExperiment}\NormalTok{(}\DataTypeTok{assays=}\KeywordTok{list}\NormalTok{(}\DataTypeTok{counts=}\NormalTok{counts),}
                     \DataTypeTok{rowRanges=}\NormalTok{rowRanges, }\DataTypeTok{colData=}\NormalTok{colData)}

\NormalTok{se}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: RangedSummarizedExperiment 
## dim: 200 6 
## metadata(0):
## assays(1): counts
## rownames: NULL
## rowData names(1): feature_id
## colnames(6): A B ... E F
## colData names(1): timepoint
\end{verbatim}

\hypertarget{subset-and-manipulate-the-summarizedexperiment-object}{%
\subsection{Subset and manipulate the SummarizedExperiment object}\label{subset-and-manipulate-the-summarizedexperiment-object}}

Now that we have a SummarizedExperiment object, we can subset it and extract/change
parts of it.

\hypertarget{extracting-parts-of-the-object}{%
\subsubsection{Extracting parts of the object}\label{extracting-parts-of-the-object}}

\texttt{colData()} and \texttt{rowData()} extract the column associated and row associated
tables. \texttt{metaData()} extracts the meta-data table if there is any table associated.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{colData}\NormalTok{(se) }\CommentTok{# extract column associated data}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## DataFrame with 6 rows and 1 column
##   timepoint
##   <integer>
## A         1
## B         2
## C         3
## D         4
## E         5
## F         6
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{rowData}\NormalTok{(se) }\CommentTok{# extrac row associated data}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## DataFrame with 200 rows and 1 column
##      feature_id
##     <character>
## 1         gene1
## 2         gene2
## 3         gene3
## 4         gene4
## 5         gene5
## ...         ...
## 196     gene196
## 197     gene197
## 198     gene198
## 199     gene199
## 200     gene200
\end{verbatim}

To extract the main table or tables that contain the values of interest such
as read counts. We must use the \texttt{assays()} function. This returns a list of
DataFrame objects associated with the object.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{assays}\NormalTok{(se) }\CommentTok{# extract list of assays}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## List of length 1
## names(1): counts
\end{verbatim}

You can use names with \texttt{\$} or \texttt{{[}{]}} notaion to extract specific tables from the list.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{assays}\NormalTok{(se)}\OperatorTok{$}\NormalTok{counts }\CommentTok{# get the table named "counts"}

\KeywordTok{assays}\NormalTok{(se)[[}\DecValTok{1}\NormalTok{]] }\CommentTok{# get the first table}
\end{Highlighting}
\end{Shaded}

\hypertarget{subsetting}{%
\subsubsection{subsetting}\label{subsetting}}

Subsetting is easy using \texttt{{[}\ {]}} notation. This is similar to the way we
subset data frames or matrices.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# subset the first five transcripts and first three samples}
\NormalTok{se[}\DecValTok{1}\OperatorTok{:}\DecValTok{5}\NormalTok{, }\DecValTok{1}\OperatorTok{:}\DecValTok{3}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: RangedSummarizedExperiment 
## dim: 5 3 
## metadata(0):
## assays(1): counts
## rownames: NULL
## rowData names(1): feature_id
## colnames(3): A B C
## colData names(1): timepoint
\end{verbatim}

One can also use \texttt{\$} operator to subset based on colData() columns. You can
extract certain samples or in our case time points.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{se[, se}\OperatorTok{$}\NormalTok{timepoint }\OperatorTok{==}\StringTok{ }\DecValTok{1}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

In addition, as \texttt{SummarizedExperiment} objects are \texttt{GRanges} objects on streoids,
they support all of the \texttt{findOverlaps()} methods and associated functions that
work on \texttt{GRanges} objects.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Subset for only rows which are in chr1:100,000-1,100,000 }
\NormalTok{roi <-}\StringTok{ }\KeywordTok{GRanges}\NormalTok{(}\DataTypeTok{seqnames=}\StringTok{"chr1"}\NormalTok{, }\DataTypeTok{ranges=}\DecValTok{100000}\OperatorTok{:}\DecValTok{1100000}\NormalTok{)}
\KeywordTok{subsetByOverlaps}\NormalTok{(se, roi)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: RangedSummarizedExperiment 
## dim: 50 6 
## metadata(0):
## assays(1): counts
## rownames: NULL
## rowData names(1): feature_id
## colnames(6): A B ... E F
## colData names(1): timepoint
\end{verbatim}

\hypertarget{visualizing-and-summarizing-genomic-intervals}{%
\section{Visualizing and summarizing genomic intervals}\label{visualizing-and-summarizing-genomic-intervals}}

Data integration and visualization is corner stone of genomic data analysis. Below, we will
show different ways of integrating and visualizing genomic intervals. These methods
can be use to visualize large amounts of data in a locus-specific or multi-loci
manner.

\hypertarget{visualizing-intervals-on-a-locus-of-interest}{%
\subsection{Visualizing intervals on a locus of interest}\label{visualizing-intervals-on-a-locus-of-interest}}

Often times, we will be interested in particular genomic locus and try to visualize
different genomic datasets over that locus. This is similar to looking at the data
over one of the genome browsers. Below we will display genes, GpG islands and read
coverage from a ChIP-seq experiment using Gviz package.For Gviz, we first need to
set the tracks to display. The tracks can be in various formats. They can be R
objects such as IRanges,GRanges and data.frame, or they can be in flat file formats
such as BigWig,BED and BAM. After the tracks are set, we can display them with
\texttt{plotTracks} function.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(Gviz)}
\CommentTok{# set tracks to display}

\CommentTok{# set CpG island track}
\NormalTok{cpgi.track=}\KeywordTok{AnnotationTrack}\NormalTok{(cpgi.gr,}
                           \DataTypeTok{name =} \StringTok{"CpG"}\NormalTok{)}

\CommentTok{# set gene track}
\CommentTok{# we will get this from EBI Biomart webservice}
\NormalTok{gene.track <-}\StringTok{ }\KeywordTok{BiomartGeneRegionTrack}\NormalTok{(}\DataTypeTok{genome =} \StringTok{"hg19"}\NormalTok{,}
                                    \DataTypeTok{chromosome =} \StringTok{"chr21"}\NormalTok{, }
                                    \DataTypeTok{start =} \DecValTok{27698681}\NormalTok{, }\DataTypeTok{end =} \DecValTok{28083310}\NormalTok{,}
                                    \DataTypeTok{name =} \StringTok{"ENSEMBL"}\NormalTok{)}


\CommentTok{# set track for ChIP-seq coverage}
\NormalTok{chipseqFile=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
                      \StringTok{"wgEncodeHaibTfbsA549.chr21.bw"}\NormalTok{,}
                      \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{cov.track=}\KeywordTok{DataTrack}\NormalTok{(chipseqFile,}\DataTypeTok{type =} \StringTok{"l"}\NormalTok{,}
                    \DataTypeTok{name=}\StringTok{"coverage"}\NormalTok{)}
  
\CommentTok{# call the display function plotTracks}
\NormalTok{track.list=}\KeywordTok{list}\NormalTok{(cpgi.track,gene.track,cov.track)}
\KeywordTok{plotTracks}\NormalTok{(track.list,}\DataTypeTok{from=}\DecValTok{27698681}\NormalTok{,}\DataTypeTok{to=}\DecValTok{28083310}\NormalTok{,}\DataTypeTok{chromsome=}\StringTok{"chr21"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/unnamed-chunk-168-1} 

}

\caption{tracks visualized using Gviz}\label{fig:unnamed-chunk-168}
\end{figure}

\hypertarget{summaries-of-genomic-intervals-on-multiple-loci}{%
\subsection{Summaries of genomic intervals on multiple loci}\label{summaries-of-genomic-intervals-on-multiple-loci}}

Looking at data one region at a time could be inefficient. One can summarize
different data sets over thousands of regions of interest and identify patterns.
This summaries can include different data types such as motifs, read coverage
and other scores associated with genomic intervals. The \texttt{genomation} package can
summarize and help identify patterns in the datasets. The datasets can have
different kinds of information and multiple file types can be used such as BED, GFF, BAM and bigWig. We will look at H3K4me3 ChIP-seq and DNAse-seq signals from H1 embryonic stem cell line. H3K4me3 is usually associated with promoters and regions with high DNAse-seq signal are associated with accessible regions, that means mostly regulatory regions. We will summarize those datasets around the transcription start sites (TSS) of genes on chromosome 20 of human hg19 assembly. We will first read the genes and extract the region around TSS, 500bp upstream and downstream. We will then create a matrix of ChIP-seq scores for those regions, each row will represent a region around a specific TSS and columns will be the scores per base. We will then plot, average enrichment values around the TSSes of genes on chromosome 20.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# get transcription start sites on chr20}
\KeywordTok{library}\NormalTok{(genomation)}
\NormalTok{transcriptFile=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
                      \StringTok{"refseq.hg19.chr20.bed"}\NormalTok{,}
                      \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{feat=}\KeywordTok{readTranscriptFeatures}\NormalTok{(transcriptFile,}
                            \DataTypeTok{remove.unusual =} \OtherTok{TRUE}\NormalTok{,}
                            \DataTypeTok{up.flank =} \DecValTok{500}\NormalTok{, }\DataTypeTok{down.flank =} \DecValTok{500}\NormalTok{)}
\NormalTok{prom=feat}\OperatorTok{$}\NormalTok{promoters }\CommentTok{# get promoters from the features}


\CommentTok{# get for H3K4me3 values around TSSes}
\CommentTok{# we use strand.aware=TRUE so - strands will}
\CommentTok{# be reversed}
\NormalTok{H3K4me3File=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
                      \StringTok{"H1.ESC.H3K4me3.chr20.bw"}\NormalTok{,}
                      \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{sm=}\KeywordTok{ScoreMatrix}\NormalTok{(H3K4me3File,prom,}
               \DataTypeTok{type=}\StringTok{"bigWig"}\NormalTok{,}\DataTypeTok{strand.aware =} \OtherTok{TRUE}\NormalTok{)}


\CommentTok{# look for the average enrichment}
\KeywordTok{plotMeta}\NormalTok{(sm, }\DataTypeTok{profile.names =} \StringTok{"H3K4me3"}\NormalTok{, }\DataTypeTok{xcoords =} \KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{500}\NormalTok{,}\DecValTok{500}\NormalTok{),}
         \DataTypeTok{ylab=}\StringTok{"H3K4me3 enrichment"}\NormalTok{,}\DataTypeTok{dispersion =} \StringTok{"se"}\NormalTok{,}
         \DataTypeTok{xlab=}\StringTok{"bases around TSS"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/unnamed-chunk-169-1} 

}

\caption{meta region plot using genomation}\label{fig:unnamed-chunk-169}
\end{figure}

The pattern we see is expected, there is a dip just around TSS and signal is more
intense on the downstream of the TSS. We can also plot a heatmap where each row is a
region around TSS and color coded by enrichment. This can show us not only the
general pattern as in the meta-region
plot but also how many of the regions produce such a pattern.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{heatMatrix}\NormalTok{(sm,}\DataTypeTok{order=}\OtherTok{TRUE}\NormalTok{,}\DataTypeTok{xcoords =} \KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{500}\NormalTok{,}\DecValTok{500}\NormalTok{),}\DataTypeTok{xlab=}\StringTok{"bases around TSS"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/unnamed-chunk-170-1} 

}

\caption{Heatmap of enrichment of H3K4me2 around TSS}\label{fig:unnamed-chunk-170}
\end{figure}

Here we saw that about half of the regions do not have any signal. In addition it seems the multi-modal profile we have observed earlier is more complicated. Certain regions seems to have signal on both sides of the TSS, whereas others have signal mostly on the downstream side.

Normally, there would be more than one experiment or we can integrate datasets from
public repositories. In this case, we can see how different signals look like on the regions we are interested in. Now, we will also use DNAse-seq data and create a list of matrices with our datasets and plot the average profile of the signals from both datasets.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{DNAseFile=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
                      \StringTok{"H1.ESC.dnase.chr20.bw"}\NormalTok{,}
                      \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}

\NormalTok{sml=}\KeywordTok{ScoreMatrixList}\NormalTok{(}\KeywordTok{c}\NormalTok{(}\DataTypeTok{H3K4me3=}\NormalTok{H3K4me3File,}
                      \DataTypeTok{DNAse=}\NormalTok{DNAseFile),prom,}
                      \DataTypeTok{type=}\StringTok{"bigWig"}\NormalTok{,}\DataTypeTok{strand.aware =} \OtherTok{TRUE}\NormalTok{)}
\KeywordTok{plotMeta}\NormalTok{(sml)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/unnamed-chunk-171-1} 

}

\caption{Average profiles of DNAse and H3K4me3 ChIP-seq}\label{fig:unnamed-chunk-171}
\end{figure}

We should now look at the heatmaps side by side and we should also cluster the rows
based on their similarity. We will be using \texttt{multiHeatMatrix} since we have multiple ScoreMatrix objects in the list. In this case, we will also use \texttt{winsorize} argument to limit extreme values,
every score above 95th percentile will be equalized the the value of the 95th percentile. In addition, \texttt{heatMatrix} and \texttt{multiHeatMatrix} can cluster the rows.
Below, we will be using k-means clustering with 3 clusters.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{1029}\NormalTok{)}
\KeywordTok{multiHeatMatrix}\NormalTok{(sml,}\DataTypeTok{order=}\OtherTok{TRUE}\NormalTok{,}\DataTypeTok{xcoords =} \KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{500}\NormalTok{,}\DecValTok{500}\NormalTok{),}
                \DataTypeTok{xlab=}\StringTok{"bases around TSS"}\NormalTok{,}\DataTypeTok{winsorize =} \KeywordTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{95}\NormalTok{),}
                \DataTypeTok{matrix.main =} \KeywordTok{c}\NormalTok{(}\StringTok{"H3K4me3"}\NormalTok{,}\StringTok{"DNAse"}\NormalTok{),}
                \DataTypeTok{column.scale=}\OtherTok{TRUE}\NormalTok{,}
                \DataTypeTok{clustfun=}\ControlFlowTok{function}\NormalTok{(x) }\KeywordTok{kmeans}\NormalTok{(x, }\DataTypeTok{centers=}\DecValTok{3}\NormalTok{)}\OperatorTok{$}\NormalTok{cluster)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/multiHeatMatrix-1} 

}

\caption{Heatmaps of H3K4me3 and DNAse data}\label{fig:multiHeatMatrix}
\end{figure}

This revealed a different picture than we have observed before. Almost half of the promoters have no signal for DNAse or H3K4me3; these regions are probably not active and associated genes are not expressed. For regions with H3K4me3 signal, there are two major patterns. One pattern where both downstream and upstream of the TSS are enriched. On the other pattern, mostly downstream of the TSS is enriched.

\hypertarget{making-karyograms-and-circos-plots}{%
\subsection{Making karyograms and circos plots}\label{making-karyograms-and-circos-plots}}

Chromosomal karyograms and circos plots are beneficial for displaying data over the
whole genome of chromosomes of interest. Although,the information that can be
displayed over these large regions are usually not very clear and only large trends
can be discerned by eye, such as loss of methylation in large regions or genome-wide.
Below, we are showing how to use \texttt{ggbio} package for plotting.
This package has a slightly different syntax than base graphics. The syntax follows
\href{http://www.springer.com/de/book/9780387245447}{grammar of graphics} logic. It is
a deconstructed way of thinking about the plot. You add your data and apply mappings
and transformations in order to achieve the final output. In ggbio, things are
relatively easy since a high-level function \texttt{autoplot} function will recognize
most of the datatypes and guess the most appropriate plot type. You can change
it is behavior by applying low-level functions. We first get the sizes of chromosomes
and make a karyogram template.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(ggbio)}
\KeywordTok{data}\NormalTok{(ideoCyto, }\DataTypeTok{package =} \StringTok{"biovizBase"}\NormalTok{)}
\NormalTok{p <-}\StringTok{ }\KeywordTok{autoplot}\NormalTok{(}\KeywordTok{seqinfo}\NormalTok{(ideoCyto}\OperatorTok{$}\NormalTok{hg19), }\DataTypeTok{layout =} \StringTok{"karyogram"}\NormalTok{)}
\NormalTok{p}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/karyo1-1} 

}

\caption{Karyogram example}\label{fig:karyo1}
\end{figure}

Next, we would like to plot CpG islands on this Karyogram. We simply do this
by adding a layer with \texttt{layout\_karyogram} function.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# read CpG islands from a generic text file}

\NormalTok{CpGiFile=filePath=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
                      \StringTok{"CpGi.hg19.table.txt"}\NormalTok{,}
                      \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{cpgi.gr=genomation}\OperatorTok{::}\KeywordTok{readGeneric}\NormalTok{(CpGiFile, }
            \DataTypeTok{chr =} \DecValTok{1}\NormalTok{, }\DataTypeTok{start =} \DecValTok{2}\NormalTok{, }\DataTypeTok{end =} \DecValTok{3}\NormalTok{,}\DataTypeTok{header=}\OtherTok{TRUE}\NormalTok{, }
          \DataTypeTok{keep.all.metadata =}\OtherTok{TRUE}\NormalTok{,}\DataTypeTok{remove.unusual=}\OtherTok{TRUE}\NormalTok{ )}

\NormalTok{p }\OperatorTok{+}\StringTok{ }\KeywordTok{layout_karyogram}\NormalTok{(cpgi.gr)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/karyo2-1} 

}

\caption{Karyogram of CpG islands over the human genome}\label{fig:karyo2}
\end{figure}

Next, we would like to plot some data over the chromosomes. This could be ChIP-seq
signal
or any other signal over the genome, we will use CpG island scores from the data set
we read earlier. We will plot a point proportional to ``obsExp'' column in the data set. We use \texttt{ylim} argument to squish the chromosomal rectangles and plot on top of those. \texttt{aes} argument defines how the data is mapped to geometry. In this case,
it says the points will have x coordinate from CpG island start positions and y coordinate from obsExp score of CpG islands.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{p }\OperatorTok{+}\StringTok{ }\KeywordTok{layout_karyogram}\NormalTok{(cpgi.gr, }\KeywordTok{aes}\NormalTok{(}\DataTypeTok{x=}\NormalTok{ start, }\DataTypeTok{y =}\NormalTok{ obsExp),}
                     \DataTypeTok{geom=}\StringTok{"point"}\NormalTok{,}
                     \DataTypeTok{ylim =} \KeywordTok{c}\NormalTok{(}\DecValTok{2}\NormalTok{,}\DecValTok{50}\NormalTok{), }\DataTypeTok{color =} \StringTok{"red"}\NormalTok{,}
                     \DataTypeTok{size=}\FloatTok{0.1}\NormalTok{,}\DataTypeTok{rect.height=}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/karyoCpG-1} 

}

\caption{Karyogram of CpG islands and their observed/expected scores over the human genome}\label{fig:karyoCpG}
\end{figure}

Another way to depict regions or quantitative signals on the chromosomes is circos plots. These are circular plots usually used for showing chromosomal rearrangements, but can also be used for depicting signals.\texttt{ggbio} package can produce all kinds of circos plots. Below, we will show how to use that for our CpG island score example.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# set the chromsome in a circle}
\CommentTok{# color set to white to look transparent }
\NormalTok{p <-}\StringTok{ }\KeywordTok{ggplot}\NormalTok{() }\OperatorTok{+}\StringTok{ }\KeywordTok{layout_circle}\NormalTok{(ideoCyto}\OperatorTok{$}\NormalTok{hg19, }\DataTypeTok{geom =} \StringTok{"ideo"}\NormalTok{, }\DataTypeTok{fill =} \StringTok{"white"}\NormalTok{,}
                              \DataTypeTok{colour=}\StringTok{"white"}\NormalTok{,}\DataTypeTok{cytoband =} \OtherTok{TRUE}\NormalTok{,}
                              \DataTypeTok{radius =} \DecValTok{39}\NormalTok{, }\DataTypeTok{trackWidth =} \DecValTok{2}\NormalTok{)}
\CommentTok{# plot the scores as points   }
\NormalTok{p <-}\StringTok{ }\NormalTok{p }\OperatorTok{+}\StringTok{ }\KeywordTok{layout_circle}\NormalTok{(cpgi.gr, }\DataTypeTok{geom =} \StringTok{"point"}\NormalTok{, }\DataTypeTok{grid=}\OtherTok{TRUE}\NormalTok{,}
                           \DataTypeTok{size =} \FloatTok{0.01}\NormalTok{, }\KeywordTok{aes}\NormalTok{(}\DataTypeTok{y =}\NormalTok{ obsExp),}\DataTypeTok{color=}\StringTok{"red"}\NormalTok{,}
                       \DataTypeTok{radius =} \DecValTok{42}\NormalTok{, }\DataTypeTok{trackWidth =} \DecValTok{10}\NormalTok{)}
\CommentTok{# set the chromosome names}
\NormalTok{p <-}\StringTok{ }\NormalTok{p }\OperatorTok{+}\StringTok{ }\KeywordTok{layout_circle}\NormalTok{(}\KeywordTok{as}\NormalTok{(}\KeywordTok{seqinfo}\NormalTok{(ideoCyto}\OperatorTok{$}\NormalTok{hg19),}\StringTok{"GRanges"}\NormalTok{), }
                       \DataTypeTok{geom =} \StringTok{"text"}\NormalTok{, }\KeywordTok{aes}\NormalTok{(}\DataTypeTok{label =}\NormalTok{ seqnames), }
                      \DataTypeTok{vjust =} \DecValTok{0}\NormalTok{, }\DataTypeTok{radius =} \DecValTok{55}\NormalTok{, }\DataTypeTok{trackWidth =} \DecValTok{7}\NormalTok{,}
                      \DataTypeTok{size=}\DecValTok{3}\NormalTok{) }

\CommentTok{# display the plot}
\NormalTok{p}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/circosCpG-1} 

}

\caption{circos plot for CpG islands scores}\label{fig:circosCpG}
\end{figure}

\hypertarget{exercises-4}{%
\section{Exercises}\label{exercises-4}}

\begin{verbatim}
> dir()
[1] "GenomicInterval.exercises.html"
\end{verbatim}

The data for the exercises is located at \texttt{GenomicIntervals\_data/data} folder.

Run the following to see the data files.

\begin{verbatim}
dir("GenomicIntervals_data/data")
\end{verbatim}

\hypertarget{operations-on-genomic-intervals-with-genomicranges-package-1}{%
\subsection{Operations on Genomic Intervals with GenomicRanges package}\label{operations-on-genomic-intervals-with-genomicranges-package-1}}

\hypertarget{section-9}{%
\subsubsection{}\label{section-9}}

Create a GRanges object using the information in the table below:

\begin{longtable}[]{@{}lllll@{}}
\toprule
chr & start & end & strand & score\tabularnewline
\midrule
\endhead
chr1 & 10000 & 10300 & + & 10\tabularnewline
chr1 & 11100 & 11500 & - & 20\tabularnewline
chr2 & 20000 & 20030 & + & 15\tabularnewline
\bottomrule
\end{longtable}

\hypertarget{section-10}{%
\subsubsection{}\label{section-10}}

use \texttt{start()}, \texttt{end()}, \texttt{strand()},\texttt{seqnames()} and \texttt{width()} functions on the GRanges
object you created. Figure out what they are doing. Can you get a subset of GRanges object for intervals that are only on + strand? If you can do that, try getting intervals that are on chr1. \emph{HINT:} GRanges objects can be subset using \protect\hyperlink{section-21}{} operator similar to data.frames but you may need
to use \texttt{start()}, \texttt{end()} and \texttt{strand()},\texttt{seqnames()} within the \protect\hyperlink{section-21}{}.

\hypertarget{section-11}{%
\subsubsection{}\label{section-11}}

Import mouse (mm9 assembly) CpG islands and refseq transcripts for chr12 from UCSC browser as \texttt{GRanges} objects using \texttt{rtracklayer} functions. HINT: Check the lecture material and modify the code there as necessary. If that somehow does not work, go to UCSC browser and download it as a BED file. The trackname for Refseq genes is ``RefSeq Genes'' and table name is ``refGene''.

\hypertarget{section-12}{%
\subsubsection{}\label{section-12}}

Following from the exercise above, get the promoters of Refseq transcripts (-1000bp and +1000 bp of the TSS) and calculate what percentage of them overlap with CpG islands. HINT: You have to get the promoter coordinates and use \texttt{findOverlaps()} or \texttt{subsetByOverlaps()} from \texttt{GenomicRanges} package. To get promoters, type \texttt{?promoters} on the R console and see how to use that function to get promoters or calculate their coordinates as shown in the lecture material.

\hypertarget{section-13}{%
\subsubsection{}\label{section-13}}

Plot the distribution of CpG island lengths for CpG islands that overlap with the
promoters.

\hypertarget{section-14}{%
\subsubsection{}\label{section-14}}

Get canonical peaks for SP1 (peaks that are in both replicates) on chr21. Peaks for each replicate are located in \texttt{GenomicIntervals\_data/data/wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep1.broadPeak.gz} and \texttt{GenomicIntervals\_data/data/wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep2.broadPeak.gz} files. HINT: You need to use \texttt{findOverlaps()} or \texttt{subsetByOverlaps()} to get the subset of peaks that occur in both replicates. You can try to read *broadPeak.gz files using genomation function \texttt{readBroadPeak}, broadPeak is just an extended BED format.
EXTRA credit: Try use \texttt{coverage()} and slice\texttt{()} functions to get canonical peaks.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(genomation)}
\NormalTok{rep1=}\KeywordTok{readBroadPeak}\NormalTok{(}\StringTok{"GenomicIntervals_data/data/wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep1.broadPeak.gz"}\NormalTok{)}
\NormalTok{rep2=}\KeywordTok{readBroadPeak}\NormalTok{(}\StringTok{"GenomicIntervals_data/data/wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep2.broadPeak.gz"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{dealing-with-mapped-high-throughput-sequencing-reads-1}{%
\subsection{Dealing with mapped high-throughput sequencing reads}\label{dealing-with-mapped-high-throughput-sequencing-reads-1}}

\hypertarget{section-15}{%
\subsubsection{}\label{section-15}}

Count the reads overlapping with canonical Sp1 peaks using the BAM file for one of the replicates: \texttt{GenomicIntervals\_data/data/wgEncodeHaibTfbsGm12878Sp1Pcr1xAlnRep1.chr21.bam}. \textbf{HINT}:
Use functions from \texttt{GenomicAlignments}, see \href{GenomicIntervals_data/lectures/genomicIntervalsTutorial.html}{lecture notes}.

\hypertarget{dealing-with-contiguous-scores-over-the-genome}{%
\subsection{Dealing with contiguous scores over the genome}\label{dealing-with-contiguous-scores-over-the-genome}}

\hypertarget{section-16}{%
\subsubsection{}\label{section-16}}

Extract Views object for the promoters on chr20 from \texttt{GenomicIntervals\_data/data/H1.ESC.H3K4me1.chr20.bw} file. Plot the first ``View'' as a line plot. \textbf{HINT}: see \href{GenomicIntervals_data/lectures/genomicIntervalsTutorial.html}{lecture notes}, adapt the code from there.

\hypertarget{section-17}{%
\subsubsection{}\label{section-17}}

Make a histogram of the maximum signal for the Views in the object you extracted above. You can use any of the view summary functions or use lapply() and write your own summary function.

\hypertarget{section-18}{%
\subsubsection{}\label{section-18}}

Get the genomic positions of maximum signal in each view and make a GRanges object. \textbf{HINT}: See ?viewRangeMaxs help page. Try to make a GRanges object out of the returned object.

\hypertarget{visualizing-and-summarizing-genomic-intervals-1}{%
\subsection{Visualizing and summarizing genomic intervals}\label{visualizing-and-summarizing-genomic-intervals-1}}

\hypertarget{section-19}{%
\subsubsection{}\label{section-19}}

Extract -500,+500 bp regions around TSSes on chr21, there are refseq files in the \texttt{GenomicIntervals\_data/data} folder or you can
pull the data out of UCSC browser. Use SP1 ChIP-seq data
(\texttt{GenomicIntervals\_data/data/wgEncodeHaibTfbsGm12878Sp1Pcr1xAlnRep1.chr21.bam} ) to create an average profile of read coverage around TSSes. Following that, visualize the read coverage with a heatmap. \textbf{HINT}: All of these possible using \texttt{genomation} package functions.

\hypertarget{section-20}{%
\subsubsection{}\label{section-20}}

Extract -500,+500 bp regions around TSSes on chr20. Use H3K4me3 (\texttt{GenomicIntervals\_data/data/H1.ESC.H3K4me3.chr20.bw}) and H3K27ac (\texttt{GenomicIntervals\_data/data/H1.ESC.H3K27ac.chr20.bw}) ChIP-seq enrichment data in the data folder and create heatmaps and average signal profiles for regions around the TSSes.

\hypertarget{section-21}{%
\subsubsection{}\label{section-21}}

Visualize one of the -500,+500 bp regions around TSS using \texttt{Gviz} functions. You should visualize both H3K4me3 and H3K27ac and the gene models.

\hypertarget{processingReads}{%
\chapter{Quality Check, Processing and Alignment of High-throughput Sequencing Reads}\label{processingReads}}

Advances in sequencing technology are helping researchers sequence the genome deeper than ever. These sequencing experiments typically yield millions of reads. These reads have to be further processed, quality checked and aligned before we can quantify the genomic signal of interest and apply statistics and/or machine learning methods. For example, you may want to count how many reads overlapping with your promoter set of interest or you may want to quantify RNA-seq reads overlapping with exons. Post-alignment operations are usually but not always similar to operations on genomic intervals. Dealing with mapped reads are described previously in chapter \ref{genomicIntervals}. In addition, we have introduced high-throughput sequencing and its applications in general in chapter \ref{intro}. In this chapter we will introduce the fundamentals of read processing and quality check, and we will show how to do those tasks in R. For a long time, quality check and mapping tasks were outside the R domain. However, nowadays certain packages in R/Bioconductor can accomplish those tasks.

\hypertarget{fasta-and-fastq-formats}{%
\section{FASTA and FASTQ formats}\label{fasta-and-fastq-formats}}

High-throughput sequencing reads are usually output from sequencing facilities as text files in a format called ``FASTQ'' or ``fastq''. This format depends on an earlier format called FASTA. The FASTA format is developed as a text-based format to represent nucleotide or protein sequences (See Figure \ref{fig:fasta} for an example).

\begin{figure}

{\centering \includegraphics[width=0.8\linewidth]{images/fastaPic} 

}

\caption{An example fasta file showing first part of PAX6 gene}\label{fig:fasta}
\end{figure}

The first line in a FASTA file usually starts with a ``\textgreater{}'' (greater-than) symbol. This first line is called the ``description line'', and can contain descriptive information about the sequence in the subsequent lines. The description can be id or name of the sequence such as gene names. However, very infrequently you may see lines starting with a ``;'' (semicolon). These lines will be taken as a comment, and can hold additional descriptive information about the sequence in subsequent lines.

An extension of the FASTA format is FASTQ format. This format is designed to handle base quality metrics output from sequencing machines. In this format, both the sequence and quality scores are represented as single ASCII characters. The format uses for lines for each sequence, and these four lines are stacked on top of each other in text files output by sequencing workflows. Each of the 4 lines will represent a read. Figure \ref{fig:fastq} shows those four lines with brief explanations for each line.

\begin{figure}

{\centering \includegraphics[width=0.8\linewidth]{images/fastqPic} 

}

\caption{FASTQ format and brief explanation of each line in the format}\label{fig:fastq}
\end{figure}

\textbf{Line 1} begins with a `@' character and is followed by a sequence identifier and an optional description. This line is utilized by the sequencing technology, and usually contains specific information for the technology. It can contain flow cell ids, lane numbers, information on read pairs. \textbf{Line 2} is the sequence letters. \textbf{Line 3} begins with a `+' character, it marks the end of sequence and is optionally followed by the same sequence identifier again in line 1. \textbf{Line 4} encodes the quality values for the sequence in Line 2, and must contain the same number of symbols as letters in the sequence. Each letter corresponds to a quality score. Although there might be different definitions of the quality scores, a \emph{de facto} standard in the field is to use ``Phred quality scores''. These scores represent the likelihood of base being called wrong. Formally, \({\displaystyle Q_{\text{phred}}=-10\log _{\text{10}}e}\), where \(e\) is probability that the base is called wrong.Since the score is in minus log scale, the higher the score, the more unlikely that the base is called wrong.

\hypertarget{quality-check-on-sequencing-reads}{%
\section{Quality check on sequencing reads}\label{quality-check-on-sequencing-reads}}

The sequencing technologies usually produce basecalls with varying quality. In addition, there could be sample specific issues in your sequencing run, such as adapter contamination. It is standard procedure to check the quality of the reads and identify problems before doing further analysis. Checking the quality and making some decisions for the downstream analysis can influence the outcome of your project.

Below, we will walk you through the quality check steps using \href{https://bioconductor.org/packages/release/bioc/html/Rqc.html}{\textbf{Rqc}} package. First, we need to feed fastq files to \texttt{rqc()} function and obtain an object with sequence quality related results. We are using example fastq files from \textbf{ShortRead} package.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(Rqc)}
\NormalTok{folder =}\StringTok{ }\KeywordTok{system.file}\NormalTok{(}\DataTypeTok{package=}\StringTok{"ShortRead"}\NormalTok{, }\StringTok{"extdata/E-MTAB-1147"}\NormalTok{)}

\CommentTok{# feeds fastq.qz files in "folder" to quality check function}
\NormalTok{qcRes=}\KeywordTok{rqc}\NormalTok{(}\DataTypeTok{path =}\NormalTok{ folder, }\DataTypeTok{pattern =} \StringTok{".fastq.gz"}\NormalTok{, }\DataTypeTok{openBrowser=}\OtherTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{sequence-quality-per-basecycle}{%
\subsection{Sequence quality per base/cycle}\label{sequence-quality-per-basecycle}}

Now that we have \texttt{qcRes} object, we can plot various sequence quality metrics for our fastq files. We will first plot " sequence quality per base/cycle". This plot depicts the quality scores across all bases at each position in the reads.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{rqcCycleQualityBoxPlot}\NormalTok{(qcRes)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.8\linewidth]{compgenomrReloaded_files/figure-latex/CycleQualityBoxPlot-1} 

}

\caption{Per base sequence quality boxplot}\label{fig:CycleQualityBoxPlot}
\end{figure}

In our case, the x-axis in the plot is labeled as ``cycle''. This is because in each sequencing ``cycle'' a fluorescently labeled nucleotide is added to complement the template sequence, and the sequencing machine identifies which nucleotide is added. Therefore, cycles corresponds to bases/nucleotides along the read, and the number of cycles is equivalent to the read length.

Long sequences can have degraded quality towards the ends of the reads. Looking at quality distribution over base positions can help us decide to do trimming towards the end of the reads or not. A good sample will have median quality scores per base above 28. If scores are below 20 towards the ends, you can think about trimming the reads.

\hypertarget{sequence-content-per-basecycle}{%
\subsection{Sequence content per base/cycle}\label{sequence-content-per-basecycle}}

Per base sequence content shows nucleotide proportions for each position. In a random sequencing library there should be no nucleotide bias and the lines should be almost parallel with each other.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{rqcCycleBaseCallsLinePlot}\NormalTok{(qcRes)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.8\linewidth]{compgenomrReloaded_files/figure-latex/baseCallFreq-1} 

}

\caption{Per base sequence quality boxplot}\label{fig:baseCallFreq}
\end{figure}

However some types of sequencing libraries can produce a biased sequence composition. For example, in RNA-Seq , it is common to have bias at the beginning of the reads. This happens because of random primers annealing to the start of reads during RNA-Seq library preparation. These primers are not truly random, and it leads to a variation at the beginning of the reads. Although RNA-seq experiments will usually have these biases, this will not affect the ability of measuring gene expression.

In addition, some libraries are inherently biased in their sequence composition. For example, in bisulfite sequencing experiments most of the cytosines will be converted to thymines. This will create a difference in C and T base compositions over the read, however this type of difference is normal for bisulfite sequencing experiments.

\hypertarget{read-frequency-plot}{%
\subsection{Read frequency plot}\label{read-frequency-plot}}

This plot shows the degree of duplication for every read in the library. A high level of duplication, non-unique reads, is likely to indicate an enrichment bias. Technical duplicates arising from PCR artefacts could cause this. PCR is a common step in library preparation which creates many copies of the sequence fragment. In RNA-seq data, non-unique read proportion can reach more than 20\%. However, these duplications may stem from simply genes being expressed at high levels. This means that there will many copies of transcripts and many copies of the same fragment. Since, we can not be sure these duplicated reads are due to PCR bias or an effect of high transcription we should not remove duplicated reads in RNA-seq analysis. However, in ChIP-seq experiments duplicated reads are more likely to be due to PCR bias.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{rqcReadFrequencyPlot}\NormalTok{(qcRes)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.8\linewidth]{compgenomrReloaded_files/figure-latex/ReadFrequencyPlot-1} 

}

\caption{Per base sequence quality boxplot}\label{fig:ReadFrequencyPlot}
\end{figure}

\hypertarget{other-quality-metrics-and-qc-tools}{%
\subsection{Other quality metrics and QC tools}\label{other-quality-metrics-and-qc-tools}}

Over-represented k-mers along the reads can be an additional check. If there are such sequences it may point to adapter contamination and should be trimmed. Adapters are known sequences that are added to the ends of the reads. This kind of contamination could also be visible at ``sequence content per base'' plots. In addition, if you know the adapter sequences you can match it to the end of the reads and trim them. The most popular tool for sequencing quality control is the fastQC tool \citep{noauthor_babraham_nodate}, which is written in Java. It produces the plots that we described above in addition to k-mer overrepresentation and adapter overrepresentation plots. The R package \href{https://cran.r-project.org/web/packages/fastqcr/index.html}{fastqcr} can run this Java tool and produce R based plots and reports. This package simply calls the Java tool and parses its results. Below, we are showing how to do that.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(fastqcr)}

\CommentTok{# install the FASTQC java tool}
\KeywordTok{fastqc_install}\NormalTok{()}

\CommentTok{# call FASTQC and record the resulting statistics}
\CommentTok{# in fastqc_results folder}
\KeywordTok{fastqc}\NormalTok{(}\DataTypeTok{fq.dir =}\NormalTok{ folder,}\DataTypeTok{qc.dir =} \StringTok{"fastqc_results"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Now that we have run FastQC on our fastq files, we can read the results to R and construct plots or reports. \texttt{gc\_report} function can create an Rmarkdown based report from FastQC output.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# view the report rendered by R functions}
\KeywordTok{qc_report}\NormalTok{(}\DataTypeTok{qc.path=}\StringTok{"fastqc_results"}\NormalTok{, }
          \DataTypeTok{result.file=}\StringTok{"reportFile"}\NormalTok{, }\DataTypeTok{preview =} \OtherTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Alternatively, we can read the results with \texttt{qc\_read()} and make specific plots we are interested in with \texttt{qc\_plot()}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# read QC results to R for one fastq file}
\NormalTok{qc <-}\StringTok{ }\KeywordTok{qc_read}\NormalTok{(}\StringTok{"fastqc_results/ERR127302_1_subset_fastqc.zip"}\NormalTok{)}

\CommentTok{# make plots, example "Per base sequence quality plot"}
\KeywordTok{qc_plot}\NormalTok{(qc, }\StringTok{"Per base sequence quality"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Apart from this, the bioconductor packages Rqc \citep{Rqc} (see \texttt{Rqc::rqcReport} function), QuasR \citep{gaidatzis_quasr:_2015} (see \texttt{QuasR::qQCReport} function), systemPipeR \citep{backman_systempiper:_2016} (see \texttt{systemPipeR::seeFastq} function), and ShortRead \citep{morgan_shortread:_2009} (see \texttt{ShortRead::report} function) packages can all generate quality reports in a similar fashion to FastQC with some differences in plot content and number.

\hypertarget{filtering-and-trimming-reads}{%
\section{Filtering and trimming reads}\label{filtering-and-trimming-reads}}

Based on the results of the quality check, you may want to trim or filter the reads. Quality check might have shown number of reads that have low quality scores. These reads will probably not align very well because of the potential mistakes in base calling, or they may align to wrong places in the genome. Therefore, you may want to remove these reads from your fastq file. Another potential scenario is that part of your reads needs to be trimmed in order align the read. In some cases, adapters will be present in either side of the read, in other cases technical errors will lead to decreasing base quality towards the ends of the reads. Both in these cases, portion of the read should be trimmed so that read can align or better align the genome. We will show how to use \textbf{QuasR} package to trim the reads. Other packages such as \textbf{ShortRead} also have capabilities to trim and filter reads. However, \texttt{QuasR::preprocessReads()} function provides a single interface to multiple preprocessing possibilities. With this function, we match adapter sequences and remove them. We can remove low-complexity reads (reads containing repetitive sequences). We can trim start or ends of the reads by a pre-defined length.

Below we will first set up the file paths to fastq files and filter them based on their length and whether or not they contain ``N'' character, which stands for unidentified base. With the same function we will also trim 3 bases from the end of the reads and also trim segments from the start of the reads if they match the ``ACCCGGGA'' sequence.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(QuasR)}

\CommentTok{# obtain a list of fastq file paths}
\NormalTok{fastqFiles <-}\StringTok{ }\KeywordTok{system.file}\NormalTok{(}\DataTypeTok{package=}\StringTok{"ShortRead"}\NormalTok{,}
                          \StringTok{"extdata/E-MTAB-1147"}\NormalTok{,}
                          \KeywordTok{c}\NormalTok{(}\StringTok{"ERR127302_1_subset.fastq.gz"}\NormalTok{,}
                            \StringTok{"ERR127302_2_subset.fastq.gz"}\NormalTok{)}
\NormalTok{)}

\CommentTok{# defined processed fastq file names}
\NormalTok{outfiles <-}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\KeywordTok{tempfile}\NormalTok{(}\DataTypeTok{pattern=}\KeywordTok{c}\NormalTok{(}\StringTok{"processed_1_"}\NormalTok{,}
                              \StringTok{"processed_2_"}\NormalTok{)),}\StringTok{".fastq"}\NormalTok{,}\DataTypeTok{sep=}\StringTok{""}\NormalTok{)}

\CommentTok{# process fastq files}
\CommentTok{# remove reads that have more than 1 N, (nBases)}
\CommentTok{# trim 3 bases from the end of the reads (truncateEndBases)}
\CommentTok{# Remove ACCCGGGA patern if it occurs at the start (Lpattern)}
\CommentTok{# remove reads shorter than 40 base-pairs (minLength)}
\KeywordTok{preprocessReads}\NormalTok{(fastqFiles, outfiles, }
                \DataTypeTok{nBases=}\DecValTok{1}\NormalTok{,}
                \DataTypeTok{truncateEndBases=}\DecValTok{3}\NormalTok{,}
                \DataTypeTok{Lpattern=}\StringTok{"ACCCGGGA"}\NormalTok{,}
                \DataTypeTok{minLength=}\DecValTok{40}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

As we have mentioned, \textbf{ShortRead} package has low-level functions, which \texttt{QuasR::preprocessReads()} also depends on. We can use these low level functions to filter reads in ways that are not possible using \texttt{QuasR::preprocessReads()} function. Below we are going to read in a fastq file and filter the reads where every quality score is below 20.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(ShortRead)}

\CommentTok{# obtain a list of fastq file paths}
\NormalTok{fastqFile <-}\StringTok{ }\KeywordTok{system.file}\NormalTok{(}\DataTypeTok{package=}\StringTok{"ShortRead"}\NormalTok{,}
                          \StringTok{"extdata/E-MTAB-1147"}\NormalTok{,}
                          \StringTok{"ERR127302_1_subset.fastq.gz"}\NormalTok{)}

\CommentTok{# read fastq file}
\NormalTok{fq =}\StringTok{ }\KeywordTok{readFastq}\NormalTok{(fastqFile)}

\CommentTok{# get quality scores per base as a matrix}
\NormalTok{qPerBase =}\StringTok{ }\KeywordTok{as}\NormalTok{(}\KeywordTok{quality}\NormalTok{(fq), }\StringTok{"matrix"}\NormalTok{)}

\CommentTok{# get number of bases per read that have quality score below 20}
\CommentTok{# we use this}
\NormalTok{qcount =}\StringTok{ }\KeywordTok{rowSums}\NormalTok{( qPerBase }\OperatorTok{<=}\StringTok{ }\DecValTok{20}\NormalTok{) }

\CommentTok{# Number of reads where all Phred scores >= 20}
\NormalTok{fq[qcount }\OperatorTok{==}\StringTok{ }\DecValTok{0}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: ShortReadQ
## length: 10699 reads; width: 72 cycles
\end{verbatim}

We can finally write out the filtered fastq file with \texttt{ShortRead::writeFastq()} function.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# write out fastq file with only reads where all }
\CommentTok{# quality scores per base are above 20}
\KeywordTok{writeFastq}\NormalTok{(fq[qcount }\OperatorTok{==}\StringTok{ }\DecValTok{0}\NormalTok{], }
           \KeywordTok{paste}\NormalTok{(fastqFile, }\StringTok{"Qfiltered"}\NormalTok{, }\DataTypeTok{sep=}\StringTok{"_"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

As fastq files can be quite large, it may not be feasible to read a 30 Gigabyte file into memory. A more memory efficient way would be to read the file piece by piece. We can do our filtering operations for each piece, write the filtered part out and read a new piece. Fortunately, this is possible by \texttt{ShortRead::FastqStreamer()} function. This function enables ``streaming'' the fastq file in pieces, which are blocks of the fastq file with a pre-defined number of reads . We can access the successive blocks with \texttt{yield()} function. Each time we call \texttt{yield()} function after opening the fastq file with \texttt{FastqStreamer()}, a new part of the file will be read to the memory.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# set up streaming with block size 1000}
\CommentTok{# every time we call the yield() function 1000 read portion}
\CommentTok{# of the file will be read successively. }
\NormalTok{f <-}\StringTok{ }\KeywordTok{FastqStreamer}\NormalTok{(fastqFile,}\DataTypeTok{readerBlockSize=}\DecValTok{1000}\NormalTok{) }

\CommentTok{# we set up a while loop to call yield() function to}
\CommentTok{# go through the file}
\ControlFlowTok{while}\NormalTok{(}\KeywordTok{length}\NormalTok{(fq <-}\StringTok{ }\KeywordTok{yield}\NormalTok{(f))) \{}
  
    \CommentTok{# remove reads where all quality scores are < 20 }
    \CommentTok{# get quality scores per base as a matrix}
\NormalTok{    qPerBase =}\StringTok{ }\KeywordTok{as}\NormalTok{(}\KeywordTok{quality}\NormalTok{(fq), }\StringTok{"matrix"}\NormalTok{)}

    \CommentTok{# get number of bases per read that have Q score < 20}
\NormalTok{    qcount =}\StringTok{ }\KeywordTok{rowSums}\NormalTok{( qPerBase }\OperatorTok{<=}\StringTok{ }\DecValTok{20}\NormalTok{) }
 
    \CommentTok{# write fastq file with mode="a", so every new block}
    \CommentTok{# is written out to the same file}
    \KeywordTok{writeFastq}\NormalTok{(fq[qcount }\OperatorTok{==}\StringTok{ }\DecValTok{0}\NormalTok{], }
               \KeywordTok{paste}\NormalTok{(fastqFile, }\StringTok{"Qfiltered"}\NormalTok{, }\DataTypeTok{sep=}\StringTok{"_"}\NormalTok{), }
               \DataTypeTok{mode=}\StringTok{"a"}\NormalTok{)}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\hypertarget{mappingaligning-reads-to-the-genome}{%
\section{Mapping/aligning reads to the genome}\label{mappingaligning-reads-to-the-genome}}

After quality check and potential pre-processing, the reads are ready to be mapped or aligned to the reference genome. This process simply finds most probable the origin of each read in the genome. Since there might be errors in sequencing and mutations in the genomes, we may not find exact matches of reads in the genomes. An important feature of the alignment algorithms is to tolerate potential mismatches between reads and the reference genome. In addition, effienct algorithms and data structures are needed for the alignment to be completed in a reasonable amount of time. Alignment methods usually create data structures to store and efficiently search the genome for matching reads. These data structures are called genome indices and creating these indices is the first step for the read alignment. Based on how indices are created, there are two major types of methods. One class of methods rely on ``hash tables'', to store and search the genomes. Hash tables are simple lookup tables, in which all possible k-mers point to locations in the genome. The general idea is that overlapping k-mers constructed from a read goes through this look up table. Each k-mer points to potential locations in the genome. Then, final location for the read is obtained by optimizing k-mer chain by their distances in the genome and in the read. This optimization process removes k-mer locations that are distant from other k-mers that map nearby each other.

Another class of algorithms build genome indices by creating Burrows-Wheeler transformation of the genome. This in essence creates a compact and searchable data structure for all reads. Although, details are out of scope for this section, these alignment tools provide faster alignment and use less memory. BWA\citep{li2009fast}, Bowtie1/2\citep{langmead2012fast} and SOAP\citep{li2009soap2} are examples of such algorithms.

The read mapping in R can be done with \texttt{gmapR} \citep{gmapR}, \texttt{QuasR} \citep{gaidatzis_quasr:_2015}, \texttt{Rsubread} \citep{liao_subread_2013}, and \texttt{systemPipeR} \citep{backman_systempiper:_2016} packages. We will demonstrate read mapping with QuasR which uses \textbf{Rbowtie} package, which wraps the Bowtie aligner. Below, we show how to map reads from a ChIP-seq experiment using QuasR/bowtie.

We will use \texttt{qAlign()} function which requires two mandatory arugments: 1) a genome file either in fasta format or as a \textbf{BSgenome} package 2) a sample file which is a text file and contains file paths to fastq files and sample names. In the case, below sample file looks like this:

\begin{verbatim}
FileName    SampleName
chip_1_1.fq.bz2 Sample1
chip_2_1.fq.bz2 Sample2
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(QuasR)}

\CommentTok{# copy example data to current working directory}
\KeywordTok{file.copy}\NormalTok{(}\KeywordTok{system.file}\NormalTok{(}\DataTypeTok{package=}\StringTok{"QuasR"}\NormalTok{, }\StringTok{"extdata"}\NormalTok{), }\StringTok{"."}\NormalTok{, }\DataTypeTok{recursive=}\OtherTok{TRUE}\NormalTok{)}

\CommentTok{# genome file in fasta format}
\NormalTok{genomeFile <-}\StringTok{ "extdata/hg19sub.fa"}

\CommentTok{# text file containing sample names and fastq file paths}
\NormalTok{sampleFile <-}\StringTok{ "extdata/samples_chip_single.txt"}

\CommentTok{# create alignments }
\NormalTok{proj <-}\StringTok{ }\KeywordTok{qAlign}\NormalTok{(sampleFile, genomeFile)}
\end{Highlighting}
\end{Shaded}

It is good to explain what is going on here as the \texttt{qAlign()} function makes things look simple. This function is designed to be easy. For example, it creates a genome index automatically if it does not exist, and will look for existing indices before it creates one. We provided only two arguments, a text file containing sample names and fastq file paths and a reference genome file. In fact, this function also has many knobs and you can change its behavior by supplying different arguments in order to affect the behavior of Bowtie. For example, you can supply parameters to Bowtie using \texttt{alignmentParameter} argument. However the \texttt{qAlign()} function is optimized for different types of alignment problems and selects alignment parameters automatically. It is designed to work with alignment and quantification tasks for RNA-seq, ChIP-seq, small-RNA sequencing, Bisulfite sequencing (DNA methylation) and allele specific analysis. If you want to change default bowtie parameters only do it for simple alignment problems such as ChIP-seq and RNA-seq.

\BeginKnitrBlock{rmdtip}
\textbf{Want to know more ?}

\begin{itemize}
\tightlist
\item
  More on hash tables and Burrows-Wheeler based aligners

  \begin{itemize}
  \tightlist
  \item
    \href{https://academic.oup.com/bib/article/11/5/473/264166}{A survey of sequence alignment algorithms for next-generation sequencing} H Li, N Homer - Briefings in bioinformatics, 2010
  \end{itemize}
\item
  \href{https://bioconductor.org/packages/release/bioc/vignettes/QuasR/inst/doc/QuasR.html}{More on QuasR} and all the alignment and post-processing capabilities.
\end{itemize}
\EndKnitrBlock{rmdtip}

\hypertarget{further-processing-of-aligned-reads}{%
\section{Further processing of aligned reads}\label{further-processing-of-aligned-reads}}

After alignment some further processing might be necessary. However, these steps are usually sequencing protocol specific. For example, for methylation C-\textgreater{}T mismatches should be counted. For gene expression measurements, reads that overlap with transcripts should be counted. These further processing tasks are either done by specialized alignment related software or can be done in R. We will explain these further processing steps when they become relevant in the context of following chapters.

\hypertarget{exercises-5}{%
\section{Exercises}\label{exercises-5}}

TBD

\hypertarget{rnaseqanalysis}{%
\chapter{RNA-seq Analysis}\label{rnaseqanalysis}}

RNA sequencing (RNA-seq) has proven as a revolutionary tool since the time it has been introduced. The throughput, accuracy, and resolution of data produced with RNA-seq has been instrumental in the study of transcriptomics in the last decade \citep{wang_rna-seq:_2009}. There is a variety of applications of transcriptome sequencing and each application may consist of different chains of tools each with many alternatives \citep{conesa_survey_2016}. In this chapter, we are going to demonstrate a common workflow of how to do differential expression analysis with downstream applications such as GO term and gene set enrichment analysis. We assume that the sequencing data was generated using one of the NGS sequencing platforms. Where applicable, we will try to provide alternatives to the reader in terms of both the tools to carry out a demonstrated analysis and also the other applications of the same sequencing data depending on the different biological questions.

\hypertarget{what-is-gene-expression}{%
\section{What is gene expression?}\label{what-is-gene-expression}}

\texttt{Gene\ expression} is a term used to describe the contribution of a gene to the overall functions and phenotype of a cell through the activity of the molecular products, which are encoded in the specific nucleotide sequence of the gene. RNA is the primary product encoded in a gene, which is transcribed in the nucleus of a cell. A class of RNA molecules, messenger RNAs, are transported from the nucleus to the cytoplasm, where the translation machinery of the cell translates the nucleotide sequence of the mRNA into proteins. The functional protein repertoire in a given cell is the primary factor that dictates the shape, function, and phenotype of a cell. Due to the prime roles of proteins for a cell's fate, most molecular biology literature is focused on protein-coding genes. However, a bigger proportion of a eukaryotic gene repertoire is reserved for non-coding genes, which code for RNA molecules that are not translated into proteins, yet carry out many important cellular functions. All in all, the term \texttt{gene\ expression} refers to the combined activity of protein-coding or non-coding products of a gene.

In a cell, there are many layers of quality controls and modifications that act upon a gene's product until the end-product attains a particular function. These layers of regulation include epigenetic, transcriptional, post-transcriptional, translational, and post-translational control mechanisms, the latter two applying only to protein-coding genes. A protein or RNA molecule, is only functional if it is produced at the right time, at the right cellular compartment, with the neccessary base or amino-acid modifications, with the correct secondary/tertiary structure (or unstructure wherever applicable), among the availability of other metabolites or molecules, which are needed to form complexes to altogether accomplish a certain cellular function. However, traditionally, the number of copies of a gene's products is considered a quantitative measure of a gene's activity. Although this approach does not reflect all of the complexity of what defines a functional molecule, quantification of the abundance of transcripts from a gene has proven to be a cost-effective method in understanding genes' functions.

\hypertarget{methods-to-detect-gene-expression}{%
\section{Methods to detect gene expression}\label{methods-to-detect-gene-expression}}

Quantification of how much expression levels of genes deviate from a baseline gives clues about which genes are actually important for, for instance, disease outcome or cell/tissue identity. The methods of detecting and quantifying gene expression has evolved from low-throughput methods such as the usage of a reporter gene with a flourescent protein product to find out if a single gene is expressed at all, to high-throughput methods such as massively parallel RNA-sequencing that can profile -at a single-nucleotide resolution- the abundance of tens of thousands of distinct transcripts encoded in the largest eukaryotic genomes.

\hypertarget{gene-expression-analysis-using-high-throughput-sequencing-technologies}{%
\section{Gene Expression Analysis Using High-throughput Sequencing Technologies}\label{gene-expression-analysis-using-high-throughput-sequencing-technologies}}

With the advent of the second-generation (a.k.a next-generation or high-throughput) sequencing technologies,
the number of genes that can be profiled for expression levels with a single experiment has increased to the order of tens of thousands of genes. Therefore, the bottleneck in this process has become the data analysis rather than the data generation. Many statistical methods and computational tools are required for getting meaningful results from the data, which comes with a lot of valuable information along with a lot of sources of noise. Fortunately, most of the steps of RNA-seq analysis have become quite mature over the years. Below we will first describe how to reach a read count table from raw fastq reads obtained from an Illumina sequencing run. We will then demonstrate in R how to process the count table, make a case-control differential expression analysis, and do some downstream functional enrichment analysis.

\hypertarget{processing-raw-data}{%
\subsection{Processing raw data}\label{processing-raw-data}}

\hypertarget{quality-check-and-read-processing}{%
\subsubsection{Quality check and read processing}\label{quality-check-and-read-processing}}

The first step in any experiment that involves high-throughput short-read sequencing should be to check the sequencing quality of the reads before starting to do any downstream analysis. The quality of the input sequences holds fundamental importance in the confidence for the biological conclusions drawn from the experiment. We have introduced quality check and processing in Chapter \ref{processingReads}, those tools and workflows also apply in RNA-seq analysis.

\hypertarget{improving-the-quality}{%
\subsubsection{Improving the quality}\label{improving-the-quality}}

The second step in the RNA-seq analysis workflow is to improve the quality of the input reads. This step could be regarded as an optional step when the sequencing quality is very good. However, even with the highest quality sequencing datasets, this step may still improve the quality of the input sequences. The most common technical artifacts that can be filtered out are the adapter sequences that contaminate the sequenced reads, and the low quality bases that are usually found at the ends of the sequences. Commonly used tools in the field (trimmomatic \citep{bolger_trimmomatic:_2014}, trimGalore \citep{noauthor_babraham_nodate-1}) are again not written in R, however there are alternative R libraries for carrying out the same functionality, for instance, QuasR \citep{gaidatzis_quasr:_2015} (see \texttt{QuasR::preprocessReads} function) and ShortRead \citep{morgan_shortread:_2009} (see \texttt{ShortRead::filterFastq} function). Some of these approaches are introduced in Chapter \ref{processingReads}.

The sequencing quality control and read pre-processing steps can be visited multiple times until achieving a satisfactory level of quality in the sequence data before moving onto the dowstream analysis steps.

\hypertarget{alignment}{%
\subsection{Alignment}\label{alignment}}

Once a decent level of quality in the sequences is reached, the expression level of the genes can be quantified by first mapping the sequences to a reference genome, and secondly matching the aligned reads to the gene annotations, in order to count the number of reads mapping to each gene. If the species under study has a well annotated transcriptome, the reads can be aligned to the transcript sequences instead of the reference genome. In cases where there is no good quality reference genome or transcriptome, it is possible to de novo assemble the transcriptome from the sequences and then quantify the expression levels of genes/transcripts.

For RNA-seq read alignments, apart from the availability of reference genomes and annotations, probably the most important factor to consider when choosing an alignment tool is whether the alignment method considers the absence of intronic regions in the sequenced reads, while the target genome may contain introns. Therefore, it is important to choose alignment tools that take into account alternative splicing. In the basic setting where a read, which originates from a cDNA sequence corresponding to an exon-exon junction, needs to be split into two parts when aligned against the genome. There are various tools that consider this factor such as STAR \citep{dobin_star:_2013}, Tophat2 \citep{kim_tophat2:_2013}, Hisat2 \citep{kim_hisat:_2015}, GSNAP \citep{wu_gmap_2016}. Most alignment tools are written in C/C++ languages because of performance concerns. There are also R libraries that can do short read alignments, these are discussed in Chapter \ref{processingReads}.

\hypertarget{quantification}{%
\subsection{Quantification}\label{quantification}}

After the reads are aligned to the target, a SAM/BAM file sorted by coordinates should have been obtained. The BAM file contains all alignment related information of all the reads that have been attempted to be aligned to the target sequence. This information consists of - most basically - the genomic coordinates (chromosome, start, end, strand) of where a sequence was matched (if at all) in the target, specific insertions/deletions/mismatches that describes the differences between the input and target sequences. These pieces of information are used along with the genomic coordinates of genome annotations such as gene/transcript models in order to count how many reads have been sequenced from a gene/transcript. As simple as it may sound, it is not a trivial task to assign reads to a gene/transcript just by comparing the genomic coordinates of the annotations and the sequences, because of the confounding factors such as overlapping gene annotations, overlapping exon annotations from different transcript isoforms of a gene, overlapping annotations from opposite DNA strands in the absence of a strand-specific sequencing protocol. Therefore, for read counting, it is important to consider:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Strand specificity of the sequencing protocol: are the reads expected to originate from the forward strand, reverse strand, or unspecific?
\item
  Counting mode:
  - when counting at the gene-level: when there are overlapping annotations, which features should the read be assigned to? Tools usually have a parameter that lets the user to select a counting mode.
  - when counting at the transcript-level: when there are multiple isoforms of a gene, which isoform should the read be assigned to? This consideration is usually an algorithmic consideration that is not modifiable by the end-user.
\end{enumerate}

Some tools can couple alignment to quantification (e.g.~STAR), while some assume the alignments are already calculated and require BAM files as input. On the other hand, in the presence of good transcriptome annotations, alignment-free methods (Salmon \citep{patro_salmon:_2017}, Kallisto \citep{bray_near-optimal_2016}, Sailfish \citep{patro_sailfish_2014}) can also be used to estimate the expression levels of transcripts/genes. There are also reference-free quantification methods that can first de novo assemble the transcriptome and estimate the expression levels based on this assembly. Such a strategy can be useful in discovering novel transcripts or may be required in cases when a good reference does not exist. If a reference transcriptome exists but of low quality, a reference-based transcriptome assembler such as Cufflinks \citep{trapnell_transcript_2010} can be used to improve the transcriptome. In case there is no available transcriptome annotation, a de novo assembler such as Trinity \citep{haas_novo_2013} or Trans-ABySS \citep{robertson_novo_2010} can be used to assemble the transcriptome from scratch.

Within R, quantification can be done using

\begin{itemize}
\tightlist
\item
  \texttt{Rsubread::featureCounts}
\item
  \texttt{QuasR::qCount}
\item
  \texttt{GenomicAlignments::summarizeOverlaps}
\end{itemize}

\hypertarget{within-sample-normalization-of-the-read-counts}{%
\subsection{Within sample normalization of the read counts}\label{within-sample-normalization-of-the-read-counts}}

The most common application after a gene's expression is quantified (as the number of reads aligned to the gene), is to compare the gene's expression in different conditions, for instance, in a case-control setting (e.g.~disease versus normal) or in a time-series (e.g.~along different developmental stages). Making such comparisons help identify the genes that might be responsible for a disease or an impaired developmental trajectory. However, there are multiple caveats that needs to be addressed before making a comparison between the read counts of a gene in different conditions \citep{maza_comparison_2013}

\begin{itemize}
\tightlist
\item
  Library size (i.e.~sequencing depth) varies between samples coming from different lanes of the flow cell of the sequencing machine.
\item
  Longer genes will have higher number of reads.\\
\item
  Library composition (i.e.~relative size of the studied transcriptome) can be different in two different biological conditions.
\item
  GC content biases across different samples may lead to a biased sampling of genes \citep{risso_gc-content_2011}.
\item
  Read coverage of a transcript can be biased and non-uniformly distributed along the transcript \citep{mortazavi_mapping_2008}.
\end{itemize}

Therefore these factors need to be taken into account before making comparisons.

The most basic normalization approaches address the sequencing depth bias. Such procedures normalize the read counts per gene by dividing each gene's read count by a certain value and multiplying it by 10\^{}6. These normalized values are usually referred to as CPM (counts per million reads):

\begin{itemize}
\tightlist
\item
  Total Counts Normalization (divide counts by the \textbf{sum} of all counts)
\item
  Upper Quartile Normalization (divide counts by the \textbf{upper quartile} value of the counts)
\item
  Median Normalization (divide counts by the \textbf{median} of all counts)
\end{itemize}

Popular metrics that improve upon CPM are RPKM/FPKM (reads/fragments per kilobase of million reads) and TPM (transcripts per million). RPKM is obtained by dividing the CPM value by another factor, which is the length of the gene per kilobases. FPKM is the same as RPKM, but is used for paired-end reads. Thus, RPKM/FPKM methods account for, firstly the \textbf{library size}, and secondly the \textbf{gene lengths}.

TPM also controls for both the library size and the gene lengths, however, with the TPM method, the read counts are first normalized by the gene length (per kilobase), and then gene-length normalized values are divided by the sum of the gene-length normalized values and multiplied by 10\^{}6. Thus, the sum of normalized values for TPM will always be equal to 10\^{}6 for each library, while the sum of RPKM/FPKM values do not sum to 10\^{}6. Therefore, it is easier to interpret TPM values than RPKM/FPKM values.

\hypertarget{computing-different-normalization-schemes-in-r}{%
\subsection{Computing different normalization schemes in R}\label{computing-different-normalization-schemes-in-r}}

Here we will assume that there is an RNA-seq count table comprising of raw-counts, meaning the number of reads counted for each gene has not been exposed to any kind of normalization and consists of integers. The rows of the count table correspond to the genes and the columns represent different samples. Here we will use a subset of the RNA-seq count table from a colorectal cancer study. We have filtered the original count table for only protein-coding genes (to improve the speed of calculation) and also selected only five metastasized colorectal cancer samples along with five normal colon samples. There is an additional column \texttt{width} that contains the length of the corresponding gene in the unit of base pairs. The length of the genes are important to compute RPKM and TPM values. The original count tables can be found from the recount2 database {[}REF{]} using the SRA project code \texttt{SRP029880} and the experimental setup along with other accessory information can be found from the NCBI Trace archive using the SRA project code \texttt{SRP029880} \href{https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP029880}{here}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#colorectal cancer}
\NormalTok{counts_file <-}\StringTok{ }\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata/rna-seq/SRP029880.raw_counts.tsv"}\NormalTok{, }\DataTypeTok{package =} \StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{coldata_file <-}\StringTok{ }\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata/rna-seq/SRP029880.colData.tsv"}\NormalTok{, }\DataTypeTok{package =} \StringTok{"compGenomRData"}\NormalTok{)}

\NormalTok{counts <-}\StringTok{ }\KeywordTok{as.matrix}\NormalTok{(}\KeywordTok{read.table}\NormalTok{(counts_file, }\DataTypeTok{header =}\NormalTok{ T, }\DataTypeTok{sep =} \StringTok{'}\CharTok{\textbackslash{}t}\StringTok{'}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\hypertarget{computing-cpm}{%
\subsubsection{Computing CPM}\label{computing-cpm}}

Let's do a summary of the counts table:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{summary}\NormalTok{(counts)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##      CASE_1             CASE_2        
##  Min.   :0.00e+00   Min.   :0.00e+00  
##  1st Qu.:5.16e+03   1st Qu.:6.46e+03  
##  Median :8.00e+04   Median :8.51e+04  
##  Mean   :2.96e+05   Mean   :2.73e+05  
##  3rd Qu.:2.52e+05   3rd Qu.:2.45e+05  
##  Max.   :2.05e+08   Max.   :1.05e+08  
##      CASE_3             CASE_4        
##  Min.   :0.00e+00   Min.   :0.00e+00  
##  1st Qu.:3.97e+03   1st Qu.:7.09e+03  
##  Median :6.41e+04   Median :9.91e+04  
##  Mean   :2.63e+05   Mean   :3.65e+05  
##  3rd Qu.:2.11e+05   3rd Qu.:3.03e+05  
##  Max.   :2.23e+08   Max.   :1.62e+08  
##      CASE_5             CTRL_1        
##  Min.   :0.00e+00   Min.   :0.00e+00  
##  1st Qu.:5.74e+03   1st Qu.:6.90e+03  
##  Median :8.55e+04   Median :8.68e+04  
##  Mean   :3.46e+05   Mean   :3.39e+05  
##  3rd Qu.:2.74e+05   3rd Qu.:2.86e+05  
##  Max.   :1.27e+08   Max.   :2.73e+08  
##      CTRL_2             CTRL_3        
##  Min.   :0.00e+00   Min.   :0.00e+00  
##  1st Qu.:4.75e+03   1st Qu.:4.76e+03  
##  Median :6.31e+04   Median :6.15e+04  
##  Mean   :2.73e+05   Mean   :2.64e+05  
##  3rd Qu.:2.15e+05   3rd Qu.:2.12e+05  
##  Max.   :3.99e+08   Max.   :2.82e+08  
##      CTRL_4             CTRL_5        
##  Min.   :0.00e+00   Min.   :0.00e+00  
##  1st Qu.:2.72e+03   1st Qu.:4.56e+03  
##  Median :3.72e+04   Median :6.23e+04  
##  Mean   :1.83e+05   Mean   :2.76e+05  
##  3rd Qu.:1.34e+05   3rd Qu.:2.19e+05  
##  Max.   :2.63e+08   Max.   :2.92e+08  
##      width        
##  Min.   :    117  
##  1st Qu.:   9668  
##  Median :  27090  
##  Mean   :  67685  
##  3rd Qu.:  70374  
##  Max.   :2473537
\end{verbatim}

To compute the CPM values for each sample (excluding the \texttt{width} column):

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cpm <-}\StringTok{ }\KeywordTok{apply}\NormalTok{(}\KeywordTok{subset}\NormalTok{(counts, }\DataTypeTok{select =} \KeywordTok{c}\NormalTok{(}\OperatorTok{-}\NormalTok{width)), }\DecValTok{2}\NormalTok{, }\ControlFlowTok{function}\NormalTok{(x) x}\OperatorTok{/}\KeywordTok{sum}\NormalTok{(}\KeywordTok{as.numeric}\NormalTok{(x)) }\OperatorTok{*}\StringTok{ }\DecValTok{10}\OperatorTok{^}\DecValTok{6}\NormalTok{)}
\KeywordTok{head}\NormalTok{(cpm)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##            CASE_1  CASE_2   CASE_3   CASE_4  CASE_5
## TSPAN6   133.0525 69.0267 118.0347 63.45320 75.1706
## TNMD       0.2541  0.1497   0.5774  0.04131  0.1603
## DPM1      62.5345 50.9434  47.9547 51.61492 47.6836
## SCYL3     17.7536 18.1283  18.9681 16.32642 14.9281
## C1ORF112  15.5608 10.9995  11.8489 15.14763  6.5660
## FGR        9.7817 27.0412   5.1668 23.44216 23.2660
##           CTRL_1  CTRL_2  CTRL_3  CTRL_4   CTRL_5
## TSPAN6   83.6818 90.8358 63.7171 66.1201 116.5649
## TNMD      0.6926  0.3495  0.8603  0.5436   0.7309
## DPM1     31.5817 22.8556 21.7976 18.6671  36.4578
## SCYL3    24.0552 19.8292 20.5182 17.9990  18.6608
## C1ORF112  7.8228  6.3511  7.3078  5.1346  10.6632
## FGR       8.9318  4.5419  4.1459  5.1712   4.3533
\end{verbatim}

Check that the sum of each column after normalization equals to 10\^{}6 (except the width column).

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{colSums}\NormalTok{(cpm)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## CASE_1 CASE_2 CASE_3 CASE_4 CASE_5 CTRL_1 CTRL_2 
##  1e+06  1e+06  1e+06  1e+06  1e+06  1e+06  1e+06 
## CTRL_3 CTRL_4 CTRL_5 
##  1e+06  1e+06  1e+06
\end{verbatim}

\hypertarget{computing-rpkm}{%
\subsubsection{Computing RPKM}\label{computing-rpkm}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# create a vector of gene lengths }
\NormalTok{geneLengths <-}\StringTok{ }\KeywordTok{as.vector}\NormalTok{(}\KeywordTok{subset}\NormalTok{(counts, }\DataTypeTok{select =} \KeywordTok{c}\NormalTok{(width)))}

\CommentTok{# compute rpkm }
\NormalTok{rpkm <-}\StringTok{ }\KeywordTok{apply}\NormalTok{(}\DataTypeTok{X =} \KeywordTok{subset}\NormalTok{(counts, }\DataTypeTok{select =} \KeywordTok{c}\NormalTok{(}\OperatorTok{-}\NormalTok{width)), }
                         \DataTypeTok{MARGIN =} \DecValTok{2}\NormalTok{, }
                         \DataTypeTok{FUN =} \ControlFlowTok{function}\NormalTok{(x) }\DecValTok{10}\OperatorTok{^}\DecValTok{9} \OperatorTok{*}\StringTok{ }\NormalTok{x }\OperatorTok{/}\StringTok{ }\NormalTok{geneLengths }\OperatorTok{/}\StringTok{ }\KeywordTok{sum}\NormalTok{(}\KeywordTok{as.numeric}\NormalTok{(x)))}
\KeywordTok{head}\NormalTok{(rpkm)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##            CASE_1   CASE_2  CASE_3   CASE_4  CASE_5
## TSPAN6   10.32776 5.357969 9.16205 4.925343 5.83487
## TNMD      0.01685 0.009922 0.03828 0.002739 0.01063
## DPM1      2.63981 2.150508 2.02434 2.178856 2.01290
## SCYL3     0.39773 0.406127 0.42494 0.365760 0.33443
## C1ORF112  0.08101 0.057267 0.06169 0.078864 0.03418
## FGR       0.42137 1.164864 0.22257 1.009828 1.00224
##           CTRL_1  CTRL_2  CTRL_3  CTRL_4  CTRL_5
## TSPAN6   6.49552 7.05083 4.94583 5.13235 9.04796
## TNMD     0.04592 0.02317 0.05704 0.03604 0.04845
## DPM1     1.33318 0.96482 0.92016 0.78801 1.53902
## SCYL3    0.53891 0.44423 0.45967 0.40323 0.41806
## C1ORF112 0.04073 0.03307 0.03805 0.02673 0.05552
## FGR      0.38476 0.19565 0.17860 0.22276 0.18753
\end{verbatim}

Check the sample sizes of RPKM. Notice that the sums of samples are all different

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{colSums}\NormalTok{(rpkm)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## CASE_1 CASE_2 CASE_3 CASE_4 CASE_5 CTRL_1 CTRL_2 
## 158291 153324 161775 173047 172761 210033 301764 
## CTRL_3 CTRL_4 CTRL_5 
## 241418 291675 252006
\end{verbatim}

\hypertarget{computing-tpm}{%
\subsubsection{Computing TPM}\label{computing-tpm}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#find gene length normalized values }
\NormalTok{rpk <-}\StringTok{ }\KeywordTok{apply}\NormalTok{( }\KeywordTok{subset}\NormalTok{(counts, }\DataTypeTok{select =} \KeywordTok{c}\NormalTok{(}\OperatorTok{-}\NormalTok{width)), }\DecValTok{2}\NormalTok{, }
              \ControlFlowTok{function}\NormalTok{(x) x}\OperatorTok{/}\NormalTok{(geneLengths}\OperatorTok{/}\DecValTok{1000}\NormalTok{))}
\CommentTok{#normalize by the sample size using rpk values}
\NormalTok{tpm <-}\StringTok{ }\KeywordTok{apply}\NormalTok{(rpk, }\DecValTok{2}\NormalTok{, }\ControlFlowTok{function}\NormalTok{(x) x }\OperatorTok{/}\StringTok{ }\KeywordTok{sum}\NormalTok{(}\KeywordTok{as.numeric}\NormalTok{(x)) }\OperatorTok{*}\StringTok{ }\DecValTok{10}\OperatorTok{^}\DecValTok{6}\NormalTok{)}
\KeywordTok{head}\NormalTok{(tpm)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##           CASE_1   CASE_2  CASE_3   CASE_4   CASE_5
## TSPAN6   65.2454 34.94535 56.6344 28.46239 33.77413
## TNMD      0.1064  0.06471  0.2366  0.01583  0.06153
## DPM1     16.6770 14.02588 12.5133 12.59109 11.65132
## SCYL3     2.5127  2.64881  2.6267  2.11364  1.93581
## C1ORF112  0.5118  0.37350  0.3813  0.45573  0.19787
## FGR       2.6620  7.59739  1.3758  5.83556  5.80130
##           CTRL_1   CTRL_2  CTRL_3   CTRL_4  CTRL_5
## TSPAN6   30.9263 23.36535 20.4865 17.59617 35.9038
## TNMD      0.2186  0.07678  0.2363  0.12356  0.1923
## DPM1      6.3475  3.19727  3.8115  2.70166  6.1071
## SCYL3     2.5658  1.47212  1.9040  1.38247  1.6589
## C1ORF112  0.1939  0.10958  0.1576  0.09165  0.2203
## FGR       1.8319  0.64836  0.7398  0.76373  0.7441
\end{verbatim}

Check the sample sizes of \texttt{tpm}. Notice that the sums of samples are all equal to 10\^{}6.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{colSums}\NormalTok{(tpm)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## CASE_1 CASE_2 CASE_3 CASE_4 CASE_5 CTRL_1 CTRL_2 
##  1e+06  1e+06  1e+06  1e+06  1e+06  1e+06  1e+06 
## CTRL_3 CTRL_4 CTRL_5 
##  1e+06  1e+06  1e+06
\end{verbatim}

None of these metrics (CPM, RPKM/FPKM, TPM) account for the other important confounding factor when comparing expression levels of genes across samples: the \textbf{library composition}, which may also be referred to as the \textbf{relative size of the compared transcriptomes}. This factor is not dependent on the sequencing technology, it is rather biological. For instance, when comparing transcriptomes of different tissues, there can be sets of genes in one tissue that consume a big chunk of the reads, while in the other tissue not expressed at all. This kind of imbalances in the composition of compared transcriptomes can lead to wrong conclusions about which genes are actually differentially expressed. This consideration is addressed in two popular R packages: DESeq2 \citep{love_moderated_2014} and edgeR \citep{robinson_edger:_2010} each with a different algorithm. edgeR uses a normalization procecure called Trimmed Mean of M-values (TMM). DESeq2 implements a normalization procedure using Median of Ratios, which is obtained by finding the ratio of log-transformed count of a gene divided by the average of log-transformed values of the gene in all samples (geometric mean), and then taking the median of these values for all genes. The raw read count of the gene is finally divided by this value (median of ratios) to obtain the normalized counts of genes. In order to see a demonstration of DESeq2's normalization method, see \texttt{insert\ section\ id} about differential expression analysis.

\hypertarget{exploratory-analysis-of-the-read-count-table}{%
\subsection{Exploratory analysis of the read count table}\label{exploratory-analysis-of-the-read-count-table}}

A typical quality control, in this case interrogating the RNA-seq experiment design, is to measure the similariy of the samples with each other in terms of the quantified expression level profiles accross a set of genes. One important observation to make is to see, whether the most similar samples to any given sample are the biological replicates of that sample. This can be computed using unsupervised clustering techniques such as hierarchical clustering and visualized as a heatmap with dendrograms. Another most commonly applied technique is a dimensionality reduction technique called Principal Component Analysis (PCA) and visualized as a two-dimensional (or in some cases three-dimensional) scatter plot. In order to find out more about the clustering methods and PCA, please refer to sections \texttt{insert\ section} and \texttt{insert\ section}, respectively.

\hypertarget{clustering}{%
\subsubsection{Clustering}\label{clustering}}

We can combine clustering and visualization of the clustering results by using heatmap functions that are available in a variety of R libraries. The basic R installation comes with the \texttt{stats::heatmap} function. However, there are other libraries available in CRAN (e.g.~pheatmap \citep{pheatmap}) or Bioconductor (e.g.~ComplexHeatmap \citep{gu_complex_2016}) that come with more flexibility and more appealing visualisations.

Here we demonstrate a heatmap using \texttt{pheatmap} package and the previously calculated \texttt{tpm} matrix.
As these matrices can be quite large, both computing the clustering and rendering the heatmaps can take a lot of resources and time. Therefore, a quick and informative way to compare samples is to select a subset of genes that are, for instance, most variable across samples, and use that subset to do the clustering and visualization.

Let's select top 100 most variable genes among the samples.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#compute the variance of each gene across samples}
\NormalTok{V <-}\StringTok{ }\KeywordTok{apply}\NormalTok{(tpm, }\DecValTok{1}\NormalTok{, var)}
\CommentTok{#sort the results by variance in decreasing order and select the top 100 genes }
\NormalTok{selectedGenes <-}\StringTok{ }\KeywordTok{names}\NormalTok{(V[}\KeywordTok{order}\NormalTok{(V, }\DataTypeTok{decreasing =}\NormalTok{ T)][}\DecValTok{1}\OperatorTok{:}\DecValTok{100}\NormalTok{])}
\end{Highlighting}
\end{Shaded}

Now we can quickly produce a heatmap where samples and genes are clustered

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(pheatmap)}
\KeywordTok{pheatmap}\NormalTok{(tpm[selectedGenes,], }\DataTypeTok{scale =} \StringTok{'row'}\NormalTok{, }\DataTypeTok{show_rownames =} \OtherTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/tpm_hier_clust_1-1} \end{center}

We can also overlay some annotation tracks to observe the clusters.
Here it is important to observe whether the replicates of the same sample cluster most closely with each other, or not. Overlaying the heatmap with such annotation and displaying sample groups with distinct colors helps quickly see if there are samples that don't cluster as expected.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{colData <-}\StringTok{ }\KeywordTok{read.table}\NormalTok{(coldata_file, }\DataTypeTok{header =}\NormalTok{ T, }\DataTypeTok{sep =} \StringTok{'}\CharTok{\textbackslash{}t}\StringTok{'}\NormalTok{, }\DataTypeTok{stringsAsFactors =} \OtherTok{TRUE}\NormalTok{)}
\KeywordTok{pheatmap}\NormalTok{(tpm[selectedGenes,], }\DataTypeTok{scale =} \StringTok{'row'}\NormalTok{, }
         \DataTypeTok{show_rownames =} \OtherTok{FALSE}\NormalTok{, }
         \DataTypeTok{annotation_col =}\NormalTok{ colData)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/tpm_hier_clust_2-1} \end{center}

\hypertarget{pca}{%
\subsubsection{PCA}\label{pca}}

Let's make a PCA plot to see the clustering of replicates as a scatter plot in two dimensions.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(stats)}
\KeywordTok{library}\NormalTok{(ggplot2)}
\CommentTok{#transpose the matrix }
\NormalTok{M <-}\StringTok{ }\KeywordTok{t}\NormalTok{(tpm[selectedGenes,])}
\CommentTok{# transform the counts to log2 scale }
\NormalTok{M <-}\StringTok{ }\KeywordTok{log2}\NormalTok{(M }\OperatorTok{+}\StringTok{ }\DecValTok{1}\NormalTok{)}
\CommentTok{#compute PCA }
\NormalTok{pcaResults <-}\StringTok{ }\KeywordTok{prcomp}\NormalTok{(M)}

\CommentTok{#plot PCA results making use of ggplot2's autoplot function}
\CommentTok{#ggfortify is needed to let ggplot2 know about PCA data structure. }
\KeywordTok{autoplot}\NormalTok{(pcaResults, }\DataTypeTok{data =}\NormalTok{ colData, }\DataTypeTok{colour =} \StringTok{'group'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/pca_1-1} \end{center}

We should observe here whether the samples from the case group (CASE) and samples from the control group (CTRL) can be split into two distinct clusters on the scatter plot of the first two largest principal components.

We can use the \texttt{summary} function to summarize the PCA results to observe the contribution of the principal components in the explained variation.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{summary}\NormalTok{(pcaResults)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Importance of components:
##                           PC1    PC2     PC3     PC4
## Standard deviation     24.396 2.5051 2.39327 1.93841
## Proportion of Variance  0.957 0.0101 0.00921 0.00604
## Cumulative Proportion   0.957 0.9671 0.97627 0.98231
##                            PC5    PC6     PC7     PC8
## Standard deviation     1.79193 1.6357 1.46059 1.30902
## Proportion of Variance 0.00516 0.0043 0.00343 0.00276
## Cumulative Proportion  0.98747 0.9918 0.99520 0.99796
##                            PC9     PC10
## Standard deviation     1.12657 4.62e-15
## Proportion of Variance 0.00204 0.00e+00
## Cumulative Proportion  1.00000 1.00e+00
\end{verbatim}

\hypertarget{correlation-plots}{%
\subsubsection{Correlation plots}\label{correlation-plots}}

Another complementary approach to see the reproducibility of the experiments is to compute the correlation scores between each pair of samples and draw a correlation plot.

Let's first compute pairwise correlation scores between every pair of samples.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(stats)}
\NormalTok{correlationMatrix <-}\StringTok{ }\KeywordTok{cor}\NormalTok{(tpm)}
\end{Highlighting}
\end{Shaded}

Let's have a look at how the correlation matrix looks:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(knitr)}
\KeywordTok{kable}\NormalTok{(correlationMatrix,}\DataTypeTok{booktabs =} \OtherTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{tabular}{lrrrrrrrrrr}
\toprule
  & CASE\_1 & CASE\_2 & CASE\_3 & CASE\_4 & CASE\_5 & CTRL\_1 & CTRL\_2 & CTRL\_3 & CTRL\_4 & CTRL\_5\\
\midrule
CASE\_1 & 1.0000 & 0.9925 & 0.9959 & 0.9934 & 0.9902 & 0.9594 & 0.9636 & 0.9550 & 0.9549 & 0.9303\\
CASE\_2 & 0.9925 & 1.0000 & 0.9887 & 0.9936 & 0.9898 & 0.9726 & 0.9794 & 0.9675 & 0.9739 & 0.9448\\
CASE\_3 & 0.9959 & 0.9887 & 1.0000 & 0.9951 & 0.9929 & 0.9649 & 0.9618 & 0.9556 & 0.9552 & 0.9401\\
CASE\_4 & 0.9934 & 0.9936 & 0.9951 & 1.0000 & 0.9923 & 0.9740 & 0.9749 & 0.9643 & 0.9710 & 0.9511\\
CASE\_5 & 0.9902 & 0.9898 & 0.9929 & 0.9923 & 1.0000 & 0.9632 & 0.9661 & 0.9502 & 0.9587 & 0.9313\\
\addlinespace
CTRL\_1 & 0.9594 & 0.9726 & 0.9649 & 0.9740 & 0.9632 & 1.0000 & 0.9880 & 0.9902 & 0.9874 & 0.9864\\
CTRL\_2 & 0.9636 & 0.9794 & 0.9618 & 0.9749 & 0.9661 & 0.9880 & 1.0000 & 0.9820 & 0.9970 & 0.9642\\
CTRL\_3 & 0.9550 & 0.9675 & 0.9556 & 0.9643 & 0.9502 & 0.9902 & 0.9820 & 1.0000 & 0.9851 & 0.9903\\
CTRL\_4 & 0.9549 & 0.9739 & 0.9552 & 0.9710 & 0.9587 & 0.9874 & 0.9970 & 0.9851 & 1.0000 & 0.9739\\
CTRL\_5 & 0.9303 & 0.9448 & 0.9401 & 0.9511 & 0.9313 & 0.9864 & 0.9642 & 0.9903 & 0.9739 & 1.0000\\
\bottomrule
\end{tabular}

We can also draw more visually appealing correlation plots using the \texttt{corrplot} package.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(corrplot)}
\CommentTok{# The correlation plot order by the results of the hierarchical clustering }
\KeywordTok{corrplot}\NormalTok{(correlationMatrix, }\DataTypeTok{order =} \StringTok{'hclust'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/corrplot_3-1} \end{center}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# The clusters are split into 2 groups by adding rectangles around the computed clusters. }
\KeywordTok{corrplot}\NormalTok{(correlationMatrix, }\DataTypeTok{order =} \StringTok{'hclust'}\NormalTok{, }\DataTypeTok{addrect =} \DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/corrplot_3-2} \end{center}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# We can add the pairwise correlation scores on the plot}
\KeywordTok{corrplot}\NormalTok{(correlationMatrix, }\DataTypeTok{order =} \StringTok{'hclust'}\NormalTok{, }\DataTypeTok{addrect =} \DecValTok{2}\NormalTok{, }\DataTypeTok{addCoef.col =} \StringTok{'white'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/corrplot_3-3} \end{center}

Here pairwise correlation levels are visualized as colored circles. \texttt{Blue} indicates positive correlation, while \texttt{Red} indicates negative correlation.

We could also plot this correlation matrix as a heatmap:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(pheatmap)}
\CommentTok{# basic correlation heatmap}
\KeywordTok{pheatmap}\NormalTok{(correlationMatrix)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/corrplot_4-1} \end{center}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# add annotations on the columns}
\KeywordTok{pheatmap}\NormalTok{(correlationMatrix, }\DataTypeTok{annotation_col =}\NormalTok{ colData)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/corrplot_4-2} \end{center}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# split the clusters into two based on the clustering similarity }
\KeywordTok{pheatmap}\NormalTok{(correlationMatrix,  }\DataTypeTok{annotation_col =}\NormalTok{ colData, }\DataTypeTok{cutree_cols =} \DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/corrplot_4-3} \end{center}

\hypertarget{differential-expression-analysis}{%
\subsection{Differential expression analysis}\label{differential-expression-analysis}}

Differential expression analysis allows to test tens of thousands of hypotheses (one test for each gene) against the null hypothesis that the activity of the gene stays the same in two different conditions. There are multiple limiting factors that influence the power of detecting genes that have real changes between two biological conditions. Among these are the limited number of biological replicates, non-normality of the distribution of the read counts, and higher uncertainty of measurements for lowly expressed genes than highly expressed genes \citep{love_moderated_2014}. Tools such as edgeR and DESeq2 address these limitations using sophisticated statistical models in order to maximize the amount of knowledge that can be extracted from such noisy datasets.\\
DESeq2:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Normalizes the read counts by computing size factors, which addresses the differences not only in the library sizes, but also the library compositions.
\item
  For each gene, estimates the dispersion value of the gene among the biological replicates. The dispersion value computed by DESeq2 is equal to the squared coefficient of variation (variation divided by the mean).\\
\item
  A line is fitted across the dispersion estimates of all genes computed in 2) versus the mean normalized counts of the genes.
\item
  Dispersion values of each gene is shrunken towards the fitted line in 3).
\item
  A Generalized Linear Model is fitted which considers additional confounding variables related to the experimental design such as sequencing batches, treatment, temperature, patient's age, sequencing technology etc.
\item
  For a given contrast (e.g.~treatment type: drug-A versus untreated), a test for differential expression is carried out against the null hypothesis that the log fold change of the normalized counts of the gene in the given pair of groups is exactly zero.
\item
  Adjusts p-values for multiple-testing.
\end{enumerate}

In order to carry out a differential expression analysis using \texttt{DESeq2}, three kinds of inputs are necessary:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  The \textbf{read count table}: must be raw read counts as integers that are not processed in any form by a normalization technique. The rows represent features (e.g.~genes, transcripts, genomic intervals) and columns represent samples.
\item
  A \textbf{colData} table: this table describes the experimental design.
\item
  A \textbf{design formula}: this formula is needed to describe the variable of interest in the analysis (e.g.~treatment status) along with (optionally) other covariates (e.g.~batch, temperature, sequencing technology).
\end{enumerate}

Let's define these inputs:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#remove the 'width' column}
\NormalTok{countData <-}\StringTok{ }\KeywordTok{as.matrix}\NormalTok{(}\KeywordTok{subset}\NormalTok{(counts, }\DataTypeTok{select =} \KeywordTok{c}\NormalTok{(}\OperatorTok{-}\NormalTok{width)))}
\CommentTok{#define the experimental setup }
\NormalTok{colData <-}\StringTok{ }\KeywordTok{read.table}\NormalTok{(coldata_file, }\DataTypeTok{header =}\NormalTok{ T, }\DataTypeTok{sep =} \StringTok{'}\CharTok{\textbackslash{}t}\StringTok{'}\NormalTok{, }\DataTypeTok{stringsAsFactors =} \OtherTok{TRUE}\NormalTok{)}
\CommentTok{#define the design formula}
\NormalTok{designFormula <-}\StringTok{ "~ group"}
\end{Highlighting}
\end{Shaded}

Now, we are ready to run DESeq2.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(DESeq2)}
\KeywordTok{library}\NormalTok{(stats)}
\CommentTok{#create a DESeq dataset object from the count matrix and the colData }
\NormalTok{dds <-}\StringTok{ }\KeywordTok{DESeqDataSetFromMatrix}\NormalTok{(}\DataTypeTok{countData =}\NormalTok{ countData, }
                              \DataTypeTok{colData =}\NormalTok{ colData, }
                              \DataTypeTok{design =} \KeywordTok{as.formula}\NormalTok{(designFormula))}
\CommentTok{#print dds object to see the contents}
\KeywordTok{print}\NormalTok{(dds)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## class: DESeqDataSet 
## dim: 19719 10 
## metadata(1): version
## assays(1): counts
## rownames(19719): TSPAN6 TNMD ... MYOCOS HSFX3
## rowData names(0):
## colnames(10): CASE_1 CASE_2 ... CTRL_4 CTRL_5
## colData names(2): source_name group
\end{verbatim}

The \texttt{DESeqDataSet} object contains all the information about the experimental setup, the read counts, and the design formulas. Certain functions can be used to access these information separately: \texttt{rownames(dds)} shows which features are used in the study (e.g.~genes), \texttt{colnames(dds)} displays the studied samples, \texttt{counts(dds)} displays the count table, \texttt{colData(dds)} displays the experimental setup.

Remove genes that have almost no information in any of the given samples.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#For each gene, we count the total number of reads for that gene in all samples }
\CommentTok{#and remove those that don't have at least 1 read. }
\NormalTok{dds <-}\StringTok{ }\NormalTok{dds[ }\KeywordTok{rowSums}\NormalTok{(}\KeywordTok{counts}\NormalTok{(dds)) }\OperatorTok{>}\StringTok{ }\DecValTok{1}\NormalTok{, ]}
\end{Highlighting}
\end{Shaded}

Now, we can use the \texttt{DESeq} function of \texttt{DESeq2}, which is a wrapper function that implements estimation of size factors to normalize the counts, estimation of dispersion values, and computing a GLM model based on the experimental design formula. This function returns a \texttt{DESeqDataSet} object, which is an updated version of the \texttt{dds} variable that we pass to the function as input.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{dds <-}\StringTok{ }\KeywordTok{DESeq}\NormalTok{(dds)}
\end{Highlighting}
\end{Shaded}

Now, we can compare and contrast the samples based on different variables of interest. In this case, we currently have only one variable, which is the \texttt{group} variable that determines if a sample belongs to the CASE group or the CTRL group.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#compute the contrast for the 'group' variable where 'CTRL' samples are used as the control group. }
\NormalTok{DEresults =}\StringTok{ }\KeywordTok{results}\NormalTok{(dds, }\DataTypeTok{contrast =} \KeywordTok{c}\NormalTok{(}\StringTok{"group"}\NormalTok{, }\StringTok{'CASE'}\NormalTok{, }\StringTok{'CTRL'}\NormalTok{))}
\CommentTok{#sort results by increasing p-value}
\NormalTok{DEresults <-}\StringTok{ }\NormalTok{DEresults[}\KeywordTok{order}\NormalTok{(DEresults}\OperatorTok{$}\NormalTok{pvalue),]}
\end{Highlighting}
\end{Shaded}

Thus we have obtained a table containing the differential expression status of case samples compared to the control samples.

It is important to note that the sequence of the elements provided in the \texttt{contrast} argument determines which group of samples are to be used as \texttt{control}. This impacts the way the results are interpreted, for instance, if a gene is found up-regulated (has a positive log2 fold change), the up-regulation status is only relative to the factor that is provided as \texttt{control}. In this case, we used samples from the \texttt{CTRL} group as \texttt{control} and contrasted the samples from the \texttt{CASE} group with respect to the \texttt{CTRL} samples. Thus genes with a positive log2 fold change are called up-regulated in the case samples with respect to the control, while genes with a negative log2 fold change are down-regulated in the case samples. Whether the deregulation is significant or not, warrants assessment of the adjusted p-values.

Let's have a look into the contents of the \texttt{DEresults} table.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#shows a summary of the results}
\KeywordTok{print}\NormalTok{(DEresults)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## log2 fold change (MLE): group CASE vs CTRL 
## Wald test p-value: group CASE vs CTRL 
## DataFrame with 19097 rows and 6 columns
##                    baseMean     log2FoldChange
##                   <numeric>          <numeric>
## CYP2E1     4829888.92231712   9.36024435305687
## FCGBP      10349993.1789843  -7.57578665890778
## ASGR2      426421.665773775   8.01830102661517
## GCKR       100182.667004114   7.82841275978615
## APOA5      438053.955454865   10.2024847632603
## ...                     ...                ...
## CCDC195    20.4981398045639 -0.215606612258398
## SPEM3      23.6369514506586  -22.1547650648436
## AC022167.5 21.8450828694606  -2.05624008484309
## BX276092.9  29.963634159758  0.407325618882749
## ETDC       22.5674885843896  -1.79527360789704
##                        lfcSE                stat
##                    <numeric>           <numeric>
## CYP2E1     0.215223173019895    43.4908761065038
## FCGBP       0.18643292339242   -40.6354549456998
## ASGR2      0.216206795652736    37.0862581002954
## GCKR       0.233376221798576    33.5441747212052
## APOA5      0.312502755796718    32.6476633373977
## ...                      ...                 ...
## CCDC195      2.8925516729234 -0.0745385516451266
## SPEM3       3.02784692796743     -7.317003003093
## AC022167.5  2.89545307238825  -0.710161772073565
## BX276092.9   2.8904762978251   0.140919895862573
## ETDC        2.89421003612133  -0.620298314735642
##                           pvalue                  padj
##                        <numeric>             <numeric>
## CYP2E1                         0                     0
## FCGBP                          0                     0
## ASGR2       4.6789767784482e-301 2.87741475285303e-297
## GCKR       1.09479013375331e-246 5.04944579440372e-243
## APOA5      8.64905801850778e-234   3.191329427669e-230
## ...                          ...                   ...
## CCDC195                       NA                    NA
## SPEM3                         NA                    NA
## AC022167.5                    NA                    NA
## BX276092.9                    NA                    NA
## ETDC                          NA                    NA
\end{verbatim}

The first three lines in this output shows the contrast and the statistical test that were used to compute these results, along with the dimensions of the resulting table (number of columns and rows). Below these lines is the actual table with 6 columns: \texttt{baseMean} represents the average normalized expression of the gene across all considered samples. \texttt{log2FoldChange} represents the base-2 logarithm of the fold change of the normalized expression of the gene in the given contrast. \texttt{lfcSE} represents the standard error of log2 fold change estimate, and \texttt{stat} is the statistic calculated in the contrast which is translated into a \texttt{pvalue} and adjusted for multiple testing in the \texttt{padj} column. To find out about the importance of adjusting for \texttt{multiple\ testing}, refer to section \texttt{insert\ section\ for\ multiple\ testing}.

\hypertarget{diagnostic-plots}{%
\subsubsection{Diagnostic plots}\label{diagnostic-plots}}

At this point, before proceeding to do any downstream analysis and jumping to conclusions about the biological insights that are reachable with the experimental data at hand, it is important to do some more diagnostic tests to improve our confidence about the quality of the data and the experimental setup.

An MA plot is useful to observe if the data normalisation worked well. MA plot is a scatterplot where x axis denotes the average of normalized counts across samples and the y axis denotes the log fold change in the given contrast. Most points are expected to be on the horizontal 0 line (most genes are expected to be not differentially expressed).

\hypertarget{ma-plot}{%
\paragraph{MA plot}\label{ma-plot}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(DESeq2)}
\NormalTok{DESeq2}\OperatorTok{::}\KeywordTok{plotMA}\NormalTok{(}\DataTypeTok{object =}\NormalTok{ dds, }\DataTypeTok{ylim =} \KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{5}\NormalTok{, }\DecValTok{5}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/DE_ma_plot-1} \end{center}

It is also important to observe the distribution of raw p-values. We expect to see a peak around low p-values and a uniform distribution at p-values above 0.1. Otherwise, adjustment for multiple testing does not work and the results are not meaningful.

\hypertarget{p-value-distribution}{%
\paragraph{p-value distribution}\label{p-value-distribution}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(ggplot2)}
\KeywordTok{ggplot}\NormalTok{(}\DataTypeTok{data =} \KeywordTok{as.data.frame}\NormalTok{(DEresults), }\KeywordTok{aes}\NormalTok{(}\DataTypeTok{x =}\NormalTok{ pvalue)) }\OperatorTok{+}\StringTok{ }\KeywordTok{geom_histogram}\NormalTok{(}\DataTypeTok{bins =} \DecValTok{100}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/DE_pval_dist-1} \end{center}

A final diagnosis is to check the biological reproducibility of the sample replicates in a PCA plot or a heatmap. To plot the PCA results, we need to extract the normalized counts from the DESeqDataSet object. It is possible to color the points in the scatterplot by the variable of interest, which helps to see if the replicates cluster well.

\hypertarget{pca-plot}{%
\paragraph{PCA plot}\label{pca-plot}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(DESeq2)}
\CommentTok{# extract normalized counts from the DESeqDataSet object}
\NormalTok{countsNormalized <-}\StringTok{ }\KeywordTok{counts}\NormalTok{(dds, }\DataTypeTok{normalized =} \OtherTok{TRUE}\NormalTok{)}

\CommentTok{# select top 500 most variable genes}
\NormalTok{selectedGenes <-}\StringTok{ }\KeywordTok{names}\NormalTok{(}\KeywordTok{sort}\NormalTok{(}\KeywordTok{apply}\NormalTok{(countsNormalized, }\DecValTok{1}\NormalTok{, var), }\DataTypeTok{decreasing =} \OtherTok{TRUE}\NormalTok{)[}\DecValTok{1}\OperatorTok{:}\DecValTok{500}\NormalTok{])}

\KeywordTok{plotPCA}\NormalTok{(countsNormalized[selectedGenes,], }\DataTypeTok{col =} \KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group))}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/DE_pca-1} \end{center}

Alternatively, the normalized counts can be transformed using \texttt{DESeq2::rlog} function and \texttt{DESeq2::plotPCA} can be readily used to plot the PCA results.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{rld <-}\StringTok{ }\KeywordTok{rlog}\NormalTok{(dds)}
\KeywordTok{plotPCA}\NormalTok{(rld, }\DataTypeTok{ntop =} \DecValTok{500}\NormalTok{, }\DataTypeTok{intgroup =} \StringTok{'group'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/DE_rld_norm-1} \end{center}

\hypertarget{relative-log-expression-rle-plot}{%
\paragraph{Relative Log Expression (RLE) plot}\label{relative-log-expression-rle-plot}}

A similar plot to the MA plot is the RLE (Relative Log Expression) plot that is useful in finding out if the data at hand needs normalization \citep{gandolfo_rle_2018}. Sometimes, even the datasets normalized using the explained methods above may need further normalization due to unforeseen sources of variation that might stem from the library preparation, the person who carries out the experiment, the date of sequencing, the temperature changes in the laboratory at the time of library preparation, and so on and sofort. RLE plot is a quick diagnostic that can be applied on the raw or normalized count matrices to see if further processing is required.

Let's do RLE plots on the raw counts and normalized counts using the \texttt{EDASeq} package \citep{risso_gc-content_2011}.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(EDASeq)}
\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow =} \KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{2}\NormalTok{))}
\KeywordTok{plotRLE}\NormalTok{(countData, }\DataTypeTok{outline=}\OtherTok{FALSE}\NormalTok{, }\DataTypeTok{ylim=}\KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{4}\NormalTok{, }\DecValTok{4}\NormalTok{), }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group), }\DataTypeTok{main =} \StringTok{'Raw Counts'}\NormalTok{)}
\KeywordTok{plotRLE}\NormalTok{(}\KeywordTok{counts}\NormalTok{(dds, }\DataTypeTok{normalized =} \OtherTok{TRUE}\NormalTok{), }\DataTypeTok{outline=}\OtherTok{FALSE}\NormalTok{, }\DataTypeTok{ylim=}\KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{4}\NormalTok{, }\DecValTok{4}\NormalTok{), }\DataTypeTok{col =} \KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group), }\DataTypeTok{ylim =} \KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{4}\NormalTok{,}\DecValTok{4}\NormalTok{), }\DataTypeTok{main =} \StringTok{'Normalized Counts'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/DE_rle_plot-1} \end{center}

Here the RLE plot is comprised of box plots, where each box-plot represents the distribution of the relative log expression of the genes expressed in the corresponding sample. Each gene's expression is divided by the median expression value of that gene across all samples. Then this is transformed to log scale, which gives the relative log expression value for a single gene. The RLE values for all the genes from a sample is visualized as a boxplot.

Ideally the boxplots are centered around the horizontal zero line and are as tightly distributed as possible \citep{risso_normalization_2014}. From the plots that we have made for the raw and normalized count data, we can observe how the normalized dataset has improved upon the raw count data for all the samples. However, in some cases, it is important to visualize RLE plots in combination with other diagnostic plots such as PCA plots, heatmaps, and correlation plots to see if there is more unwanted variation in the data, which can be further accounted for using packages such as \texttt{RUVseq} \citep{risso_normalization_2014} and \texttt{sva} \citep{leek_sva_2012}. See about how to use RUVSeq package to account for unwanted sources of noise in the count tables.

\hypertarget{functional-enrichment-analysis}{%
\subsection{Functional Enrichment Analysis}\label{functional-enrichment-analysis}}

\hypertarget{go-term-analysis}{%
\subsubsection{GO term analysis}\label{go-term-analysis}}

In a typical differential expression analysis, thousands of genes are found differentially expressed between two groups of samples. While prior knowledge of the functions of individual genes can give some clues about what kind of cellular processes have been affected, e.g.~by a drug treatment, manually going through the whole list of thousands of genes would be very cumbersome and not be very informative in the end. Therefore a commonly used tool to address this problem is to do enrichment analyses of functional terms that appear associated to the given set of differentially expressed genes more often than expected by chance. The functional terms usually are associated to multiple genes. Thus, genes can be grouped into sets by shared functional terms. However, it is important to have an agreed upon controlled vocabulary on the list of terms used to describe the functions of genes. Otherwise, it would be impossible to exchange scientific results globally. That's why initiatives such as Gene Ontology Consortium have collated a list of Gene Ontology (GO) terms for each gene. GO term enrichment analysis is probably the most common analysis applied after a differential expression analysis. GO term analysis helps quickly find out systematic changes that can describe differences between groups of samples.

In R, the simplest way to do enrichment analysis is via the \texttt{gProfileR} package.

Let's select the genes that are significantly differentially expressed between the case and control samples.
Let's remove genes that have an adjusted p-value below 0.1 and that show a 2-fold change (either negative or positive) in the case compared to control.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(DESeq2)}
\KeywordTok{library}\NormalTok{(gProfileR)}
\KeywordTok{library}\NormalTok{(knitr)}
\CommentTok{# extract differential expression results}
\NormalTok{DEresults <-}\StringTok{ }\KeywordTok{results}\NormalTok{(dds, }\DataTypeTok{contrast =} \KeywordTok{c}\NormalTok{(}\StringTok{'group'}\NormalTok{, }\StringTok{'CASE'}\NormalTok{, }\StringTok{'CTRL'}\NormalTok{))}

\CommentTok{#remove genes with NA values }
\NormalTok{DE <-}\StringTok{ }\NormalTok{DEresults[}\OperatorTok{!}\KeywordTok{is.na}\NormalTok{(DEresults}\OperatorTok{$}\NormalTok{padj),]}
\CommentTok{#select genes with adjusted p-values below 0.1}
\NormalTok{DE <-}\StringTok{ }\NormalTok{DE[DE}\OperatorTok{$}\NormalTok{padj }\OperatorTok{<}\StringTok{ }\FloatTok{0.1}\NormalTok{,]}
\CommentTok{#select genes with absolute log2 fold change above 1 (two-fold change)}
\NormalTok{DE <-}\StringTok{ }\NormalTok{DE[}\KeywordTok{abs}\NormalTok{(DE}\OperatorTok{$}\NormalTok{log2FoldChange) }\OperatorTok{>}\StringTok{ }\DecValTok{1}\NormalTok{,]}

\CommentTok{#get the list of genes of interest}
\NormalTok{genesOfInterest <-}\StringTok{ }\KeywordTok{rownames}\NormalTok{(DE)}

\CommentTok{#calculate enriched GO terms}
\NormalTok{goResults <-}\StringTok{ }\KeywordTok{gprofiler}\NormalTok{(}\DataTypeTok{query =}\NormalTok{ genesOfInterest, }
                     \DataTypeTok{organism =} \StringTok{'hsapiens'}\NormalTok{, }
                     \DataTypeTok{src_filter =} \StringTok{'GO'}\NormalTok{, }
                     \DataTypeTok{hier_filtering =} \StringTok{'moderate'}\NormalTok{)}
\CommentTok{# sort the enriched GO terms by pvalue and print the top 10 terms }
\CommentTok{#  for the selected columns from the go results}
\KeywordTok{kable}\NormalTok{(goResults[}\KeywordTok{order}\NormalTok{(goResults}\OperatorTok{$}\NormalTok{p.value), }
                \KeywordTok{c}\NormalTok{(}\DecValTok{3}\OperatorTok{:}\DecValTok{4}\NormalTok{, }\DecValTok{7}\NormalTok{, }\DecValTok{10}\NormalTok{, }\DecValTok{12}\NormalTok{)][}\DecValTok{1}\OperatorTok{:}\DecValTok{10}\NormalTok{,],}\DataTypeTok{booktabs =} \OtherTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{tabular}{lrrrll}
\toprule
  & p.value & term.size & precision & domain & term.name\\
\midrule
55 & 0 & 2740 & 0.223 & CC & plasma membrane part\\
24 & 0 & 1609 & 0.136 & BP & ion transport\\
30 & 0 & 3656 & 0.258 & BP & regulation of biological quality\\
28 & 0 & 385 & 0.042 & BP & extracellular structure organization\\
34 & 0 & 7414 & 0.452 & BP & multicellular organismal process\\
\addlinespace
89 & 0 & 1069 & 0.090 & MF & transmembrane transporter activity\\
41 & 0 & 1073 & 0.090 & BP & organic acid metabolic process\\
16 & 0 & 975 & 0.083 & BP & response to drug\\
1 & 0 & 1351 & 0.107 & BP & biological adhesion\\
11 & 0 & 4760 & 0.302 & BP & system development\\
\bottomrule
\end{tabular}

\hypertarget{gene-set-enrichment-analysis}{%
\subsubsection{Gene set enrichment analysis}\label{gene-set-enrichment-analysis}}

A gene set is a collection of genes with some common property. This shared property among a set of genes could be a GO term, a common biological pathway, a shared interaction partner, or any biologically relevant commonality that is meaningful in the context of the pursued experiment. Gene set enrichment analysis (GSEA) is a valuable exploratory analysis tool that can associate systematic changes to a high-level function rather than individual genes. Analysis of coordinated changes of expression levels of gene sets can provide complementary benefits on top of per-gene based differential expression analyses. For instance, consider a gene set belonging to a biological pathway where each member of the pathway displays a slight deregulation in a disease sample compared to a normal sample. In such a case individual genes might not be picked up by the per-gene based differential expression analysis. Thus, the GO/Pathway enrichment on the differentially expressed list of genes would not show an enrichment of this pathway. However, the additive effect of slight changes of the genes could amount to a large effect at the level of the gene set, thus the pathway could be detected as a significant pathway that could explain the mechanistic problems in the disease sample.

We use the bioconductor package \texttt{gage} \citep{luo_gage:_2009} to demonstrate how to do GSEA using normalized expression data of the samples as input. Here we are using only two gene sets: one from the top GO term discovered from the previous GO analysis, one that we compile by randomly selecting a list of genes. However, annotated gene sets can be used from databases such as MSIGDB \citep{subramanian_gene_2005}, which compile gene sets from a variety of resources such as KEGG \citep{kanehisa_kegg_2016} and REACTOME \citep{fabregat_reactome_2018}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Let's define the first gene set as the list of genes from one of the significant GO terms  }
\CommentTok{# found in the GO analysis. }
\CommentTok{# order go results by pvalue}
\NormalTok{goResults <-}\StringTok{ }\NormalTok{goResults[}\KeywordTok{order}\NormalTok{(goResults}\OperatorTok{$}\NormalTok{p.value),]}
\CommentTok{#restrict the terms that have at most 100 genes overlapping with the query}
\NormalTok{go <-}\StringTok{ }\NormalTok{goResults[goResults}\OperatorTok{$}\NormalTok{overlap.size }\OperatorTok{<}\StringTok{ }\DecValTok{100}\NormalTok{,]}
\CommentTok{# use the top term from this table to create a gene set }
\NormalTok{geneSet1 <-}\StringTok{ }\KeywordTok{unlist}\NormalTok{(}\KeywordTok{strsplit}\NormalTok{(go[}\DecValTok{1}\NormalTok{,]}\OperatorTok{$}\NormalTok{intersection, }\StringTok{','}\NormalTok{))}

\CommentTok{#Define another gene set by just randomly selecting 25 genes from the counts table}
\CommentTok{#get normalized counts from DESeq2 results}
\NormalTok{normalizedCounts <-}\StringTok{ }\KeywordTok{counts}\NormalTok{(dds, }\DataTypeTok{normalized =} \OtherTok{TRUE}\NormalTok{)}
\NormalTok{geneSet2 <-}\StringTok{ }\KeywordTok{sample}\NormalTok{(}\KeywordTok{rownames}\NormalTok{(normalizedCounts), }\DecValTok{25}\NormalTok{)}

\NormalTok{geneSets <-}\StringTok{ }\KeywordTok{list}\NormalTok{(}\StringTok{'top_GO_term'}\NormalTok{ =}\StringTok{ }\NormalTok{geneSet1,}
                 \StringTok{'random_set'}\NormalTok{ =}\StringTok{ }\NormalTok{geneSet2)}
\end{Highlighting}
\end{Shaded}

Using the defined gene sets, we'd like to do a group comparison between the case samples with respect to the control samples.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(gage)}
\CommentTok{#use the normalized counts to carry out a GSEA. }
\NormalTok{gseaResults <-}\StringTok{ }\KeywordTok{gage}\NormalTok{(}\DataTypeTok{exprs =} \KeywordTok{log2}\NormalTok{(normalizedCounts}\OperatorTok{+}\DecValTok{1}\NormalTok{), }
           \DataTypeTok{ref =} \KeywordTok{match}\NormalTok{(}\KeywordTok{rownames}\NormalTok{(colData[colData}\OperatorTok{$}\NormalTok{group }\OperatorTok{==}\StringTok{ 'CTRL'}\NormalTok{,]), }\KeywordTok{colnames}\NormalTok{(normalizedCounts)), }
           \DataTypeTok{samp =} \KeywordTok{match}\NormalTok{(}\KeywordTok{rownames}\NormalTok{(colData[colData}\OperatorTok{$}\NormalTok{group }\OperatorTok{==}\StringTok{ 'CASE'}\NormalTok{,]), }\KeywordTok{colnames}\NormalTok{(normalizedCounts)),}
           \DataTypeTok{gsets =}\NormalTok{ geneSets, }\DataTypeTok{compare =} \StringTok{'as.group'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

We can observe if there is a significant upregulation or downregulation of the gene set in the case group compared to the controls by accessing \texttt{gseaResults\$greater} or \texttt{gseaResults\$less}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{knitr}\OperatorTok{::}\KeywordTok{kable}\NormalTok{(gseaResults}\OperatorTok{$}\NormalTok{greater,}
      \DataTypeTok{booktabs =} \OtherTok{TRUE}\NormalTok{, }
      \DataTypeTok{caption =} \StringTok{"Up-regulation statistics"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{table}[t]

\caption{\label{tab:gseaPost1}Up-regulation statistics}
\centering
\begin{tabular}{lrrrrrr}
\toprule
  & p.geomean & stat.mean & p.val & q.val & set.size & exp1\\
\midrule
top\_GO\_term & 0.0000 & 7.199 & 0.0000 & 0.0000 & 32 & 0.0000\\
random\_set & 0.1578 & 1.014 & 0.1578 & 0.1578 & 25 & 0.1578\\
\bottomrule
\end{tabular}
\end{table}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{kable}\NormalTok{(gseaResults}\OperatorTok{$}\NormalTok{less, }
      \DataTypeTok{booktabs =} \OtherTok{TRUE}\NormalTok{,}\DataTypeTok{caption =} \StringTok{'Down-regulation statistics'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{table}[t]

\caption{\label{tab:gseaPost2}Down-regulation statistics}
\centering
\begin{tabular}{lrrrrrr}
\toprule
  & p.geomean & stat.mean & p.val & q.val & set.size & exp1\\
\midrule
random\_set & 0.8422 & 1.014 & 0.8422 & 1 & 25 & 0.8422\\
top\_GO\_term & 1.0000 & 7.199 & 1.0000 & 1 & 32 & 1.0000\\
\bottomrule
\end{tabular}
\end{table}

We can see that the random gene set shows no significant up or down-regulation, while the gene set we defined using the top GO term shows a significant down-regulation (adjusted p-value \textless{} 0.0007). It is worthwhile to visualize these systematic changes in a heatmap:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(pheatmap)}
\CommentTok{# get the expression data for the gene set of interest}
\NormalTok{M <-}\StringTok{ }\NormalTok{normalizedCounts[}\KeywordTok{rownames}\NormalTok{(normalizedCounts) }\OperatorTok{%in%}\StringTok{ }\NormalTok{geneSet1, ]}
\CommentTok{# log transform the counts for visualization }
\CommentTok{# scaling by row helps visualizing relative change of expression of a gene in multiple conditions}
\KeywordTok{pheatmap}\NormalTok{(}\KeywordTok{log2}\NormalTok{(M}\OperatorTok{+}\DecValTok{1}\NormalTok{), }
         \DataTypeTok{annotation_col =}\NormalTok{ colData, }
         \DataTypeTok{show_rownames =} \OtherTok{FALSE}\NormalTok{, }
         \DataTypeTok{scale =} \StringTok{'row'}\NormalTok{, }
         \DataTypeTok{cutree_cols =} \DecValTok{2}\NormalTok{, }
         \DataTypeTok{cutree_rows =} \DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/gseaPost3-1} \end{center}

We can see that about two thirds of the genes from this gene set have been up-regulated in the case samples
compared to the controls.

\hypertarget{accounting-for-additional-sources-of-variation}{%
\subsection{Accounting for additional sources of variation}\label{accounting-for-additional-sources-of-variation}}

When doing a differential expression analysis in a case-control setting, the variable of interest, i.e.~the variable that explains the separation of the case samples from the control is usually the treatment, genotypic differences, a certain phenotype so on and sofort. However, in reality, depending on how the experiment and the sequencing was designed, there may be additional factors that might contribute to the variation between the compared samples. Sometimes, such variables are known, for instance, the date of the sequencing for each sample (batch information), the temperature under which samples were kept, etc. Such variables are not necessarily biological but rather technical, however, they still impact the measurements obtained from an RNA-seq experiment. Such variables can introduces systematic shifts in the obtained measurements. Here, we will demonstrate: firstly how to account for such variables using DESeq2, when the possibles sources of variation are actually known; secondly, how to account for such variables when all we have is just a count table but we observe that the variable of interest only explains a small proportion of the differences between case and control samples.

\hypertarget{accounting-for-covariates-using-deseq2}{%
\subsubsection{Accounting for covariates using DESeq2}\label{accounting-for-covariates-using-deseq2}}

For demonstration purposes, we will use a subset of the count table obtained for a heart disease study, where there are RNA-seq samples from subjects with normal and failing hearts. We again use a subset of the samples, focusing on 6 case and 6 control samples and we only consider protein-coding genes (for speed concerns).

Let's import count and colData for this experiment.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{counts_file <-}\StringTok{ }\KeywordTok{system.file}\NormalTok{(}\StringTok{'extdata/rna-seq/SRP021193.raw_counts.tsv'}\NormalTok{, }\DataTypeTok{package =} \StringTok{'compGenomRData'}\NormalTok{)}
\NormalTok{colData_file <-}\StringTok{ }\KeywordTok{system.file}\NormalTok{(}\StringTok{'extdata/rna-seq/SRP021193.colData.tsv'}\NormalTok{, }\DataTypeTok{package =} \StringTok{'compGenomRData'}\NormalTok{)}

\NormalTok{counts <-}\StringTok{ }\KeywordTok{read.table}\NormalTok{(counts_file)}
\NormalTok{colData <-}\StringTok{ }\KeywordTok{read.table}\NormalTok{(colData_file, }\DataTypeTok{header =}\NormalTok{ T, }\DataTypeTok{sep =} \StringTok{'}\CharTok{\textbackslash{}t}\StringTok{'}\NormalTok{, }\DataTypeTok{stringsAsFactors =} \OtherTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Let's take a look at how the samples cluster by calculating the TPM counts.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(pheatmap)}
\CommentTok{#find gene length normalized values }
\NormalTok{geneLengths <-}\StringTok{ }\NormalTok{counts}\OperatorTok{$}\NormalTok{width}
\NormalTok{rpk <-}\StringTok{ }\KeywordTok{apply}\NormalTok{( }\KeywordTok{subset}\NormalTok{(counts, }\DataTypeTok{select =} \KeywordTok{c}\NormalTok{(}\OperatorTok{-}\NormalTok{width)), }\DecValTok{2}\NormalTok{, }
              \ControlFlowTok{function}\NormalTok{(x) x}\OperatorTok{/}\NormalTok{(geneLengths}\OperatorTok{/}\DecValTok{1000}\NormalTok{))}
\CommentTok{#normalize by the sample size using rpk values}
\NormalTok{tpm <-}\StringTok{ }\KeywordTok{apply}\NormalTok{(rpk, }\DecValTok{2}\NormalTok{, }\ControlFlowTok{function}\NormalTok{(x) x }\OperatorTok{/}\StringTok{ }\KeywordTok{sum}\NormalTok{(}\KeywordTok{as.numeric}\NormalTok{(x)) }\OperatorTok{*}\StringTok{ }\DecValTok{10}\OperatorTok{^}\DecValTok{6}\NormalTok{)}

\NormalTok{selectedGenes <-}\StringTok{ }\KeywordTok{names}\NormalTok{(}\KeywordTok{sort}\NormalTok{(}\KeywordTok{apply}\NormalTok{(tpm, }\DecValTok{1}\NormalTok{, var), }\DataTypeTok{decreasing =}\NormalTok{ T)[}\DecValTok{1}\OperatorTok{:}\DecValTok{100}\NormalTok{])}
\KeywordTok{pheatmap}\NormalTok{(tpm[selectedGenes,], }
             \DataTypeTok{scale =} \StringTok{'row'}\NormalTok{,}
             \DataTypeTok{annotation_col =}\NormalTok{ colData, }
             \DataTypeTok{show_rownames =} \OtherTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/batch_effects_2-1} \end{center}

Here we can see from the clusters that the dominating variable is the `Library Selection' variable rather than the `diagnosis' variable that determines the state of the organ from which the sample was taken. Case and control samples are all mixed in both two major clusters. However, ideally, we'd like to see a separation of the case and control samples regardless of the additional covariates. Luckily, DESeq2 can account for such confounding variables. Below is a demonstration of how we instruct DESeq2 to account for the `library selection' variable:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(DESeq2)}
\CommentTok{# remove the 'width' column from the counts matrix}
\NormalTok{countData <-}\StringTok{ }\KeywordTok{as.matrix}\NormalTok{(}\KeywordTok{subset}\NormalTok{(counts, }\DataTypeTok{select =} \KeywordTok{c}\NormalTok{(}\OperatorTok{-}\NormalTok{width)))}
\CommentTok{# set up a DESeqDataSet object}
\NormalTok{dds <-}\StringTok{ }\KeywordTok{DESeqDataSetFromMatrix}\NormalTok{(}\DataTypeTok{countData =}\NormalTok{ countData, }
                              \DataTypeTok{colData =}\NormalTok{ colData, }
                              \DataTypeTok{design =} \OperatorTok{~}\StringTok{ }\NormalTok{LibrarySelection }\OperatorTok{+}\StringTok{ }\NormalTok{group)}
\end{Highlighting}
\end{Shaded}

When constructing the design formula, it is very important to pay attention to the sequence of variables. We leave the variable of interest to the last and we can add as many covariates as we want to the beginning of the design formula. Please refer to the DESeq2 vignette if you'd like to learn more about how to construct design formulas.

Now, we can run the differential expression analysis as has been demonstrated previously.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# run DESeq}
\NormalTok{dds <-}\StringTok{ }\KeywordTok{DESeq}\NormalTok{(dds)}
\CommentTok{# extract results}
\NormalTok{DEresults <-}\StringTok{ }\KeywordTok{results}\NormalTok{(dds, }\DataTypeTok{contrast =} \KeywordTok{c}\NormalTok{(}\StringTok{'group'}\NormalTok{, }\StringTok{'CASE'}\NormalTok{, }\StringTok{'CTRL'}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\hypertarget{accounting-for-estimated-covariates-using-ruvseq}{%
\subsubsection{Accounting for estimated covariates using RUVSeq}\label{accounting-for-estimated-covariates-using-ruvseq}}

In cases when the sources of potential variation are not known, it is worthwhile to use tools such as \texttt{RUVSeq} or \texttt{sva} that can estimate potential sources of variation and clean up the counts table from those sources of variation. Later on, the estimated covariates can be integrated into DESeq2's design formula.

Let's see how to utilize RUVseq package to first diagnose the problem and then solve it. Here, for demonstration purposes, we'll use a count table from a lung carcinoma study in which a transcription factor (Ets homologous factor - EHF) is overexpressed and compared to the control samples with baseline EHF expression. Again, we only consider protein coding genes and use only five case and five control samples. The original data can be found on the \texttt{recount2} database with the accession `SRP049988'.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{counts_file <-}\StringTok{ }\KeywordTok{system.file}\NormalTok{(}\StringTok{'extdata/rna-seq/SRP049988.raw_counts.tsv'}\NormalTok{, }\DataTypeTok{package =} \StringTok{'compGenomRData'}\NormalTok{)}
\NormalTok{colData_file <-}\StringTok{ }\KeywordTok{system.file}\NormalTok{(}\StringTok{'extdata/rna-seq/SRP049988.colData.tsv'}\NormalTok{, }\DataTypeTok{package =} \StringTok{'compGenomRData'}\NormalTok{)}

\NormalTok{counts <-}\StringTok{ }\KeywordTok{read.table}\NormalTok{(counts_file)}
\NormalTok{colData <-}\StringTok{ }\KeywordTok{read.table}\NormalTok{(colData_file, }\DataTypeTok{header =}\NormalTok{ T, }\DataTypeTok{sep =} \StringTok{'}\CharTok{\textbackslash{}t}\StringTok{'}\NormalTok{, }\DataTypeTok{stringsAsFactors =} \OtherTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Let's start by making heatmaps of the samples using TPM counts

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#find gene length normalized values }
\NormalTok{geneLengths <-}\StringTok{ }\NormalTok{counts}\OperatorTok{$}\NormalTok{width}
\NormalTok{rpk <-}\StringTok{ }\KeywordTok{apply}\NormalTok{( }\KeywordTok{subset}\NormalTok{(counts, }\DataTypeTok{select =} \KeywordTok{c}\NormalTok{(}\OperatorTok{-}\NormalTok{width)), }\DecValTok{2}\NormalTok{, }
              \ControlFlowTok{function}\NormalTok{(x) x}\OperatorTok{/}\NormalTok{(geneLengths}\OperatorTok{/}\DecValTok{1000}\NormalTok{))}
\CommentTok{#normalize by the sample size using rpk values}
\NormalTok{tpm <-}\StringTok{ }\KeywordTok{apply}\NormalTok{(rpk, }\DecValTok{2}\NormalTok{, }\ControlFlowTok{function}\NormalTok{(x) x }\OperatorTok{/}\StringTok{ }\KeywordTok{sum}\NormalTok{(}\KeywordTok{as.numeric}\NormalTok{(x)) }\OperatorTok{*}\StringTok{ }\DecValTok{10}\OperatorTok{^}\DecValTok{6}\NormalTok{)}
\NormalTok{selectedGenes <-}\StringTok{ }\KeywordTok{names}\NormalTok{(}\KeywordTok{sort}\NormalTok{(}\KeywordTok{apply}\NormalTok{(tpm, }\DecValTok{1}\NormalTok{, var), }\DataTypeTok{decreasing =}\NormalTok{ T)[}\DecValTok{1}\OperatorTok{:}\DecValTok{100}\NormalTok{])}
\KeywordTok{pheatmap}\NormalTok{(tpm[selectedGenes,], }
             \DataTypeTok{scale =} \StringTok{'row'}\NormalTok{,}
             \DataTypeTok{annotation_col =}\NormalTok{ colData, }
             \DataTypeTok{cutree_cols =} \DecValTok{2}\NormalTok{, }
             \DataTypeTok{show_rownames =} \OtherTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/ruv_diagnose_1-1} \end{center}

We can see that the overal clusters look fine except for one of the case samples (CASE\_5) clusters more closely with the control samples than the other case samples. This mis-clustering could be a result of some batch effect, or any other technical preparation steps. However, the \texttt{colData} object doesn't contain any variables that we can directly distinguish this sample from others. So, let's use RUVSeq to estimate potential covariates to see if the clustering results can be improved.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(EDASeq)}
\CommentTok{# remove 'width' column from counts}
\NormalTok{countData <-}\StringTok{ }\KeywordTok{as.matrix}\NormalTok{(}\KeywordTok{subset}\NormalTok{(counts, }\DataTypeTok{select =} \KeywordTok{c}\NormalTok{(}\OperatorTok{-}\NormalTok{width)))}
\CommentTok{# create a seqExpressionSet object using EDASeq package }
\NormalTok{set <-}\StringTok{ }\KeywordTok{newSeqExpressionSet}\NormalTok{(}\DataTypeTok{counts =}\NormalTok{ countData,}
                           \DataTypeTok{phenoData =}\NormalTok{ colData)}

\CommentTok{# make an RLE plot and a PCA plot on raw count data and color samples by group }
\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow =} \KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\KeywordTok{plotRLE}\NormalTok{(set, }\DataTypeTok{outline=}\OtherTok{FALSE}\NormalTok{, }\DataTypeTok{ylim=}\KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{4}\NormalTok{, }\DecValTok{4}\NormalTok{), }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group))}
\KeywordTok{plotPCA}\NormalTok{(set, }\DataTypeTok{col =} \KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group))}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/ruv_diagnose_2-1} \end{center}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{## make RLE and PCA plots on TPM matrix }
\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow =} \KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\KeywordTok{plotRLE}\NormalTok{(tpm, }\DataTypeTok{outline=}\OtherTok{FALSE}\NormalTok{, }\DataTypeTok{ylim=}\KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{4}\NormalTok{, }\DecValTok{4}\NormalTok{), }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group))}
\KeywordTok{plotPCA}\NormalTok{(tpm, }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group), }\DataTypeTok{cex=}\FloatTok{1.2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/ruv_diagnose_2-2} \end{center}

Both RLE and PCA plots look better on normalized data compared to raw data, but still suggest the necessity of further improvement, because the the CASE\_5 sample still clusters with the control samples. We haven't yet accounted for the source of unwanted variation.

\hypertarget{removing-unwanted-variation-from-the-data}{%
\subsubsection{Removing unwanted variation from the data}\label{removing-unwanted-variation-from-the-data}}

RUVSeq has three main functions for removing unwanted variation: \texttt{RUVg}, \texttt{RUVs}, and \texttt{RUVr}. Here, we will demonstrate how to use \texttt{RUVg} and \texttt{RUVs}. \texttt{RUVr} will be left as an exercise for the reader.

\hypertarget{using-ruvg}{%
\paragraph{Using RUVg}\label{using-ruvg}}

One way of removing unwanted variation is dependent on using a set of reference genes that are not expected to change by the sources of technical variation. One strategy along this line is to use spike-in genes, which are artifically introduced into the sequencing run \citep{jiang_synthetic_2011}. However, there are many sequencing datasets that don't have this spike-in data available. In such cases, an emprical set of genes can be collected from the expression data by doing a differential expression analysis and discovering genes that are unchanged in the given conditions. These unchanged genes are used to center the complete expression data to clean up the data from systematic shifts in expression due to the unwanted sources of variation. Another stragety could be to use a set of house-keeping genes as negative controls, and use them as a reference to correct the systematic biases in the data. Let's use a list of \textasciitilde{}500 house-keeping genes compiled here: \url{http://www.stat.berkeley.edu/~johann/ruv/resources/hk.txt}.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(RUVSeq)}
\CommentTok{#source for house-keeping genes collection: https://m.tau.ac.il/~elieis/HKG/HK_genes.txt}
\NormalTok{house_keeping_genes <-}\StringTok{ }\KeywordTok{read.table}\NormalTok{(}\DataTypeTok{file =} \KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata/rna-seq/HK_genes.txt"}\NormalTok{, }
                                                     \DataTypeTok{package =} \StringTok{'compGenomRData'}\NormalTok{), }
                                  \DataTypeTok{header =} \OtherTok{FALSE}\NormalTok{)}
\CommentTok{# let's take an intersection of the house-keeping genes with the genes available in the count table }
\NormalTok{house_keeping_genes <-}\StringTok{ }\KeywordTok{intersect}\NormalTok{(}\KeywordTok{rownames}\NormalTok{(set), house_keeping_genes}\OperatorTok{$}\NormalTok{V1)}

\CommentTok{# now, we use these genes as the emprical set of genes as input to RUVg.}
\CommentTok{# we try different values of k and see how the PCA plots look }
\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow =} \KeywordTok{c}\NormalTok{(}\DecValTok{3}\NormalTok{, }\DecValTok{2}\NormalTok{))}
\ControlFlowTok{for}\NormalTok{(k }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{6}\NormalTok{) \{}
\NormalTok{  set_g <-}\StringTok{ }\KeywordTok{RUVg}\NormalTok{(}\DataTypeTok{x =}\NormalTok{ set, }\DataTypeTok{cIdx =}\NormalTok{ house_keeping_genes, }\DataTypeTok{k =}\NormalTok{ k)}
  \KeywordTok{plotPCA}\NormalTok{(set_g, }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group), }\DataTypeTok{cex=}\FloatTok{1.2}\NormalTok{, }\DataTypeTok{main =} \KeywordTok{paste0}\NormalTok{(}\StringTok{'with RUVg, k = '}\NormalTok{,k))}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/ruv_g-1} \end{center}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# choose k = 1}
\NormalTok{set_g <-}\StringTok{ }\KeywordTok{RUVg}\NormalTok{(}\DataTypeTok{x =}\NormalTok{ set, }\DataTypeTok{cIdx =}\NormalTok{ house_keeping_genes, }\DataTypeTok{k =} \DecValTok{1}\NormalTok{)}

\CommentTok{# now let's do diagnostics: compare the count matrices with or without RUVg processing}
\CommentTok{# RLE plots}
\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow =} \KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\KeywordTok{plotRLE}\NormalTok{(set, }\DataTypeTok{outline=}\OtherTok{FALSE}\NormalTok{, }\DataTypeTok{ylim=}\KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{4}\NormalTok{, }\DecValTok{4}\NormalTok{), }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group), }\DataTypeTok{main =} \StringTok{'without RUVg'}\NormalTok{)}
\KeywordTok{plotRLE}\NormalTok{(set_g, }\DataTypeTok{outline=}\OtherTok{FALSE}\NormalTok{, }\DataTypeTok{ylim=}\KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{4}\NormalTok{, }\DecValTok{4}\NormalTok{), }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group), }\DataTypeTok{main =} \StringTok{'with RUVg'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/ruv_g-2} \end{center}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# PCA plots}
\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow =} \KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\KeywordTok{plotPCA}\NormalTok{(set, }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group), }\DataTypeTok{cex=}\FloatTok{1.2}\NormalTok{,  }\DataTypeTok{main =} \StringTok{'without RUVg'}\NormalTok{)}
\KeywordTok{plotPCA}\NormalTok{(set_g, }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group), }\DataTypeTok{cex=}\FloatTok{1.2}\NormalTok{, }\DataTypeTok{main =} \StringTok{'with RUVg'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/ruv_g-3} \end{center}

We can observe that using RUVg using house-keeping genes as reference has improved the clusters, however not ideal. There is another strategy of RUVSeq that works better in the presence of replicates without a confounded design, which is the \texttt{RUVs} function. Let's see how that performs with this data. This time we don't use the house-keeping genes. We rather use all genes as input to RUVs.

\hypertarget{using-ruvs}{%
\paragraph{Using RUVs}\label{using-ruvs}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# make a table of sample groups from colData }
\NormalTok{differences <-}\StringTok{ }\KeywordTok{makeGroups}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group)}
\CommentTok{## looking for two different sources of unwanted variation (k = 2)}
\CommentTok{## use information from all genes in the expression object}
\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow =} \KeywordTok{c}\NormalTok{(}\DecValTok{3}\NormalTok{, }\DecValTok{2}\NormalTok{))}
\ControlFlowTok{for}\NormalTok{(k }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{6}\NormalTok{) \{}
\NormalTok{  set_s <-}\StringTok{ }\KeywordTok{RUVs}\NormalTok{(set, }\KeywordTok{unique}\NormalTok{(}\KeywordTok{rownames}\NormalTok{(set)), }\DataTypeTok{k=}\NormalTok{k, differences) }\CommentTok{#all genes}
  \KeywordTok{plotPCA}\NormalTok{(set_s, }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group), }\DataTypeTok{cex=}\FloatTok{1.2}\NormalTok{, }\DataTypeTok{main =} \KeywordTok{paste0}\NormalTok{(}\StringTok{'with RUVs, k = '}\NormalTok{,k))}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/ruv_s-1} \end{center}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# choose k = 3}
\NormalTok{set_s <-}\StringTok{ }\KeywordTok{RUVs}\NormalTok{(set, }\KeywordTok{unique}\NormalTok{(}\KeywordTok{rownames}\NormalTok{(set)), }\DataTypeTok{k=}\DecValTok{3}\NormalTok{, differences) }\CommentTok{#}

\CommentTok{# repeat the RLE and PCA plots }
\CommentTok{## compare the initial and processed objects}
\CommentTok{## RLE plots}
\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow =} \KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\KeywordTok{plotRLE}\NormalTok{(set, }\DataTypeTok{outline=}\OtherTok{FALSE}\NormalTok{, }\DataTypeTok{ylim=}\KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{4}\NormalTok{, }\DecValTok{4}\NormalTok{), }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group), }\DataTypeTok{main =} \StringTok{'without RUVs'}\NormalTok{)}
\KeywordTok{plotRLE}\NormalTok{(set_s, }\DataTypeTok{outline=}\OtherTok{FALSE}\NormalTok{, }\DataTypeTok{ylim=}\KeywordTok{c}\NormalTok{(}\OperatorTok{-}\DecValTok{4}\NormalTok{, }\DecValTok{4}\NormalTok{), }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group), }\DataTypeTok{main =} \StringTok{'with RUVs'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/ruv_s-2} \end{center}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{## PCA plots}
\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow =} \KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{))}
\KeywordTok{plotPCA}\NormalTok{(set, }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group), }\DataTypeTok{cex=}\FloatTok{1.2}\NormalTok{,  }\DataTypeTok{main =} \StringTok{'without RUVs'}\NormalTok{)}
\KeywordTok{plotPCA}\NormalTok{(set_s, }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group), }\DataTypeTok{cex=}\FloatTok{1.2}\NormalTok{, }\DataTypeTok{main =} \StringTok{'with RUVs'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/ruv_s-3} \end{center}

Let's compare PCA results from RUVs and RUVg with the initial raw counts matrix:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{par}\NormalTok{(}\DataTypeTok{mfrow =} \KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{3}\NormalTok{))}
\KeywordTok{plotPCA}\NormalTok{(countData, }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group), }\DataTypeTok{main =} \StringTok{'without RUV - raw counts'}\NormalTok{)}
\KeywordTok{plotPCA}\NormalTok{(set_g, }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group), }\DataTypeTok{main =} \StringTok{'with RUVg'}\NormalTok{)}
\KeywordTok{plotPCA}\NormalTok{(set_s, }\DataTypeTok{col=}\KeywordTok{as.numeric}\NormalTok{(colData}\OperatorTok{$}\NormalTok{group), }\DataTypeTok{main =} \StringTok{'with RUVs'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/ruv_compare-1} \end{center}

It looks like RUVs has performed better than RUVg in this case. So, let's use count data that is processed by RUVs to re-do the initial heatmap.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(EDASeq)}
\KeywordTok{library}\NormalTok{(pheatmap)}
\CommentTok{# extract normalized counts that are cleared from unwanted variation using RUVs}
\NormalTok{normCountData <-}\StringTok{ }\KeywordTok{normCounts}\NormalTok{(set_s)}
\NormalTok{selectedGenes <-}\StringTok{ }\KeywordTok{names}\NormalTok{(}\KeywordTok{sort}\NormalTok{(}\KeywordTok{apply}\NormalTok{(normCountData, }\DecValTok{1}\NormalTok{, var), }\DataTypeTok{decreasing =} \OtherTok{TRUE}\NormalTok{))[}\DecValTok{1}\OperatorTok{:}\DecValTok{500}\NormalTok{]}
\KeywordTok{pheatmap}\NormalTok{(normCountData[selectedGenes,], }
         \DataTypeTok{annotation_col =}\NormalTok{ colData, }
         \DataTypeTok{show_rownames =} \OtherTok{FALSE}\NormalTok{, }
         \DataTypeTok{cutree_cols =} \DecValTok{2}\NormalTok{,}
         \DataTypeTok{scale =} \StringTok{'row'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/ruv_post-1} \end{center}

As can be observed the replicates from different groups cluster much better with each other after processing with \texttt{RUVs}. It is important to note that RUVs uses information from replicates to shift the expression data and it would not work in a confounding design where the replicates of case samples and replicates of the control samples are sequenced in different batches.

\hypertarget{re-run-deseq2-with-the-computed-covariates}{%
\subsubsection{Re-run DESeq2 with the computed covariates}\label{re-run-deseq2-with-the-computed-covariates}}

Having computed the sources of variation using RUVs, we can actually integrate these variables with DESeq2 to re-do the differential expression analysis.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(DESeq2)}
\CommentTok{#set up DESeqDataSet object}
\NormalTok{dds <-}\StringTok{ }\KeywordTok{DESeqDataSetFromMatrix}\NormalTok{(}\DataTypeTok{countData =}\NormalTok{ countData,}
                              \DataTypeTok{colData =}\NormalTok{ colData, }
                              \DataTypeTok{design =} \OperatorTok{~}\StringTok{ }\NormalTok{group)}
\CommentTok{# filter for low count genes}
\NormalTok{dds <-}\StringTok{ }\NormalTok{dds[}\KeywordTok{rowSums}\NormalTok{(}\KeywordTok{counts}\NormalTok{(dds)) }\OperatorTok{>}\StringTok{ }\DecValTok{10}\NormalTok{]}

\CommentTok{# insert the covariates computed using RUVs into DESeqDataSet object}
\KeywordTok{colData}\NormalTok{(dds) <-}\StringTok{ }\KeywordTok{cbind}\NormalTok{(}\KeywordTok{colData}\NormalTok{(dds), }\KeywordTok{pData}\NormalTok{(set_s)[}\KeywordTok{rownames}\NormalTok{(}\KeywordTok{colData}\NormalTok{(dds)), }\KeywordTok{grep}\NormalTok{(}\StringTok{'W_[0-9]'}\NormalTok{, }\KeywordTok{colnames}\NormalTok{(}\KeywordTok{pData}\NormalTok{(set_s)))])}

\CommentTok{# update the design formula for the DESeq analysis (save the variable of interest to the last!)}
\KeywordTok{design}\NormalTok{(dds) <-}\StringTok{ }\ErrorTok{~}\StringTok{ }\NormalTok{W_}\DecValTok{1} \OperatorTok{+}\StringTok{ }\NormalTok{W_}\DecValTok{2} \OperatorTok{+}\StringTok{ }\NormalTok{W_}\DecValTok{3} \OperatorTok{+}\StringTok{ }\NormalTok{group }
\CommentTok{# repeat the analysis }
\NormalTok{dds <-}\StringTok{ }\KeywordTok{DESeq}\NormalTok{(dds)}
\CommentTok{# extract deseq results }
\NormalTok{res <-}\StringTok{ }\KeywordTok{results}\NormalTok{(dds, }\DataTypeTok{contrast =} \KeywordTok{c}\NormalTok{(}\StringTok{'group'}\NormalTok{, }\StringTok{'CASE'}\NormalTok{, }\StringTok{'CTRL'}\NormalTok{))}
\NormalTok{res <-}\StringTok{ }\NormalTok{res[}\KeywordTok{order}\NormalTok{(res}\OperatorTok{$}\NormalTok{padj),]}
\end{Highlighting}
\end{Shaded}

\hypertarget{other-applications-of-rna-seq}{%
\section{Other applications of RNA-seq}\label{other-applications-of-rna-seq}}

RNA-seq generates valuable data that contains information not only at the gene level but also at the level of exons and transcripts. Moreover, the kind of information that we can extract from RNA-seq is not limited to expression quantification. It is possible to detect alternative splicing events such as novel isoforms \citep{trapnell_transcript_2010}, differential usage of exons \citep{anders_detecting_2012} . It is also possible to observe sequence variants (substitutions, insertions, deletions, RNA-editing) that may change the translated protein product \citep{mckenna_genome_2010}. In the context of cancer genomes, gene-fusion events can be detected with RNA-seq \citep{mcpherson_defuse:_2011}. Finally, for the purposes of gene prediction or improving existing gene predictions, RNA-seq is a valuable method \citep{stanke_augustus:_2005}. In order to learn more about how to implement these, it is recommended to go through the tutorials of the cited tools.

\hypertarget{exercises-6}{%
\section{Exercises}\label{exercises-6}}

\hypertarget{exploring-the-count-tables}{%
\subsection{Exploring the count tables}\label{exploring-the-count-tables}}

Here, import an example count table and do some exploration of the expression data.

\begin{verbatim}
counts_file <- system.file("extdata/rna-seq/SRP029880.raw_counts.tsv", package = "compGenomRData")
coldata_file <- system.file("extdata/rna-seq/SRP029880.colData.tsv", package = "compGenomRData")
\end{verbatim}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Basic:
\end{enumerate}

\begin{itemize}
\tightlist
\item
  Normalize the counts using the TPM approach.
\item
  Plot a heatmap of the top 500 most variable genes. Compare with the heatmap obtained using 100 most variable genes.
\item
  Re-do the heatmaps setting \texttt{scale} argument to \texttt{none}, and \texttt{column}. Compare the results with \texttt{scale\ =\ \textquotesingle{}row\textquotesingle{}}.
\item
  Draw a correlation plot for the samples depicting the sample differences as `ellipses', drawing only the upper end of the matrix, order samples by hierarchical clustering results based on \texttt{average} linkage clustering method.
\end{itemize}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{1}
\tightlist
\item
  Intermediate:
\end{enumerate}

\begin{itemize}
\tightlist
\item
  How else could the count matrix be subsetted to obtain quick and accurate clusters? Try selecting top 100 genes that have the highest total expression in all samples and re-draw the cluster heatmaps and PCA plots.
\item
  Add an additional column to the annotation data.frame object to annotate the samples and use the updated annotation data.frame to plot the heatmaps. (Hint: assign different batch values to CASE and CTRL samples). Make a PCA plot and color samples by the added variable (e.g.~batch).
\end{itemize}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{2}
\tightlist
\item
  Advanced:
\end{enumerate}

\begin{itemize}
\tightlist
\item
  Try making the heatmaps using all the genes in the count table, rather than sub-selecting.
\item
  Use \href{https://cran.r-project.org/web/packages/tsne/tsne.pdf}{tsne package} to draw a t-SNE plot of the expression values. Color the points by sample group. Compare the results with the PCA plots.
\end{itemize}

\hypertarget{differential-expression-analysis-1}{%
\subsection{Differential expression analysis}\label{differential-expression-analysis-1}}

Firstly, carry out a differential expression analysis starting from raw counts

Use the below datasets:

\begin{verbatim}
counts_file <- system.file("extdata/rna-seq/SRP029880.raw_counts.tsv", package = "compGenomRData")
coldata_file <- system.file("extdata/rna-seq/SRP029880.colData.tsv", package = "compGenomRData")
\end{verbatim}

\begin{itemize}
\tightlist
\item
  Import the read counts and colData tables
\item
  Set up a DESeqDataSet object
\item
  Filter out genes with low counts
\item
  Run DESeq2 contrasting the \texttt{CASE} sample with \texttt{CONTROL} samples.
\end{itemize}

Now, you are ready to do the following exercises:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Basic
\end{enumerate}

\begin{itemize}
\tightlist
\item
  Make a volcano plot using the differential expression analysis results. (Hint: x axis denotes the log2FoldChange and the y-axis represents the -log10(pvalue)).\\
\item
  Use DESeq2::plotDispEsts to make a dispersion plot and find out the meaning of this plot. (Hint: type ?DESeq2::plotDispEsts)
\end{itemize}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{1}
\tightlist
\item
  Intermediate
\end{enumerate}

\begin{itemize}
\item
  Find out about \texttt{lfcThreshold} argument of \texttt{DESeq2::results} function. What is its default value? What does it mean to change the default value to, for instance, \texttt{1}?
\item
  What is independent filtering? What happens if we don't use it? Google \texttt{independent\ filtering\ statquest} and watch the online video about independent filtering.
\end{itemize}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{2}
\tightlist
\item
  Advanced
\end{enumerate}

\begin{itemize}
\item
  Re-do the differential expression analysis using edgeR package. Find out how much DESeq2 and edgeR agree on the list of differentially expressed genes.
\item
  Use \texttt{compcodeR} package to run the differential expression analysis using at least three different tools and compare and contrast the results following the compcodeR vignette.
\end{itemize}

\hypertarget{functional-enrichment-analysis-1}{%
\subsection{Functional enrichment analysis}\label{functional-enrichment-analysis-1}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Basic:
\end{enumerate}

\begin{itemize}
\item
  Re-run gProfileR, this time using pathway annotations such as KEGG, REACTOME, and protein complex databases such as CORUM, in addition to the GO terms. Sort the resulting tables by columns \texttt{precision} and/or \texttt{recall}. How do the top GO terms change when sorted for \texttt{precision}, \texttt{recall}, or \texttt{p.value}?
\item
  Repeat the gene set enrichment analysis by trying different options for the \texttt{compare} argument of \texttt{GAGE:gage}
  function. How do the results differ?
\end{itemize}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{1}
\tightlist
\item
  Intermediate:
\end{enumerate}

\begin{itemize}
\item
  Make a scatter plot of GO term sizes and obtained p-values by setting the \texttt{gProfiler::gprofiler} argument \texttt{significant\ =\ FALSE}. Is there a correlation of term sizes and p-values? (Hint: take -log10 of p-values). If so, how can this bias be mitigated?
\item
  Do a gene-set enrichment analysis using gene sets from top 10 GO terms.
\item
  What are the other available R packages that can carry out gene set enrichment analysis for RNA-seq datasets?
\end{itemize}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{2}
\tightlist
\item
  Advanced:
\end{enumerate}

\begin{itemize}
\tightlist
\item
  Use the topGO package (\url{https://bioconductor.org/packages/release/bioc/html/topGO.html}) to re-do the GO term analysis. Compare and contrast the results with what has been obtained using gProfileR package. Which tool is faster? gProfileR or topGO? Why?
\item
  Given a gene set annotated for human, how can it be utilized to work on C. elegans data? (Hint: see \texttt{biomaRt::getLDS} {[}REF{]} function or \texttt{RCAS::retrieveOrthologs} {[}REF{]} function).
\item
  Import curated pathway gene sets with Entrez identifiers from the \href{http://software.broadinstitute.org/gsea/msigdb/collections.jsp}{MSIGDB database} and re-do the GSEA for all curated gene sets.
\end{itemize}

\hypertarget{removing-unwanted-variation-from-the-expression-data}{%
\subsection{Removing unwanted variation from the expression data}\label{removing-unwanted-variation-from-the-expression-data}}

For the exercises below, use the datasets at:

\begin{verbatim}
counts_file <- system.file('extdata/rna-seq/SRP049988.raw_counts.tsv', package = 'compGenomRData')
colData_file <- system.file('extdata/rna-seq/SRP049988.colData.tsv', package = 'compGenomRData')
\end{verbatim}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Basic:
\end{enumerate}

\begin{itemize}
\tightlist
\item
  Run RUVSeq using multiple values of \texttt{k} from 1 to 10 and compare and contrast the PCA plots obtained from the normalized counts of each RUVSeq run.
\end{itemize}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{1}
\tightlist
\item
  Intermediate:
\end{enumerate}

\begin{itemize}
\item
  Re-run RUVSeq using the \texttt{RUVr} function. Compare PCA plots from \texttt{RUVs}, \texttt{RUVg} and \texttt{RUVr} using the same \texttt{k} values and find out which one performs the best.
\item
  Do the necessary diagnostic plots using the differential expression results from the EHF count table.
\end{itemize}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{2}
\tightlist
\item
  Advanced:
\end{enumerate}

\begin{itemize}
\tightlist
\item
  Use the \texttt{sva\ package} to discover sources of unwanted variation and re-do the differential expression analysis using variables from the output of \texttt{sva} and compare the results with DESeq2 results using RUVSeq corrected normalization counts.
\end{itemize}

\hypertarget{chipseq}{%
\chapter{ChIP-seq Analysis}\label{chipseq}}

in progress\ldots{}

\hypertarget{bsseq}{%
\chapter{DNA methylation analysis using bisulfite sequencing data}\label{bsseq}}

\hypertarget{what-is-dna-methylation}{%
\subsection{what is DNA methylation ?}\label{what-is-dna-methylation}}

Cytosine methylation (5-methylcytosine, 5mC) is one of the main covalent base modifications in eukaryotic genomes, generally observed on CpG dinucleotides. Methylation can also rarely occur in non-CpG context, but this was mainly observed in human embryonic stem and neuronal cells \citep[\citet{Lister2013-vs}]{Lister2009-sd}. DNA methylation is a part of the epigenetic regulation mechanism of gene expression. It is cell-type specific DNA modification. It is reversible but mostly remains stable through cell division. There are roughly 28 million CpGs in the human genome, 60--80\% are generally methylated. Less than 10\% of CpGs occur in CG-dense regions that are termed CpG islands in the human genome \citep{Smith2013-jh}. It has been demonstrated that DNA methylation is also not uniformly distributed over the genome, but rather is associated with CpG density. In vertebrate genomes, cytosine bases are usually unmethylated in CpG-rich regions such as CpG islands and tend to be methylated in CpG-deficient regions. Vertebrate genomes are largely CpG deficient except at CpG islands. Conversely, invertebrates such as Drosophila melanogaster and Caenorhabditis elegans do not exhibit cytosine methylation and consequently do not have CpG rich and poor regions but rather a steady CpG frequency over their genomes \citep{Deaton2011-pm}.

\hypertarget{how-dna-methylation-is-set}{%
\subsection{How DNA methylation is set ?}\label{how-dna-methylation-is-set}}

DNA methylation is established by DNA methyltransferases DNMT3A and DNMT3B in combination with DNMT3L and maintained through cell division by the methyltransferase DNMT1 and associated proteins. DNMT3a and DNMT3b are in charge of the de novo methylation during early development. Loss of 5mC can be achieved passively by dilution during replication or exclusion of DNMT1 from the nucleus. Recent discoveries of ten-eleven translocation (TET) family of proteins and their ability to convert 5-methylcytosine (5mC) into 5-hydroxymethylcytosine (5hmC) in vertebrates provide a path for catalysed active DNA demethylation \citep{Tahiliani2009-ar}. Iterative oxidations of 5hmC catalysed by TET result in 5-formylcytosine (5fC) and 5-carboxylcytosine (5caC). 5caC mark is excised from DNA by G/T mismatch-specific thymine-DNA glycosylase (TDG), which as a result returns cytosine residue back to its unmodified state \citep{He2011-pw}. Apart from these, mainly bacteria but possibly higher eukaryotes contain base modifications on bases other than cytosine, such as methylated adenine or guanine \citep{Clark2011-sc}.

\hypertarget{how-to-measure-dna-methylation-with-bisulfite-sequencing}{%
\subsection{How to measure DNA methylation with bisulfite-sequencing}\label{how-to-measure-dna-methylation-with-bisulfite-sequencing}}

One of the most reliable and popular ways to measure DNA methylation is high-throughput bisulfite sequencing. This method, and the related ones, allow measurement of DNA methylation at the single nucleotide resolution. The bisulfite conversion turns unmethylated Cs to Ts and methylated Cs remain intact. Then, the only thing to do is to align the reads with those C-\textgreater{}T conversions and count C-\textgreater{}T mutations to calculate fraction of methylated bases. In the end, we can get quantitative genome-wide measurements for DNA methylation.

\hypertarget{analyzing-dna-methylation-data}{%
\subsection{Analyzing DNA methylation data}\label{analyzing-dna-methylation-data}}

For the remainder of this chapter, we will explain how to do DNA methylation analysis using R. The analysis process is somewhat similar to the analysis patterns observed in other sequencing data analyses. The process can be chunked to four main parts with further sub-chunks:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Processing raw data
\end{enumerate}

\begin{itemize}
\tightlist
\item
  Quality check
\item
  Alignment and post-alignment processing
\item
  Methylation calling
\item
  Filtering bases
\end{itemize}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{1}
\tightlist
\item
  Exploratory analysis
\end{enumerate}

\begin{itemize}
\tightlist
\item
  Clustering
\item
  PCA
\end{itemize}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{2}
\tightlist
\item
  Finding interesting regions
\end{enumerate}

\begin{itemize}
\tightlist
\item
  Differential methylation
\item
  Methylation segmentation
\end{itemize}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{3}
\tightlist
\item
  Annotating interesting regions
\end{enumerate}

\begin{itemize}
\tightlist
\item
  Nearest genes
\item
  Annotation with other genomic features
\item
  Integration with other quantitative genomics data
\end{itemize}

\hypertarget{processing-raw-data-and-getting-data-into-r}{%
\section{Processing raw data and getting data into R}\label{processing-raw-data-and-getting-data-into-r}}

The rawest form of data that most users get is probably in the form of fastq files obtained from the sequencing experiments. We will describe necessary steps and the tools that can be used for raw data processing and if exists we will mention their R equivalents. However, the data processing is usually done outside of the R framework and for the following sections we will assume that the data processing is done and our analysis is starting from methylation call files.

Typical data processing step starts with data quality check. The fastq files are first ran through a quality check software that shows the quality of the sequencing run. We would typically use \href{https://www.bioinformatics.babraham.ac.uk/projects/fastqc/}{fastQC} for this. However, there are several bioconductor packages that could be of use, such as \href{https://bioconductor.org/packages/release/bioc/html/Rqc.html}{Rqc} and \href{https://bioconductor.org/packages/release/bioc/html/QuasR.html}{QuasR}. Following the quality check, provided everything is OK, the reads can be aligned to the genome. Before the alignment adapters or low quality ends of reads can be trimmed to increase number of alignments. Low quality ends mostly likely have poor basecalls which will lead to many mismatches. Reads with non-trimmed adapters will also not align to the genome. We would use adapter trimming tools such as \href{}{cutadapt} or \href{}{flexbar} for this purpose, although there are a bunch of them to be chosen from. Following this, reads are aligned to the genome with a bisulfite treatment aware aligner. For our own purposes, we use \href{}{Bismark}, however there are other equally accurate aligners, some are reviewed \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3378906/}{here}. In addition, Bioconductor package \href{https://bioconductor.org/packages/release/bioc/html/QuasR.html}{QuasR} can align BS-seq reads within the R framework.

After alignment, we need to call C-\textgreater{}T conversions and calculate fraction/percentage of methylation. Most of the time aligners come with auxiliary tools that calculate per base methylation values. Normally, they output a tabular format containing the location of the Cs and methylation value and strand. Within R, \href{https://bioconductor.org/packages/release/bioc/html/QuasR.html}{QuasR} and \href{https://bioconductor.org/packages/release/bioc/html/methylKit.html}{methylKit} can call methylation values from BAM files albeit with some limitations. In essence, these methylation call files can be easily read into R and downstream analysis within R starts from that point. An important quality measure at this stage is to look at conversion rate. This simply means how many unmethylated Cs are converted to Ts. Since we expect, non-CpG methylation to be rare. We can simply count number of C-\textgreater{}T conversions in non-CpG context and calculate conversion rate. The best way to do this would be via spike-in sequences where we expect no methylation at all. Since, non-CpG methylation is tissue specific, calculating conversion rate using non-CpG Cs might be misleading in some cases.

\hypertarget{data-filtering-and-exploratory-analysis}{%
\section{Data filtering and exploratory analysis}\label{data-filtering-and-exploratory-analysis}}

We assume that we start the analysis in R with the methylation call files. We will read those files in and carry out exploratory analysis, we will show how to filter bases or regions from the data and in what circumstances we might need to do so. We will use \href{https://bioconductor.org/packages/release/bioc/html/methylKit.html}{methylKit}\citep{Akalin2012-af} package for the bulk of the analysis.

\hypertarget{reading-methylation-call-files}{%
\subsection{Reading methylation call files}\label{reading-methylation-call-files}}

A typical methylation call file looks like this:

\begin{verbatim}
##         chrBase   chr    base strand coverage freqC
## 1 chr21.9764539 chr21 9764539      R       12 25.00
## 2 chr21.9764513 chr21 9764513      R       12  0.00
## 3 chr21.9820622 chr21 9820622      F       13  0.00
## 4 chr21.9837545 chr21 9837545      F       11  0.00
## 5 chr21.9849022 chr21 9849022      F      124 72.58
##    freqT
## 1  75.00
## 2 100.00
## 3 100.00
## 4 100.00
## 5  27.42
\end{verbatim}

Most of the time bisulfite sequencing experiments have test and control samples. The test samples can be from a disease tissue while the control samples can be from a healthy tissue. You can read a set of methylation call files that have test/control conditions giving \texttt{treatment} vector option. The treatment vector defines the sample groups and it is very important for the differential methylation analysis. For sake of subsequent analysis, file.list, sample.id and treatment option should have the same order. In the following example, first two files have the sample ids ``test1'' and ``test2'' and as determined by treatment vector they belong to the same group. The third and fourth files have sample ids ``ctrl1'' and ``ctrl2'' and they belong to the same group as indicated by the treatment vector. We will first get a list of file paths and have a look at the content.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(methylKit)}
\NormalTok{file.list=}\KeywordTok{list}\NormalTok{( }\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{, }
                            \StringTok{"test1.myCpG.txt"}\NormalTok{, }\DataTypeTok{package =} \StringTok{"methylKit"}\NormalTok{),}
                \KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}
                            \StringTok{"test2.myCpG.txt"}\NormalTok{, }\DataTypeTok{package =} \StringTok{"methylKit"}\NormalTok{),}
                \KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{, }
                            \StringTok{"control1.myCpG.txt"}\NormalTok{, }\DataTypeTok{package =} \StringTok{"methylKit"}\NormalTok{),}
                \KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{, }
                            \StringTok{"control2.myCpG.txt"}\NormalTok{, }\DataTypeTok{package =} \StringTok{"methylKit"}\NormalTok{) )}

\NormalTok{file.list}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [[1]]
## [1] "/Users/aakalin/Rlibs/methylKit/extdata/test1.myCpG.txt"
## 
## [[2]]
## [1] "/Users/aakalin/Rlibs/methylKit/extdata/test2.myCpG.txt"
## 
## [[3]]
## [1] "/Users/aakalin/Rlibs/methylKit/extdata/control1.myCpG.txt"
## 
## [[4]]
## [1] "/Users/aakalin/Rlibs/methylKit/extdata/control2.myCpG.txt"
\end{verbatim}

As you can see \texttt{file.list} variable is a simple list of file paths. Each file contains methylation calls for a given sample. Now, we can read the files with \texttt{methRead()} function.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# read the files to a methylRawList object: myobj}
\NormalTok{myobj=}\KeywordTok{methRead}\NormalTok{(file.list,}
           \DataTypeTok{sample.id=}\KeywordTok{list}\NormalTok{(}\StringTok{"test1"}\NormalTok{,}\StringTok{"test2"}\NormalTok{,}\StringTok{"ctrl1"}\NormalTok{,}\StringTok{"ctrl2"}\NormalTok{),}
           \DataTypeTok{assembly=}\StringTok{"hg18"}\NormalTok{,}
           \DataTypeTok{treatment=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{),}
           \DataTypeTok{context=}\StringTok{"CpG"}
\NormalTok{           )}
\end{Highlighting}
\end{Shaded}

tab-separated bedgraph like formats from Bismark methylation caller can also be read in by methylkit. In those cases, we have to provide either \texttt{pipeline="bismarkCoverage"} or \texttt{pipeline="bismarkCytosineReport"} to \texttt{methRead} function. In addition to the options we mentioned above,
any tab separated text file with a generic format can be read in using methylKit,
such as methylation ratio files from \href{http://code.google.com/p/bsmap/}{BSMAP}.
See \href{http://zvfak.blogspot.com/2012/10/how-to-read-bsmap-methylation-ratio.html}{here} for an example.

Before we move on, let us have a look at what kind of information is stored in \texttt{myobj}. This is technically a \texttt{methylRawList} object, which is essentially a list of \texttt{methylRaw} objects. These objects hold
the information for location of Cs, and methylated Cs and unmethylated Cs.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{## inside the methylRawList object}
\KeywordTok{length}\NormalTok{(myobj)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 4
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{head}\NormalTok{(myobj[[}\DecValTok{1}\NormalTok{]])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##     chr   start     end strand coverage numCs numTs
## 1 chr21 9764539 9764539      -       12     3     9
## 2 chr21 9764513 9764513      -       12     0    12
## 3 chr21 9820622 9820622      +       13     0    13
## 4 chr21 9837545 9837545      +       11     0    11
## 5 chr21 9849022 9849022      +      124    90    34
## 6 chr21 9853326 9853326      +       17    12     5
\end{verbatim}

\hypertarget{further-quality-check}{%
\subsection{Further quality check}\label{further-quality-check}}

It is always a good idea to check how the data looks like before proceeding further. For example, the methylation values should have bimodal distribution generally. This can be checked via
\texttt{getMethylationStats} function. Normally, we should see a bimodal
distributions. Strong deviations from the bimodality may be due poor experimental quality, such as problems with bisulfite treatment.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{getMethylationStats}\NormalTok{(myobj[[}\DecValTok{2}\NormalTok{]],}\DataTypeTok{plot=}\OtherTok{TRUE}\NormalTok{,}\DataTypeTok{both.strands=}\OtherTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/methStats-1} \end{center}

In addition, we might want to see coverage values. By default, methylkit handles bases with at least 10X coverage by that can be changed. The bases with unusually high coverage is usually alarming. It might indicate a PCR bias issue in the experimental procedure. The general coverage statistics can be checked with
\texttt{getCoverageStats} function.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{getCoverageStats}\NormalTok{(myobj[[}\DecValTok{2}\NormalTok{]],}\DataTypeTok{plot=}\OtherTok{TRUE}\NormalTok{,}\DataTypeTok{both.strands=}\OtherTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/coverageStats-1} \end{center}

It might be useful to filter samples based on coverage. Particularly, if our samples are suffering from PCR bias it would be useful to discard bases with very high read coverage. Furthermore, we would also like to discard bases that have low read coverage, a high enough read coverage will increase the power of the statistical tests. The code below filters a \texttt{methylRawList} and discards bases that have coverage below 10X and also discards the bases that have more than 99.9th percentile of coverage in each sample.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{filtered.myobj=}\KeywordTok{filterByCoverage}\NormalTok{(myobj,}\DataTypeTok{lo.count=}\DecValTok{10}\NormalTok{,}\DataTypeTok{lo.perc=}\OtherTok{NULL}\NormalTok{,}
                                      \DataTypeTok{hi.count=}\OtherTok{NULL}\NormalTok{,}\DataTypeTok{hi.perc=}\FloatTok{99.9}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{merging-samples-into-a-single-table}{%
\subsection{Merging samples into a single table}\label{merging-samples-into-a-single-table}}

When we first read the files, each file is stored as its own entity. If we want compare samples in any way, we need to make a unified data structure that contains CpGs that are covered in most samples. The \texttt{unite} function creates a new object using the CpGs covered in each sample. This means

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{meth=}\KeywordTok{unite}\NormalTok{(myobj, }\DataTypeTok{destrand=}\OtherTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Let us take a look at the data content of methylBase object:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{head}\NormalTok{(meth)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##     chr   start     end strand coverage1 numCs1 numTs1
## 1 chr21 9853296 9853296      +        17     10      7
## 2 chr21 9853326 9853326      +        17     12      5
## 3 chr21 9860126 9860126      +        39     38      1
## 4 chr21 9906604 9906604      +        68     42     26
## 5 chr21 9906616 9906616      +        68     52     16
## 6 chr21 9906619 9906619      +        68     59      9
##   coverage2 numCs2 numTs2 coverage3 numCs3 numTs3
## 1       333    268     65        18     16      2
## 2       329    249     79        16     14      2
## 3        83     78      5        83     83      0
## 4       111     97     14        23     18      5
## 5       111    104      7        23     14      9
## 6       111    109      2        22     18      4
##   coverage4 numCs4 numTs4
## 1       395    341     54
## 2       379    284     95
## 3        41     40      1
## 4        37     33      4
## 5        37     27     10
## 6        37     29      8
\end{verbatim}

By default, \texttt{unite} function produces bases/regions covered in all samples. That requirement can be relaxed using ``min.per.group'' option in \texttt{unite} function.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# creates a methylBase object, }
\CommentTok{# where only CpGs covered with at least 1 sample per group will be returned}

\CommentTok{# there were two groups defined by the treatment vector, }
\CommentTok{# given during the creation of myobj: treatment=c(1,1,0,0)}
\NormalTok{meth.min=}\KeywordTok{unite}\NormalTok{(myobj,}\DataTypeTok{min.per.group=}\NormalTok{1L)}
\end{Highlighting}
\end{Shaded}

\hypertarget{filtering-cpgs}{%
\subsection{Filtering CpGs}\label{filtering-cpgs}}

We might need to filter the CpGs further before exploratory analysis or even before the downstream analysis such as differential methylation . For exploratory analysis, it is of general interest to see how samples relate to each other and we might want to remove CpGs that are not variable before doing that. Or we might remove Cs that are potentially C-\textgreater{}T mutations. First, we show how to
filter based on variation. Below, we extract percent methylation values from CpGs as a matrix. Calculate standard deviation for each CpG and filter based on standard deviation. We also plot the the distribution of per CpG standard deviations.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{pm=}\KeywordTok{percMethylation}\NormalTok{(meth) }\CommentTok{# get percent methylation matrix}
\NormalTok{mds=matrixStats}\OperatorTok{::}\KeywordTok{rowSds}\NormalTok{(pm) }\CommentTok{# calculate standard deviation of CpGs}
\KeywordTok{head}\NormalTok{(meth[mds}\OperatorTok{>}\DecValTok{20}\NormalTok{,])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##      chr   start     end strand coverage1 numCs1
## 11 chr21 9906681 9906681      +        21     12
## 12 chr21 9906694 9906694      +        21      9
## 13 chr21 9906700 9906700      +        13      6
## 14 chr21 9906714 9906714      +        14      3
## 18 chr21 9906873 9906873      +        12      8
## 23 chr21 9927527 9927527      +        17      5
##    numTs1 coverage2 numCs2 numTs2 coverage3 numCs3
## 11      9        60     56      4        37     14
## 12     12        60     53      7        39     16
## 13      7        53     43     10        30      8
## 14     11        41     37      4        25     19
## 18      4        41     33      8        15      4
## 23     12        40     22     18        32     32
##    numTs3 coverage4 numCs4 numTs4
## 11     23        26     11     15
## 12     23        26     15     11
## 13     22        23     10     13
## 14      6        21     19      2
## 18     11        22      7     15
## 23      0        14     11      3
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{hist}\NormalTok{(mds,}\DataTypeTok{col=}\StringTok{"cornflowerblue"}\NormalTok{,}\DataTypeTok{xlab=}\StringTok{"Std. dev. per CpG"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/methVar-1} \end{center}

Now, let's assume we know the locations of C-\textgreater{}T mutations. These locations should be removed from the analysis as they do not represent
bisulfite treatment associated conversions. Mutation locations are
stored in a \texttt{GRanges} object, and we can use that to remove CpGs
overlapping with mutations. In order to do overlap operation, we will convert the methylKit object to a \texttt{GRanges} object and do the filtering with \texttt{\%over\%} function within \texttt{{[}\ {]}}. The returned object will still be a methylKit object.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(GenomicRanges)}
\NormalTok{mut=}\KeywordTok{GRanges}\NormalTok{(}\DataTypeTok{seqnames=}\KeywordTok{c}\NormalTok{(}\StringTok{"chr21"}\NormalTok{,}\StringTok{"chr21"}\NormalTok{),}
            \DataTypeTok{ranges=}\KeywordTok{IRanges}\NormalTok{(}\DataTypeTok{start=}\KeywordTok{c}\NormalTok{(}\DecValTok{9853296}\NormalTok{, }\DecValTok{9853326}\NormalTok{),}
                           \DataTypeTok{end=}\KeywordTok{c}\NormalTok{( }\DecValTok{9853296}\NormalTok{,}\DecValTok{9853326}\NormalTok{)))}

\CommentTok{# select CpGs that do not overlap with mutations}
\NormalTok{sub.meth=meth[}\OperatorTok{!}\StringTok{ }\KeywordTok{as}\NormalTok{(meth,}\StringTok{"GRanges"}\NormalTok{) }\OperatorTok{%over%}\StringTok{ }\NormalTok{mut,]}
\KeywordTok{nrow}\NormalTok{(meth)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 963
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{nrow}\NormalTok{(sub.meth)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 961
\end{verbatim}

\hypertarget{clustering-samples}{%
\subsection{Clustering samples}\label{clustering-samples}}

Clustering is used for grouping data points by their similarity. It is a general concept that can be achieved by many different algorithms. In the context of DNA methylation we are trying to find samples that are similar to each other. For example, if we sequenced three heart samples and 4 liver samples, we would expect liver samples will be more similar to each other than heart samples on the DNA methylation space.

The following function will cluster the samples and draw a dendrogram.
It will use correlation distance which is \(1-\rho\) , where \(\rho\) is the correlation coefficient between two pairs of samples. The cluster tree will be drawn using the ``ward'' method. This specific variant uses a ``bottom up'' approach: each data point starts in its own cluster, and pairs of clusters are merged as one moves up the hierarchy. In Ward's method, two clusters are merged if the variance is minimized compared to other possible merge operations. This bottom up approach helps build the dendrogram showing the relationship between clusters.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{clusterSamples}\NormalTok{(meth, }\DataTypeTok{dist=}\StringTok{"correlation"}\NormalTok{, }\DataTypeTok{method=}\StringTok{"ward"}\NormalTok{, }\DataTypeTok{plot=}\OtherTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/unnamed-chunk-191-1} \end{center}

\begin{verbatim}
## 
## Call:
## hclust(d = d, method = HCLUST.METHODS[hclust.method])
## 
## Cluster method   : ward.D 
## Distance         : pearson 
## Number of objects: 4
\end{verbatim}

Setting the \texttt{plot=FALSE} will return a dendrogram object which can be manipulated by users or fed in to other user functions that can work with dendrograms.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{hc =}\StringTok{ }\KeywordTok{clusterSamples}\NormalTok{(meth, }\DataTypeTok{dist=}\StringTok{"correlation"}\NormalTok{, }\DataTypeTok{method=}\StringTok{"ward"}\NormalTok{, }\DataTypeTok{plot=}\OtherTok{FALSE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{principal-component-analysis-1}{%
\subsection{Principal component analysis}\label{principal-component-analysis-1}}

Principal component analysis (PCA) is a mathematical transformation of possibly(correlated) variables into a number of uncorrelated variables called principal components. The resulting components from this transformation is defined in such a way that the first principal component has the highest variance and accounts for as most of the variability in the data. The following function will plot a scree plot for importance of components.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{PCASamples}\NormalTok{(meth, }\DataTypeTok{screeplot=}\OtherTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/unnamed-chunk-193-1} \end{center}

We can also plot PC1 and PC2 axis and a scatter plot of our samples on those axis which will reveal how they cluster within these new dimensions. Similar to clustering dendrogram, we would like to see samples that are similar to be close to each other on the scatter plot. If they are not, it might indicate problems with the experiment such as batch effects.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{pc=}\KeywordTok{PCASamples}\NormalTok{(meth,}\DataTypeTok{obj.return =} \OtherTok{TRUE}\NormalTok{, }\DataTypeTok{adj.lim=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/unnamed-chunk-194-1} \end{center}

In this case, we also returned an object from the plotting function. this is the output of R \texttt{prcomp} function, which includes loadings and eigen vectors which might be useful. You can also do your own PCA analysis using \texttt{percMethylation} and \texttt{prcomp}. In the case above, the methylation matrix is transponsed. This allows us to compare distances between samples on the PCA scatterplot.

\hypertarget{extracting-interesting-regions-segmentation-and-differential-methylation}{%
\section{Extracting interesting regions: segmentation and differential methylation}\label{extracting-interesting-regions-segmentation-and-differential-methylation}}

When analyzing DNA methylation data, we usually look for regions that are different than the rest of the methylome or different from a reference methylome. These regions are so called ``interesting regions''. They usually mark important genomic features that are related to gene regulation which in turn defines the cell type. Therefore, it is a general interest to find such regions and analyze them further to understand our biological sample or to answer specific research questions. Below we will describe two ways of defining ``regions of interest''.

\hypertarget{differential-methylation}{%
\subsection{Differential methylation}\label{differential-methylation}}

Once methylation proportions per base are obtained, generally, the differences between methylation profiles are considered next. When there are multiple sample groups where each group defines a separate biological entity or treatment, it is usually of interest to locate bases or regions with different methylation proportions across the sample groups. The bases or regions with different methylation proportions across samples are called differentially methylated CpG sites (DMCs) and differentially methylated regions (DMRs). They have been shown to play a role in many different diseases due to their association with epigenetic control of gene regulation. In addition, DNA methylation profiles can be highly tissue-specific due to their role in gene regulation \citep{Schubeler2015-ai}. DNA methylation is highly informative when studying normal and diseased cells, because it can also act as a biomarker. For example, the presence of large-scale abnormally methylated genomic regions is a hallmark feature of many types of cancers \citep{Ehrlich2002-hv}. Because of aforementioned reasons, investigating differential methylation is usually one of the primary goals of doing bisulfite sequencing.

\hypertarget{fishers-exact-test}{%
\subsubsection{Fisher's exact test}\label{fishers-exact-test}}

Differential DNA methylation is usually calculated by comparing the proportion of methylated Cs in a test sample relative to a control. In simple comparisons between such pairs of samples (i.e.~test and control), methods such as Fisher's Exact Test can be used. if there are replicates, replicates can be pooled within groups to a single sample per group. This strategy, however, does not take into account biological variability between replicates. We will now show how to compare pairs of samples via \texttt{calculateDiffMeth} function in \texttt{methylKit}. When there are only one sample per sample group, \texttt{calculateDiffMeth} automatically applies Fisher's exact test. We will not extract one sample from each group and run \texttt{calculateDiffMeth}, which will automatically run Fisher's exact test.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{getSampleID}\NormalTok{(meth)}
\NormalTok{new.meth=}\KeywordTok{reorganize}\NormalTok{(meth,}\DataTypeTok{sample.ids=}\KeywordTok{c}\NormalTok{(}\StringTok{"test1"}\NormalTok{,}\StringTok{"ctrl1"}\NormalTok{),}\DataTypeTok{treatment=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{0}\NormalTok{))}
\NormalTok{dmf=}\KeywordTok{calculateDiffMeth}\NormalTok{(new.meth)}
\end{Highlighting}
\end{Shaded}

As mentioned, we can also pool the samples from the same group by adding up the number of Cs and Ts per group. This way even if we have replicated experiments we treat them as single experiments, and can apply Fisher's exact test. We will now pool the samples and apply

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{pooled.meth=}\KeywordTok{pool}\NormalTok{(meth,}\DataTypeTok{sample.ids=}\KeywordTok{c}\NormalTok{(}\StringTok{"test"}\NormalTok{,}\StringTok{"control"}\NormalTok{))}
\NormalTok{dm.pooledf=}\KeywordTok{calculateDiffMeth}\NormalTok{(pooled.meth)}
\end{Highlighting}
\end{Shaded}

\texttt{calculateDiffMeth} function returns the P-values for all bases or regions in the input methylBase object. We need to filter to get differentially methylated CpGs. This can be done via \texttt{getMethlyDiff} function or simple filtering via \texttt{{[}\ {]}} notation. Below we show how to filter the \texttt{methylDiff} object output by \texttt{calculateDiffMeth()} function in order to get differentially methylated CpGs. The function arguments defines cutoff values for the methylation difference between groups and Q-value. In these cases, we require a methylation difference of 25\% and Q-value of at least \(0.01\).

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# get differentially methylated bases/regions with specific cutoffs}
\NormalTok{all.diff=}\KeywordTok{getMethylDiff}\NormalTok{(dm.pooledf,}\DataTypeTok{difference=}\DecValTok{25}\NormalTok{,}\DataTypeTok{qvalue=}\FloatTok{0.01}\NormalTok{,}\DataTypeTok{type=}\StringTok{"all"}\NormalTok{)}

\CommentTok{# get hyper-methylated}
\NormalTok{hyper=}\KeywordTok{getMethylDiff}\NormalTok{(dm.pooledf,}\DataTypeTok{difference=}\DecValTok{25}\NormalTok{,}\DataTypeTok{qvalue=}\FloatTok{0.01}\NormalTok{,}\DataTypeTok{type=}\StringTok{"hyper"}\NormalTok{)}

\CommentTok{# get hypo-methylated}
\NormalTok{hypo=}\KeywordTok{getMethylDiff}\NormalTok{(dm.pooledf,}\DataTypeTok{difference=}\DecValTok{25}\NormalTok{,}\DataTypeTok{qvalue=}\FloatTok{0.01}\NormalTok{,}\DataTypeTok{type=}\StringTok{"hypo"}\NormalTok{)}

\CommentTok{#using [ ] notation}
\NormalTok{hyper2=dm.pooledf[dm.pooledf}\OperatorTok{$}\NormalTok{qvalue }\OperatorTok{<}\StringTok{ }\FloatTok{0.01} \OperatorTok{&}\StringTok{ }\NormalTok{dm.pooledf}\OperatorTok{$}\NormalTok{meth.diff }\OperatorTok{>}\StringTok{ }\DecValTok{25}\NormalTok{,]}
\end{Highlighting}
\end{Shaded}

\hypertarget{logistic-regression-based-tests}{%
\subsubsection{Logistic regression based tests}\label{logistic-regression-based-tests}}

Regression-based methods are generally used to model methylation levels in relation to the sample groups and variation between replicates. Differences between currently available regression methods stem from the choice of distribution to model the data and the variation associated with it. In the simplest case, linear regression can be used to model methylation per given CpG or loci across sample groups. The model fits regression coefficients to model the expected methylation proportion values for each CpG site across sample groups. Hence, the null hypothesis of the model coefficients being zero could be tested using t-statistics. However, linear regression based methods might produce fitted methylation levels outside the range {[}0,1{]} unless the values are transformed before regression. An alternative is logistic regression, which can deal with data strictly bounded between 0 and 1 and with non-constant variance, such as methylation proportion/fraction values. In the logistic regression, it is assumed that fitted values have variation \(np(1-p)\), where \(p\) is the fitted methylation proportion for a given sample and n is the read coverage. If the observed variance is larger or smaller than assumed by the model, one speaks of under- or over-dispersion. This over/under-dispersion can be corrected by calculating a scaling factor and using that factor to adjust the variance estimates as in \(np(1-p)s\), where \(s\) is the scaling factor. MethylKit can apply logistic regression to test the methylation difference with or without the over-dispersion correction. In this case, Chi-square or F-test can be used to compare the difference in the deviances of the null model and the alternative model. The null model assumes there is no relationship between sample groups and methylation, and the alternative model assumes that there is a relationship where sample groups are predictive of methylation values for a given CpG or region for which the model is constructed. Next, we are going to use the logistic regression based model with over-dispersion correction and Chi-square test.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{dm.lr=}\KeywordTok{calculateDiffMeth}\NormalTok{(meth,}\DataTypeTok{overdispersion =} \StringTok{"MN"}\NormalTok{,}\DataTypeTok{test =}\StringTok{"Chisq"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{betabinomial-distribution-based-tests}{%
\subsubsection{Betabinomial distribution based tests}\label{betabinomial-distribution-based-tests}}

More complex regression models use beta binomial distribution and are particularly useful for better modeling the variance. Similar to logistic regression, their observation follows binomial distribution (number of reads), but methylation proportion itself can vary across samples, according to a beta distribution. It can deal with fitting values in {[}0,1{]} range and performs better when there is greater variance than expected by the simple logistic model. In essence, these models have a different way of calculating a scaling factor when there is over-dispersion in the model. Further enhancements are made to these models by using the Empirical Bayes methods that can better estimate hyper parameters of the beta distribution (variance-related parameters) by borrowing information between loci or regions within the genome to aid with inference about each individual loci or region. We are now going to use a beta-binomial based model called DSS\citep{Feng2014-pd} to calculate differential methylation.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{dm.dss=}\KeywordTok{calculateDiffMethDSS}\NormalTok{(meth)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Using internal DSS code...
\end{verbatim}

\hypertarget{differential-methylation-for-regions-rather-than-base-pairs}{%
\subsubsection{Differential methylation for regions rather than base-pairs}\label{differential-methylation-for-regions-rather-than-base-pairs}}

Until now, we worked on differentially methylated cytosines. However,
working with base-pair resolution data has its problems. Not all the CpGs will be covered in all samples, if covered they may have low coverage which reduces the power of the tests. Instead of base-pairs, we can choose to work with regions. So, it might be desirable to summarize methylation information over pre-defined regions rather than doing base-pair resolution analysis. \texttt{methylKit} provides functionality to do such analysis. We can either tile the whole genome to tiles with predefined length, or we can use pre-defined regions such as promoters or CpG islands. This kind of regional analysis is carried out by adding up C and T counts from each covered cytosine and returning a total C and T count for each region.

The function below tiles the genome with windows 1000bp length and 1000bp step-size and summarizes the methylation information on those tiles. In this case, it returns a \texttt{methylRawList} object which can be fed into \texttt{unite} and \texttt{calculateDiffMeth} functions consecutively to get differentially methylated regions.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{tiles=}\KeywordTok{tileMethylCounts}\NormalTok{(myobj,}\DataTypeTok{win.size=}\DecValTok{1000}\NormalTok{,}\DataTypeTok{step.size=}\DecValTok{1000}\NormalTok{)}
\KeywordTok{head}\NormalTok{(tiles[[}\DecValTok{1}\NormalTok{]],}\DecValTok{3}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##     chr   start     end strand coverage numCs numTs
## 1 chr21 9764001 9765000      *       24     3    21
## 2 chr21 9820001 9821000      *       13     0    13
## 3 chr21 9837001 9838000      *       11     0    11
\end{verbatim}

In addition, if we are interested in particular regions, we can also get those regions as methylKit objects after summarizing the methylation information as described above. The code below summarizes the methylation information over a given set of promoter regions and outputs a \texttt{methylRaw} or \texttt{methylRawList} object depending on the input. We are using the output of
\texttt{genomation} functions used above to provide the locations of promoters. For regional summary functions, we need to
provide regions of interest as GRanges object.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(genomation)}

\CommentTok{# read the gene BED file}
\NormalTok{gene.obj=}\KeywordTok{readTranscriptFeatures}\NormalTok{(}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{, }\StringTok{"refseq.hg18.bed.txt"}\NormalTok{, }
                                           \DataTypeTok{package =} \StringTok{"methylKit"}\NormalTok{))}
\NormalTok{promoters=}\KeywordTok{regionCounts}\NormalTok{(myobj,gene.obj}\OperatorTok{$}\NormalTok{promoters)}

\KeywordTok{head}\NormalTok{(promoters[[}\DecValTok{1}\NormalTok{]])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##     chr    start      end strand coverage numCs numTs
## 1 chr21 10011791 10013791      -     7953  6662  1290
## 2 chr21 10119796 10121796      -     1725  1171   554
## 3 chr21 10119808 10121808      -     1725  1171   554
## 4 chr21 13903368 13905368      +       10    10     0
## 5 chr21 14273636 14275636      -      282   220    62
## 6 chr21 14509336 14511336      +     1058    55  1003
\end{verbatim}

In addition, it is possible to cluster DMCs based on their proximity and direction of differential methylation. This can be achieved by \texttt{methSeg} function in methylKit. We will see more about \texttt{methSeg} function in the following section.
But it can take the output of \texttt{getMethylDiff} function therefore can work on DMCs to get differentially methylated regions.

\hypertarget{adding-covariates}{%
\subsubsection{Adding covariates}\label{adding-covariates}}

Covariates can be included in the analysis as well in methylKit. The \texttt{calculateDiffMeth} function will then try to
separate the influence of the covariates from the
treatment effect via the logistic regression model. In this case, we will test
if full model (model with treatment and covariates) is better than the model with
the covariates only. If there is no effect due to the treatment (sample groups),
the full model will not explain the data better than the model with covariates
only. In \texttt{calculateDiffMeth}, this is achieved by
supplying the \texttt{covariates} argument in the format of a \texttt{data.frame}.
Below, we simulate methylation data and add make a \texttt{data.frame} for the age.
The data frame can include more columns, and those columns can also be
\texttt{factor} variables. The row order of the data.frame should match the order
of samples in the \texttt{methylBase} object. Below we are showing an example
of this using a simulated data set where methylation values of CpGs will be affected by the age of the sample.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{covariates=}\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{age=}\KeywordTok{c}\NormalTok{(}\DecValTok{30}\NormalTok{,}\DecValTok{80}\NormalTok{,}\DecValTok{34}\NormalTok{,}\DecValTok{30}\NormalTok{,}\DecValTok{80}\NormalTok{,}\DecValTok{40}\NormalTok{))}
\NormalTok{sim.methylBase=}\KeywordTok{dataSim}\NormalTok{(}\DataTypeTok{replicates=}\DecValTok{6}\NormalTok{,}\DataTypeTok{sites=}\DecValTok{1000}\NormalTok{,}
                        \DataTypeTok{treatment=}\KeywordTok{c}\NormalTok{(}\KeywordTok{rep}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{3}\NormalTok{),}\KeywordTok{rep}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{3}\NormalTok{)),}
                        \DataTypeTok{covariates=}\NormalTok{covariates,}
                        \DataTypeTok{sample.ids=}\KeywordTok{c}\NormalTok{(}\KeywordTok{paste0}\NormalTok{(}\StringTok{"test"}\NormalTok{,}\DecValTok{1}\OperatorTok{:}\DecValTok{3}\NormalTok{),}\KeywordTok{paste0}\NormalTok{(}\StringTok{"ctrl"}\NormalTok{,}\DecValTok{1}\OperatorTok{:}\DecValTok{3}\NormalTok{)))}

\NormalTok{my.diffMeth3=}\KeywordTok{calculateDiffMeth}\NormalTok{(sim.methylBase,}
                               \DataTypeTok{covariates=}\NormalTok{covariates,}
                               \DataTypeTok{overdispersion=}\StringTok{"MN"}\NormalTok{,}
                               \DataTypeTok{test=}\StringTok{"Chisq"}\NormalTok{,}\DataTypeTok{mc.cores=}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{methylation-segmentation}{%
\subsection{Methylation segmentation}\label{methylation-segmentation}}

The analysis of methylation dynamics is not exclusively restricted to differentially methylated regions across samples, apart from this there is also an interest in examining the methylation profiles within the same sample. Usually, depressions in methylation profiles pinpoint regulatory regions like gene promoters that co-localize with CG-dense CpG islands. On the other hand, many gene-body regions are extensively methylated and CpG-poor \citep{Bock2012-oh}. These observations would describe a bimodal model of either hyper- or hypomethylated regions dependent on the local density of CpGs \citep{Lovkvist2016-ky}. However, given the detection of CpG-poor regions with locally reduced levels of methylation (on average 30\%) in pluripotent embryonic stem cells and in neuronal progenitors in both mouse and human, a different model seems also reasonable \citep{Stadler2011-iu}. These low-methylated regions (LMRs) are located distal to promoters, have little overlap with CpG islands and associated with enhancer marks such as p300 binding sites and H3K27ac enrichment.

Now we are going to try to segment portion for the H1 human embryonic stem cell line. MethylKit uses change-point analysis to segment the methylome. In change-point analysis, the change-points of a genome-wide methylation signal are recorded and the genome is partitioned into regions between consecutive change points. CpGs in each segment is similar to eachoter more than the following segment.
After segmentation, methylKit function \texttt{methSeg} identifies segments that are further clustered into segment classes using a mixture modeling approach. This clustering is based on only the average methylation level of the segments and allows the detection of distinct methylome features comparable to unmethylated regions (UMRs), lowly methylated regions (LMRs) and fully methylated regions (FMRs) mentioned at Stadler et. al (2012). The code snippet below reads the methylation data from H1 cell line as a \texttt{GRanges} object, and runs the segmentation with potentially up to classes of segments. Mixture modelling determines the optimal number of segments using a statistic called bayesian information criterion (BIC). BIC is a statistic based on model likelihood and helps us select the model that fits the data better. We have set the number of segment classes to try using \texttt{G=1:4} argument.The \texttt{minSeg} arguments are related to minimum number of CpGs in the segments.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# read methylation data}

\NormalTok{methFile=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{,}\StringTok{"H1.chr21.chr22.rds"}\NormalTok{,}
                     \DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{mbw=}\KeywordTok{readRDS}\NormalTok{(methFile)}

\CommentTok{# segment the methylation data}
\NormalTok{res=}\KeywordTok{methSeg}\NormalTok{(mbw,}\DataTypeTok{minSeg=}\DecValTok{10}\NormalTok{,}\DataTypeTok{G=}\DecValTok{1}\OperatorTok{:}\DecValTok{4}\NormalTok{,}
            \DataTypeTok{join.neighbours =} \OtherTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/unnamed-chunk-198-1} \end{center}

In this case, we know that BIC does not improve much after 4 segment classes. Now, we will not have a look at the characteristics of the segment classes. We are going to plot mean methylation value and the length of the segment as a scatter plot.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# plot }
\KeywordTok{plot}\NormalTok{(res}\OperatorTok{$}\NormalTok{seg.mean,}
     \KeywordTok{log10}\NormalTok{(}\KeywordTok{width}\NormalTok{(res)),}\DataTypeTok{pch=}\DecValTok{20}\NormalTok{,}
     \DataTypeTok{col=}\NormalTok{scales}\OperatorTok{::}\KeywordTok{alpha}\NormalTok{(}\KeywordTok{rainbow}\NormalTok{(}\DecValTok{4}\NormalTok{)[}\KeywordTok{as.numeric}\NormalTok{(res}\OperatorTok{$}\NormalTok{seg.group)], }\FloatTok{0.2}\NormalTok{),}
     \DataTypeTok{ylab=}\StringTok{"log10(length)"}\NormalTok{,}
     \DataTypeTok{xlab=}\StringTok{"methylation proportion"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{compgenomrReloaded_files/figure-latex/segplot-1} \end{center}

The highly methylated segment classes that have more than 70\% methylation are usually longer, median length is 17889 bp. The segment class that has the lowest methylation values have median length of 1376 bp and the shortest segment class has low to medium methylation level, has median length of 412 bp.

\hypertarget{working-with-large-files}{%
\subsection{Working with large files}\label{working-with-large-files}}

We might want to perform differential methylation analysis in R using whole genome methylation data of multiple samples. The problem is that for genome-wide experiments, file sizes can easily range from hundreds of megabytes to gigabytes and processing multiple instances of those files in memory (RAM) might become unfeasible unless we have access to a high performance cluster (HPC) with extensive RAM. If we want to use a desktop computer or laptop with limited RAM, we either need to restrict our analysis to a subset of the data or use packages that can handle this situation.

The methylKit package provides the capability of dealing large files and high number of samples by exploiting flat file databases to substitute in-memory objects. The internal data apart from meta information has a tabular structure storing chromosome, start/end position, strand information of the associated CpG base just like many other biological formats like BED, GFF or SAM. By exporting this tabular data into a TAB-delimited file and making sure it is accordingly position-sorted it can be indexed using the generic \href{http://www.htslib.org/doc/tabix.html}{tabix tool}. In general tabix indexing is a generalization of BAM indexing for generic TAB-delimited files. It inherits all the advantages of BAM indexing, including data compression and efficient random access in terms of few seek function calls per query \citep{Li2011-wc}. MethylKit relies on {[}Rsamtools{]} (\url{http://bioconductor.org/packages/release/bioc/html/Rsamtools.html}) which implements tabix functionality for R and this way internal methylKit objects can be efficiently stored as compressed file on the disk and still be fast accessed. Another advantage is that existing compressed files can be loaded in interactive sessions, allowing the backup and transfer of intermediate analysis results.

methylKit provides the capability for storing objects in tabix format within various functions. Every methylKit object has their tabix-based flat-file database equivalent. For example, when reading a methylation call file the \texttt{dbtype} argument can be provided, this will create tabix based objects.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ myobj=}\KeywordTok{methRead}\NormalTok{( file.list,}
               \DataTypeTok{sample.id=}\KeywordTok{list}\NormalTok{(}\StringTok{"test1"}\NormalTok{,}\StringTok{"test2"}\NormalTok{,}\StringTok{"ctrl1"}\NormalTok{,}\StringTok{"ctrl2"}\NormalTok{),}
               \DataTypeTok{assembly=}\StringTok{"hg18"}\NormalTok{,}\DataTypeTok{treatment=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{),}
               \DataTypeTok{dbtype=}\StringTok{"tabix"}\NormalTok{) }
\end{Highlighting}
\end{Shaded}

The advantage of tabix-based objects is of course saving memory and more efficient parallelization for differential methylation calculation. However, since the data is written to a file and indexed whenever a new object is created, working with tabix-based objects will be slower at certain steps of the analysis compared to in-memory objects.

\hypertarget{annotation-of-dmrsdmcs-and-segments}{%
\section{Annotation of DMRs/DMCs and segments}\label{annotation-of-dmrsdmcs-and-segments}}

The regions of interest obtained through differential methylation or segmentation analysis often need to be integrated with genome annotation datasets. Without this type of integration, differential methylation or segmentation results will be hard to interpret in biological terms. The most common annotation task is to see where regions of interest land in relation to genes and gene parts and regulatory regions: Do they mostly occupy promoter, intronic or exonic regions? Do they overlap with repeats? Do they overlap with other epigenomic markers or long-range regulatory regions? These questions are not specific to methylation nearly all regions of interest obtained via genome-wide studies have to deal with such questions. Thus, there are already multiple software tools that can produce such annotations. One is the Bioconductor package \href{http://bioconductor.org/packages/release/bioc/html/genomation.html}{genomation}\citep{Akalin2015-yk}. It can be used to annotate DMRs/DMCs and it can also be used to integrate methylation proportions over the genome with other quantitative information and produce meta-gene plots or heatmaps. Below, we are reading a BED file for transcripts and using that to annotate DMCs with promoter/intron/exon/intergenic annotation.\texttt{genomation::readTranscriptFeatures} function reads a BED12 file, calculates the coordinates of promoters, exons and introns and the subsequent function uses that information for annotation.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(genomation)}

\CommentTok{# read the gene BED file}
\NormalTok{transcriptBED=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{, }\StringTok{"refseq.hg18.bed.txt"}\NormalTok{, }
                                           \DataTypeTok{package =} \StringTok{"methylKit"}\NormalTok{)}
\NormalTok{gene.obj=}\KeywordTok{readTranscriptFeatures}\NormalTok{(transcriptBED)}
\CommentTok{#}
\CommentTok{# annotate differentially methylated CpGs with }
\CommentTok{# promoter/exon/intron using annotation data}
\CommentTok{#}
\KeywordTok{annotateWithGeneParts}\NormalTok{(}\KeywordTok{as}\NormalTok{(all.diff,}\StringTok{"GRanges"}\NormalTok{),gene.obj)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##   promoter       exon     intron intergenic 
##      28.24      15.27      33.59      58.02 
##   promoter       exon     intron intergenic 
##      28.24       0.00      13.74      58.02 
## promoter     exon   intron 
##     0.29     0.03     0.17 
##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##       5     815   49918   52410   94644  313528
\end{verbatim}

Similarly, we can read the CpG island annotation and annotate our differentially methylated bases/regions with them.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# read the shores and flanking regions and name the flanks as shores }
\CommentTok{# and CpG islands as CpGi}
\NormalTok{cpg.file=}\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{, }\StringTok{"cpgi.hg18.bed.txt"}\NormalTok{, }
                                        \DataTypeTok{package =} \StringTok{"methylKit"}\NormalTok{)}
\NormalTok{cpg.obj=}\KeywordTok{readFeatureFlank}\NormalTok{(cpg.file,}
                           \DataTypeTok{feature.flank.name=}\KeywordTok{c}\NormalTok{(}\StringTok{"CpGi"}\NormalTok{,}\StringTok{"shores"}\NormalTok{))}
\CommentTok{#}
\CommentTok{# convert methylDiff object to GRanges and annotate}
\NormalTok{diffCpGann=}\KeywordTok{annotateWithFeatureFlank}\NormalTok{(}\KeywordTok{as}\NormalTok{(all.diff,}\StringTok{"GRanges"}\NormalTok{),}
\NormalTok{                                    cpg.obj}\OperatorTok{$}\NormalTok{CpGi,cpg.obj}\OperatorTok{$}\NormalTok{shores,}
                         \DataTypeTok{feature.name=}\StringTok{"CpGi"}\NormalTok{,}\DataTypeTok{flank.name=}\StringTok{"shores"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{other-r-packages-that-can-be-used-for-methylation-analysis}{%
\section{Other R packages that can be used for methylation analysis}\label{other-r-packages-that-can-be-used-for-methylation-analysis}}

\begin{itemize}
\tightlist
\item
  \href{http://bioconductor.org/packages/release/bioc/html/genomation.html}{DSS} beta-binomial models with Empirical Bayes for moderating dispersion.
\item
  \href{http://bioconductor.org/packages/release/bioc/html/BSseq.html}{BSseq} Regional differential methylation analysis using smoothing and linear regression based tests.
\item
  \href{http://bioconductor.org/packages/release/bioc/html/BiSeq.html}{BiSeq} Regional differential methylation analysis using beta-binomial models
\item
  \href{http://bioconductor.org/packages/release/bioc/html/MethylSeekR.html}{MethylSeekR}: Methylome segmentation using HMM and cutoffs.
\item
  \href{http://bioconductor.org/packages/release/bioc/html/QuasR.html}{QuasR}: Methylation aware alignment and methylation calling. As well as fastQC-like fastq raw data quality check features.
\end{itemize}

\hypertarget{exercises-7}{%
\section{Exercises}\label{exercises-7}}

\hypertarget{exercise-1}{%
\subsection{Exercise 1}\label{exercise-1}}

The main objective of this exercise is getting differential methylated cytosines between two groups of samples: IDH-mut (AML patients with IDH mutations) vs NBM (normal bone marrow samples).

\begin{itemize}
\tightlist
\item
  Download methylation call files from GEO. These files are readable by methlKit using default \texttt{methRead} arguments.
\end{itemize}

\begin{longtable}[]{@{}ll@{}}
\toprule
samples & Link\tabularnewline
\midrule
\endhead
IDH1\_rep1 & \href{https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM919990\&format=file\&file=GSM919990\%5FIDH\%2Dmut\%5F1\%5FmyCpG\%2Etxt\%2Egz}{link}\tabularnewline
IDH1\_rep2 & \href{https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM919991\&format=file\&file=GSM919991\%5FIDH\%5Fmut\%5F2\%5FmyCpG\%2Etxt\%2Egz}{link}\tabularnewline
NBM\_rep1 & \href{https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM919982\&format=file\&file=GSM919982\%5FNBM\%5F1\%5FmyCpG\%2Etxt\%2Egz}{link}\tabularnewline
NBM\_rep2 & \href{https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM919984\&format=file\&file=GSM919984\%5FNBM\%5F2\%5FRep1\%5FmyCpG\%2Etxt\%2Egz}{link}\tabularnewline
\bottomrule
\end{longtable}

\begin{itemize}
\tightlist
\item
  Find differentially methylated cytosines. Use chr1 and chr2 only if you need to save time. You can subset it after you download the files either in R or unix. The files are for hg18 assembly of human genome.
\item
  Describe the differential methylation trend, what is the main effect ?
\item
  Annotate differentially methylated cytosines (DMCs) promoter/intron/exon ?
\item
  Which genes are the nearest to DMCs ? Can you do gene set analysis either in R or via web-based tools ?
\end{itemize}

Example code for reading a file:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(methylKit)}
\NormalTok{m=}\KeywordTok{methRead}\NormalTok{(}\StringTok{"~/Downloads/GSM919982_NBM_1_myCpG.txt.gz"}\NormalTok{,}\DataTypeTok{sample.id =} \StringTok{"idh"}\NormalTok{,}\DataTypeTok{assembly=}\StringTok{"hg18"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{exercise-2}{%
\subsection{Exercise 2}\label{exercise-2}}

The main objective of this exercise is to learn how to do methylome segmentation and the downstream analysis for annotation and data integration.

\begin{itemize}
\tightlist
\item
  Download the human embryonic stem-cell (H1 Cell Line) methylation bigWig files from \href{http://egg2.wustl.edu/roadmap/web_portal/processed_data.html\#MethylData}{Roadmap Epigenomics website}. It may take a while to understand how the website is structured and which bigWig file to use. That is part of the exercise. The files you will download are for hg19 assembly unless stated otherwise.
\item
  Do segmentation on hESC methylome. you can only do chr1 if it takes too much time.
\item
  Annotate segments, what kind of gene-based features each segment class overlaps with (promoter/exon/intron)
\item
  For each segment type, annotate the segments with chromHMM annotations from Roadmap Epigenome database available (here){[}\url{https://egg2.wustl.edu/roadmap/web_portal/chr_state_learning.html\#core_15state}{]}, the specific file you should use is (here){[}\url{https://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/coreMarks/jointModel/final/E003_15_coreMarks_mnemonics.bed.gz}{]}. This is a bed file with chromHMM annotations. chromHMM annotations are parts of the genome identified by a hidden-markov-model based machine-learning algorithm. The segments correspond to active promoters, enhancers, active transcription, insulators. etc. The chromHMM model uses histone modification ChIP-seq and potentially other ChIP-seq data sets to annotate the genome.
\end{itemize}

\hypertarget{references}{%
\section{References}\label{references}}

\hypertarget{multiomics}{%
\chapter{Multi-omics Analysis}\label{multiomics}}

Living cells are a symphony of complex processes. Modern sequencing technology has lead to many comprehensive assays being routinely available to experimenters, giving us different ways to peek at the internal doings of the cells, each experiment revealing a different part of some underlying processes. As an example, most cells have the same DNA, but sequencing the genome of a cell allows us to find mutations and structural alterations that drive tumerogenesis in cancer. If we treat the DNA with bisulfite prior to sequencing, cytosine residues are converted to uracil, but 5-methylcytosine residues are unaffected. This allows us to probe the methylation patterns of the genome, or its methylome. By sequencing the mRNA molecules in a cell, we can calculate the abundance, in different samples, of different mRNA transcripts, or uncover its transcriptome. Performing different experiments on the same samples, for instance RNA-seq, DNA-seq, and BS-seq, results in multi-dimensional omics datasets, which enable the study of relationships between different biological processes, e.g.~DNA methylation, mutations, and gene expression, and the leveraging of multiple data types to draw inferences about biological systems. This chapter provides an overview of some of the available methods for such analyses, focusing on matrix factorization approaches. In the examples in this chapter we will demonstrate how these methods are applicable to cancer subtyping, i.e.~finding tumors which are driven by the same oncogenic processes.

\hypertarget{use-case-multi-omics-data-from-colorectal-cancer}{%
\subsection{Use case: Multi-omics data from colorectal cancer}\label{use-case-multi-omics-data-from-colorectal-cancer}}

The examples in this chapter will use the following data: a set of 121 tumors from the TCGA \citep{tcga_pan_cancer} cohorts of Colon and Rectum adenocarcinoma. The tumors have been profiled for gene expression using RNA-seq, mutations using Exome-seq, and copy number variations using genotyping arrays. The subset of tumors used in this chapter belong to two distinct molecular subtypes defined by the Colorectal Cancer Subtyping Consortium \citep{cmscc}, \emph{CMS1} and \emph{CMS3}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{csvfile <-}\StringTok{ }\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{, }\StringTok{"multi-omics"}\NormalTok{, }\StringTok{"COREAD_CMS13_gex.csv"}\NormalTok{,}\DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{x1 <-}\StringTok{ }\KeywordTok{read.csv}\NormalTok{(csvfile, }\DataTypeTok{row.names=}\DecValTok{1}\NormalTok{)}
\KeywordTok{rownames}\NormalTok{(x1) <-}\StringTok{ }\KeywordTok{sapply}\NormalTok{(}\KeywordTok{strsplit}\NormalTok{(}\KeywordTok{rownames}\NormalTok{(x1), }\StringTok{"}\CharTok{\textbackslash{}\textbackslash{}}\StringTok{|"}\NormalTok{), }\ControlFlowTok{function}\NormalTok{(x) x[}\DecValTok{1}\NormalTok{])}
\NormalTok{knitr}\OperatorTok{::}\KeywordTok{kable}\NormalTok{(}\KeywordTok{head}\NormalTok{(}\KeywordTok{t}\NormalTok{(}\KeywordTok{head}\NormalTok{(x1))), }\DataTypeTok{caption=}\StringTok{"Gene expression data (head)"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{table}[t]

\caption{\label{tab:moloadMultiomicsGE}Gene expression data (head)}
\centering
\begin{tabular}{l|r|r|r|r|r|r}
\hline
  & RNF113A & S100A13 & AP3D1 & ATP6V1G1 & UBQLN4 & TPPP3\\
\hline
TCGA.A6.2672 & 21.20 & 19.73 & 11.53 & 0.00 & 15.36 & 12.77\\
\hline
TCGA.A6.3809 & 21.51 & 18.66 & 12.99 & 14.13 & 19.62 & 0.00\\
\hline
TCGA.A6.5661 & 20.08 & 18.97 & 10.84 & 15.31 & 0.00 & 0.00\\
\hline
TCGA.A6.5665 & 0.00 & 11.88 & 10.24 & 19.79 & 0.00 & 0.00\\
\hline
TCGA.A6.6653 & 0.00 & 12.08 & 0.00 & 0.00 & 0.00 & 0.00\\
\hline
TCGA.A6.6780 & 0.00 & 12.99 & 0.00 & 19.97 & 13.18 & 11.59\\
\hline
\end{tabular}
\end{table}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{csvfile <-}\StringTok{ }\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{, }\StringTok{"multi-omics"}\NormalTok{, }\StringTok{"COREAD_CMS13_muts.csv"}\NormalTok{,}\DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{x2 <-}\StringTok{ }\KeywordTok{read.csv}\NormalTok{(csvfile, }\DataTypeTok{row.names=}\DecValTok{1}\NormalTok{)}
\NormalTok{x2[x2}\OperatorTok{>}\DecValTok{0}\NormalTok{]=}\DecValTok{1}
\NormalTok{knitr}\OperatorTok{::}\KeywordTok{kable}\NormalTok{(}\KeywordTok{head}\NormalTok{(}\KeywordTok{t}\NormalTok{(}\KeywordTok{head}\NormalTok{(x2))), }\DataTypeTok{caption=}\StringTok{"Mutation data (head)"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{table}[t]

\caption{\label{tab:moloadMultiomicsMUT}Mutation data (head)}
\centering
\begin{tabular}{l|r|r|r|r|r|r}
\hline
  & TTN & TP53 & APC & KRAS & SYNE1 & MUC16\\
\hline
TCGA.A6.2672 & 1 & 0 & 0 & 0 & 1 & 1\\
\hline
TCGA.A6.3809 & 1 & 0 & 0 & 0 & 0 & 0\\
\hline
TCGA.A6.5661 & 1 & 0 & 0 & 0 & 1 & 1\\
\hline
TCGA.A6.5665 & 1 & 0 & 0 & 0 & 1 & 1\\
\hline
TCGA.A6.6653 & 1 & 0 & 0 & 1 & 0 & 0\\
\hline
TCGA.A6.6780 & 1 & 0 & 0 & 0 & 0 & 1\\
\hline
\end{tabular}
\end{table}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{csvfile <-}\StringTok{ }\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{, }\StringTok{"multi-omics"}\NormalTok{, }\StringTok{"COREAD_CMS13_cnv.csv"}\NormalTok{,}\DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{x3 <-}\StringTok{ }\KeywordTok{read.csv}\NormalTok{(csvfile, }\DataTypeTok{row.names=}\DecValTok{1}\NormalTok{)}
\NormalTok{knitr}\OperatorTok{::}\KeywordTok{kable}\NormalTok{(}\KeywordTok{head}\NormalTok{(}\KeywordTok{t}\NormalTok{(}\KeywordTok{head}\NormalTok{(x3))), }\DataTypeTok{caption=}\StringTok{"Copy number data (head)"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{table}[t]

\caption{\label{tab:moloadMultiomicsCNV}Copy number data (head)}
\centering
\begin{tabular}{l|r|r|r|r|r|r}
\hline
  & 8p23.2 & 8p23.3 & 8p23.1 & 8p21.3 & 8p12 & 8p22\\
\hline
TCGA.A6.2672 & 0 & 0 & 0 & 0 & 0 & 0\\
\hline
TCGA.A6.3809 & 0 & 0 & 0 & 0 & 0 & 0\\
\hline
TCGA.A6.5661 & 0 & 0 & 0 & 0 & 0 & 0\\
\hline
TCGA.A6.5665 & 0 & 0 & 0 & 0 & 0 & 0\\
\hline
TCGA.A6.6653 & 0 & 0 & 0 & 0 & 0 & 0\\
\hline
TCGA.A6.6780 & 0 & 0 & 0 & 0 & 0 & 0\\
\hline
\end{tabular}
\end{table}

Each of the data types (gene expresion, mutations, copy number variation) on its own, provides some signal which allows to somewhat separate the samples into the two different subtypes:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{csvfile <-}\StringTok{ }\KeywordTok{system.file}\NormalTok{(}\StringTok{"extdata"}\NormalTok{, }\StringTok{"multi-omics"}\NormalTok{, }\StringTok{"COREAD_CMS13_subtypes.csv"}\NormalTok{,}\DataTypeTok{package=}\StringTok{"compGenomRData"}\NormalTok{)}
\NormalTok{covariates <-}\StringTok{ }\KeywordTok{read.csv}\NormalTok{(csvfile, }\DataTypeTok{row.names=}\DecValTok{1}\NormalTok{)}
\KeywordTok{rownames}\NormalTok{(covariates) <-}\StringTok{ }\KeywordTok{gsub}\NormalTok{(}\DataTypeTok{pattern =} \StringTok{'-'}\NormalTok{, }\DataTypeTok{replacement =} \StringTok{'}\CharTok{\textbackslash{}\textbackslash{}}\StringTok{.'}\NormalTok{, }\KeywordTok{rownames}\NormalTok{(covariates))}
\NormalTok{covariates <-}\StringTok{ }\NormalTok{covariates[}\KeywordTok{colnames}\NormalTok{(x1),]}
\NormalTok{anno_col <-}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{cms=}\KeywordTok{as.factor}\NormalTok{(covariates}\OperatorTok{$}\NormalTok{cms_label))}
\KeywordTok{rownames}\NormalTok{(anno_col) <-}\StringTok{ }\KeywordTok{rownames}\NormalTok{(covariates)}
\end{Highlighting}
\end{Shaded}

We will first check the gene expression data in relation to the subtypes.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{pheatmap}\OperatorTok{::}\KeywordTok{pheatmap}\NormalTok{(x1,}
                   \DataTypeTok{annotation_col =}\NormalTok{ anno_col,}
                   \DataTypeTok{show_colnames =} \OtherTok{FALSE}\NormalTok{,}
                   \DataTypeTok{show_rownames =} \OtherTok{FALSE}\NormalTok{,}
                   \DataTypeTok{main=}\StringTok{"Gene expression data"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics[width=0.6\linewidth]{compgenomrReloaded_files/figure-latex/mogeneExpressionHeatmap-1} 

}

\caption{Heatmap of gene expression data for colorectal cancers}\label{fig:mogeneExpressionHeatmap}
\end{figure}

Next up is mutation data. We are going to again make heatmap with subtype annotation.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{pheatmap}\OperatorTok{::}\KeywordTok{pheatmap}\NormalTok{(x2,}
                   \DataTypeTok{annotation_col =}\NormalTok{ anno_col,}
                   \DataTypeTok{show_colnames =} \OtherTok{FALSE}\NormalTok{,}
                   \DataTypeTok{show_rownames =} \OtherTok{FALSE}\NormalTok{,}
                   \DataTypeTok{main=}\StringTok{"Mutation data"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/momutationsHeatmap-1} 

}

\caption{Heatmap of mutation data for colorectal cancers}\label{fig:momutationsHeatmap}
\end{figure}

Next, we are looking into copy number variation data and again tyring to see if clustered samples are in concordance with subtypes.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{pheatmap}\OperatorTok{::}\KeywordTok{pheatmap}\NormalTok{(x3,}
                   \DataTypeTok{annotation_col =}\NormalTok{ anno_col,}
                   \DataTypeTok{show_colnames =} \OtherTok{FALSE}\NormalTok{,}
                   \DataTypeTok{show_rownames =} \OtherTok{FALSE}\NormalTok{,}
                   \DataTypeTok{main=}\StringTok{"Copy number data"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/moCNVHeatmap-1} 

}

\caption{Heatmap of copy number variation data, colorectal cancers}\label{fig:moCNVHeatmap}
\end{figure}

\hypertarget{latent-variable-models-for-multi-omics-integration}{%
\section{Latent variable models for multi-omics integration}\label{latent-variable-models-for-multi-omics-integration}}

Unupervised multi-omics integration methods are methods that look for patterns within and across data types, in a label-agnostic fashion, i.e.~without knowledge of the identity or label of the analyzed samples (e.g.~cell type, tumor/normal). This chapter focuses on latent variable models, a form of dimensionality reduction technique (see Chapter 4, Unsupervised Learning). Latent variable models make an assumption that the high dimensional data we observe (e.g.~counts of tens of thousands of mRNA molecules) arrise from a lower dimension description. The variables in that lower dimensional description are termed \emph{Latent Variables}, as they are believed to be latent in the data, but not directly observable through experimentation. Therefore, there is a need for methods to infer the latent variables from the data. For instance, (see Chapter 8, RNA-seq analysis) the relative abundance of different mRNA molecules in a cell is largely determined by the cell type. There are other experiments which may be used to discern the cell type of cells (e.g.~looking at them under a microscope), but an RNA-seq experiment does not, directly, reveal whether the analyzed sample was taken from one organ or another. A latent variable model would set the cell type as a latent variable, and the observable abundance of mRNA molecules to be dependent on the value of the latent variable (e.g.~if the latent variable is ``Regulatory T-cell'', we would expect to find high expression of CD4, FOXP3, and CD25).

\hypertarget{matrix-factorization-methods-for-unsupervised-multi-omics-data-integration}{%
\section{Matrix factorization methods for unsupervised multi-omics data integration}\label{matrix-factorization-methods-for-unsupervised-multi-omics-data-integration}}

Matrix factorization techniques attempt to infer a set of latent variables from the data by finding factors of a data matrix. Principal Component Analysis (Chapter 4) is a form of matrix factorization which finds factors based on the covariance structure of the data. Generally, matrix factorization methods may be formulated as

\[
X = WH,
\]
where \(X\) is the \emph{data matrix}, \([M \times N]\) where \(M\) is the number of features (typically genes), and \(N\) is the number of samples. \(W\) is an \([M \times K]\) \emph{factors} matrix, and \(H\) is the \([K \times N]\) \emph{latent variable coefficient matrix}. Tying this back to PCA, where \(X = U \Sigma V^T\), we may formulate the factorization in the same terms by setting \(W=U\Sigma\) and \(H=V^T\). If \(K=rank(X)\), this factorization is lossless, i.e.~\(X=WH\). However if we choose \(K<rank(X)\), the factorization is lossy, i.e.~\(X \approx WH\). In that case, matrix factorization methods normally opt to minimize the error

\[
min~\|X-WH\|,
\]

which may be further subject to some constraints or regularization terms. As we normally seek a latent variable model with a considerably lower dimensionality than \(X\), this is the more common case.

\begin{figure}

{\centering \includegraphics[width=1\linewidth]{images/matrix_factorization} 

}

\caption{General matrix factorization framework. The data matrix on the left hand side is decomposed into factors on the right hand side. The equality may be an approximation as some matrix factorization methods are lossless (exact), while others are an approximation.}\label{fig:momatrixFactorization}
\end{figure}

In Figure \ref{fig:momatrixFactorization}, the \(5 \times 4\) data matrix \(X\) is decomposed to a 2-dimensional latent variable model.

\hypertarget{multiple-factor-analysis}{%
\subsection{Multiple Factor Analysis}\label{multiple-factor-analysis}}

Multiple factor analysis is a natural starting point for a discussion about matrix factorization methods for intergrating multiple data types. It is a straightforward extension of PCA into the domain of multiple data types \footnote{When dealing with categorical variables, MFA uses MCA (Multipe Correspondence Analysis). This is less relevant to biological data analysis and will not be discussed here}.

Consider the figure below, a naive extension of PCA to a multi-omics context:

\begin{figure}

{\centering \includegraphics[width=1\linewidth]{images/mfa} 

}

\caption{A naive extension of PCA to multi-omics; data matrices from different platforms are stacked, before applying PCA.}\label{fig:moMFA}
\end{figure}

Formally, we have
\[
X = \begin{bmatrix}
           X_{1} \\
           X_{2} \\
           \vdots \\
           X_{L}
         \end{bmatrix} = WH,
\]
a joint decomposition of the different data matrices (\(X_i\)) into the factor matrix \(W\) and the latent variable matrix \(H\). This way, we can leverage the ability of PCA to find the highest variance decomposition of the data, when the data consists of different omics types. As a reminder, PCA finds the linear combinations of the features which, when the data is projected onto them, preserve the most variance of any \(K\) dimensional space. But because measurements from different experients have different scales, they will also have variance (and co-variance) at different scales.

Multiple Factor Analysis addresses this issue and achieves balance among the data types by normalizing each of the data types, before stacking them and passing them on to PCA. Formally, MFA is given by

\[
X = \begin{bmatrix}
           X_{1} / \lambda^{(1)}_1 \\
           X_{2} / \lambda^{(2)}_1 \\
           \vdots \\
           X_{L} / \lambda^{(L)}_1
         \end{bmatrix} = WH,
\]
where \(\lambda^{(i)}_1\) is the first eigenvalue of the principal component decomposition of \(X_i\).

\hypertarget{mfa-in-r}{%
\subsubsection{MFA in R}\label{mfa-in-r}}

MFA is available through the CRAN package \texttt{FactoMineR}. The code snippet below shows how to run it:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{r.mfa <-}\StringTok{ }\NormalTok{FactoMineR}\OperatorTok{::}\KeywordTok{MFA}\NormalTok{(}\KeywordTok{t}\NormalTok{(}\KeywordTok{rbind}\NormalTok{(x1,x2,x3)), }\KeywordTok{c}\NormalTok{(}\KeywordTok{dim}\NormalTok{(x1)[}\DecValTok{1}\NormalTok{], }\KeywordTok{dim}\NormalTok{(x2)[}\DecValTok{1}\NormalTok{], }\KeywordTok{dim}\NormalTok{(x3)[}\DecValTok{1}\NormalTok{]), }\DataTypeTok{graph=}\OtherTok{FALSE}\NormalTok{)}
\NormalTok{mfa.h <-}\StringTok{ }\NormalTok{r.mfa}\OperatorTok{$}\NormalTok{global.pca}\OperatorTok{$}\NormalTok{ind}\OperatorTok{$}\NormalTok{coord}
\NormalTok{mfa.w <-}\StringTok{ }\NormalTok{r.mfa}\OperatorTok{$}\NormalTok{quanti.var}\OperatorTok{$}\NormalTok{coord}

\NormalTok{mfa_df <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(mfa.h)}
\NormalTok{mfa_df}\OperatorTok{$}\NormalTok{subtype <-}\StringTok{ }\KeywordTok{factor}\NormalTok{(covariates[}\KeywordTok{rownames}\NormalTok{(mfa_df),]}\OperatorTok{$}\NormalTok{cms_label)}
\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{ggplot}\NormalTok{(mfa_df, ggplot2}\OperatorTok{::}\KeywordTok{aes}\NormalTok{(}\DataTypeTok{x=}\NormalTok{Dim}\FloatTok{.1}\NormalTok{, }\DataTypeTok{y=}\NormalTok{Dim}\FloatTok{.2}\NormalTok{, }\DataTypeTok{color=}\NormalTok{subtype)) }\OperatorTok{+}\StringTok{ }\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{geom_point}\NormalTok{() }\OperatorTok{+}\StringTok{ }\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{ggtitle}\NormalTok{(}\StringTok{"Scatter plot of MFA"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/momultiomicsFactoMineR-1} 

}

\caption{MFA of multi-omics data creates seperability between the tumpr sub-types CMS1 and CMS3. }\label{fig:momultiomicsFactoMineR1}
\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{pheatmap}\OperatorTok{::}\KeywordTok{pheatmap}\NormalTok{(}\KeywordTok{t}\NormalTok{(mfa.h)[}\DecValTok{1}\OperatorTok{:}\DecValTok{2}\NormalTok{,], }\DataTypeTok{annotation_col =}\NormalTok{ anno_col, }\DataTypeTok{show_colnames =} \OtherTok{FALSE}\NormalTok{,}
         \DataTypeTok{main=}\StringTok{"MFA for multi-omics integration"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/momultiomicsFactoMineR-2} 

}

\caption{MFA of multi-omics data creates seperability between the tumpr sub-types CMS1 and CMS3. }\label{fig:momultiomicsFactoMineR2}
\end{figure}

\hypertarget{further-reading}{%
\subsubsection{Further Reading}\label{further-reading}}

\begin{itemize}
\tightlist
\item
  \href{http://factominer.free.fr/}{The FactoMineR website}
\item
  \href{https://en.wikipedia.org/wiki/Multiple_factor_analysis}{The Wikipedia page}
\end{itemize}

\hypertarget{joint-non-negative-matrix-factorization}{%
\subsection{Joint Non-negative Matrix Factorization}\label{joint-non-negative-matrix-factorization}}

NMF (Non-negative Matrix Factorization) is an algorithm from 2000 that seeks to find a non-negative additive decomposition for a non-negative data matrix. It takes the familiar form \(X \approx WH\), with \(X \ge 0\), \(W \ge 0\), and \(H \ge 0\). The non-negative constraints make a lossless decomposition (i.e.~\(X=WH\)) generally impossible. Hence, NMF attempts to find a solution which minimizes the Frobenius norm of the reconstruction:

\[
min~\|X-WH\|_F \\
W \ge 0, \\
H \ge 0,
\]

where the Frobenius norm \(\|\cdot\|_F\) is the matrix equivalent of the Euclidean distance:

\[
\|X\|_F = \sqrt{\sum_i\sum_jx_{ij}^2}.
\]

This is typically solved for \(W\) and \(H\) using random initializations followed by iterations of a multiplicative update rule:

\begin{align}
    W_{t+1} &= W_t^T \frac{XH_t^T}{XH_tH_t^T} \\
    H_{t+1} &= H_t \frac{W_t^TX}{W^T_tW_tX}.
\end{align}

Since this algorithm is guaranteed only to converge to a local minima, it is typically run several times with random initializations, and the best result is kept.

In the multi-omics context, we will, as in the MFA case, wish to find a decomposition for an integrated data matrix of the form

\[
X = \begin{bmatrix}
    X_{1} \\
    X_{2} \\
    \vdots \\
    X_{L}
\end{bmatrix},
\]

with \(X_i\)s denoting data from different omics platforms.

As NMF seeks to minimize the reconstruction error \(\|X-WH\|_F\), some care needs to be taken with regards to data normalization. Different omics platforms may produce data with different scales (i.e.~real-valued gene expression quantification, binary mutation data, etc.), and so will have different baseline Frobenius norms. To address this, when doing Joint NMF, we first feature-normalize each data matrix, and then normalize by the Frobenius norm of the data matrix. Formally, we run NMF on

\[
X = \begin{bmatrix}
    X_{1}^N / \alpha_1 \\
    X_{2}^N / \alpha_2 \\
    \vdots \\
    X_{L}^N / \alpha_L
\end{bmatrix},
\]

where \(X_i^N\) is the feature-normalized data matrix \(X_i^N = \frac{x^{ij}}{\sum_jx^{ij}}\), and \(\alpha_i = \|X_{i}^N\|_F\).

Another consideration with NMF is the non-negativity constraint. Different omics data types may have negative values, for instance, copy-number variations (CNVs) may be positive, indicating gains, or negative, indicating losses. In order to turn such data into a non-negative form, we will split each feature into two features, one new feature holding all the non-negative values of the original feature, and another feature holding the absolute value of the negative ones:

TODO table of CNVs before and after splitting column.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{knitr}\OperatorTok{::}\KeywordTok{kable}\NormalTok{(cnvs_with_neg, }\DataTypeTok{caption=}\StringTok{"Example copy number data. Data can be both positive (amplified regions) or negative (deleted regions)."}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\textbackslash{}begin\{table\}{[}t{]}

\textbackslash{}caption\{(\#tab:mocnvsplitcol\_show1)Example copy number data. Data can be both positive (amplified regions) or negative (deleted regions).\}
\centering

\begin{tabular}{l|r|r}
\hline
  & seg1 & seg2\\
\hline
samp1 & 1 & 0\\
\hline
samp2 & 2 & 1\\
\hline
samp3 & 1 & -2\\
\hline
\end{tabular}

\textbackslash{}end\{table\}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{knitr}\OperatorTok{::}\KeywordTok{kable}\NormalTok{(cnvs_split_pos, }\DataTypeTok{caption=}\StringTok{"Example copy number data after splitting each column to a column representing copy number gains (+) and a column representing deletions (-). This data matrix is non-negative, and thus suitable for NMF algorithms."}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\textbackslash{}begin\{table\}{[}t{]}

\textbackslash{}caption\{(\#tab:mocnvsplitcol\_show2)Example copy number data after splitting each column to a column representing copy number gains (+) and a column representing deletions (-). This data matrix is non-negative, and thus suitable for NMF algorithms.\}
\centering

\begin{tabular}{l|r|r|r|r}
\hline
  & seg1+ & seg1- & seg2+ & seg2-\\
\hline
samp1 & 1 & 0 & 0 & 0\\
\hline
samp2 & 2 & 0 & 1 & 0\\
\hline
samp3 & 1 & 0 & 0 & 2\\
\hline
\end{tabular}

\textbackslash{}end\{table\}

\hypertarget{nmf-in-r}{%
\subsubsection{NMF in R}\label{nmf-in-r}}

Many NMF algorithms are available through the CRAN package \texttt{NMF}:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Feature-normalize the data}
\NormalTok{x1.featnorm <-}\StringTok{ }\NormalTok{x1 }\OperatorTok{/}\StringTok{ }\KeywordTok{rowSums}\NormalTok{(x1)}
\NormalTok{x2.featnorm <-}\StringTok{ }\NormalTok{x2 }\OperatorTok{/}\StringTok{ }\KeywordTok{rowSums}\NormalTok{(x2)}
\NormalTok{x3.featnorm <-}\StringTok{ }\NormalTok{x3 }\OperatorTok{/}\StringTok{ }\KeywordTok{rowSums}\NormalTok{(x3)}

\CommentTok{# Normalize by each omics type's frobenius norm}
\NormalTok{x1.featnorm.frobnorm <-}\StringTok{ }\NormalTok{x1.featnorm }\OperatorTok{/}\StringTok{ }\KeywordTok{norm}\NormalTok{(}\KeywordTok{as.matrix}\NormalTok{(x1.featnorm), }\DataTypeTok{type=}\StringTok{"F"}\NormalTok{)}
\NormalTok{x2.featnorm.frobnorm <-}\StringTok{ }\NormalTok{x2.featnorm }\OperatorTok{/}\StringTok{ }\KeywordTok{norm}\NormalTok{(}\KeywordTok{as.matrix}\NormalTok{(x2.featnorm), }\DataTypeTok{type=}\StringTok{"F"}\NormalTok{)}
\NormalTok{x3.featnorm.frobnorm <-}\StringTok{ }\NormalTok{x3.featnorm }\OperatorTok{/}\StringTok{ }\KeywordTok{norm}\NormalTok{(}\KeywordTok{as.matrix}\NormalTok{(x3.featnorm), }\DataTypeTok{type=}\StringTok{"F"}\NormalTok{)}

\CommentTok{# Split the features of the CNV matrix into two non-negative features each}
\NormalTok{x3.featnorm.frobnorm.nonneg <-}\StringTok{ }\KeywordTok{t}\NormalTok{(}\KeywordTok{split_neg_columns}\NormalTok{(}\KeywordTok{t}\NormalTok{(x3.featnorm.frobnorm)))}

\CommentTok{# this hack is here because of an NMF / DelayedArray incompatibility}
\CommentTok{# https://github.com/renozao/NMF/issues/85}
\KeywordTok{detach}\NormalTok{(}\StringTok{"package:Rqc"}\NormalTok{)}
\KeywordTok{detach}\NormalTok{(}\StringTok{"package:RUVSeq"}\NormalTok{)}
\KeywordTok{detach}\NormalTok{(}\StringTok{"package:EDASeq"}\NormalTok{)}
\KeywordTok{detach}\NormalTok{(}\StringTok{"package:ShortRead"}\NormalTok{)}
\KeywordTok{detach}\NormalTok{(}\StringTok{"package:GenomicAlignments"}\NormalTok{)}
\KeywordTok{detach}\NormalTok{(}\StringTok{"package:DESeq2"}\NormalTok{)}
\KeywordTok{detach}\NormalTok{(}\StringTok{"package:SummarizedExperiment"}\NormalTok{)}
\KeywordTok{detach}\NormalTok{(}\StringTok{"package:DelayedArray"}\NormalTok{)}
\CommentTok{# end hack}

\KeywordTok{require}\NormalTok{(NMF)}
\NormalTok{r.nmf <-}\StringTok{ }\KeywordTok{nmf}\NormalTok{(}\KeywordTok{t}\NormalTok{(}\KeywordTok{rbind}\NormalTok{(x1.featnorm.frobnorm, x2.featnorm.frobnorm, x3.featnorm.frobnorm.nonneg)), }\DecValTok{2}\NormalTok{, }\DataTypeTok{method=}\StringTok{'Frobenius'}\NormalTok{)}
\NormalTok{nmf.h <-}\StringTok{ }\NormalTok{NMF}\OperatorTok{::}\KeywordTok{basis}\NormalTok{(r.nmf)}
\NormalTok{nmf.w <-}\StringTok{ }\NormalTok{NMF}\OperatorTok{::}\KeywordTok{coef}\NormalTok{(r.nmf)}

\NormalTok{nmf_df <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(nmf.h)}
\KeywordTok{colnames}\NormalTok{(nmf_df) <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"dim1"}\NormalTok{, }\StringTok{"dim2"}\NormalTok{)}
\NormalTok{nmf_df}\OperatorTok{$}\NormalTok{subtype <-}\StringTok{ }\KeywordTok{factor}\NormalTok{(covariates[}\KeywordTok{rownames}\NormalTok{(nmf_df),]}\OperatorTok{$}\NormalTok{cms_label)}
\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{ggplot}\NormalTok{(nmf_df, ggplot2}\OperatorTok{::}\KeywordTok{aes}\NormalTok{(}\DataTypeTok{x=}\NormalTok{dim1, }\DataTypeTok{y=}\NormalTok{dim2, }\DataTypeTok{color=}\NormalTok{subtype)) }\OperatorTok{+}\StringTok{ }\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{geom_point}\NormalTok{() }\OperatorTok{+}\StringTok{ }\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{ggtitle}\NormalTok{(}\StringTok{"Scatterplot of 2-component NMF for multi-omics integration"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/monmfFeatureNormalization-1} 

}

\caption{NMF creates a disentangled representation of the data using two components which allow for separation between tumor sub-types CMS1 and CMS3 based on NMF factors learned form multi-omics data.}\label{fig:monmfFeatureNormalization1}
\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{pheatmap}\OperatorTok{::}\KeywordTok{pheatmap}\NormalTok{(}\KeywordTok{t}\NormalTok{(nmf_df[,}\DecValTok{1}\OperatorTok{:}\DecValTok{2}\NormalTok{]), }\DataTypeTok{annotation_col =}\NormalTok{ anno_col, }\DataTypeTok{show_colnames=}\OtherTok{FALSE}\NormalTok{, }\DataTypeTok{main=}\StringTok{"Heatmap of 2-component NMF"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/monmfFeatureNormalization-2} 

}

\caption{NMF creates a disentangled representation of the data using two components which allow for separation between tumor sub-types CMS1 and CMS3 based on NMF factors learned form multi-omics data.}\label{fig:monmfFeatureNormalization2}
\end{figure}

\hypertarget{further-reading-1}{%
\subsubsection{Further reading}\label{further-reading-1}}

\begin{itemize}
\tightlist
\item
  Zhang S., Li Q., Liu J., Zhou X. J. (2011). A novel computational framework for simultaneous integration of multiple types of genomic data to identify microRNA-gene regulatory modules. Bioinformatics 27, i401--i409. 10.1093/bioinformatics/btr206 \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3117336/}{PMC free article}
\item
  Zhang S., Liu C.-C., Li W., Shen H., Laird P. W., Zhou X. J. (2012). Discovery of multi-dimensional modules by integrative analysis of cancer genomic data. Nucleic Acids Res. 40, 9379--9391. 10.1093/nar/gks725 \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3479191/}{PMC free article}
\end{itemize}

\hypertarget{icluster}{%
\subsection{iCluster}\label{icluster}}

iCluster takes a Bayesian approach to the latent variable model. In Bayesian statistics, we infer distributions over model parameters, rather than finding a single maximum-likelihood parameter estimate. In iCluster, we model the data as

\[
X_{(i)} = W_{(i)}Z + \epsilon_i,
\]

where \(X_{(i)}\) is a data matrix from a single omics platform, \(W_{(i)}\) are model parameters, \(Z\) is a latent variable matrix, and is shared between the different omics platforms, and \(\epsilon_i\) is a ``noise'' random variable, \(\epsilon \sim N(0,\Psi)\), with \(\Psi = diag(\psi_1,\dots \psi_M)\) is a diagonal covariance matrix.

\begin{figure}

{\centering \includegraphics[width=1\linewidth]{images/icluster} 

}

\caption{Sketch of iCluster model. Each omics datatype is decomposed to a coefficient matrix and a shared latent variable matrix, plus noise.}\label{fig:moiCluster}
\end{figure}

Note that with this construction, the omics measurements \(X\) are expected to be the same for samples with the same latent variable representation, up to gaussian noise. Further, we assume a Gaussian prior distribution on the latent variables \(Z \sim N(0,I)\), which means we assume \(X_{(i)} \sim N \big( 0,W_{(i)} W_{(i)}^T + \Psi_{(i)} \big)\). In order to find suitable values for \(W\), \(Z\), and \(\Psi\), we can write down the multivariate normal log-likelihood function and optimize it. For a multivariate normal distribution with mean \(0\) and covariance \(\Sigma\), the log-likelihood function is given by

\[
\ell = -\frac{1}{2} \bigg( \ln (|\Sigma|) + X^T \Sigma^{-1} X + k\ln (2 \pi) \bigg)
\]

(this is simply the log of the Probability Density Function of a multivariate gaussian). For the multi-omics iCluster case, we have \(X=\big( X_{(1)}, \dots, X_{(L)} \big)^T\), \(W = \big( W_{(1)}, \dots, W_{(L)} \big)^T\), where \(X\) is a multivariate normal with \(0\)-mean and \(\Sigma = W W^T + \Psi\) covariance. Hence, the log-likelihood function for the iCluster modle is given by:

\[
\ell_{iC}(W,\Sigma) = -\frac{1}{2} \bigg( \sum_{i=1}^L \ln (|\Sigma|) + X^T\Sigma^{-1}X + p_i \ln (2 \pi) \bigg)
\]

where \(p_i\) is the number of features in omics data type \(i\). Because this model has more parameters than we typically have samples, we need to push the model to use fewer parameters than it has at its disposal, by using regularization. iCluster uses Lasso regularization, which is a direct penalty on the absolute value of the parameters. I.e., instead of optimizing \(\ell_{iC}(W,\Sigma)\), we will optimize the regularized log-likelihood:

\[
\ell = \ell_{iC}(W,\Sigma) - \lambda\|W\|_1.
\]

The parameter \(\lambda\) acts as a dial to weigh the tradeoff between better model fits (higher log-likelihood) and a sparser model, with more \(w_{ij}\)s set to \(0\), which gives models which generalize better and are more interpretable.

In order to solve this problem, iCluster employs the Expectation Maximization (EM) algorithm. The full details are beyond the scope of this textbook. We will introduce a short sketch instead. The intuition behind the EM algorithm is a more general case of the k-means clustering algorithm (Chapter 4).

\hypertarget{em-algorithm-sketch}{%
\paragraph{EM algorithm sketch}\label{em-algorithm-sketch}}

\begin{itemize}
\tightlist
\item
  Initialize \(W\) and \(\Psi\)
\item
  \textbf{Until convergence of \(W\), \(\Psi\)}

  \begin{itemize}
  \tightlist
  \item
    E-step: calculate the expected value of \(Z\) given the current estimates of \(W\) and \(\Psi\) and the data \(X\)
  \item
    M-step: calculate maximum likelihood estimates for the parameters \(W\) and \(\Psi\) based on the current estimate of \(Z\) and the data \(X\).
  \end{itemize}
\end{itemize}

\hypertarget{icluster-1}{%
\subsubsection{iCluster+}\label{icluster-1}}

iCluster+ is an extension of the iCluster framework, which allows for omics types to arrise from other distributions than a gaussian. While normal distributions are a good assumption for log-transformed, centered gene expression data, it is a poor model for binary mutations data, or for copy number variation data, which can typically take the values \((-2, 1, 0, 1, 2)\) for for heterozygous / monozygous deletions or amplifications. iCluster+ allows the different \(X\)s to have different distributions:

\begin{itemize}
\tightlist
\item
  for binary mutations, \(X\) is drawn from a multivariate binomial
\item
  for normal, continuous data, \(X\) is drawn from a multivariate gaussian
\item
  for copy number variations, \(X\) is drawn from a multinomial
\item
  for count data, \(X\) is drawn from a poisson.
\end{itemize}

In that way, iCluster+ allows us to explicitly model our assumptions about the distributions of our different omics data types, and leverage the strengths of Bayesian inferrence.

Both iCluster and iCluster+ make use of sophisticated Bayesian inference algorithms (EM for iCluster, Metropolis-Hastings MCMC for iCluster+), which means they do not scale up trivially. Therefore, it is recommended to filter down the features to a manageable size before inputing data to the algorithm. The exact size of ``manageable'' data depends on your hardware, but a rule of thumb is that dimensions in the thousands are ok, but in the tens of thousands might be too slow.

\hypertarget{running-icluster}{%
\subsubsection{running iCluster}\label{running-icluster}}

iCluster+ is available through the BioConductor package \texttt{iClusterPlus}:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{r.icluster <-}\StringTok{ }\NormalTok{iClusterPlus}\OperatorTok{::}\KeywordTok{iClusterPlus}\NormalTok{(}\KeywordTok{t}\NormalTok{(x1), }\KeywordTok{t}\NormalTok{(x2), }\KeywordTok{t}\NormalTok{(x3), }\DataTypeTok{type=}\KeywordTok{c}\NormalTok{(}\StringTok{"gaussian"}\NormalTok{, }\StringTok{"binomial"}\NormalTok{, }\StringTok{"multinomial"}\NormalTok{), }\DataTypeTok{K=}\DecValTok{2}\NormalTok{, }\DataTypeTok{alpha=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{,}\DecValTok{1}\NormalTok{), }\DataTypeTok{lambda=}\KeywordTok{c}\NormalTok{(.}\DecValTok{03}\NormalTok{,.}\DecValTok{03}\NormalTok{,.}\DecValTok{03}\NormalTok{))}

\NormalTok{icluster.z <-}\StringTok{ }\NormalTok{r.icluster}\OperatorTok{$}\NormalTok{meanZ}
\NormalTok{icluster.ws <-}\StringTok{ }\NormalTok{r.icluster}\OperatorTok{$}\NormalTok{beta}
\NormalTok{icluster.clusters <-}\StringTok{ }\NormalTok{r.icluster}\OperatorTok{$}\NormalTok{clusters}

\NormalTok{icp_df <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(icluster.z)}
\KeywordTok{colnames}\NormalTok{(icp_df) <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"dim1"}\NormalTok{, }\StringTok{"dim2"}\NormalTok{)}
\KeywordTok{rownames}\NormalTok{(icp_df) <-}\StringTok{ }\KeywordTok{colnames}\NormalTok{(x1)}
\NormalTok{icp_df}\OperatorTok{$}\NormalTok{subtype <-}\StringTok{ }\KeywordTok{factor}\NormalTok{(covariates[}\KeywordTok{rownames}\NormalTok{(icp_df),]}\OperatorTok{$}\NormalTok{cms_label)}
\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{ggplot}\NormalTok{(icp_df, ggplot2}\OperatorTok{::}\KeywordTok{aes}\NormalTok{(}\DataTypeTok{x=}\NormalTok{dim1, }\DataTypeTok{y=}\NormalTok{dim2, }\DataTypeTok{color=}\NormalTok{subtype)) }\OperatorTok{+}\StringTok{ }\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{geom_point}\NormalTok{() }\OperatorTok{+}\StringTok{ }\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{ggtitle}\NormalTok{(}\StringTok{"Scatter plot of iCluster+ factors"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/momultiOmicsiclusterplus-1} 

}

\caption{iCluster+ learns factors which allow tumor sub-types CMS1 and CMS3 to be discriminated.}\label{fig:momultiOmicsiclusterplus1}
\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{pheatmap}\OperatorTok{::}\KeywordTok{pheatmap}\NormalTok{(}\KeywordTok{t}\NormalTok{(icp_df[,}\DecValTok{1}\OperatorTok{:}\DecValTok{2}\NormalTok{]), }\DataTypeTok{annotation_col =}\NormalTok{ anno_col, }\DataTypeTok{show_colnames =} \OtherTok{FALSE}\NormalTok{, }\DataTypeTok{main=}\StringTok{"Heatmap of iCluster+ factors"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/momultiOmicsiclusterplus-2} 

}

\caption{iCluster+ learns factors which allow tumor sub-types CMS1 and CMS3 to be discriminated.}\label{fig:momultiOmicsiclusterplus2}
\end{figure}

\hypertarget{further-reading-2}{%
\subsubsection{Further reading}\label{further-reading-2}}

\begin{itemize}
\tightlist
\item
  Shen R., Olshen A. B., Ladanyi M. (2009). Integrative clustering of multiple genomic data types using a joint latent variable model with application to breast and lung cancer subtype analysis. Bioinformatics 25, 2906--2912. 10.1093/bioinformatics/btp543 \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2800366/}{PMC free article}
\item
  Shen R., Mo Q., Schultz N., Seshan V. E., Olshen A. B., Huse J., et al.~. (2012). Integrative subtype discovery in glioblastoma using iCluster. PLoS ONE 7:e35236. 10.1371/journal.pone.0035236 \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3335101/}{PMC free article}
\item
  \href{http://www-stat.stanford.edu/\%7Etibs/lasso/lasso.pdf}{Tibshirani, R. (1996). Regression shrinkage and selection via the lasso. J. Royal. Statist. Soc B., Vol. 58, No.~1, pages 267-288)}
\item
  \href{www.jstor.org/stable/2984875}{Dempster, A. P., et al.~``Maximum Likelihood from Incomplete Data via the EM Algorithm.'' Journal of the Royal Statistical Society. Series B (Methodological), vol.~39, no. 1, 1977, pp.~1--38. JSTOR, JSTOR}
\item
  \href{https://www.jstor.org/stable/2334940}{Hastings, W.K. (1970). ``Monte Carlo Sampling Methods Using Markov Chains and Their Applications''. Biometrika. 57 (1): 97--109. doi:10.1093/biomet/57.1.97}
\end{itemize}

\hypertarget{clustering-using-latent-factors}{%
\section{Clustering using latent factors}\label{clustering-using-latent-factors}}

A common analysis in biological investigations is clustering. This is often interesting in cancer studies as one hopes to find groups of tumors (clusters) which behave similarly, i.e.~have similar risks and/or respond to the same drugs. PCA is a common step in clustering analyses, and so it is easy to see how the latent variable models above may all be a useful pre-processing step before clustering. In the examples below, we will use the latent variables inferred by the algorithms in the previous section on the set of Colorectal cancer tumors from the TCGA.

\hypertarget{one-hot-clustering}{%
\subsection{One-hot clustering}\label{one-hot-clustering}}

A specific clustering method for NMF data is to assume each sample is driven by one component, i.e.~that the number of clusters \(K\) is the same as the number of latent variables in the model and that each sample may be associated to one of those components. We assign each sample a cluster label based on the latent variable which affects it the most. The figure above (Heatmap of 2-component NMF) shows the latent variable values for the two latent variables, for the 72 tumors, obtained by Joint NMF.

The two rows are the two latent variables, and the columns are the 72 tumors. We can observe that most tumors are indeed driven mainly by one of the factors, and not a combination of the two. We can use this to assign each tumor a cluster label based on its dominant factor:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{nmf.clusters <-}\StringTok{ }\KeywordTok{max.col}\NormalTok{(nmf.h)}
\KeywordTok{names}\NormalTok{(nmf.clusters) <-}\StringTok{ }\KeywordTok{rownames}\NormalTok{(nmf.h)}
\NormalTok{anno_nmf_cl <-}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{nmf.cluster=}\KeywordTok{factor}\NormalTok{(nmf.clusters), }\DataTypeTok{cms.subtype=}\KeywordTok{factor}\NormalTok{(covariates[}\KeywordTok{rownames}\NormalTok{(nmf.h),]}\OperatorTok{$}\NormalTok{cms_label))}
\NormalTok{pheatmap}\OperatorTok{::}\KeywordTok{pheatmap}\NormalTok{(}\KeywordTok{t}\NormalTok{(nmf.h[}\KeywordTok{order}\NormalTok{(nmf.clusters),]), }\DataTypeTok{cluster_cols=}\OtherTok{FALSE}\NormalTok{, }\DataTypeTok{cluster_rows=}\OtherTok{FALSE}\NormalTok{,}
         \DataTypeTok{annotation_col =}\NormalTok{ anno_nmf_cl,}
         \DataTypeTok{show_colnames =} \OtherTok{FALSE}\NormalTok{,}
         \DataTypeTok{main=}\StringTok{"Joint NMF factors with clusters and molecular subtypes"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/moNMFClustering-1} 

}

\caption{Joint NMF factors with clusters, and molecular sub-types. One-hot clustering assigns one cluser per dimension, where each sample is assigned a cluster based on its dominant component. The clusters largely recapitulate the CMS sub-types.}\label{fig:moNMFClustering}
\end{figure}

We see that using one-hot clustering with Joint NMF, we were able to find two clusters in the data which correspond fairly well with the molecular subtype of the tumors.

The one-hot clustering method does not lend itself very well to the other methods discussed above, i.e.~iCluster and MFA. The latent variables produced by those other methods may be negative, and further, in the case of iCluster, are going to assume a multivariate gaussian shape. As such, it is not trivial to pick one ``dominant factor'' for them. For NMF variants, this is a very common way to assign clusters.

\hypertarget{k-means-clustering-1}{%
\subsection{K-means clustering}\label{k-means-clustering-1}}

K-means clustering was introduced in Chapter 4. K-means is a special case of the EM algorithm, and indeed iCluster was originally conceived as an extension of K-means from binary cluster assignments to real-valued latent variables. The iCluster algorithm, as it is so named, calls for application of K-means clustering on its latent variables, after the inference step.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{rownames}\NormalTok{(icluster.z) <-}\StringTok{ }\KeywordTok{rownames}\NormalTok{(covariates)}
\NormalTok{icluster.clusters <-}\StringTok{ }\KeywordTok{kmeans}\NormalTok{(icluster.z, }\DecValTok{2}\NormalTok{)}\OperatorTok{$}\NormalTok{cluster}
\KeywordTok{names}\NormalTok{(icluster.clusters) <-}\StringTok{ }\KeywordTok{rownames}\NormalTok{(icluster.z)}
\NormalTok{anno_icluster_cl <-}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{iCluster=}\KeywordTok{factor}\NormalTok{(icluster.clusters), }\DataTypeTok{cms.subtype=}\KeywordTok{factor}\NormalTok{(covariates}\OperatorTok{$}\NormalTok{cms_label))}
\NormalTok{pheatmap}\OperatorTok{::}\KeywordTok{pheatmap}\NormalTok{(}\KeywordTok{t}\NormalTok{(icluster.z[}\KeywordTok{order}\NormalTok{(icluster.clusters),]), }\DataTypeTok{cluster_cols=}\OtherTok{FALSE}\NormalTok{, }\DataTypeTok{cluster_rows=}\OtherTok{FALSE}\NormalTok{,}
         \DataTypeTok{show_colnames =} \OtherTok{FALSE}\NormalTok{,}
         \DataTypeTok{annotation_col =}\NormalTok{ anno_icluster_cl,}
         \DataTypeTok{main=}\StringTok{"iCluster factors with clusters and molecular subtypes"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/moiClusterHeatmap-1} 

}

\caption{K-means clustering on iCluster+ factors largely recapitulates the CMS sub-types.}\label{fig:moiClusterHeatmap}
\end{figure}

This demonstrates the ability of iClusterPlus to find clusters which correspond to molecular subtypes, based on multi-omics data.

\hypertarget{biological-interpretation-of-latent-factors}{%
\section{Biological interpretation of latent factors}\label{biological-interpretation-of-latent-factors}}

\hypertarget{inspection-of-feature-weights-in-loading-vectors}{%
\subsection{Inspection of feature weights in loading vectors}\label{inspection-of-feature-weights-in-loading-vectors}}

The most straightforward way to go about interpreting the latent factors in a biological context, is to look at the coefficients which are associated with them. The latent variable models introduced above all take the linear form \(X \approx WH\), where \(W\) is a factor matrix, with coefficients tying each latent variable with each of the features in the \(L\) original multi-omics data matrices. By inspecting these coefficients, we can get a sense of which multi-omics features are co-regulated. The figure below shows the coefficients for the Joint NMF analysis above:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{nmfw <-}\StringTok{ }\KeywordTok{t}\NormalTok{(nmf.w)}

\NormalTok{data_anno <-}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{omics=}\KeywordTok{c}\NormalTok{(}\KeywordTok{rep}\NormalTok{(}\StringTok{'expression'}\NormalTok{,}\KeywordTok{dim}\NormalTok{(x1)[}\DecValTok{1}\NormalTok{]), }\KeywordTok{rep}\NormalTok{(}\StringTok{'mut'}\NormalTok{,}\KeywordTok{dim}\NormalTok{(x2)[}\DecValTok{1}\NormalTok{]), }\KeywordTok{rep}\NormalTok{(}\StringTok{'cnv'}\NormalTok{,}\KeywordTok{dim}\NormalTok{(x3.featnorm.frobnorm.nonneg)[}\DecValTok{1}\NormalTok{])))}
\KeywordTok{rownames}\NormalTok{(data_anno) <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\KeywordTok{rownames}\NormalTok{(x1), }\KeywordTok{paste0}\NormalTok{(}\StringTok{"mut:"}\NormalTok{, }\KeywordTok{rownames}\NormalTok{(x2)), }\KeywordTok{rownames}\NormalTok{(x3.featnorm.frobnorm.nonneg))}
\KeywordTok{rownames}\NormalTok{(nmfw) <-}\StringTok{ }\KeywordTok{rownames}\NormalTok{(data_anno)}

\NormalTok{pheatmap}\OperatorTok{::}\KeywordTok{pheatmap}\NormalTok{(nmfw,}
                   \DataTypeTok{cluster_cols =} \OtherTok{FALSE}\NormalTok{,}
                   \DataTypeTok{annotation_row =}\NormalTok{ data_anno,}
                   \DataTypeTok{main=}\StringTok{"NMF coefficients"}\NormalTok{,}
                   \DataTypeTok{clustering_distance_rows =} \StringTok{"manhattan"}\NormalTok{,}
                   \DataTypeTok{fontsize_row =} \DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/moNMFHeatmap-1} 

}

\caption{Heatmap showing the association of input features from multi-omics data (gene expression, copy number variation, and mutations), with JNMF factors. Gene expression features dominate both factors, but copy numbers and mutations mostly affect only one factor each.}\label{fig:moNMFHeatmap}
\end{figure}

Inspection of the factor coefficients in the heatmap above reveals that Joint NMF has found two nearly orthogonal non-negative factors. One is associated with high expression of the HOXC11, ZIC5, and XIRP1 genes, frequent mutations in the BRAF, PCDHGA6, and DNAH5 genes, as well as losses in the 18q12.2 and gains in 8p21.1 cytobands. The other factor is associated with high expression of the SOX1 gene, more frequent mutations in the APC, KRAS, and TP53 genes, and a weak association with some CNVs.

\hypertarget{disentangled-representations}{%
\subsubsection{Disentangled representations}\label{disentangled-representations}}

The property displayed above, where each feature is predominantly associated with only a single factor, is termed \emph{disentangledness}, i.e., it leads to \emph{disentangled} latent variable representations, as changing one input feature only affects a single latent variable. This property is very desirable as it greatly simplifies the biological interpretation of modules. Here, we have two modules with a set of co-occurring molecular signatures which merit deeper investigation into the mechanisms by which these different omics features are related. For this reason, NMF is widely used in computational biology today.

\hypertarget{making-sense-of-factors-using-enrichment-analysis}{%
\subsection{Making sense of factors using enrichment analysis}\label{making-sense-of-factors-using-enrichment-analysis}}

In order to investigate the oncogenic processes that drive the differences between tumors, we may draw upon biological prior knowledge by looking for overlaps between genes that drive certain tumors, and genes involved in familiar biological processes.

\hypertarget{enrichment-analysis}{%
\subsubsection{Enrichment analysis}\label{enrichment-analysis}}

The recent decades of genomics have uncovered many of the ways in which genes cooperate to perform biological functions in concert. This work has resulted in rich annotations of genes, groups of genes, and the different functions they carry out. Examples of such annotations include the Gene Ontology Consortium's \emph{GO terms} \citep[\citet{go_latest_paper}]{go_first_paper}, the \emph{Reactome pathways database} \citep{reactome_latent_paper}, and the \emph{Kyoto Encyclopaedia of Genes and Genomes} \citep{kegg_latest_paper}. These resources, as well as others, publish lists of so-called \emph{gene sets}, or \emph{pathways}, which are a set of genes which are known to operate together in some biological function, e.g.~protein synthesis, DNA mismatch repair, cellular adhesion, and many other functions. Gene set enrichment analysis is a method which looks for overlaps between genes which we have found to be of interest, e.g.~by them being implicated in a certain tumor type, and the a-priori gene sets discussed above.

In the context of making sense of latent factors, the question we will be asking is whether the genes which drive the value of a latent factor (the genes with the highest factor coefficients) also belong to any interesting annotated gene sets, and whether the overlap greater than we would expect by chance. If there are \(N\) genes in total, \(K\) of which belong to a gene set, the probability that \(k\) out of the \(n\) genes associated with a latent factor are also associated with a gene set is given by the hypergeometric distribution:

\[
P(k) = \frac{{\binom{K}{k}} - \binom{N-K}{n-k}}{\binom{N}{n}}.
\]

the \textbf{hypergeometric test} uses the hypergeometric distribution to assess the statistical significance of the presense of genes belonging to a gene set in the latent factor. The null hypothesis is that there is no relationship between genes in a gene set, and genes in a latent factor. When testing for over-representation of gene set genes in a latent factor, the P value from the hypergeometric test is the probability of getting \(k\) or more genes from a gene set in a latent factor

\[
p = \sum_{i=k}^K P(k=i).
\]

The hypergeometric enrichment test is also referred to as \emph{Fisher's one-sided exact test}. This way, we can determine if the genes associated with a factor significantly overlap (beyond chance) the genes involved in a biological process. Because we will typically be testing many gene sets, we will also need to apply multiple testing correction, such as Benjamini-Hochberg correction (see Chapter 3, multiple testing correction).

\hypertarget{example-in-r}{%
\subsubsection{Example in R}\label{example-in-r}}

In R, we can do this analysis using the \texttt{enrichR} package, which gives us access to many gene set libraries. In the example below, we'll find the genes associated with preferentially NMF factor 1 or NMF factor 2, by the contribution of those genes' expression values to the factor. Then, we'll use \texttt{enrichR} to query the Gene Ontology terms which might be overlapping. \texttt{enrichR} performs

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{genes.factor}\FloatTok{.1}\NormalTok{ <-}\StringTok{ }\KeywordTok{names}\NormalTok{(}\KeywordTok{which}\NormalTok{(nmfw[}\DecValTok{1}\OperatorTok{:}\KeywordTok{dim}\NormalTok{(x1)[}\DecValTok{1}\NormalTok{],}\DecValTok{1}\NormalTok{] }\OperatorTok{>}\StringTok{ }\NormalTok{nmfw[}\DecValTok{1}\OperatorTok{:}\KeywordTok{dim}\NormalTok{(x1)[}\DecValTok{1}\NormalTok{],}\DecValTok{2}\NormalTok{]))}
\NormalTok{genes.factor}\FloatTok{.2}\NormalTok{ <-}\StringTok{ }\KeywordTok{names}\NormalTok{(}\KeywordTok{which}\NormalTok{(nmfw[}\DecValTok{1}\OperatorTok{:}\KeywordTok{dim}\NormalTok{(x1)[}\DecValTok{1}\NormalTok{],}\DecValTok{1}\NormalTok{] }\OperatorTok{<}\StringTok{ }\NormalTok{nmfw[}\DecValTok{1}\OperatorTok{:}\KeywordTok{dim}\NormalTok{(x1)[}\DecValTok{1}\NormalTok{],}\DecValTok{2}\NormalTok{]))}

\NormalTok{go.factor}\FloatTok{.1}\NormalTok{ <-}\StringTok{ }\NormalTok{enrichR}\OperatorTok{::}\KeywordTok{enrichr}\NormalTok{(genes.factor}\FloatTok{.1}\NormalTok{,}
                                \DataTypeTok{databases =} \KeywordTok{c}\NormalTok{(}\StringTok{"GO_Biological_Process_2018"}\NormalTok{)}
\NormalTok{                                )}\OperatorTok{$}\NormalTok{GO_Biological_Process_}\DecValTok{2018}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Uploading data to Enrichr... Done.
##   Querying GO_Biological_Process_2018... Done.
## Parsing results... Done.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{go.factor}\FloatTok{.2}\NormalTok{ <-}\StringTok{ }\NormalTok{enrichR}\OperatorTok{::}\KeywordTok{enrichr}\NormalTok{(genes.factor}\FloatTok{.2}\NormalTok{,}
                                \DataTypeTok{databases =} \KeywordTok{c}\NormalTok{(}\StringTok{"GO_Biological_Process_2018"}\NormalTok{)}
\NormalTok{                                )}\OperatorTok{$}\NormalTok{GO_Biological_Process_}\DecValTok{2018}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Uploading data to Enrichr... Done.
##   Querying GO_Biological_Process_2018... Done.
## Parsing results... Done.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{ggplot}\NormalTok{(}\KeywordTok{head}\NormalTok{(go.factor}\FloatTok{.1}\NormalTok{), ggplot2}\OperatorTok{::}\KeywordTok{aes}\NormalTok{(}\DataTypeTok{y=}\OperatorTok{-}\KeywordTok{log10}\NormalTok{(Adjusted.P.value), }\DataTypeTok{x=}\NormalTok{Term)) }\OperatorTok{+}\StringTok{ }\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{geom_bar}\NormalTok{(}\DataTypeTok{stat=}\StringTok{'identity'}\NormalTok{) }\OperatorTok{+}\StringTok{ }\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{coord_flip}\NormalTok{() }\OperatorTok{+}\StringTok{ }\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{ggtitle}\NormalTok{(}\StringTok{"Top GO-terms}\CharTok{\textbackslash{}n}\StringTok{associated with}\CharTok{\textbackslash{}n}\StringTok{NMF factor 1"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/moNMFGOTerms-1} 

}

\caption{Top GO-terms associated with NMF factor 1.}\label{fig:moNMFGOTerms}
\end{figure}

\hypertarget{interpretation-using-additional-covariates}{%
\subsection{Interpretation using additional covariates}\label{interpretation-using-additional-covariates}}

Another way to ascribe biological significance to the latent variables is by correlating them with additional covariates we might have about the samples. In our example, the Colorectal cancer tumors have also been characterized for Microsattelite Instability status, using an external test (typically PCR-based). By examining the latent variable values as they relate to a tumor's MSI status, we might discover that we've learned latent factors that are related to it.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{a <-}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{age=}\NormalTok{covariates}\OperatorTok{$}\NormalTok{age, }\DataTypeTok{gender=}\KeywordTok{as.numeric}\NormalTok{(covariates}\OperatorTok{$}\NormalTok{gender), }\DataTypeTok{msi=}\NormalTok{covariates}\OperatorTok{$}\NormalTok{msi)}
\NormalTok{b <-}\StringTok{ }\NormalTok{nmf.h}
\KeywordTok{colnames}\NormalTok{(b) <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{'factor1'}\NormalTok{, }\StringTok{'factor2'}\NormalTok{)}
\NormalTok{cov_factor <-}\StringTok{ }\KeywordTok{cbind}\NormalTok{(a,b)}
\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{ggplot}\NormalTok{(cov_factor, ggplot2}\OperatorTok{::}\KeywordTok{aes}\NormalTok{(}\DataTypeTok{x=}\NormalTok{msi, }\DataTypeTok{y=}\NormalTok{factor1, }\DataTypeTok{group=}\NormalTok{msi)) }\OperatorTok{+}\StringTok{ }\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{geom_boxplot}\NormalTok{() }\OperatorTok{+}\StringTok{ }\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{ggtitle}\NormalTok{(}\StringTok{"NMF factor 1 microsatellite instability"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/moNMFClinicalCovariates-1} 

}

\caption{Box plots show values of NMF factors 1 and 2 for tumors separated by their microsatellite instability status. Factor 1 is high in microsatellite-stable tumors, while factor 2 is high in microsatellite instable tumors.}\label{fig:moNMFClinicalCovariates1}
\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{ggplot}\NormalTok{(cov_factor, ggplot2}\OperatorTok{::}\KeywordTok{aes}\NormalTok{(}\DataTypeTok{x=}\NormalTok{msi, }\DataTypeTok{y=}\NormalTok{factor2, }\DataTypeTok{group=}\NormalTok{msi)) }\OperatorTok{+}\StringTok{ }\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{geom_boxplot}\NormalTok{() }\OperatorTok{+}\StringTok{ }\NormalTok{ggplot2}\OperatorTok{::}\KeywordTok{ggtitle}\NormalTok{(}\StringTok{"NMF factor 2 and microsatellite instability"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{figure}

{\centering \includegraphics{compgenomrReloaded_files/figure-latex/moNMFClinicalCovariates-2} 

}

\caption{Box plots show values of NMF factors 1 and 2 for tumors separated by their microsatellite instability status. Factor 1 is high in microsatellite-stable tumors, while factor 2 is high in microsatellite instable tumors.}\label{fig:moNMFClinicalCovariates2}
\end{figure}

The figures above show that NMF factor 1 values are higher both microsattelite instable tumors, while NMF factor 2 is higher for microsatellite stable tumors. This is an example of multi-omics integration methods finding latent variables which are clinically significant.

\hypertarget{exercises-8}{%
\section{Exercises}\label{exercises-8}}

\hypertarget{easy}{%
\subsubsection{Easy}\label{easy}}

\begin{itemize}
\tightlist
\item
  Find features associated with iCluster and MFA factors, and visualize the feature weights.
\end{itemize}

\hypertarget{intermediate}{%
\subsubsection{Intermediate}\label{intermediate}}

\begin{itemize}
\item
  Normalizing the data matrices by their \(\lambda_1\)'s as in MFA supposes we wish to assign each data type the same importance in the down-stream analysis. This leads to a natural generalizaiton whereby the different data types may be differently weighed. Provide an implementation of weighed-MFA where the different data types may be assigned individual weights.
\item
  In order to use NMF algorithms on data which can be negative, we need to split each feature into two new features, one positive and one negative. Implement the following function, and see that the included test does not fail:
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{split_neg_columns <-}\StringTok{ }\ControlFlowTok{function}\NormalTok{(x) \{}
    \CommentTok{# your code here}
\NormalTok{\}}

\CommentTok{# a test that shows the function above works}
\NormalTok{test_split_neg_columns <-}\StringTok{ }\ControlFlowTok{function}\NormalTok{() \{}
\NormalTok{    input <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(}\KeywordTok{cbind}\NormalTok{(}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{1}\NormalTok{),}\KeywordTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{,}\OperatorTok{-}\DecValTok{2}\NormalTok{)))}
\NormalTok{    output <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(}\KeywordTok{cbind}\NormalTok{(}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{1}\NormalTok{), }\KeywordTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{), }\KeywordTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{,}\DecValTok{0}\NormalTok{), }\KeywordTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{2}\NormalTok{)))}
    \KeywordTok{stopifnot}\NormalTok{(}\KeywordTok{all}\NormalTok{(output }\OperatorTok{==}\StringTok{ }\KeywordTok{split_neg_columns}\NormalTok{(input)))}
\NormalTok{\}}
\KeywordTok{test_split_neg_columns}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{itemize}
\item
  The iCluster+ algorithm has some parameters which may be tuned for maximum performance. The \texttt{iClusterPlus} package has a method, \texttt{iClusterPlus::tune.iClusterPlus}, which does this automatically based on the Bayesian Information Criterion (BIC). Run the example above and find the optimal lambda and alpha values.
\item
  Another covariate in the metadata of these tumors is their \emph{CpG island methylator Phenotype} (CIMP). This is a phenotype carried by a group of colorectal cancers that display hypermethylation of promoter CpG island sites, resulting in the inactivation of some tumor suppressors. This is also assayed using an external test. Do any of the multi-omics methods surveyed find a latent variable that is associated with the tumor's CIMP phenotype?
\end{itemize}

\hypertarget{advanced}{%
\subsubsection{Advanced}\label{advanced}}

\begin{itemize}
\tightlist
\item
  Does MFA give a disentangled representation?
\item
  Does iCluster give disentangled representations? Why do you think that is?
\end{itemize}

\bibliography{book.bib}

\backmatter
\printindex

\end{document}
